<!doctype html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Bin filtering and normalization &#8212; TADbit 1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/forkme_nature.css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Compartments and TADs detection" href="tutorial_8-Compartments_and_TADs_detection.html" />
    <link rel="prev" title="Filtering" href="tutorial_6-Filtering_mapped_reads.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_8-Compartments_and_TADs_detection.html" title="Compartments and TADs detection"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_6-Filtering_mapped_reads.html" title="Filtering"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">TADbit 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Bin filtering and normalization</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="bin-filtering-and-normalization">
<h1>Bin filtering and normalization<a class="headerlink" href="#bin-filtering-and-normalization" title="Permalink to this headline">¶</a></h1>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.parsers.hic_parser</span> <span class="kn">import</span> <span class="n">load_hic_data_from_bam</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell</span> <span class="o">=</span> <span class="s1">&#39;mouse_B&#39;</span>
<span class="n">repl</span> <span class="o">=</span> <span class="s1">&#39;rep1&#39;</span>

<span class="n">reso</span> <span class="o">=</span> <span class="mi">100000</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span>
    <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">.bam&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">repl</span><span class="p">),</span>
    <span class="n">reso</span><span class="p">,</span> <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">(Matrix size 27269x27269)                                                    [2020-02-06 11:17:21]

- Parsing BAM (122 chunks)                                                   [2020-02-06 11:17:21]
   .......... .......... .......... .......... ..........     50/122
   .......... .......... .......... .......... ..........    100/122
   .......... .......... ..                                  122/122

- Getting matrices                                                           [2020-02-06 11:17:53]
   .......... .......... .......... .......... ..........     50/122
   .......... .......... .......... .......... ..........    100/122
   .......... .......... ..                                  122/122</pre>
<section id="filter-bins-with-low-interaction-counts">
<h2>Filter bins with low interaction counts<a class="headerlink" href="#filter-bins-with-low-interaction-counts" title="Permalink to this headline">¶</a></h2>
<p>In this case, the filtering we are applying is divided in two steps: 1.
First we ask to remove all columns having a total of less than 10
interactions (<code class="docutils literal notranslate"><span class="pre">min_count=10</span></code>). This is a relatively loose filtering.
2. We apply a more complex filter. We will remove bins with a total
number of counts much lower than the mean of the interaction counts.</p>
<p><strong>Note</strong><em>: For this (very sparse) dataset, we required very few (10)
interactions per bin. In normal cases, to filter bins with too low
interaction counts one should require that each bin contains a minimum
percentage (default 75%) of cells with no-zero counts, or a minimum
number of interactions in total (for good quality data the recommended
value could be 2500).</em></p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">filter_columns</span><span class="p">(</span><span class="n">draw_hist</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">min_count</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">by_mean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">silent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">WARNING: Using twice min_count as the matrix was symmetricized and contains twice as many interactions as the original</pre>
<img alt="../_images/tutorial_7-Bin_filtering_and_normalization_6_1.png" src="../_images/tutorial_7-Bin_filtering_and_normalization_6_1.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;removed </span><span class="si">{:,}</span><span class="s1"> columns out of </span><span class="si">{:,}</span><span class="s1"> (</span><span class="si">{:.1f}</span><span class="s1">%)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">hic_data</span><span class="o">.</span><span class="n">bads</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">hic_data</span><span class="p">),</span>
                                                           <span class="nb">len</span><span class="p">(</span><span class="n">hic_data</span><span class="o">.</span><span class="n">bads</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">hic_data</span><span class="p">))</span> <span class="o">*</span> <span class="mi">100</span><span class="p">))</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">removed 2,660 columns out of 27,269 (9.8%)</pre>
<p>At 100 kb it makes no sense to view the full matrix (your matrix will
have more cells than your screen have pixels), so it is much more
informative to plot a specific region of the genome only using the
<code class="docutils literal notranslate"><span class="pre">focus</span></code> parameter.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_7-Bin_filtering_and_normalization_10_0.png" src="../_images/tutorial_7-Bin_filtering_and_normalization_10_0.png" />
</section>
<section id="normalization-algorithms">
<h2>Normalization algorithms<a class="headerlink" href="#normalization-algorithms" title="Permalink to this headline">¶</a></h2>
<p><strong>Note</strong><em>: if columns with a lot of zeroes are present the ICE
normalization will last very long to converge, and these low-coverage
columns will present, at the end of the normalization, few cells with
very high values of interaction</em></p>
<section id="iterative-correction-and-eigenvector-decomposition-ice">
<h3>Iterative Correction and Eigenvector decomposition (ICE)<a class="headerlink" href="#iterative-correction-and-eigenvector-decomposition-ice" title="Permalink to this headline">¶</a></h3>
<p>ICE normalization <a class="reference external" href="#cite-Imakaev2012a">(Imakaev et al., 2012)</a>
assumes an equal experimental visibility of each bin and seeks
iteratively for biases that equalize the sum of counts per bin in the
matrix. At each iteration, a new matrix is generated by dividing each
cell by the product of the sum of counts in its row times the sum of
counts in its column. The process converges to a matrix in which all
bins have identical sum.</p>
<p>If <span class="math notranslate nohighlight">\(W\)</span> is the raw matrix, <span class="math notranslate nohighlight">\(N\)</span> is its size, and
<span class="math notranslate nohighlight">\(i\)</span>(<span class="math notranslate nohighlight">\(j\)</span>) the index of the columns(rows), the normalized
matrix <span class="math notranslate nohighlight">\(M\)</span> is iteratively computed as:</p>
<div class="math notranslate nohighlight">
\[M_{i,j} = \frac{W_{i,j}}{\sqrt{\sum_{n=0}^N{W_{i,n}} \times \sum_{n=0}^N{W_{n,j}}}}\]</div>
<p>This normalization has usually a quite strong effect, and visually the
matrices look very smooth and regular.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">(</span><span class="n">iterations</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">max_dev</span><span class="o">=</span><span class="mf">0.00001</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">iterative correction
  - copying matrix
  - computing biases
          1083.000        5384.660      101635.000    0  17.87492
          4407.395        5483.780        8251.728    1   0.50475
          4560.550        5512.003        6238.145    2   0.17261
          5116.644        5521.481        6119.412    3   0.10829
          5173.833        5524.944        5781.389    4   0.06355
          5374.323        5526.241        5761.161    5   0.04251
          5380.193        5526.744        5621.079    6   0.02652
          5470.168        5526.942        5623.500    7   0.01747
          5465.095        5527.021        5561.686    8   0.01120
          5506.121        5527.053        5567.812    9   0.00737
          5500.600        5527.066        5539.771   10   0.00479
          5519.386        5527.071        5544.417   11   0.00314
          5515.770        5527.073        5531.733   12   0.00205
          5524.250        5527.074        5534.470   13   0.00134
          5522.248        5527.075        5528.790   14   0.00087
          5526.033        5527.075        5530.231   15   0.00057
          5525.014        5527.075        5527.710   16   0.00037
          5526.688        5527.075        5528.422   17   0.00024
          5526.195        5527.075        5527.312   18   0.00016
          5526.929        5527.075        5527.650   19   0.00010
          5526.699        5527.075        5527.165   20   0.00007
          5527.019        5527.075        5527.321   21   0.00004
          5526.915        5527.075        5527.110   22   0.00003
          5527.053        5527.075        5527.180   23   0.00002
          5527.006        5527.075        5527.089   24   0.00001
          5527.066        5527.075        5527.120   25   0.00001
rescaling to factor 1
  - getting the sum of the matrix
    =&gt; 24609.000
  - rescaling biases</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>

<span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_7-Bin_filtering_and_normalization_16_0.png" src="../_images/tutorial_7-Bin_filtering_and_normalization_16_0.png" />
</section>
<section id="save-biases">
<h3>Save biases<a class="headerlink" href="#save-biases" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">save_biases</span><span class="p">(</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">_ICE_</span><span class="si">{2}</span><span class="s1">kb.biases&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
    <span class="n">cell</span><span class="p">,</span> <span class="n">repl</span><span class="p">,</span> <span class="n">reso</span> <span class="o">//</span> <span class="mi">1000</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="vanilla-coverage-normalization">
<h3>Vanilla coverage normalization<a class="headerlink" href="#vanilla-coverage-normalization" title="Permalink to this headline">¶</a></h3>
<p>The vanilla normalization <a class="reference external" href="#cite-Rao2014">(Rao et al., 2014)</a> is a
variation of the ICE where a single iteration is performed.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">(</span><span class="n">iterations</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_dev</span><span class="o">=</span><span class="mf">0.00001</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">iterative correction
  - copying matrix
  - computing biases
rescaling to factor 1
  - getting the sum of the matrix
    =&gt; 25061.997
  - rescaling biases</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>

<span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_7-Bin_filtering_and_normalization_22_0.png" src="../_images/tutorial_7-Bin_filtering_and_normalization_22_0.png" />
</section>
<section id="id1">
<h3>Save biases<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">save_biases</span><span class="p">(</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">_Vanilla_</span><span class="si">{2}</span><span class="s1">kb.biases&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
    <span class="n">cell</span><span class="p">,</span> <span class="n">repl</span><span class="p">,</span> <span class="n">reso</span> <span class="o">//</span> <span class="mi">1000</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="square-root-vanilla-coverage-sqrt-normalization">
<h3>Square root vanilla coverage (SQRT) normalization<a class="headerlink" href="#square-root-vanilla-coverage-sqrt-normalization" title="Permalink to this headline">¶</a></h3>
<p>The SQRT vanilla normalization <a class="reference external" href="#cite-Rao2014">(Rao et al., 2014)</a>
is a variation of the Vanilla coverage where each element in the matrix
is divided by the square root of the product of sums of counts.</p>
<div class="math notranslate nohighlight">
\[M_{i,j} = \frac{W_{i,j}}{\sqrt{\sum_{n=0}^N{W_{i,n}} \times \sum_{n=0}^N{W_{n,j}}}}\]</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">(</span><span class="n">iterations</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sqrt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">max_dev</span><span class="o">=</span><span class="mf">0.00001</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">iterative correction
  - copying matrix
  - computing biases
rescaling to factor 1
  - getting the sum of the matrix
    =&gt; 1790593.806
  - rescaling biases</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>

<span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_7-Bin_filtering_and_normalization_28_0.png" src="../_images/tutorial_7-Bin_filtering_and_normalization_28_0.png" />
</section>
<section id="id2">
<h3>Save biases<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">save_biases</span><span class="p">(</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">_SQRT_</span><span class="si">{2}</span><span class="s1">kb.biases&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
    <span class="n">cell</span><span class="p">,</span> <span class="n">repl</span><span class="p">,</span> <span class="n">reso</span> <span class="o">//</span> <span class="mi">1000</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="oned-normalization">
<h3>OneD normalization<a class="headerlink" href="#oned-normalization" title="Permalink to this headline">¶</a></h3>
<p>OneD normalization <a class="reference external" href="#cite-Vidal2018">(Vidal et al., 2018)</a> is based
on fitting a non-linear model between the total amount of contacts per
bin and the known biases: - GC content - number of RE sites (the most
important bias, the more cut sites, the more mapped reads) - read
mappability</p>
<p>As the estimation of each of this statistics is very important for the
normalization, they are left outside the normalization function, in
order to allow user to modify them.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.parsers.bed_parser</span>          <span class="kn">import</span> <span class="n">parse_mappability_bedGraph</span>
<span class="kn">from</span> <span class="nn">pytadbit.utils.normalize_hic</span>         <span class="kn">import</span> <span class="n">oneD</span>
<span class="kn">from</span> <span class="nn">pytadbit.mapping.restriction_enzymes</span> <span class="kn">import</span> <span class="n">RESTRICTION_ENZYMES</span>
<span class="kn">from</span> <span class="nn">pytadbit.parsers.genome_parser</span>       <span class="kn">import</span> <span class="n">parse_fasta</span><span class="p">,</span> <span class="n">get_gc_content</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fasta</span> <span class="o">=</span> <span class="s1">&#39;genome/Mus_musculus-GRCm38.p6/Mus_musculus-GRCm38.p6.fa&#39;</span>
<span class="n">genome</span> <span class="o">=</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="n">fasta</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>get mappability ~2 min</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f_mappability</span> <span class="o">=</span> <span class="s1">&#39;genome/Mus_musculus-GRCm38.p6/Mus_musculus-GRCm38.p6.50mer.bedGraph&#39;</span>

<span class="n">mappability</span> <span class="o">=</span> <span class="n">parse_mappability_bedGraph</span><span class="p">(</span><span class="n">f_mappability</span><span class="p">,</span> <span class="n">reso</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">reduce</span>

<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">mappability</span><span class="p">:</span>
        <span class="n">mappability</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">//</span> <span class="n">reso</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mappability</span><span class="p">[</span><span class="n">c</span><span class="p">])</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">//</span> <span class="n">reso</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">mappability</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span>
            <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">//</span> <span class="n">reso</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">mappability</span><span class="p">[</span><span class="n">c</span><span class="p">]))</span>
<span class="n">mappability</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">+</span> <span class="n">y</span><span class="p">,</span>
                     <span class="p">(</span><span class="n">mappability</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">))</span>
</pre></div>
</div>
<p>get GC content ~ 30 sec</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gc_content</span> <span class="o">=</span> <span class="n">get_gc_content</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">reso</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<p>compute r_sites ~30 sec</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">re_site</span> <span class="o">=</span> <span class="n">RESTRICTION_ENZYMES</span><span class="p">[</span><span class="s1">&#39;MboI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>

<span class="n">n_rsites</span>  <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">crm</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="n">crm</span><span class="p">]),</span> <span class="n">reso</span><span class="p">):</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">crm</span><span class="p">][</span><span class="n">pos</span><span class="p">:</span><span class="n">pos</span> <span class="o">+</span> <span class="n">reso</span> <span class="o">+</span> <span class="mi">400</span><span class="p">]</span>
        <span class="n">n_rsites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">re_site</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sum_cols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">hic_data</span><span class="o">.</span><span class="n">bads</span> <span class="k">else</span>
            <span class="nb">sum</span><span class="p">(</span><span class="mf">0.0</span> <span class="k">if</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">hic_data</span><span class="o">.</span><span class="n">bads</span> <span class="k">else</span> <span class="n">hic_data</span><span class="p">[</span><span class="n">c</span><span class="p">,</span> <span class="n">l</span><span class="p">]</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">hic_data</span><span class="p">)))</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">hic_data</span><span class="p">))]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">biases</span> <span class="o">=</span> <span class="n">oneD</span><span class="p">(</span><span class="n">tot</span><span class="o">=</span><span class="n">sum_cols</span><span class="p">,</span> <span class="nb">map</span><span class="o">=</span><span class="n">mappability</span><span class="p">,</span> <span class="n">res</span><span class="o">=</span><span class="n">n_rsites</span><span class="p">,</span> <span class="n">cg</span><span class="o">=</span><span class="n">gc_content</span><span class="p">)</span>
</pre></div>
</div>
<div class="ansi-block highlight-default notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">bias</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">biases</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>

<span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_7-Bin_filtering_and_normalization_45_0.png" src="../_images/tutorial_7-Bin_filtering_and_normalization_45_0.png" />
</section>
<section id="id3">
<h3>Save biases<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span><span class="o">.</span><span class="n">save_biases</span><span class="p">(</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">_oneD_</span><span class="si">{2}</span><span class="s1">kb.biases&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
    <span class="n">cell</span><span class="p">,</span> <span class="n">repl</span><span class="p">,</span> <span class="n">reso</span> <span class="o">//</span> <span class="mi">1000</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="comparison">
<h3>Comparison<a class="headerlink" href="#comparison" title="Permalink to this headline">¶</a></h3>
<p>By eye, the effect of choosing one normalization over an other is often
very small</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">pytadbit.utils.extraviews</span> <span class="kn">import</span> <span class="n">plot_HiC_matrix</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">22</span><span class="p">,</span> <span class="mi">21</span><span class="p">))</span>
<span class="n">axe</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Raw data&#39;</span><span class="p">)</span>
<span class="n">matrix</span> <span class="o">=</span> <span class="n">hic_data</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3:10000000-40000000&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">plot_HiC_matrix</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">matrix</span><span class="p">),</span> <span class="n">axe</span><span class="o">=</span><span class="n">axe</span><span class="p">,</span><span class="c1">#  vmin=0.5,</span>
                <span class="n">triangular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">axe</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">num</span><span class="p">,</span> <span class="n">norm</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">([</span><span class="s1">&#39;SQRT&#39;</span><span class="p">,</span> <span class="s1">&#39;oneD&#39;</span><span class="p">,</span> <span class="s1">&#39;Vanilla&#39;</span><span class="p">,</span> <span class="s1">&#39;ICE&#39;</span><span class="p">],</span> <span class="mi">3</span><span class="p">):</span>
    <span class="c1"># load biases</span>
    <span class="n">hic_data</span><span class="o">.</span><span class="n">load_biases</span><span class="p">(</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">_</span><span class="si">{3}</span><span class="s1">_</span><span class="si">{2}</span><span class="s1">kb.biases&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
    <span class="n">cell</span><span class="p">,</span> <span class="n">repl</span><span class="p">,</span> <span class="n">reso</span> <span class="o">//</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">norm</span><span class="p">))</span>
    <span class="c1"># retrieve data as matrix</span>
    <span class="n">matrix</span> <span class="o">=</span> <span class="n">hic_data</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3:10000000-40000000&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">masked</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># define subplot</span>
    <span class="n">axe</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">num</span><span class="p">)</span>
    <span class="c1"># draw plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">norm</span><span class="p">)</span>
    <span class="n">plot_HiC_matrix</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">matrix</span><span class="p">),</span> <span class="n">axe</span><span class="o">=</span><span class="n">axe</span><span class="p">,</span>
                    <span class="n">vmin</span><span class="o">=-</span><span class="mf">0.5</span> <span class="k">if</span> <span class="n">norm</span><span class="o">==</span><span class="s1">&#39;oneD&#39;</span> <span class="k">else</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># manually adjust color scale depending on the normalization</span>
                    <span class="n">bad_color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">triangular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">axe</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">/home/dcastillo/miniconda2/envs/py3_tadbit/lib/python3.7/site-packages/ipykernel_launcher.py:5: RuntimeWarning: divide by zero encountered in log2
  &quot;&quot;&quot;
/home/dcastillo/miniconda2/envs/py3_tadbit/lib/python3.7/site-packages/ipykernel_launcher.py:18: RuntimeWarning: divide by zero encountered in log2</pre>
<img alt="../_images/tutorial_7-Bin_filtering_and_normalization_51_1.png" src="../_images/tutorial_7-Bin_filtering_and_normalization_51_1.png" />
</section>
<section id="other-normalizations">
<h3>Other normalizations<a class="headerlink" href="#other-normalizations" title="Permalink to this headline">¶</a></h3>
<p>ICE and Vanilla normalizations are widely used however other, more
convoluted, normalizations <a class="reference external" href="#cite-hu2012hicnorm">(Hu et al., 2012)</a>
<a class="reference external" href="#cite-Yaffe2011">(Yaffe and Tanay, 2011)</a> can be used outside
TADbit and then loaded in TADbit as normalized matrices for further
analysis.</p>
</section>
<section id="best-normalization">
<h3>Best normalization<a class="headerlink" href="#best-normalization" title="Permalink to this headline">¶</a></h3>
<p>Which is the best normalization to use is a question that can not be
answered easily because it depends on the type of data and the type of
analysis.</p>
<p>Most of the time Hi-C experiments are conducted in different conditions
and, for each, in several replicates. A good way to find the best
normalization method may be to select the one that achieve to minimize
the differences between replicates and maximize the differences between
conditions (this in the context of the analysis to be performed).</p>
<!--bibtex
@article{hu2012hicnorm,
  title={HiCNorm: removing biases in Hi-C data via Poisson regression},
  author={Hu, Ming and Deng, Ke and Selvaraj, Siddarth and Qin, Zhaohui and Ren, Bing and Liu, Jun S},
  journal={Bioinformatics},
  volume={28},
  number={23},
  pages={3131--3133},
  year={2012},
  publisher={Oxford Univ Press}
}
@article{Yaffe2011,
abstract = {Hi-C experiments measure the probability of physical proximity between pairs of chromosomal loci on a genomic scale. We report on several systematic biases that substantially affect the Hi-C experimental procedure, including the distance between restriction sites, the GC content of trimmed ligation junctions and sequence uniqueness. To address these biases, we introduce an integrated probabilistic background model and develop algorithms to estimate its parameters and renormalize Hi-C data. Analysis of corrected human lymphoblast contact maps provides genome-wide evidence for interchromosomal aggregation of active chromatin marks, including DNase-hypersensitive sites and transcriptionally active foci. We observe extensive long-range (up to 400 kb) cis interactions at active promoters and derive asymmetric contact profiles next to transcription start sites and CTCF binding sites. Clusters of interacting chromosomal domains suggest physical separation of centromere-proximal and centromere-distal regions. These results provide a computational basis for the inference of chromosomal architectures from Hi-C experiments.},
author = {Yaffe, Eitan and Tanay, Amos},
doi = {10.1038/ng.947},
file = {:home/fransua/.local/share/data/Mendeley Ltd./Mendeley Desktop/Downloaded/Yaffe, Tanay - 2011 - Probabilistic modeling of Hi-C contact maps eliminates systematic biases to characterize global chromosomal archit.pdf:pdf},
issn = {1546-1718},
journal = {Nature genetics},
keywords = {Binding Sites,Chromosomes,Cluster Analysis,Epigenesis,Genetic,Human,Humans,Lymphocytes,Lymphocytes: ultrastructure,Models,Probability},
mendeley-groups = {Research articles},
month = {nov},
number = {11},
pages = {1059--65},
pmid = {22001755},
title = {{Probabilistic modeling of Hi-C contact maps eliminates systematic biases to characterize global chromosomal architecture.}},
url = {http://www.ncbi.nlm.nih.gov/pubmed/22001755},
volume = {43},
year = {2011}
}
@article{Imakaev2012a,
abstract = {Extracting biologically meaningful information from chromosomal interactions obtained with genome-wide chromosome conformation capture (3C) analyses requires the elimination of systematic biases. We present a computational pipeline that integrates a strategy to map sequencing reads with a data-driven method for iterative correction of biases, yielding genome-wide maps of relative contact probabilities. We validate this ICE (iterative correction and eigenvector decomposition) technique on published data obtained by the high-throughput 3C method Hi-C, and we demonstrate that eigenvector decomposition of the obtained maps provides insights into local chromatin states, global patterns of chromosomal interactions, and the conserved organization of human and mouse chromosomes.},
author = {Imakaev, Maxim V and Fudenberg, Geoffrey and McCord, Rachel Patton and Naumova, Natalia and Goloborodko, Anton and Lajoie, Bryan R and Dekker, Job and Mirny, Leonid A},
doi = {10.1038/nmeth.2148},
file = {:home/fransua/.local/share/data/Mendeley Ltd./Mendeley Desktop/Downloaded/Imakaev et al. - 2012 - Iterative correction of Hi-C data reveals hallmarks of chromosome organization.pdf:pdf},
issn = {1548-7105},
journal = {Nature methods},
keywords = {Hi-C},
mendeley-groups = {stats/Hi-C,Research articles},
mendeley-tags = {Hi-C},
month = {oct},
number = {10},
pages = {999--1003},
pmid = {22941365},
title = {{Iterative correction of Hi-C data reveals hallmarks of chromosome organization.}},
url = {http://www.ncbi.nlm.nih.gov/pubmed/22941365},
volume = {9},
year = {2012}
}
@article{Rao2014,
author = {Rao, Suhas S P and Huntley, Miriam H and Durand, Neva C and Stamenova, Elena K and Bochkov, Ivan D. and {James T. Robinson} and Sanborn, Adrian L. and Machol, Ido and Omer, Arina D. and Lander, Eric S. and Lieberman-Aiden, Erez},
doi = {10.1016/j.cell.2014.11.021},
file = {:home/fransua/.local/share/data/Mendeley Ltd./Mendeley Desktop/Downloaded/Rao et al. - 2014 - A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping.pdf:pdf},
issn = {0092-8674},
journal = {Cell},
keywords = {Hi-C},
mendeley-groups = {Research articles,projects/GEVO/CTCF},
mendeley-tags = {Hi-C},
number = {7},
pages = {1665--1680},
pmid = {25497547},
publisher = {Elsevier Inc.},
title = {{A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping}},
url = {http://dx.doi.org/10.1016/j.cell.2014.11.021},
volume = {159},
year = {2014}
}

--></section>
<section id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<p>[^](#ref-1) Imakaev, Maxim V and Fudenberg, Geoffrey and McCord, Rachel
Patton and Naumova, Natalia and Goloborodko, Anton and Lajoie, Bryan R
and Dekker, Job and Mirny, Leonid A. 2012. <em>Iterative correction of Hi-C
data reveals hallmarks of chromosome organization.</em>.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/22941365">URL</a></p>
<p>[^](#ref-2) Rao, Suhas S P and Huntley, Miriam H and Durand, Neva C and
Stamenova, Elena K and Bochkov, Ivan D. and James T. Robinson and
Sanborn, Adrian L. and Machol, Ido and Omer, Arina D. and Lander, Eric
S. and Lieberman-Aiden, Erez. 2014. <em>A 3D Map of the Human Genome at
Kilobase Resolution Reveals Principles of Chromatin Looping</em>.
<a class="reference external" href="http://dx.doi.org/10.1016/j.cell.2014.11.021">URL</a></p>
<p>[^](#ref-3) Enrique Vidal, François le Dily, Javier Quilez, Ralph
Stadhouders, Yasmina Cuartero, Thomas Graf, Marc A Marti-Renom, Miguel
Beato, Guillaume J Filion. 2018. <em>OneD: increasing reproducibility of
Hi-C samples with abnormal karyotypes</em>.
<a class="reference external" href="https://doi.org/10.1093/nar/gky064">URL</a></p>
<p>[^](#ref-4) Hu, Ming and Deng, Ke and Selvaraj, Siddarth and Qin,
Zhaohui and Ren, Bing and Liu, Jun S. 2012. <em>HiCNorm: removing biases in
Hi-C data via Poisson regression</em>.</p>
<p>[^](#ref-5) Yaffe, Eitan and Tanay, Amos. 2011. <em>Probabilistic modeling
of Hi-C contact maps eliminates systematic biases to characterize global
chromosomal architecture.</em>.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/22001755">URL</a></p>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_1-Retrieve_published_HiC_datasets.html">Retrieve HiC dataset from NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_2-Preparation_of_the_reference_genome.html">Preparation of the reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3-Hi-C_quality_check.html">Hi-C quality check</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html">Iterative vs fragment-based mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html">Load mapped read-ends and generate a matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html#quality-check-of-the-hi-c-experiment">Quality check of the Hi-C experiment</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_6-Filtering_mapped_reads.html">Filtering</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bin filtering and normalization</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_6-Filtering_mapped_reads.html"
                        title="previous chapter">Filtering</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_8-Compartments_and_TADs_detection.html"
                        title="next chapter">Compartments and TADs detection</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_8-Compartments_and_TADs_detection.html" title="Compartments and TADs detection"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_6-Filtering_mapped_reads.html" title="Filtering"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">TADbit 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Bin filtering and normalization</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Sep 25, 2021.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.2.0.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>