<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Mapping paired-end reads from NGS experiment (with GEM) &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Tutorials" href="../tutorial.html" />
    <link rel="next" title="Getting started" href="tutorial_1_general.html" />
    <link rel="prev" title="Tutorials" href="../tutorial.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="mapping-paired-end-reads-from-ngs-experiment-with-gem">
<h1>Mapping paired-end reads from NGS experiment (with GEM)<a class="headerlink" href="#mapping-paired-end-reads-from-ngs-experiment-with-gem" title="Permalink to this headline">¶</a></h1>
<div class="section" id="download-and-index-reference-genome">
<h2>Download and index reference genome<a class="headerlink" href="#download-and-index-reference-genome" title="Permalink to this headline">¶</a></h2>
<p>Download last human reference genome from ucsc:
<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/">http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/</a> It&#8217;s better to
download the file that contains each chromosome separately (like this we
could only keep those which we are interested in): hg38.chromFa.tar.gz
Uncompress the genome</p>
<div class="highlight-python"><div class="highlight"><pre>tar -xzf hg38.chromFa.tar.gz
</pre></div>
</div>
<p>Check the header of the fasta files (it&#8217;s important to know how
chromosomes are called in these files):</p>
<div class="highlight-python"><div class="highlight"><pre>head chr1.fa
</pre></div>
</div>
<p>Create a single file with all chromosomes together:</p>
<div class="highlight-python"><div class="highlight"><pre>cat chr1.fa chr2.fa chr3.fa chr4.fa chr5.fa chr6.fa chr7.fa chr8.fa chr9.fa chr10.fa chr11.fa chr12.fa chr13.fa chr14.fa  chr15.fa  chr16.fa  chr17.fa  chr18.fa  chr19.fa chr20.fa  chr21.fa  chr22.fa chrX.fa  chrY.fa chrM.fa &gt; hg38.fa
</pre></div>
</div>
<p>Index reference genome:</p>
<div class="highlight-python"><div class="highlight"><pre>gemtools index -i hg38.fa -t 8
</pre></div>
</div>
<p>Creates a file hg38.gem that we will be used for the mapping.</p>
</div>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<div class="section" id="retrieving-hi-c-experiment-sra-format">
<h3>Retrieving Hi-C experiment SRA format<a class="headerlink" href="#retrieving-hi-c-experiment-sra-format" title="Permalink to this headline">¶</a></h3>
<p>Along this tutorial we are going to use the dataset from [Dixon2012]_
that can be downloaded at:
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE35156">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE35156</a></p>
<p>For this part of the tutorial we are going to work with the experiments
conducted on <em>Homo sapiens</em> IMR90 and ESC (replicate 1 in both cases).</p>
<p>Each of this replica consists of severa SRA files:</p>
<div class="code python highlight-python"><div class="highlight"><pre>! ls ~/db/FASTQs/hsap/dixon_2012/*/*/*.sra
</pre></div>
</div>
<pre class="ansi-block literal-block">
ls: cannot access /home/fransua/db/FASTQs/hsap/dixon_2012/*/*/*.sra: No such file or directory
</pre>
</div>
<div class="section" id="first-quality-check-on-the-data">
<h3>First quality check on the data<a class="headerlink" href="#first-quality-check-on-the-data" title="Permalink to this headline">¶</a></h3>
<p>First we need to convert SRA files in FASTQ format, as TADbit does not
support reading directly from SRA. For this purpose we may want to use
the SRA-toolkit from the NCBI (<a class="reference external" href="https://github.com/ncbi/sratoolkit">https://github.com/ncbi/sratoolkit</a>). We
finally shall endup with a single FASTQ file per experimental replica,
with commands like (can last for hours):</p>
<div class="code python highlight-python"><div class="highlight"><pre>%%bash
fastq-dump -DQ &#39;+&#39; --defline-seq &#39;@$ac.$si&#39; --split-files -A SRX116345 --gzip -O /scratch/db/FASTQs/hsap/dixon_2012/
</pre></div>
</div>
<pre class="ansi-block literal-block">
2016-04-12T13:43:58 fastq-dump.2.5.7 err: item not found while constructing within virtual database module - the path 'SRX116345' cannot be opened as database or table
</pre>
<p>fastq-dump -A SRX116344 &#8211;gzip SRX116344/<em>/</em>.sra</p>
<p>the compressed FASTQ files should be:</p>
<div class="code python highlight-python"><div class="highlight"><pre>! ls ~/db/FASTQs/hsap/dixon_2012/
</pre></div>
</div>
<pre class="ansi-block literal-block">
SRX116344  SRX116345
</pre>
<p>Usig these files directly we can infer the quality of the Hi-C
experiment with TADbit:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.utils.fastq_utils</span> <span class="kn">import</span> <span class="n">quality_plot</span>

<span class="n">quality_plot</span><span class="p">(</span><span class="s1">&#39;/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116344.fastq.gz&#39;</span><span class="p">,</span>
             <span class="n">nreads</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">paired</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_15_0.png" src="../_images/tutorial_0_mapping_15_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.utils.fastq_utils</span> <span class="kn">import</span> <span class="n">quality_plot</span>

<span class="n">quality_plot</span><span class="p">(</span><span class="s1">&#39;/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116345.fastq.gz&#39;</span><span class="p">,</span>
             <span class="n">nreads</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">paired</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_16_0.png" src="../_images/tutorial_0_mapping_16_0.png" />
</div>
<div class="section" id="iterative-mapping">
<h3>Iterative mapping<a class="headerlink" href="#iterative-mapping" title="Permalink to this headline">¶</a></h3>
<p>TADbit implements the iterative mapping as proposed by [Imakaev2012]_</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.mapper</span> <span class="kn">import</span> <span class="n">iterative_mapping</span>

<span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;SRX116344_200&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.full_mapper</span> <span class="kn">import</span> <span class="n">full_mapping</span>

<span class="n">mapped_r1</span> <span class="o">=</span> <span class="n">full_mapping</span><span class="p">(</span><span class="s1">&#39;ref_genome/hg38.gem&#39;</span><span class="p">,</span> <span class="s1">&#39;/scratch/db/FASTQs/hsap/dixon_2012/</span><span class="si">%s</span><span class="s1">.fastq&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
                         <span class="s1">&#39;mapped_read1&#39;</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="mi">25</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">75</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">50</span><span class="p">),(</span><span class="mi">1</span><span class="p">,</span><span class="mi">100</span><span class="p">)),</span>
                         <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">frag_map</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">nthreads</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">clean</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;tmp&#39;</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Preparing FASTQ file
  - conversion to MAP format
  - trimming reads 1-25
Mapping reads in window 1-25...
TO GEM /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_XS30yn
Parsing result...
   x removing GEM input /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_XS30yn
   x removing map /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_XS30yn_full_1-25.map
Preparing MAP file
  - trimming reads 1-75
   x removing original input /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_XS30yn_filt_1-25.map
Mapping reads in window 1-75...
TO GEM /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_UZgsV7
Parsing result...
   x removing GEM input /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_UZgsV7
   x removing map /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_UZgsV7_full_1-75.map
Preparing MAP file
  - trimming reads 1-50
   x removing original input /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_UZgsV7_filt_1-75.map
Mapping reads in window 1-50...
TO GEM /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_EFl7SN
Parsing result...
   x removing GEM input /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_EFl7SN
   x removing map /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_EFl7SN_full_1-50.map
Preparing MAP file
  - trimming reads 1-100
   x removing original input /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_EFl7SN_filt_1-50.map
Mapping reads in window 1-100...
TO GEM /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_1EnI_f
Parsing result...
   x removing GEM input /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_1EnI_f
   x removing map /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_1EnI_f_full_1-100.map
  - splitting into restriction enzyme (RE) fragments using ligation sites
  - ligation sites are replaced by RE sites to match the reference genome
    * enzyme: HindIII, ligation site: AAGCTAGCTT, RE site: AAGCTT
Preparing MAP file
  - trimming reads 1-100
Mapping fragments of remaining reads...
TO GEM /home/fransua/Box/tadbits/tadbit-doc/doc/notebooks/tmp/SRX116344_200_kX6mt9
Parsing result...
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;Mapping Read 1&#39;</span>

<span class="n">sams1</span> <span class="o">=</span> <span class="n">iterative_mapping</span><span class="p">(</span>
            <span class="n">gem_index_path</span>       <span class="o">=</span> <span class="s1">&#39;ref_genome/hg38.gem&#39;</span><span class="p">,</span>
            <span class="n">fastq_path</span>           <span class="o">=</span> <span class="s1">&#39;/scratch/db/FASTQs/hsap/dixon_2012/</span><span class="si">%s</span><span class="s1">.fastq&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">out_sam_path</span>         <span class="o">=</span> <span class="s1">&#39;/scratch/mapped_reads/</span><span class="si">%s</span><span class="s1">_r1.sam&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">temp_dir</span>             <span class="o">=</span> <span class="s1">&#39;/scratch/tmp_dir_</span><span class="si">%s</span><span class="s1">/&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">range_start</span>          <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">],</span> <span class="c1"># starts with a flag sequence</span>
            <span class="n">range_stop</span>           <span class="o">=</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">35</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">55</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">65</span><span class="p">,</span> <span class="mi">70</span><span class="p">,</span> <span class="mi">75</span><span class="p">],</span>
            <span class="n">nthreads</span>             <span class="o">=</span> <span class="mi">8</span><span class="p">,</span>  <span class="c1"># number of CPUs to use</span>
            <span class="n">max_reads_per_chunk</span>  <span class="o">=</span> <span class="mi">40000000</span><span class="p">,</span> <span class="c1"># this is for computers with not too much RAM</span>
            <span class="n">single_end</span>           <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="k">print</span> <span class="s1">&#39;  created </span><span class="si">%s</span><span class="s1"> SAM files:&#39;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">sams1</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping Read 1
Split input file /scratch/db/FASTQs/hsap/dixon_2012/SRX116344_200.fastq into chunks
2 chunks obtained
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
  created 24 SAM files:
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Mapping Read 2&#39;</span>

<span class="n">sams2</span> <span class="o">=</span> <span class="n">iterative_mapping</span><span class="p">(</span>
            <span class="n">gem_index_path</span>       <span class="o">=</span> <span class="s1">&#39;ref_genome/hg38.gem&#39;</span><span class="p">,</span>
            <span class="n">fastq_path</span>           <span class="o">=</span> <span class="s1">&#39;/scratch/db/FASTQs/hsap/dixon_2012/</span><span class="si">%s</span><span class="s1">.fastq&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">out_sam_path</span>         <span class="o">=</span> <span class="s1">&#39;/scratch/mapped_reads/</span><span class="si">%s</span><span class="s1">_r2.sam&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">temp_dir</span>             <span class="o">=</span> <span class="s1">&#39;/scratch/tmp_dir_</span><span class="si">%s</span><span class="s1">/&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">range_start</span>          <span class="o">=</span> <span class="p">[</span><span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">],</span>
            <span class="n">range_stop</span>           <span class="o">=</span> <span class="p">[</span><span class="mi">121</span><span class="p">,</span> <span class="mi">126</span><span class="p">,</span> <span class="mi">131</span><span class="p">,</span> <span class="mi">136</span><span class="p">,</span> <span class="mi">141</span><span class="p">,</span> <span class="mi">146</span><span class="p">,</span> <span class="mi">151</span><span class="p">,</span> <span class="mi">156</span><span class="p">,</span> <span class="mi">161</span><span class="p">,</span> <span class="mi">166</span><span class="p">,</span> <span class="mi">171</span><span class="p">,</span> <span class="mi">176</span><span class="p">],</span>
            <span class="n">nthreads</span>             <span class="o">=</span> <span class="mi">8</span><span class="p">,</span>
            <span class="n">max_reads_per_chunk</span>  <span class="o">=</span> <span class="mi">40000000</span><span class="p">,</span>
            <span class="n">single_end</span>           <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>

<span class="k">print</span> <span class="s1">&#39;  created </span><span class="si">%s</span><span class="s1"> SAM files:&#39;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">sams2</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping Read 2
Split input file /scratch/db/FASTQs/hsap/dixon_2012/SRX116344_200.fastq into chunks
2 chunks obtained
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
  created 24 SAM files:
</pre>
</div>
<div class="section" id="map-parsing">
<h3>Map parsing<a class="headerlink" href="#map-parsing" title="Permalink to this headline">¶</a></h3>
<p>In case we lost the sams1/sams2 lists of paths we can &#8220;reload&#8221; them like
this:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">os</span>

<span class="n">sams1</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">sams2</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">sam</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="s1">&#39;/scratch/mapped_reads/&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="s1">&#39;_r1.sam&#39;</span> <span class="ow">in</span> <span class="n">sam</span><span class="p">:</span>
        <span class="n">sams1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;/scratch/mapped_reads/&#39;</span> <span class="o">+</span> <span class="n">sam</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sams2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;/scratch/mapped_reads/&#39;</span> <span class="o">+</span> <span class="n">sam</span><span class="p">)</span>
<span class="k">print</span> <span class="n">sams1</span>
<span class="k">print</span> <span class="n">sams2</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
['/scratch/mapped_reads/SRX116344_200_r1.sam.1.3:1-30', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.6:1-45', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.7:1-50', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.8:1-55', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.12:1-75', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.12:1-75', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.6:1-45', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.1:1-20', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.1:1-20', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.9:1-60', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.4:1-35', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.10:1-65', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.4:1-35', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.9:1-60', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.5:1-40', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.2:1-25', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.8:1-55', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.7:1-50', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.11:1-70', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.2:1-25', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.3:1-30', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.5:1-40', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.11:1-70', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.10:1-65']
['/scratch/mapped_reads/SRX116344_200_r2.sam.1.11:102-171', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.2:102-126', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.9:102-161', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.12:102-176', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.5:102-141', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.2:102-126', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.8:102-156', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.10:102-166', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.5:102-141', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.9:102-161', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.7:102-151', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.6:102-146', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.4:102-136', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.1:102-121', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.11:102-171', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.4:102-136', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.7:102-151', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.6:102-146', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.3:102-131', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.8:102-156', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.10:102-166', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.3:102-131', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.1:102-121', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.12:102-176']
</pre>
<div class="section" id="mapping-analysis">
<h4>Mapping analysis<a class="headerlink" href="#mapping-analysis" title="Permalink to this headline">¶</a></h4>
<p>We collect mapped reads at all window sizes into a single file (a single
file for read1, and a single file for read2). These 2 files also contain
the placement of the restriction enzyme sites in the genome.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.parsers.sam_parser</span>    <span class="kn">import</span> <span class="n">parse_sam</span>
<span class="kn">from</span> <span class="nn">pytadbit.parsers.genome_parser</span> <span class="kn">import</span> <span class="n">parse_fasta</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c1"># loads the genome</span>
<span class="n">genome_seq</span> <span class="o">=</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="s1">&#39;/scratch/db/index_files/Homo_sapiens-79/Homo_sapiens.fa&#39;</span><span class="p">,</span>
                         <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;MT&#39;</span><span class="p">]])</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Parsing 1 as chr1
Parsing 2 as chr2
Parsing 3 as chr3
Parsing 4 as chr4
Parsing 5 as chr5
Parsing 6 as chr6
Parsing 7 as chr7
Parsing 8 as chr8
Parsing 9 as chr9
Parsing 10 as chr10
Parsing 11 as chr11
Parsing 12 as chr12
Parsing 13 as chr13
Parsing 14 as chr14
Parsing 15 as chr15
Parsing 16 as chr16
Parsing 17 as chr17
Parsing 18 as chr18
Parsing 19 as chr19
Parsing 20 as chr20
Parsing 21 as chr21
Parsing 22 as chr22
Parsing X as chrX
Parsing Y as chrY
Parsing MT as chrMT
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c1"># new file with info of each &quot;read1&quot; and its placement with respect to RE sites</span>
<span class="n">reads1</span> <span class="o">=</span> <span class="s1">&#39;/scratch/results/</span><span class="si">%s</span><span class="s1">_r1_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span>
<span class="c1"># new file with info of each &quot;read2&quot; and its placement with respect to RE sites</span>
<span class="n">reads2</span> <span class="o">=</span> <span class="s1">&#39;/scratch/results/</span><span class="si">%s</span><span class="s1">_r2_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;Parse SAM files...&#39;</span>
<span class="n">parse_sam</span><span class="p">(</span><span class="n">sams1</span><span class="p">,</span> <span class="n">sams2</span><span class="p">,</span> <span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">genome_seq</span><span class="p">,</span> <span class="s1">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Parse SAM files...
Searching and mapping RE sites to the reference genome
Found 859855 RE sites
Loading read1
Loading read2
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.3:1-30
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.6:1-45
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.7:1-50
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.8:1-55
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.12:1-75
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.12:1-75
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.6:1-45
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.1:1-20
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.2:1-25
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.9:1-60
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.10:1-65
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.1:1-20
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.4:1-35
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.4:1-35
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.9:1-60
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.2:1-25
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.3:1-30
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.11:1-70
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.8:1-55
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.11:1-70
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.5:1-40
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.7:1-50
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.10:1-65
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.5:1-40
Sorting reads
Removing temporary files...loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.11:102-171
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.2:102-126
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.9:102-161
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.12:102-176
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.5:102-141
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.2:102-126
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.10:102-166
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.8:102-156
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.11:102-171
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.6:102-146
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.5:102-141
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.9:102-161
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.7:102-151
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.1:102-121
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.4:102-136
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.6:102-146
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.4:102-136
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.3:102-131
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.7:102-151
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.8:102-156
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.3:102-131
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.10:102-166
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.12:102-176
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.1:102-121
</pre>
</div>
</div>
<div class="section" id="plot-iterative-mapping">
<h3>Plot iterative mapping<a class="headerlink" href="#plot-iterative-mapping" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">plot_iterative_mapping</span>

<span class="n">lengths</span> <span class="o">=</span> <span class="n">plot_iterative_mapping</span><span class="p">(</span><span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">total_reads</span><span class="o">=</span><span class="mi">51493359</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_33_0.png" src="../_images/tutorial_0_mapping_33_0.png" />
</div>
<div class="section" id="merging-mapped-read1-and-read2">
<h3>Merging mapped &#8220;read1&#8221; and &#8220;read2&#8221;<a class="headerlink" href="#merging-mapped-read1-and-read2" title="Permalink to this headline">¶</a></h3>
<p>We create a new file that will contain only the reads mapped in both
ends (&#8220;read1&#8221; and &#8220;read2&#8221; uniquely mapped)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.mapper</span> <span class="kn">import</span> <span class="n">get_intersection</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">reads</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/results/</span><span class="si">%s</span><span class="s1">_both_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">get_intersection</span><span class="p">(</span><span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">reads</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Found 41123290 pair of reads mapping uniquely
</pre>
<div class="section" id="descriptive-statistics">
<h4>Descriptive statistics<a class="headerlink" href="#descriptive-statistics" title="Permalink to this headline">¶</a></h4>
<p>In order to confirm the size of the inserts fed to the sequencer, we can
look at the distribution of genomic distances between the mapped read1
and read2 of dangling-ends. From this analysis we can extract the
maximum insert size, that is an important value to classify reads during
the filtering process.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">insert_sizes</span>

<span class="n">insert_sizes</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_41_0.png" src="../_images/tutorial_0_mapping_41_0.png" />
</div>
</div>
<div class="section" id="simple-descriptive-stats">
<h3>Simple descriptive stats<a class="headerlink" href="#simple-descriptive-stats" title="Permalink to this headline">¶</a></h3>
<div class="section" id="how-the-count-in-interaction-falls-as-the-genomic-distance-is-larger">
<h4>How the count in interaction falls as the genomic distance is larger<a class="headerlink" href="#how-the-count-in-interaction-falls-as-the-genomic-distance-is-larger" title="Permalink to this headline">¶</a></h4>
<p>Here we want to see how the interaction between to two genomic region
decays as the distance between these two loci is larger. Theexpectation
is that at distances between 700 kb and 10 Mb the decay in logarithm
scale is -1.</p>
<p>In the example below are represented the interactions in between genomic
regions that, each, spans over 10 kb (resolution parameter).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">plot_distance_vs_interactions</span>

<span class="n">plot_distance_vs_interactions</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">max_diff</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_45_0.png" src="../_images/tutorial_0_mapping_45_0.png" />
</div>
<div class="section" id="genomic-coverage-of-our-reads">
<h4>Genomic coverage of our reads<a class="headerlink" href="#genomic-coverage-of-our-reads" title="Permalink to this headline">¶</a></h4>
<p>In the plot above we want to see the distribution of the reads mapped in
the genome (regardless of their interactions). Here, te expecation is to
see a minimum number of reads mapping in all positions of the genome
with falls around centromeres and telomeres.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">plot_genomic_distribution</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_49_0.png" src="../_images/tutorial_0_mapping_49_0.png" />
<p><em>The picks in the plot correspond to PCR artifact that we will remove in
the filtering step (see bellow)</em></p>
<p>This plot can be zoomed in the y axis in order to avoid depending on
these artifacts. The plot can also be generated only for a given number
of chromosomes</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">),</span>
                          <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr8&#39;</span><span class="p">],</span> <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_52_0.png" src="../_images/tutorial_0_mapping_52_0.png" />
</div>
<div class="section" id="interaction-matrix">
<h4>Interaction matrix<a class="headerlink" href="#interaction-matrix" title="Permalink to this headline">¶</a></h4>
<p>The plot above is probablythe most informative, in order to infer the
qualtity of an Hi-C experiment. This plot represents the matrix of
interaction, the distribution of these interaction as an histogram or as
a function of genomic distance. Some statistics on the specificity of
these interaction, like the cis-to-trans ratio (expected to be between
40 and 60%), and the 3 first eigen vectors of the matrix highlighting
the principal structural features of the matrix (in non-normalized
matrices eigen-vectors are not very informative however).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
/usr/lib/python2.7/dist-packages/numpy/core/_methods.py:55: RuntimeWarning: Mean of empty slice.
  warnings.warn(&quot;Mean of empty slice.&quot;, RuntimeWarning)
/usr/lib/python2.7/dist-packages/numpy/core/_methods.py:67: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
</pre>
<img alt="../_images/tutorial_0_mapping_56_1.png" src="../_images/tutorial_0_mapping_56_1.png" />
</div>
</div>
</div>
<div class="section" id="filter-reads">
<h2>Filter reads<a class="headerlink" href="#filter-reads" title="Permalink to this headline">¶</a></h2>
<p>In order to remove interactions between reads that are experimental
artifacts, or just uninoformative, a series of adjustable filters can be
applied:</p>
<ol class="arabic simple">
<li>self-circle : reads are comming from a single RE fragment and point
to the outside (—-&lt;===—===&gt;—)</li>
<li>dangling-end : reads are comming from a single RE fragment and point
to the inside (—-===&gt;—&lt;===—)</li>
<li>error : reads are comming from a single RE fragment and point in the
same direction</li>
<li>extra dangling-end : reads are comming from different RE fragment
but are close enough (&lt; max_molecule length) and point to the
inside</li>
<li>too close from RES : semi-dangling-end filter, start position of one
of the read is too close (5 bp by default) from RE cutting site.</li>
<li>too short : remove reads comming from small restriction less than
100 bp (default) because they are comparable to the read length</li>
<li>too large : remove reads comming from large restriction fragments
(default: 100 Kb, P &lt; 10-5 to occur in a randomized genome) as they
likely represent poorly assembled or repeated regions</li>
<li>over-represented : reads coming from the top 0.5% most frequently
detected restriction fragments, they may be prone to PCR artifacts
or represent fragile regions of the genome or genome assembly errors</li>
<li>duplicated : the combination of the start positions of the reads is
repeated -&gt; PCR artifact (only keep one copy)</li>
<li>random breaks : start position of one of the read is too far (more
than min_dist_to_re) from RE cutting site. Non-canonical enzyme
activity or random physical breakage of the chromatin.</li>
</ol>
<p>The function <code class="docutils literal"><span class="pre">filter_reads</span></code> works in parallel (4 threads), and creates
one file per filter (10 files, which path are an extension of the input
file containing the reads).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.filter</span> <span class="kn">import</span> <span class="n">filter_reads</span>

<span class="n">masked</span> <span class="o">=</span> <span class="n">filter_reads</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">max_molecule_length</span><span class="o">=</span><span class="mi">505</span><span class="p">,</span> <span class="n">min_dist_to_re</span><span class="o">=</span><span class="mi">760</span><span class="p">,</span>
                      <span class="n">over_represented</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">max_frag_size</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span>
                      <span class="n">min_frag_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">re_proximity</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Filtered reads (and percentage of total):

     TOTAL mapped              :     41123290 (100.00%)
  -----------------------------------------------------
   1- self-circle               :       653029 (  1.59%)
   2- dangling-end              :     17054088 ( 41.47%)
   3- error                     :       250158 (  0.61%)
   4- extra dangling-end        :      4115590 ( 10.01%)
   5- too close from RES        :      5761851 ( 14.01%)
   6- too short                 :       307169 (  0.75%)
   7- too large                 :        63151 (  0.15%)
   8- over-represented          :      1334209 (  3.24%)
   9- duplicated                :      9132048 ( 22.21%)
  10- random breaks             :     18282481 ( 44.46%)
</pre>
<p>Previous function creates one file per filter. Each containing the list
of IDs of the reads falling into the corresponding filter. In order to
apply filter, the function <code class="docutils literal"><span class="pre">apply_filter</span></code> will create a new file
without the reads contained in the files. By default all filters are
applied.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.filter</span> <span class="kn">import</span> <span class="n">apply_filter</span>
<span class="n">filt_reads</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/results/</span><span class="si">%s</span><span class="s1">_filtered_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">apply_filter</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">filt_reads</span><span class="p">,</span> <span class="n">masked</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
9403143 reads written to file
</pre>
<p>An example to apply only the 9 first filters:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">apply_filter</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">filt_reads</span><span class="p">,</span> <span class="n">masked</span><span class="p">,</span> <span class="n">filters</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">])</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
13666735 reads written to file
</pre>
<p>Filters can also be applied in a &#8220;reverse&#8221; way in order to select only
&#8220;bad&#8221; reads.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">sc_de</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/results/</span><span class="si">%s</span><span class="s1">_self_circles_and_dangling-ends.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span>

<span class="n">apply_filter</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">sc_de</span><span class="p">,</span> <span class="n">masked</span><span class="p">,</span> <span class="n">filters</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
17707117 reads written to file
</pre>
<p>This can be used for example to analyze the distribution of
dangling-ends and self-circle along the genome</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">sc_de</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr8&#39;</span><span class="p">])</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_69_0.png" src="../_images/tutorial_0_mapping_69_0.png" />
<p>Once filtered the peaks previously seen should disapeear:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr8&#39;</span><span class="p">],</span>
                         <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">250</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_71_0.png" src="../_images/tutorial_0_mapping_71_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr8&#39;</span><span class="p">],</span>
                         <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">250</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_72_0.png" src="../_images/tutorial_0_mapping_72_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_73_0.png" src="../_images/tutorial_0_mapping_73_0.png" />
<p>These maps can be zoomed to a given region, like first chromosome:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">250</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_75_0.png" src="../_images/tutorial_0_mapping_75_0.png" />
<p>Same as above, calling the focus using directly chromosome name and
using a smaller resolution (100 kb):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr1&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_77_0.png" src="../_images/tutorial_0_mapping_77_0.png" />
</div>
<div class="section" id="filtering-and-normalization">
<h2>Filtering and normalization<a class="headerlink" href="#filtering-and-normalization" title="Permalink to this headline">¶</a></h2>
<div class="section" id="removal-of-columns-having-to-few-data">
<h3>Removal of columns having to few data<a class="headerlink" href="#removal-of-columns-having-to-few-data" title="Permalink to this headline">¶</a></h3>
<p>Depending on the normalization method, the presence of columns with high
proportion of zeros can prevent to converge into a satisfactory result.</p>
<p>For this part of the processing of the data we will start to work on
full matrices. This step is critical in the sense that <strong>we have to
decide at which resolution we are going to analyze the data</strong>. For this
tutorial we will use a resolution of 1 Mb.</p>
<p><em>Note</em> : as all previous steps ended in the generation of a single file,
we just need to load the name of the saved file with the filtered reads:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;SRX116344_200&#39;</span>
<span class="n">filt_reads</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/results/</span><span class="si">%s</span><span class="s1">_filtered_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="kn">import</span> <span class="n">load_hic_data_from_reads</span>

<span class="n">hic_data</span> <span class="o">=</span> <span class="n">load_hic_data_from_reads</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">)</span>
</pre></div>
</div>
<p>We can visualize the matrix using the same function as before, with the
file of reads:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>

<span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_85_0.png" src="../_images/tutorial_0_mapping_85_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_data</span><span class="o">.</span><span class="n">filter_columns</span><span class="p">(</span><span class="n">draw_hist</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
WARNING: removing columns having more than 2327 zeroes:
   123   124   125   126   127   128   129   130   131   132   133   134   135   136   137   138   139   140   141   142
   143   145   340   492   585   742   882   930   931   932   953  1124  1294  1295  1296  1440  1441  1541  1585  1586
  1587  1588  1589  1590  1591  1592  1593  1594  1595  1596  1597  1598  1599  1600  1601  1602  1603  1607  1680  1721
  1866  1867  1868  1950  1986  2085  2086  2087  2088  2089  2090  2091  2092  2093  2094  2095  2096  2097  2098  2099
  2100  2101  2102  2200  2201  2202  2203  2204  2205  2206  2207  2208  2209  2210  2211  2212  2213  2214  2215  2216
  2217  2218  2307  2308  2309  2310  2311  2312  2313  2314  2315  2316  2317  2318  2319  2320  2321  2322  2323  2324
  2325  2326  2327  2329  2446  2447  2448  2449  2450  2451  2452  2453  2454  2455  2500  2525  2601  2602  2603  2604
  2605  2691  2692  2752  2790  2791  2792  2793  2794  2795  2796  2797  2801  2802  2837  2838  2839  2840  2841  2842
  2843  2844  2845  2846  2847  2850  2851  2852  2888  2889  2890  2947  2948  2949  3044  3045  3046  3055  3067  3068
  3069  3070  3071  3072  3073  3074  3075  3076  3077  3078  3079  3080  3081  3082  3083  3084  3085  3086  3087  3088
  3089  3090  3091  3092  3093  3094  3095  3096  3097  3098  3099  3100  3101  3102  3103
/usr/lib/python2.7/dist-packages/numpy/core/numeric.py:460: ComplexWarning: Casting complex values to real discards the imaginary part
  return array(a, dtype, copy=False, order=order)
</pre>
<img alt="../_images/tutorial_0_mapping_86_1.png" src="../_images/tutorial_0_mapping_86_1.png" />
<pre class="ansi-block literal-block">
/usr/local/lib/python2.7/dist-packages/pytadbit/utils/hic_filtering.py:145: ComplexWarning: Casting complex values to real discards the imaginary part
  round(root, 3), ' '.join(

WARNING: removing columns having less than 1381.618 counts:
   123   124   125   126   127   128   129   130   131   132   133   134   135   136   137   138   139   140   141   142
   143   144   145   340   492   585   691   742   882   930   931   932   953  1124  1294  1295  1296  1440  1441  1541
  1585  1586  1587  1588  1589  1590  1591  1592  1593  1594  1595  1596  1597  1598  1599  1600  1601  1602  1603  1604
  1607  1609  1680  1721  1866  1867  1868  1950  1986  2085  2086  2087  2088  2089  2090  2091  2092  2093  2094  2095
  2096  2097  2098  2099  2100  2101  2102  2200  2201  2202  2203  2204  2205  2206  2207  2208  2209  2210  2211  2212
  2213  2214  2215  2216  2217  2218  2307  2308  2309  2310  2311  2312  2313  2314  2315  2316  2317  2318  2319  2320
  2321  2322  2323  2324  2325  2326  2327  2329  2446  2447  2448  2449  2450  2451  2452  2453  2454  2455  2500  2525
  2584  2601  2602  2603  2604  2605  2691  2692  2752  2790  2791  2792  2793  2794  2795  2796  2797  2801  2802  2837
  2838  2839  2840  2841  2842  2843  2844  2845  2846  2847  2850  2851  2852  2888  2889  2890  2947  2948  2949  3044
  3045  3046  3047  3055  3067  3068  3069  3070  3071  3072  3073  3074  3075  3076  3077  3078  3079  3080  3081  3082
  3083  3084  3085  3086  3087  3088  3089  3090  3091  3092  3093  3094  3095  3096  3097  3098  3099  3100  3101  3102
  3103
</pre>
<p>Filtered columns (to high count of zeroes, or to low mean value) will be
skipped in most of analysis available and are now shaded in the matrix
representation:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
/usr/lib/python2.7/dist-packages/numpy/core/_methods.py:55: RuntimeWarning: Mean of empty slice.
  warnings.warn(&quot;Mean of empty slice.&quot;, RuntimeWarning)
/usr/lib/python2.7/dist-packages/numpy/core/_methods.py:67: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
</pre>
<img alt="../_images/tutorial_0_mapping_88_1.png" src="../_images/tutorial_0_mapping_88_1.png" />
</div>
<div class="section" id="normalization">
<h3>Normalization<a class="headerlink" href="#normalization" title="Permalink to this headline">¶</a></h3>
<p>TADbit implements ICE normalization strategy [Imakaev2012]_ which
basically consists constructing a new in dividing each cell</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_data</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">(</span><span class="n">iterations</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">max_dev</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
iterative correction
          1392.000        9454.176       13609.000    0   0.85276
          8583.102        9678.681       18479.813    1   0.90933
          7057.620        9725.242       10177.855    2   0.27430
          9531.512        9733.579       11274.686    3   0.15833
          9076.352        9735.375        9827.806    4   0.06769
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_92_0.png" src="../_images/tutorial_0_mapping_92_0.png" />
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="">Mapping paired-end reads from NGS experiment (with GEM)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads">Identification and Analysis of TADs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../tutorial.html"
                        title="previous chapter">Tutorials</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_1_general.html"
                        title="next chapter">Getting started</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             >next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright .
      Last updated on Apr 12, 2016.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>
