<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Experimental data assessment and model parameters optimisation &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Modelling a region of the chromatin" href="tutorial_11-3D_Models_production_and_analysis.html" />
    <link rel="prev" title="Comparison of Hi-C experiments" href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_11-3D_Models_production_and_analysis.html" title="Modelling a region of the chromatin"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" title="Comparison of Hi-C experiments"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="experimental-data-assessment-and-model-parameters-optimisation">
<h1>Experimental data assessment and model parameters optimisation<a class="headerlink" href="#experimental-data-assessment-and-model-parameters-optimisation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="data-preparation">
<h2>Data preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">¶</a></h2>
<p>The first step to generate three-dimensional (3D) models of a specific
genomic regions is to filter columns with low counts and with no
diagonal count in order to remove outliers or problematic columns from
the interaction matrix. The particles associated with the filtered
columns will be modelled, but will have no experimental data applied.</p>
<p>Here we load the data previous data already normalised.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="kn">import</span> <span class="n">load_chromosome</span>
<span class="kn">from</span> <span class="nn">pytadbit.parsers.hic_parser</span> <span class="kn">import</span> <span class="n">load_hic_data_from_bam</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span> <span class="o">=</span> <span class="n">load_chromosome</span><span class="p">(</span><span class="s1">&#39;results/fragment/chr3.tdb&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">B</span><span class="p">,</span> <span class="n">PSC</span> <span class="o">=</span> <span class="n">crm</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">B</span><span class="p">,</span> <span class="n">PSC</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">(Experiment mouse_B (resolution: 100 kb, TADs: 96, Hi-C rows: 1601, normalized: None),
 Experiment mouse_PSC (resolution: 100 kb, TADs: 118, Hi-C rows: 1601, normalized: None))</pre>
<p>Load raw data matrices, and normalized matrices</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">base_path</span> <span class="o">=</span> <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">.bam&#39;</span>
<span class="n">bias_path</span> <span class="o">=</span> <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/04_normalizing/biases_</span><span class="si">{0}</span><span class="s1">_both_</span><span class="si">{1}</span><span class="s1">kb.biases&#39;</span>
<span class="n">reso</span> <span class="o">=</span> <span class="mi">100000</span>
<span class="n">chrname</span> <span class="o">=</span> <span class="s1">&#39;chr3&#39;</span>
<span class="n">cel1</span> <span class="o">=</span> <span class="s1">&#39;mouse_B&#39;</span>
<span class="n">cel2</span> <span class="o">=</span> <span class="s1">&#39;mouse_PSC&#39;</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data1</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel1</span><span class="p">),</span>
                                   <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                   <span class="n">region</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span>
                                   <span class="n">biases</span><span class="o">=</span><span class="n">bias_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel1</span><span class="p">,</span> <span class="n">reso</span> <span class="o">//</span> <span class="mi">1000</span><span class="p">),</span>
                                   <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">hic_data2</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel2</span><span class="p">),</span>
                                   <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                   <span class="n">region</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span>
                                   <span class="n">biases</span><span class="o">=</span><span class="n">bias_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel2</span><span class="p">,</span> <span class="n">reso</span> <span class="o">//</span> <span class="mi">1000</span><span class="p">),</span>
                                   <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">(Matrix size 1601x1601)                                                      [2020-02-06 16:39:39]

- Parsing BAM (101 chunks)                                                   [2020-02-06 16:39:39]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101

- Getting matrices                                                           [2020-02-06 16:39:42]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101


(Matrix size 1601x1601)                                                      [2020-02-06 16:39:46]

- Parsing BAM (101 chunks)                                                   [2020-02-06 16:39:46]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101

- Getting matrices                                                           [2020-02-06 16:39:50]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">B</span><span class="o">.</span><span class="n">load_hic_data</span><span class="p">([</span><span class="n">hic_data1</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">)])</span>
<span class="n">B</span><span class="o">.</span><span class="n">load_norm_data</span><span class="p">([</span><span class="n">hic_data1</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">)])</span>

<span class="n">PSC</span><span class="o">.</span><span class="n">load_hic_data</span><span class="p">([</span><span class="n">hic_data2</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">)])</span>
<span class="n">PSC</span><span class="o">.</span><span class="n">load_norm_data</span><span class="p">([</span><span class="n">hic_data2</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">)])</span>
</pre></div>
</div>
<p>It is a good practice to check that the data is there:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">visualize</span><span class="p">([</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">],</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_11_0.png" src="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_11_0.png" />
<p>Focus on the genomic region to model.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">visualize</span><span class="p">([</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">],</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">360</span><span class="p">))</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_13_0.png" src="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_13_0.png" />
</div>
<div class="section" id="data-modellability-assessment-via-mmp-score">
<h2>Data modellability assessment via MMP score<a class="headerlink" href="#data-modellability-assessment-via-mmp-score" title="Permalink to this headline">¶</a></h2>
<p>We can use the Matrix Modeling Potential (MMP) score (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4402535/">Trussart M. et
al. Nature Communication,
2017</a>) to
identify a priori whether the interaction matrices have the potential of
being use for modeling. The MMP score ranges from 0 to 1 and combines
three different measures: the contribution of the significant
eigenvectors, the skewness and the kurtosis of the distribution of
Z-scores.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.utils.three_dim_stats</span> <span class="kn">import</span> <span class="n">mmp_score</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mmp_score</span><span class="p">(</span><span class="n">hic_data1</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3:30000000-36000000&#39;</span><span class="p">),</span>
          <span class="n">savefig</span><span class="o">=</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/mmp_score.png&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel1</span><span class="p">))</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">(0.8049308283731964,
 0.7885888244416531,
 0.6736346044021908,
 0.9006021702003049)</pre>
<p><img alt="image0" src="tutorial/../nbpictures/Contact_map.png" /></p>
</div>
<div class="section" id="data-transformation-and-scoring-function">
<h2>Data Transformation and scoring function<a class="headerlink" href="#data-transformation-and-scoring-function" title="Permalink to this headline">¶</a></h2>
<p>This step is automatically done in TADbit. A a weight is generated for
each pair of interactions proportional to their interaction count as in
formula:</p>
<div class="math notranslate nohighlight">
\[weight(I, J) = \frac{\sum^N_{i=0}{\sum^N_{j=0}{(matrix(i, j))}}}{\sum^N_{i=0}{(matrix(i, J))} \times \sum^N_{j=0}{(matrix(I, j))}}\]</div>
<p>The raw data are then multiplied by this weight. In the case that
multiple experiments are used, the weighted interaction values are
normalised using a factor (default set as 1) in order to compare between
experiments. Then, a Z-score of the off-diagonal normalised/weighted
interaction is calculated as in formula:</p>
<div class="math notranslate nohighlight">
\[zscore(I, J) = \frac{log_{10}(weight(I, J) \times matrix(I, J)) - mean(log_{10}(weight \times matrix))}{stddev(log_{10}(weight \times matrix))}\]</div>
<p>The Z-scores are then transformed to distance restraints. To define the
type of restraints between each pair of particles. we need to identified
empirically three optimal parameters (i) a maximal distance between two
non-interacting particles (maxdist), (ii) a lower-bound cutoff to define
particles that do not interact frequently (lowfreq) and (iii) an
upper-bound cutoff to define particles that do interact frequently
(upfreq). In TADbit this is done via a grid search approach.</p>
<p>The following picture shows the different component of the scoring
funtion that is optimised during the Monte Carlo simulated annealing
sampling protocol. Two consecutive particles are spatially restrained by
a harmonic oscillator with an equilibrium distance that corresponds to
the sum of their radii. Non-consecutive particles with contact
frequencies above the upper-bound cutoff are restrained by a harmonic
oscillator at an equilibrium distance, while those below the lower-bound
cutoff are maintained further than an equilibrium distance by a lower
bound harmonic oscillator.</p>
<p><img alt="image0" src="tutorial/../nbpictures/Contact_map.png" /></p>
<div class="section" id="optimization-of-parameters">
<h3>Optimization of parameters<a class="headerlink" href="#optimization-of-parameters" title="Permalink to this headline">¶</a></h3>
<p>We need to identified empirically (via a grid-search optimisation) the
optimal parameters for the mdoelling procedure:</p>
<ul class="simple">
<li><p>maxdist: maximal distance assosiated two interacting particles.</p></li>
<li><p>upfreq: to define particles that do interact frequently (defines
attraction)</p></li>
<li><p>lowfreq: to define particles that do not interact frequently (
defines repulsion)</p></li>
<li><p>dcutoff: the definition of “contact” in units of bead diameter. Value
of 2 means that a contact will occur when 2 beads are closer than 2
times their diameter. This will be used to compare 3D models with
Hi-C interaction maps.</p></li>
</ul>
<p>Pairs of beads interacting less than lowfreq (left dashed line) are
penalized if they are closer than their assigned minimum distance
(Harmonic lower bound). Pairs of beads interacting more than ufreq
(right dashed line) are penalized if they are further apart than their
assigned maximum distance (Harmonic upper bound). Pairs of beads which
interaction fall in between lowfreq and upfreq are not penalized except
if they are neighbours (Harmonic)</p>
<p><img alt="image0" src="tutorial/../nbpictures/Contact_map.png" /></p>
<p>In the parameter optimization step we are going to give a set of ranges
for the different search parameters. For each possible combination
TADbit will produce a set of models.</p>
<p>In each individual model we consider that two beads are in contact if
their distance in 3D space is lower than the specified distance cutoff.
TADbit builds a cumulative contact map for each set of models as shown
in the schema below. The contact map is then compared with the Hi-C
interaction experiment by means of a Spearman correlation coefficient.
The sets having higher correlation coefficients are those that best
represents the original data.</p>
<p><img alt="image0" src="tutorial/../nbpictures/Contact_map.png" /></p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span> <span class="o">=</span> <span class="n">B</span><span class="o">.</span><span class="n">optimal_imp_parameters</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">360</span><span class="p">,</span> <span class="n">n_models</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">n_keep</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
                                   <span class="n">upfreq_range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">),</span>
                                   <span class="n">lowfreq_range</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">),</span>
                                   <span class="n">maxdist_range</span><span class="o">=</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">,</span> <span class="mi">500</span><span class="p">),</span>
                                   <span class="n">dcutoff_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">Optimizing 61 particles
  num scale kbending        maxdist lowfreq upfreq  dcutoff correlation</pre>
<pre class="ansi-block literal-block">1   0.01  0               1000    -0.9    0       4      0.2021
1   0.01  0               1000    -0.9    0       3      0.5706
1   0.01  0               1000    -0.9    0       2      0.8769
2   0.01  0               1000    -0.9    0.3     4      0.3995
2   0.01  0               1000    -0.9    0.3     3      0.6799
2   0.01  0               1000    -0.9    0.3     2      0.8864
3   0.01  0               1000    -0.9    0.6     4      0.62
3   0.01  0               1000    -0.9    0.6     3      0.7709
3   0.01  0               1000    -0.9    0.6     2      0.8531
4   0.01  0               1000    -0.6    0       4      0.205
4   0.01  0               1000    -0.6    0       3      0.58
4   0.01  0               1000    -0.6    0       2      0.8773
5   0.01  0               1000    -0.6    0.3     4      0.4014
5   0.01  0               1000    -0.6    0.3     3      0.6827
5   0.01  0               1000    -0.6    0.3     2      0.8869
6   0.01  0               1000    -0.6    0.6     4      0.6181
6   0.01  0               1000    -0.6    0.6     3      0.776
6   0.01  0               1000    -0.6    0.6     2      0.8502
7   0.01  0               1000    -0.3    0       4      0.2084
7   0.01  0               1000    -0.3    0       3      0.5789
7   0.01  0               1000    -0.3    0       2      0.877
8   0.01  0               1000    -0.3    0.3     4      0.3999
8   0.01  0               1000    -0.3    0.3     3      0.6746
8   0.01  0               1000    -0.3    0.3     2      0.8824
9   0.01  0               1000    -0.3    0.6     4      0.6224
9   0.01  0               1000    -0.3    0.6     3      0.7678
9   0.01  0               1000    -0.3    0.6     2      0.8425
10  0.01  0               1000    0       0       4      0.19
10  0.01  0               1000    0       0       3      0.5669
10  0.01  0               1000    0       0       2      0.8763
11  0.01  0               1000    0       0.3     4      0.404
11  0.01  0               1000    0       0.3     3      0.6753
11  0.01  0               1000    0       0.3     2      0.8883
12  0.01  0               1000    0       0.6     4      0.6052
12  0.01  0               1000    0       0.6     3      0.7689
12  0.01  0               1000    0       0.6     2      0.8443
13  0.01  0               1500    -0.9    0       4      0.3574
13  0.01  0               1500    -0.9    0       3      0.6476
13  0.01  0               1500    -0.9    0       2      0.8926
14  0.01  0               1500    -0.9    0.3     4      0.4787
14  0.01  0               1500    -0.9    0.3     3      0.7314
14  0.01  0               1500    -0.9    0.3     2      0.8977
15  0.01  0               1500    -0.9    0.6     4      0.6623
15  0.01  0               1500    -0.9    0.6     3      0.7958
15  0.01  0               1500    -0.9    0.6     2      0.8401
16  0.01  0               1500    -0.6    0       4      0.3513
16  0.01  0               1500    -0.6    0       3      0.646
16  0.01  0               1500    -0.6    0       2      0.8951
17  0.01  0               1500    -0.6    0.3     4      0.4728
17  0.01  0               1500    -0.6    0.3     3      0.7341
17  0.01  0               1500    -0.6    0.3     2      0.8985
18  0.01  0               1500    -0.6    0.6     4      0.6491
18  0.01  0               1500    -0.6    0.6     3      0.7945
18  0.01  0               1500    -0.6    0.6     2      0.8407
19  0.01  0               1500    -0.3    0       4      0.367
19  0.01  0               1500    -0.3    0       3      0.6571
19  0.01  0               1500    -0.3    0       2      0.8897
20  0.01  0               1500    -0.3    0.3     4      0.4695
20  0.01  0               1500    -0.3    0.3     3      0.7355
20  0.01  0               1500    -0.3    0.3     2      0.9004
21  0.01  0               1500    -0.3    0.6     4      0.6568
21  0.01  0               1500    -0.3    0.6     3      0.7947
21  0.01  0               1500    -0.3    0.6     2      0.8391
22  0.01  0               1500    0       0       4      0.3722
22  0.01  0               1500    0       0       3      0.6566
22  0.01  0               1500    0       0       2      0.8943
23  0.01  0               1500    0       0.3     4      0.4745
23  0.01  0               1500    0       0.3     3      0.7278
23  0.01  0               1500    0       0.3     2      0.8946
24  0.01  0               1500    0       0.6     4      0.6529
24  0.01  0               1500    0       0.6     3      0.7951
24  0.01  0               1500    0       0.6     2      0.8404
25  0.01  0               2000    -0.9    0       4      0.4613
25  0.01  0               2000    -0.9    0       3      0.7304
25  0.01  0               2000    -0.9    0       2      0.8991
26  0.01  0               2000    -0.9    0.3     4      0.5551
26  0.01  0               2000    -0.9    0.3     3      0.7915
26  0.01  0               2000    -0.9    0.3     2      0.8977
27  0.01  0               2000    -0.9    0.6     4      0.6894
27  0.01  0               2000    -0.9    0.6     3      0.8012
27  0.01  0               2000    -0.9    0.6     2      0.8377
28  0.01  0               2000    -0.6    0       4      0.4717
28  0.01  0               2000    -0.6    0       3      0.7292
28  0.01  0               2000    -0.6    0       2      0.9036
29  0.01  0               2000    -0.6    0.3     4      0.5526
29  0.01  0               2000    -0.6    0.3     3      0.7815
29  0.01  0               2000    -0.6    0.3     2      0.8998
30  0.01  0               2000    -0.6    0.6     4      0.7032
30  0.01  0               2000    -0.6    0.6     3      0.8049
30  0.01  0               2000    -0.6    0.6     2      0.8387
31  0.01  0               2000    -0.3    0       4      0.4678
31  0.01  0               2000    -0.3    0       3      0.715
31  0.01  0               2000    -0.3    0       2      0.8939
32  0.01  0               2000    -0.3    0.3     4      0.5525
32  0.01  0               2000    -0.3    0.3     3      0.7861
32  0.01  0               2000    -0.3    0.3     2      0.8994
33  0.01  0               2000    -0.3    0.6     4      0.7005
33  0.01  0               2000    -0.3    0.6     3      0.804
33  0.01  0               2000    -0.3    0.6     2      0.8399
34  0.01  0               2000    0       0       4      0.4594
34  0.01  0               2000    0       0       3      0.7167
34  0.01  0               2000    0       0       2      0.8992
35  0.01  0               2000    0       0.3     4      0.5481
35  0.01  0               2000    0       0.3     3      0.7883
35  0.01  0               2000    0       0.3     2      0.8981
36  0.01  0               2000    0       0.6     4      0.692
36  0.01  0               2000    0       0.6     3      0.8032
36  0.01  0               2000    0       0.6     2      0.8541</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">show_best</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_23_0.png" src="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_23_0.png" />
<p>Refine optimization in a small region:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">run_grid_search</span><span class="p">(</span><span class="n">upfreq_range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">),</span> <span class="n">lowfreq_range</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.9</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">),</span>
                       <span class="n">maxdist_range</span><span class="o">=</span><span class="p">[</span><span class="mi">1750</span><span class="p">],</span>
                       <span class="n">dcutoff_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
                       <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">Optimizing 61 particles
  num scale kbending        maxdist lowfreq upfreq  dcutoff correlation</pre>
<pre class="ansi-block literal-block">1   0.01  0               1750    -0.9    0       3      0.6859
1   0.01  0               1750    -0.9    0       2      0.8985
2   0.01  0               1750    -0.9    0.3     3      0.7587
2   0.01  0               1750    -0.9    0.3     2      0.9
3   0.01  0               1750    -0.6    0       3      0.6879
3   0.01  0               1750    -0.6    0       2      0.8937
4   0.01  0               1750    -0.6    0.3     3      0.7521
4   0.01  0               1750    -0.6    0.3     2      0.8993
5   0.01  0               1750    -0.3    0       3      0.6857
5   0.01  0               1750    -0.3    0       2      0.9
6   0.01  0               1750    -0.3    0.3     3      0.7644
6   0.01  0               1750    -0.3    0.3     2      0.8981</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">show_best</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_26_0.png" src="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_26_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">run_grid_search</span><span class="p">(</span><span class="n">upfreq_range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">),</span> <span class="n">lowfreq_range</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
                       <span class="n">maxdist_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2000</span><span class="p">,</span> <span class="mi">2250</span><span class="p">],</span>
                       <span class="n">dcutoff_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                       <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">xx        0.01    0               2000    -0.3    0       4.0     0.4678
xx        0.01    0               2000    -0.3    0.3     4.0     0.5525</pre>
<pre class="ansi-block literal-block">Optimizing 61 particles
  num scale kbending        maxdist lowfreq upfreq  dcutoff correlation</pre>
<pre class="ansi-block literal-block">1   0.01  0               2000    -0.2    0       2      0.9001
2   0.01  0               2000    -0.2    0.3     2      0.8989
3   0.01  0               2000    -0.1    0       2      0.901
4   0.01  0               2000    -0.1    0.3     2      0.8967
xx        0.01    0               2000    0       0       4.0     0.4594
xx        0.01    0               2000    0       0.3     4.0     0.5481
5   0.01  0               2250    -0.3    0       2      0.8962
6   0.01  0               2250    -0.3    0.3     2      0.8964
7   0.01  0               2250    -0.2    0       2      0.9024
8   0.01  0               2250    -0.2    0.3     2      0.9004
9   0.01  0               2250    -0.1    0       2      0.8992
10  0.01  0               2250    -0.1    0.3     2      0.9009
11  0.01  0               2250    0       0       2      0.9001
12  0.01  0               2250    0       0.3     2      0.8967</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">show_best</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_28_0.png" src="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_28_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">run_grid_search</span><span class="p">(</span><span class="n">upfreq_range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span> <span class="n">lowfreq_range</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
                       <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
                       <span class="n">maxdist_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2000</span><span class="p">,</span> <span class="mi">2250</span><span class="p">],</span>
                       <span class="n">dcutoff_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">xx        0.01    0               2000    -0.3    0       4.0     0.4678</pre>
<pre class="ansi-block literal-block">Optimizing 61 particles
  num scale kbending        maxdist lowfreq upfreq  dcutoff correlation</pre>
<pre class="ansi-block literal-block">1   0.01  0               2000    -0.3    0.1     2      0.898
2   0.01  0               2000    -0.3    0.2     2      0.8979
xx        0.01    0               2000    -0.3    0.3     4.0     0.5525
xx        0.01    0               2000    -0.2    0       2.0     0.9001
3   0.01  0               2000    -0.2    0.1     2      0.8908
4   0.01  0               2000    -0.2    0.2     2      0.8944
xx        0.01    0               2000    -0.2    0.3     2.0     0.8989
xx        0.01    0               2000    -0.1    0       2.0     0.901
5   0.01  0               2000    -0.1    0.1     2      0.8967
6   0.01  0               2000    -0.1    0.2     2      0.8941
xx        0.01    0               2000    -0.1    0.3     2.0     0.8967
xx        0.01    0               2000    0       0       4.0     0.4594
7   0.01  0               2000    0       0.1     2      0.9006
8   0.01  0               2000    0       0.2     2      0.8914
xx        0.01    0               2000    0       0.3     4.0     0.5481
xx        0.01    0               2250    -0.3    0       2.0     0.8962
9   0.01  0               2250    -0.3    0.1     2      0.8991
10  0.01  0               2250    -0.3    0.2     2      0.8957
xx        0.01    0               2250    -0.3    0.3     2.0     0.8964
xx        0.01    0               2250    -0.2    0       2.0     0.9024
11  0.01  0               2250    -0.2    0.1     2      0.8984
12  0.01  0               2250    -0.2    0.2     2      0.8993
xx        0.01    0               2250    -0.2    0.3     2.0     0.9004
xx        0.01    0               2250    -0.1    0       2.0     0.8992
13  0.01  0               2250    -0.1    0.1     2      0.9024
14  0.01  0               2250    -0.1    0.2     2      0.8959
xx        0.01    0               2250    -0.1    0.3     2.0     0.9009
xx        0.01    0               2250    0       0       2.0     0.9001
15  0.01  0               2250    0       0.1     2      0.9033
16  0.01  0               2250    0       0.2     2      0.895
xx        0.01    0               2250    0       0.3     2.0     0.8967</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">show_best</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_30_0.png" src="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_30_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_B</span><span class="o">.</span><span class="n">get_best_parameters_dict</span><span class="p">()</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">{'scale': 0.01,
 'kbending': 0.0,
 'maxdist': 2000.0,
 'lowfreq': -0.6,
 'upfreq': 0.0,
 'dcutoff': 2.0,
 'reference': '',
 'kforce': 5}</pre>
<p>For the other replicate, we can reduce the space of search:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_PSC</span> <span class="o">=</span> <span class="n">PSC</span><span class="o">.</span><span class="n">optimal_imp_parameters</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">360</span><span class="p">,</span> <span class="n">n_models</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">n_keep</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
                                     <span class="n">upfreq_range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
                                     <span class="n">lowfreq_range</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.3</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
                                     <span class="n">maxdist_range</span><span class="o">=</span><span class="p">(</span><span class="mi">2000</span><span class="p">,</span> <span class="mi">2250</span><span class="p">,</span> <span class="mi">250</span><span class="p">),</span>
                                     <span class="n">dcutoff_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">Optimizing 61 particles
  num scale kbending        maxdist lowfreq upfreq  dcutoff correlation</pre>
<pre class="ansi-block literal-block">1   0.01  0               2000    -0.3    0       2      0.9309
2   0.01  0               2000    -0.3    0.1     2      0.9237
3   0.01  0               2000    -0.3    0.2     2      0.9177
4   0.01  0               2000    -0.3    0.3     2      0.9084
5   0.01  0               2000    -0.2    0       2      0.9268
6   0.01  0               2000    -0.2    0.1     2      0.9245
7   0.01  0               2000    -0.2    0.2     2      0.9159
8   0.01  0               2000    -0.2    0.3     2      0.9095
9   0.01  0               2000    -0.1    0       2      0.9283
10  0.01  0               2000    -0.1    0.1     2      0.9239
11  0.01  0               2000    -0.1    0.2     2      0.9136
12  0.01  0               2000    -0.1    0.3     2      0.9071
13  0.01  0               2250    -0.3    0       2      0.9285
14  0.01  0               2250    -0.3    0.1     2      0.9233
15  0.01  0               2250    -0.3    0.2     2      0.9165
16  0.01  0               2250    -0.3    0.3     2      0.9068
17  0.01  0               2250    -0.2    0       2      0.9294
18  0.01  0               2250    -0.2    0.1     2      0.9214
19  0.01  0               2250    -0.2    0.2     2      0.9154
20  0.01  0               2250    -0.2    0.3     2      0.9041
21  0.01  0               2250    -0.1    0       2      0.9304
22  0.01  0               2250    -0.1    0.1     2      0.9219
23  0.01  0               2250    -0.1    0.2     2      0.9157
24  0.01  0               2250    -0.1    0.3     2      0.9076</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_PSC</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">show_best</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_34_0.png" src="tutorial/../nbpictures//tutorial_10-Modeling_parameters_optimization_34_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opt_PSC</span><span class="o">.</span><span class="n">get_best_parameters_dict</span><span class="p">()</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">{'scale': 0.01,
 'kbending': 0.0,
 'maxdist': 2000.0,
 'lowfreq': -0.3,
 'upfreq': 0.0,
 'dcutoff': 2.0,
 'reference': '',
 'kforce': 5}</pre>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Experimental data assessment and model parameters optimisation</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_11-3D_Models_production_and_analysis.html">Modelling a region of the chromatin</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html"
                        title="previous chapter">Comparison of Hi-C experiments</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_11-3D_Models_production_and_analysis.html"
                        title="next chapter">Modelling a region of the chromatin</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_11-3D_Models_production_and_analysis.html" title="Modelling a region of the chromatin"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" title="Comparison of Hi-C experiments"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Feb 07, 2020.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>