<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>The HiC_data object &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Compartments" href="tutorial_8-Compartments_and_TADs_detection.html" />
    <link rel="prev" title="Filtering" href="tutorial_6-Filtering_mapped_reads.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_8-Compartments_and_TADs_detection.html" title="Compartments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_6-Filtering_mapped_reads.html" title="Filtering"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1><p>Table of Contents</p>
</h1><div class="toc"><ul class="toc-item"><li><p>The HiC_data object</p>
</li><li><p>Filter bins with low interaction counts</p>
</li><li><p>Normalization algorithms</p>
<ul class="toc-item"><li><p>Iterative Correction and Eigenvector decomposition (ICE)</p>
<ul class="toc-item"><li><p>Save biases</p>
</li></ul></li><li><p>Vanilla coverage normalization</p>
<ul class="toc-item"><li><p>Save biases</p>
</li></ul></li><li><p>Square root vanilla coverage (SQRT) normalization</p>
<ul class="toc-item"><li><p>Save biases</p>
</li></ul></li><li><p>OneD normalization</p>
<ul class="toc-item"><li><p>Save biases</p>
</li></ul></li><li><p>Comparison</p>
</li><li><p>Other normalizations</p>
</li><li><p>Best normalization</p>
</li></ul></li><li><p>References</p>
</li></ul></div><div class="section" id="the-hic-data-object">
<h1>The HiC_data object<a class="headerlink" href="#the-hic-data-object" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="filter-bins-with-low-interaction-counts">
<h1>Filter bins with low interaction counts<a class="headerlink" href="#filter-bins-with-low-interaction-counts" title="Permalink to this headline">¶</a></h1>
<p>In this case, the filtering we are applying is divided in two steps: 1.
First we ask to remove all columns having a total of less than 10
interactions (<code class="docutils literal notranslate"><span class="pre">min_count=10</span></code>). This is a relatively loose filtering.
2. We apply a more complex filter. We will remove bins with a total
number of counts much lower than the mean of the interaction counts.</p>
<p><strong>*Note</strong>: For this (very sparse) dataset, we required very few (10)
interactions per bin. In normal cases, to filter bins with too low
interaction counts one should require that each bin contains a minimum
percentage (default 75%) of cells with no-zero counts, or a minimum
number of interactions in total (for good quality data the recommended
value could be 2500).*</p>
<img alt="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_7_1.png" src="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_7_1.png" />
<p>At 100 kb it makes no sense to view the full matrix (your matrix will
have more cells than your screen have pixels), so it is much more
informative to plot a specific region of the genome only using the
<code class="docutils literal notranslate"><span class="pre">focus</span></code> parameter.</p>
<img alt="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_12_1.png" src="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_12_1.png" />
</div>
<div class="section" id="normalization-algorithms">
<h1>Normalization algorithms<a class="headerlink" href="#normalization-algorithms" title="Permalink to this headline">¶</a></h1>
<p><strong>*Note</strong>: if columns with a lot of zeroes are present the ICE
normalization will last very long to converge, and these low-coverage
columns will present, at the end of the normalization, few cells with
very high values of interaction*</p>
<div class="section" id="iterative-correction-and-eigenvector-decomposition-ice">
<h2>Iterative Correction and Eigenvector decomposition (ICE)<a class="headerlink" href="#iterative-correction-and-eigenvector-decomposition-ice" title="Permalink to this headline">¶</a></h2>
<p>ICE normalization <a class="reference external" href="#cite-Imakaev2012a">(Imakaev et al., 2012)</a>
assumes an equal experimental visibility of each bin and seeks
iteratively for biases that equalize the sum of counts per bin in the
matrix. At each iteration, a new matrix is generated by dividing each
cell by the product of the sum of counts in its row times the sum of
counts in its column. The process converges to a matrix in which all
bins have identical sum.</p>
<p>If <span class="math notranslate nohighlight">\(W\)</span> is the raw matrix, <span class="math notranslate nohighlight">\(N\)</span> is its size, and
<span class="math notranslate nohighlight">\(i\)</span>(<span class="math notranslate nohighlight">\(j\)</span>) the index of the columns(rows), the normalized
matrix <span class="math notranslate nohighlight">\(M\)</span> is iteratively computed as:</p>
<div class="math notranslate nohighlight">
\[M_{i,j} = \frac{W_{i,j}}{\sqrt{\sum_{n=0}^N{W_{i,n}} \times \sum_{n=0}^N{W_{n,j}}}}\]</div>
<p>This normalization has usually a quite strong effect, and visually the
matrices look very smooth and regular.</p>
<img alt="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_18_0.png" src="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_18_0.png" />
<div class="section" id="save-biases">
<h3>Save biases<a class="headerlink" href="#save-biases" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="vanilla-coverage-normalization">
<h2>Vanilla coverage normalization<a class="headerlink" href="#vanilla-coverage-normalization" title="Permalink to this headline">¶</a></h2>
<p>The vanilla normalization <a class="reference external" href="#cite-Rao2014">(Rao et al., 2014)</a> is a
variation of the ICE where a single iteration is performed.</p>
<img alt="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_24_0.png" src="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_24_0.png" />
<div class="section" id="id1">
<h3>Save biases<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="square-root-vanilla-coverage-sqrt-normalization">
<h2>Square root vanilla coverage (SQRT) normalization<a class="headerlink" href="#square-root-vanilla-coverage-sqrt-normalization" title="Permalink to this headline">¶</a></h2>
<p>The SQRT vanilla normalization <a class="reference external" href="#cite-Rao2014">(Rao et al., 2014)</a>
is a variation of the Vanilla coverage where each element in the matrix
is divided by the square root of the product of sums of counts.</p>
<div class="math notranslate nohighlight">
\[M_{i,j} = \frac{W_{i,j}}{\sqrt{\sum_{n=0}^N{W_{i,n}} \times \sum_{n=0}^N{W_{n,j}}}}\]</div>
<img alt="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_30_0.png" src="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_30_0.png" />
<div class="section" id="id2">
<h3>Save biases<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="oned-normalization">
<h2>OneD normalization<a class="headerlink" href="#oned-normalization" title="Permalink to this headline">¶</a></h2>
<p>OneD normalization <a class="reference external" href="#cite-Vidal2018">(Vidal et al., 2018)</a> is based
on fitting a non-linear model between the total amount of contacts per
bin and the known biases: - GC content - number of RE sites (the most
important bias, the more cut sites, the more mapped reads) - read
mappability</p>
<p>As the estimation of each of this statistics is very important for the
normalization, they are left outside the normalization function, in
order to allow user to modify them.</p>
<p>get mappability ~2 min</p>
<p>get GC content ~ 30 sec</p>
<p>compute r_sites ~30 sec</p>
<img alt="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_47_0.png" src="tutorial/../nbpictures//tutorial_7-Bin_filtering_and_normalization_47_0.png" />
<div class="section" id="id3">
<h3>Save biases<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="comparison">
<h2>Comparison<a class="headerlink" href="#comparison" title="Permalink to this headline">¶</a></h2>
<p>By eye, the effect of choosing one normalization over an other is often
very small</p>
</div>
<div class="section" id="other-normalizations">
<h2>Other normalizations<a class="headerlink" href="#other-normalizations" title="Permalink to this headline">¶</a></h2>
<p>ICE and Vanilla normalizations are widely used however other, more
convoluted, normalizations <a class="reference external" href="#cite-hu2012hicnorm">(Hu et al., 2012)</a>
<a class="reference external" href="#cite-Yaffe2011">(Yaffe and Tanay, 2011)</a> can be used outside
TADbit and then loaded in TADbit as normalized matrices for further
analysis.</p>
</div>
<div class="section" id="best-normalization">
<h2>Best normalization<a class="headerlink" href="#best-normalization" title="Permalink to this headline">¶</a></h2>
<p>Which is the best normalization to use is a question that can not be
answered easily because it depends on the type of data and the type of
analysis.</p>
<p>Most of the time Hi-C experiments are conducted in different conditions
and, for each, in several replicates. A good way to find the best
normalization method may be to select the one that achieve to minimize
the differences between replicates and maximize the differences between
conditions (this in the context of the analysis to be performed).</p>
<!--bibtex
@article{hu2012hicnorm,
  title={HiCNorm: removing biases in Hi-C data via Poisson regression},
  author={Hu, Ming and Deng, Ke and Selvaraj, Siddarth and Qin, Zhaohui and Ren, Bing and Liu, Jun S},
  journal={Bioinformatics},
  volume={28},
  number={23},
  pages={3131--3133},
  year={2012},
  publisher={Oxford Univ Press}
}
@article{Yaffe2011,
abstract = {Hi-C experiments measure the probability of physical proximity between pairs of chromosomal loci on a genomic scale. We report on several systematic biases that substantially affect the Hi-C experimental procedure, including the distance between restriction sites, the GC content of trimmed ligation junctions and sequence uniqueness. To address these biases, we introduce an integrated probabilistic background model and develop algorithms to estimate its parameters and renormalize Hi-C data. Analysis of corrected human lymphoblast contact maps provides genome-wide evidence for interchromosomal aggregation of active chromatin marks, including DNase-hypersensitive sites and transcriptionally active foci. We observe extensive long-range (up to 400 kb) cis interactions at active promoters and derive asymmetric contact profiles next to transcription start sites and CTCF binding sites. Clusters of interacting chromosomal domains suggest physical separation of centromere-proximal and centromere-distal regions. These results provide a computational basis for the inference of chromosomal architectures from Hi-C experiments.},
author = {Yaffe, Eitan and Tanay, Amos},
doi = {10.1038/ng.947},
file = {:home/fransua/.local/share/data/Mendeley Ltd./Mendeley Desktop/Downloaded/Yaffe, Tanay - 2011 - Probabilistic modeling of Hi-C contact maps eliminates systematic biases to characterize global chromosomal archit.pdf:pdf},
issn = {1546-1718},
journal = {Nature genetics},
keywords = {Binding Sites,Chromosomes,Cluster Analysis,Epigenesis,Genetic,Human,Humans,Lymphocytes,Lymphocytes: ultrastructure,Models,Probability},
mendeley-groups = {Research articles},
month = {nov},
number = {11},
pages = {1059--65},
pmid = {22001755},
title = {{Probabilistic modeling of Hi-C contact maps eliminates systematic biases to characterize global chromosomal architecture.}},
url = {http://www.ncbi.nlm.nih.gov/pubmed/22001755},
volume = {43},
year = {2011}
}
@article{Imakaev2012a,
abstract = {Extracting biologically meaningful information from chromosomal interactions obtained with genome-wide chromosome conformation capture (3C) analyses requires the elimination of systematic biases. We present a computational pipeline that integrates a strategy to map sequencing reads with a data-driven method for iterative correction of biases, yielding genome-wide maps of relative contact probabilities. We validate this ICE (iterative correction and eigenvector decomposition) technique on published data obtained by the high-throughput 3C method Hi-C, and we demonstrate that eigenvector decomposition of the obtained maps provides insights into local chromatin states, global patterns of chromosomal interactions, and the conserved organization of human and mouse chromosomes.},
author = {Imakaev, Maxim V and Fudenberg, Geoffrey and McCord, Rachel Patton and Naumova, Natalia and Goloborodko, Anton and Lajoie, Bryan R and Dekker, Job and Mirny, Leonid A},
doi = {10.1038/nmeth.2148},
file = {:home/fransua/.local/share/data/Mendeley Ltd./Mendeley Desktop/Downloaded/Imakaev et al. - 2012 - Iterative correction of Hi-C data reveals hallmarks of chromosome organization.pdf:pdf},
issn = {1548-7105},
journal = {Nature methods},
keywords = {Hi-C},
mendeley-groups = {stats/Hi-C,Research articles},
mendeley-tags = {Hi-C},
month = {oct},
number = {10},
pages = {999--1003},
pmid = {22941365},
title = {{Iterative correction of Hi-C data reveals hallmarks of chromosome organization.}},
url = {http://www.ncbi.nlm.nih.gov/pubmed/22941365},
volume = {9},
year = {2012}
}
@article{Rao2014,
author = {Rao, Suhas S P and Huntley, Miriam H and Durand, Neva C and Stamenova, Elena K and Bochkov, Ivan D. and {James T. Robinson} and Sanborn, Adrian L. and Machol, Ido and Omer, Arina D. and Lander, Eric S. and Lieberman-Aiden, Erez},
doi = {10.1016/j.cell.2014.11.021},
file = {:home/fransua/.local/share/data/Mendeley Ltd./Mendeley Desktop/Downloaded/Rao et al. - 2014 - A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping.pdf:pdf},
issn = {0092-8674},
journal = {Cell},
keywords = {Hi-C},
mendeley-groups = {Research articles,projects/GEVO/CTCF},
mendeley-tags = {Hi-C},
number = {7},
pages = {1665--1680},
pmid = {25497547},
publisher = {Elsevier Inc.},
title = {{A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping}},
url = {http://dx.doi.org/10.1016/j.cell.2014.11.021},
volume = {159},
year = {2014}
}

--></div>
</div>
<div class="section" id="references">
<h1>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<p>[^](#ref-1) Imakaev, Maxim V and Fudenberg, Geoffrey and McCord, Rachel
Patton and Naumova, Natalia and Goloborodko, Anton and Lajoie, Bryan R
and Dekker, Job and Mirny, Leonid A. 2012. <em>Iterative correction of Hi-C
data reveals hallmarks of chromosome organization.</em>.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/22941365">URL</a></p>
<p>[^](#ref-2) Rao, Suhas S P and Huntley, Miriam H and Durand, Neva C and
Stamenova, Elena K and Bochkov, Ivan D. and James T. Robinson and
Sanborn, Adrian L. and Machol, Ido and Omer, Arina D. and Lander, Eric
S. and Lieberman-Aiden, Erez. 2014. <em>A 3D Map of the Human Genome at
Kilobase Resolution Reveals Principles of Chromatin Looping</em>.
<a class="reference external" href="http://dx.doi.org/10.1016/j.cell.2014.11.021">URL</a></p>
<p>[^](#ref-3) Enrique Vidal, François le Dily, Javier Quilez, Ralph
Stadhouders, Yasmina Cuartero, Thomas Graf, Marc A Marti-Renom, Miguel
Beato, Guillaume J Filion. 2018. <em>OneD: increasing reproducibility of
Hi-C samples with abnormal karyotypes</em>.
<a class="reference external" href="https://doi.org/10.1093/nar/gky064">URL</a></p>
<p>[^](#ref-4) Hu, Ming and Deng, Ke and Selvaraj, Siddarth and Qin,
Zhaohui and Ren, Bing and Liu, Jun S. 2012. <em>HiCNorm: removing biases in
Hi-C data via Poisson regression</em>.</p>
<p>[^](#ref-5) Yaffe, Eitan and Tanay, Amos. 2011. <em>Probabilistic modeling
of Hi-C contact maps eliminates systematic biases to characterize global
chromosomal architecture.</em>.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/22001755">URL</a></p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux under conda</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html#software">Software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html#installing-tadbit-on-gnu-linux">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_1-Retrieve_published_HiC_datasets.html">Retrieve HiC dataset from NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_1-Retrieve_published_HiC_datasets.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_2-Preparation_of_the_reference_genome.html">Search for the reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_2-Preparation_of_the_reference_genome.html#download-from-the-ncbi">Download from the NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_2-Preparation_of_the_reference_genome.html#creation-of-an-index-file-for-gem-mapper">Creation of an index file for GEM mapper</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3-Hi-C_quality_check.html">FASTQ format</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3-Hi-C_quality_check.html#quality-check-before-mapping">Quality check before mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html">Iterative vs fragment-based mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html">Load mapped read-ends and generate a matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html#quality-check-of-the-hi-c-experiment">Quality check of the Hi-C experiment</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_6-Filtering_mapped_reads.html">Filtering</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">The HiC_data object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filter-bins-with-low-interaction-counts">Filter bins with low interaction counts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#normalization-algorithms">Normalization algorithms</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_6-Filtering_mapped_reads.html"
                        title="previous chapter">Filtering</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_8-Compartments_and_TADs_detection.html"
                        title="next chapter">Compartments</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_8-Compartments_and_TADs_detection.html" title="Compartments"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_6-Filtering_mapped_reads.html" title="Filtering"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Jun 21, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>