<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Search for the reference genome &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="FASTQ format" href="tutorial_3-Hi-C_quality_check.html" />
    <link rel="prev" title="Retrieve HiC dataset from NCBI" href="tutorial_1-Retrieve_published_HiC_datasets.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_3-Hi-C_quality_check.html" title="FASTQ format"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_1-Retrieve_published_HiC_datasets.html" title="Retrieve HiC dataset from NCBI"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1><p>Table of Contents</p>
</h1><div class="toc"><ul class="toc-item"><li><p>Search for the reference genome</p>
<ul class="toc-item"><li><p>Mus musculus’s reference genome sequence</p>
</li></ul></li><li><p>Download from the NCBI</p>
<ul class="toc-item"><li><p>List of chromosomes/contigs</p>
</li><li><p>Sequences of each chromosome/contig</p>
</li><li><p>Concatenate all contigs/chromosomes into single files</p>
</li></ul></li><li><p>Creation of an index file for GEM mapper</p>
<ul class="toc-item"><li><p>Compute mappability values needed for bias specific normalizations</p>
</li><li><p>Cleanup</p>
</li></ul></li></ul></div><div class="section" id="search-for-the-reference-genome">
<h1>Search for the reference genome<a class="headerlink" href="#search-for-the-reference-genome" title="Permalink to this headline">¶</a></h1>
<p>Usually NGS reads are mapped against a reference genome containing only
the assembled chromosomes, and not the remaining contigs. And this
methodology is perfectly valid. However in order to decrease the
probability of having mapping errors, adding all unassembled contigs may
help:</p>
<blockquote>
<div><p>For variant discovery, RNA-seq and ChIP-seq, it is recommended to
use the entire primary assembly, including assembled chromosomes AND
unlocalized/unplaced contigs, for the purpose of read mapping. Not
including unlocalized and unplaced contigs potentially leads to more
mapping errors.</p>
<p><em>from: http://lh3lh3.users.sourceforge.net/humanref.shtml</em></p>
</div></blockquote>
<p>We are thus going to download full chromosomes and unassembled contigs.
From these sequences we are then going to create two reference genomes:
- one “classic” reference genome with only assembled chromosomes, used
to compute statistics on the genome (GC content, number of restriction
sites or mappability) - one that would contain all chromosomes and
unassembled contigs, used exclusively for mapping.</p>
<div class="section" id="mus-musculus-s-reference-genome-sequence">
<h2><em>Mus musculus</em>’s reference genome sequence<a class="headerlink" href="#mus-musculus-s-reference-genome-sequence" title="Permalink to this headline">¶</a></h2>
<p>We search for the most recent reference genome corresponding to Mouse
(<a class="reference external" href="https://www.ncbi.nlm.nih.gov/genome?term=mus%20musculus">https://www.ncbi.nlm.nih.gov/genome?term=mus%20musculus</a>).</p>
<p>From there we obtain these identifiers:</p>
<p><em>The variables defined above can be modified for any other species,
resulting in new results for the following commands.</em></p>
</div>
</div>
<div class="section" id="download-from-the-ncbi">
<h1>Download from the NCBI<a class="headerlink" href="#download-from-the-ncbi" title="Permalink to this headline">¶</a></h1>
<div class="section" id="list-of-chromosomes-contigs">
<h2>List of chromosomes/contigs<a class="headerlink" href="#list-of-chromosomes-contigs" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="sequences-of-each-chromosome-contig">
<h2>Sequences of each chromosome/contig<a class="headerlink" href="#sequences-of-each-chromosome-contig" title="Permalink to this headline">¶</a></h2>
<p>For each contig/chromosome download the corresponding FASTA file from
NCBI</p>
</div>
<div class="section" id="concatenate-all-contigs-chromosomes-into-single-files">
<h2>Concatenate all contigs/chromosomes into single files<a class="headerlink" href="#concatenate-all-contigs-chromosomes-into-single-files" title="Permalink to this headline">¶</a></h2>
<p>Remove all the other files (with single chromosome/contig)</p>
</div>
</div>
<div class="section" id="creation-of-an-index-file-for-gem-mapper">
<h1>Creation of an index file for GEM mapper<a class="headerlink" href="#creation-of-an-index-file-for-gem-mapper" title="Permalink to this headline">¶</a></h1>
<p>The path to the index file will be:
<strong>``{dirname}/{species}-{assembly}/{species}_contigs.gem``</strong></p>
<div class="section" id="compute-mappability-values-needed-for-bias-specific-normalizations">
<h2>Compute mappability values needed for bias specific normalizations<a class="headerlink" href="#compute-mappability-values-needed-for-bias-specific-normalizations" title="Permalink to this headline">¶</a></h2>
<p>In this case we can use the FASTA of the genome whithout contigs and
follow these step:</p>
</div>
<div class="section" id="cleanup">
<h2>Cleanup<a class="headerlink" href="#cleanup" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux under conda</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html#software">Software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html#installing-tadbit-on-gnu-linux">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_1-Retrieve_published_HiC_datasets.html">Retrieve HiC dataset from NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_1-Retrieve_published_HiC_datasets.html#references">References</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Search for the reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#download-from-the-ncbi">Download from the NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="#creation-of-an-index-file-for-gem-mapper">Creation of an index file for GEM mapper</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3-Hi-C_quality_check.html">FASTQ format</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3-Hi-C_quality_check.html#quality-check-before-mapping">Quality check before mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html">Iterative vs fragment-based mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html">Load mapped read-ends and generate a matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html#quality-check-of-the-hi-c-experiment">Quality check of the Hi-C experiment</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_6-Filtering_mapped_reads.html">Filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_7-Bin_filtering_and_normalization.html">The HiC_data object</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_7-Bin_filtering_and_normalization.html#filter-bins-with-low-interaction-counts">Filter bins with low interaction counts</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_7-Bin_filtering_and_normalization.html#normalization-algorithms">Normalization algorithms</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_7-Bin_filtering_and_normalization.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_1-Retrieve_published_HiC_datasets.html"
                        title="previous chapter">Retrieve HiC dataset from NCBI</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_3-Hi-C_quality_check.html"
                        title="next chapter">FASTQ format</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_3-Hi-C_quality_check.html" title="FASTQ format"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_1-Retrieve_published_HiC_datasets.html" title="Retrieve HiC dataset from NCBI"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Jun 21, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>