<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>TAD clustering &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Tutorials" href="../tutorial.html" />
    <link rel="next" title="Hi-C data processing" href="tutorial_4_data_normalization.html" />
    <link rel="prev" title="Alignment of TAD boundaries" href="tutorial_2_boundary_alignment.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_4_data_normalization.html" title="Hi-C data processing"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_2_boundary_alignment.html" title="Alignment of TAD boundaries"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="../tutorial.html" accesskey="U">Tutorials</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="tad-clustering">
<h1><a class="toc-backref" href="#id2">TAD clustering</a><a class="headerlink" href="#tad-clustering" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#tad-clustering" id="id2">TAD clustering</a><ul>
<li><a class="reference internal" href="#compare-two-tads" id="id3">Compare two TADs</a></li>
</ul>
</li>
</ul>
</div>
<p>TADbit allows to compare the topology of TADs using directly a Hi-C matrix. This comparison is done using
the same methodology used for protein structure comparison with contact map overlap (CMO) <a class="reference internal" href="../biblio.html#dilena2010" id="id1">[DiLena2010]</a>.</p>
<p>The function that allows such comparison is <a class="reference internal" href="../reference/reference_clustering.html#pytadbit.tad_clustering.tad_cmo.optimal_cmo" title="pytadbit.tad_clustering.tad_cmo.optimal_cmo"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.tad_clustering.tad_cmo.optimal_cmo()</span></tt></a>. The
comparisons are pairwise; the main idea of using it is to pull together several of these pairwise comparisons
in order to find group of TADs with specific characteristics.</p>
<div class="section" id="compare-two-tads">
<h2><a class="toc-backref" href="#id3">Compare two TADs</a><a class="headerlink" href="#compare-two-tads" title="Permalink to this headline">¶</a></h2>
<p>To compare TADs, a Chromosome with defined TADs is needed. Thus, following with the example chromosome used
in <a class="reference internal" href="tutorial_1_general.html#getting-start"><em>Getting started</em></a> (following the example up to <a class="reference internal" href="tutorial_1_general.html#run-tadbit"><em>Find Topologically Associating Domains</em></a>), these re the steps to follow:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="kn">import</span> <span class="n">Chromosome</span>
<span class="n">my_chrom</span> <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s">&#39;My first chromosome&#39;</span><span class="p">)</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s">&#39;First Hi-C experiment&#39;</span><span class="p">,</span> <span class="n">hic_data</span><span class="o">=</span><span class="s">&quot;../../scripts/sample_data/HIC_k562_chr19_chr19_100000_obs.txt&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">find_tad</span><span class="p">(</span><span class="s">&#39;First Hi-C experiment&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Once done, all the TADs for this chromosome will be defined. To select two TADs:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">tad1</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">my_chrom</span><span class="o">.</span><span class="n">iter_tads</span><span class="p">(</span><span class="s">&#39;First Hi-C experiment&#39;</span><span class="p">))[</span><span class="mi">41</span><span class="p">]</span>
<span class="n">tad2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">my_chrom</span><span class="o">.</span><span class="n">iter_tads</span><span class="p">(</span><span class="s">&#39;First Hi-C experiment&#39;</span><span class="p">))[</span><span class="mi">39</span><span class="p">]</span>
</pre></div>
</div>
<p>And to align them:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.tad_clustering.tad_cmo</span> <span class="kn">import</span> <span class="n">optimal_cmo</span>
<span class="n">align1</span><span class="p">,</span> <span class="n">align2</span><span class="p">,</span> <span class="n">score</span> <span class="o">=</span> <span class="n">optimal_cmo</span><span class="p">(</span><span class="n">tad1</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">tad2</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">max_num_v</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">long_nw</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">long_dist</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s">&#39;frobenius&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The optimal_cmo function returns two alignments corresponding to the sequence of gaps needed for each TAD to
be aligned with the other. The score element contains three values, an alignment score that depends on the
method used to align, and the p-value and rho value of the Spearman correlation between the two Hi-C matrices.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.tad_clustering.tad_cmo</span> <span class="kn">import</span> <span class="n">optimal_cmo</span>
</pre></div>
</div>
<p>Following is the output of the example script <em>&#8216;clustering.py&#8217;</em>:</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="../_images/clustering.png"><img alt="../_images/clustering.png" src="../_images/clustering.png" style="width: 900px;" /></a>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads">Identification and Analysis of TADs</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_1_general.html">Getting started</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_1_general.html#find-topologically-associating-domains">Find Topologically Associating Domains</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_2_boundary_alignment.html">Alignment of TAD boundaries</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="">TAD clustering</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_2_boundary_alignment.html"
                        title="previous chapter">Alignment of TAD boundaries</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_4_data_normalization.html"
                        title="next chapter">Hi-C data processing</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_4_data_normalization.html" title="Hi-C data processing"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_2_boundary_alignment.html" title="Alignment of TAD boundaries"
             >previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="../tutorial.html" >Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright .
      Last updated on Jan 10, 2014.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit"><img
            style="position: fixed; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>