<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Preparation of the reference genome &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Hi-C quality check" href="tutorial_3-Hi-C_quality_check.html" />
    <link rel="prev" title="Retrieve HiC dataset from NCBI" href="tutorial_1-Retrieve_published_HiC_datasets.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_3-Hi-C_quality_check.html" title="Hi-C quality check"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_1-Retrieve_published_HiC_datasets.html" title="Retrieve HiC dataset from NCBI"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="preparation-of-the-reference-genome">
<h1>Preparation of the reference genome<a class="headerlink" href="#preparation-of-the-reference-genome" title="Permalink to this headline">¶</a></h1>
<p>Usually NGS reads are mapped against a reference genome containing only
the assembled chromosomes, and not the remaining contigs. And this
methodology is perfectly valid. However in order to decrease the
probability of having mapping errors, adding all unassembled contigs may
help:</p>
<blockquote>
<div><p>For variant discovery, RNA-seq and ChIP-seq, it is recommended to use
the entire primary assembly, including assembled chromosomes AND
unlocalized/unplaced contigs, for the purpose of read mapping. Not
including unlocalized and unplaced contigs potentially leads to more
mapping errors.</p>
<p><em>from: http://lh3lh3.users.sourceforge.net/humanref.shtml</em></p>
</div></blockquote>
<p>We are thus going to download full chromosomes and unassembled contigs.
From these sequences we are then going to create two reference genomes:
- one “classic” reference genome with only assembled chromosomes, used
to compute statistics on the genome (GC content, number of restriction
sites or mappability) - one that would contain all chromosomes and
unassembled contigs, used exclusively for mapping.</p>
<div class="section" id="mus-musculuss-reference-genome-sequence">
<h2><em>Mus musculus</em>’s reference genome sequence<a class="headerlink" href="#mus-musculuss-reference-genome-sequence" title="Permalink to this headline">¶</a></h2>
<p>We search for the most recent reference genome corresponding to Mouse
(<a class="reference external" href="https://www.ncbi.nlm.nih.gov/genome?term=mus%20musculus">https://www.ncbi.nlm.nih.gov/genome?term=mus%20musculus</a>).</p>
<p>From there we obtain these identifiers:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">species</span>  <span class="o">=</span> <span class="s1">&#39;Mus_musculus&#39;</span>
<span class="n">taxid</span>    <span class="o">=</span> <span class="s1">&#39;10090&#39;</span>
<span class="n">assembly</span> <span class="o">=</span> <span class="s1">&#39;GRCm38.p6&#39;</span>
<span class="n">genbank</span>  <span class="o">=</span> <span class="s1">&#39;GCF_000001635.26&#39;</span>
</pre></div>
</div>
<p><em>The variables defined above can be modified for any other species,
resulting in new results for the following commands.</em></p>
<div class="section" id="download-from-the-ncbi">
<h3>Download from the NCBI<a class="headerlink" href="#download-from-the-ncbi" title="Permalink to this headline">¶</a></h3>
<div class="section" id="list-of-chromosomes-contigs">
<h4>List of chromosomes/contigs<a class="headerlink" href="#list-of-chromosomes-contigs" title="Permalink to this headline">¶</a></h4>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sumurl</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;ftp://ftp.ncbi.nlm.nih.gov/genomes/all/</span><span class="si">{0}</span><span class="s1">/</span><span class="si">{1}</span><span class="s1">/</span><span class="si">{2}</span><span class="s1">/</span><span class="si">{3}</span><span class="s1">/</span><span class="si">{4}</span><span class="s1">_</span><span class="si">{5}</span><span class="s1">/&#39;</span>
          <span class="s1">&#39;</span><span class="si">{4}</span><span class="s1">_</span><span class="si">{5}</span><span class="s1">_assembly_report.txt&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">genbank</span><span class="p">[:</span><span class="mi">3</span><span class="p">],</span> <span class="n">genbank</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">7</span><span class="p">],</span> <span class="n">genbank</span><span class="p">[</span><span class="mi">7</span><span class="p">:</span><span class="mi">10</span><span class="p">],</span>
                                                <span class="n">genbank</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">13</span><span class="p">],</span> <span class="n">genbank</span><span class="p">,</span> <span class="n">assembly</span><span class="p">)</span>

<span class="n">crmurl</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi&#39;</span>
          <span class="s1">&#39;?db=nuccore&amp;id=</span><span class="si">%s</span><span class="s1">&amp;rettype=fasta&amp;retmode=text&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span> <span class="n">sumurl</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.26_GRCm38.p6/GCF_000001635.26_GRCm38.p6_assembly_report.txt</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span> wget -q <span class="nv">$sumurl</span> -O chromosome_list.txt
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span> head chromosome_list.txt
</pre></div>
</div>
<pre class="ansi-block literal-block"># Assembly name:  GRCm38.p6
# Description:    Genome Reference Consortium Mouse Build 38 patch release 6 (GRCm38.p6)
# Organism name:  Mus musculus (house mouse)
# Infraspecific name:  strain=C57BL/6J
# Taxid:          10090
# BioProject:     PRJNA20689
# Submitter:      Genome Reference Consortium
# Date:           2017-9-15
# Assembly type:  haploid-with-alt-loci
# Release type:   patch</pre>
</div>
</div>
</div>
<div class="section" id="sequences-of-each-chromosome-contig">
<h2>Sequences of each chromosome/contig<a class="headerlink" href="#sequences-of-each-chromosome-contig" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dirname</span> <span class="o">=</span> <span class="s1">&#39;genome&#39;</span>
<span class="o">!</span> mkdir -p <span class="o">{</span>dirname<span class="o">}</span>
</pre></div>
</div>
<p>For each contig/chromosome download the corresponding FASTA file from
NCBI</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">contig</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;chromosome_list.txt&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
        <span class="k">continue</span>
    <span class="n">seq_name</span><span class="p">,</span> <span class="n">seq_role</span><span class="p">,</span> <span class="n">assigned_molecule</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">genbank</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">refseq</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="mi">7</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">seq_role</span> <span class="o">==</span> <span class="s1">&#39;assembled-molecule&#39;</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;chr</span><span class="si">%s</span><span class="s1">.fasta&#39;</span> <span class="o">%</span> <span class="n">assigned_molecule</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;chr</span><span class="si">%s</span><span class="s1">_</span><span class="si">%s</span><span class="s1">.fasta&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">assigned_molecule</span><span class="p">,</span> <span class="n">seq_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">))</span>
    <span class="n">contig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>

    <span class="n">outfile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">outfile</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">outfile</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="n">error_code</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;wget &quot;</span><span class="si">%s</span><span class="s1">&quot; --no-check-certificate -O </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">crmurl</span> <span class="o">%</span> <span class="p">(</span><span class="n">genbank</span><span class="p">),</span> <span class="n">outfile</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">error_code</span><span class="p">:</span>
        <span class="n">error_code</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;wget &quot;</span><span class="si">%s</span><span class="s1">&quot; --no-check-certificate -O </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">crmurl</span> <span class="o">%</span> <span class="p">(</span><span class="n">refseq</span><span class="p">),</span> <span class="n">outfile</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">error_code</span><span class="p">:</span>
        <span class="nb">print</span> <span class="n">genbank</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">CM001008.2</pre>
</div>
<div class="section" id="concatenate-all-contigs-chromosomes-into-single-files">
<h2>Concatenate all contigs/chromosomes into single files<a class="headerlink" href="#concatenate-all-contigs-chromosomes-into-single-files" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">write_to_fasta</span><span class="p">(</span><span class="n">line</span><span class="p">):</span>
    <span class="n">contig_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">write_to_fastas</span><span class="p">(</span><span class="n">line</span><span class="p">):</span>
    <span class="n">contig_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
    <span class="n">simple_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;mkdir -p </span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="n">assembly</span><span class="p">))</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">0</pre>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">contig_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0}</span><span class="s1">/</span><span class="si">{1}</span><span class="s1">-</span><span class="si">{2}</span><span class="s1">/</span><span class="si">{1}</span><span class="s1">-</span><span class="si">{2}</span><span class="s1">_contigs.fa&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="n">assembly</span><span class="p">),</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
<span class="n">simple_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0}</span><span class="s1">/</span><span class="si">{1}</span><span class="s1">-</span><span class="si">{2}</span><span class="s1">/</span><span class="si">{1}</span><span class="s1">-</span><span class="si">{2}</span><span class="s1">.fa&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="n">assembly</span><span class="p">),</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">molecule</span> <span class="ow">in</span> <span class="n">contig</span><span class="p">:</span>
    <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0}</span><span class="s1">/</span><span class="si">{1}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="n">molecule</span><span class="p">))</span>
    <span class="n">oline</span> <span class="o">=</span> <span class="s1">&#39;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">))</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">fh</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
    <span class="c1"># if molecule is an assembled chromosome we write to both files, otherwise only to the *_contigs one</span>
    <span class="n">write</span> <span class="o">=</span> <span class="n">write_to_fasta</span> <span class="k">if</span> <span class="s1">&#39;_&#39;</span> <span class="ow">in</span> <span class="n">molecule</span> <span class="k">else</span> <span class="n">write_to_fastas</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>
        <span class="n">write</span><span class="p">(</span><span class="n">oline</span><span class="p">)</span>
        <span class="n">oline</span> <span class="o">=</span> <span class="n">line</span>
    <span class="c1"># last line usually empty...</span>
    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
        <span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
<span class="n">contig_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="n">simple_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Remove all the other files (with single chromosome/contig)</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span> rm -f <span class="o">{</span>dirname<span class="o">}</span>/*.fasta
</pre></div>
</div>
<div class="section" id="creation-of-an-index-file-for-gem-mapper">
<h3>Creation of an index file for GEM mapper<a class="headerlink" href="#creation-of-an-index-file-for-gem-mapper" title="Permalink to this headline">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span> gem-indexer -T <span class="m">8</span> -i <span class="o">{</span>dirname<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>_contigs.fa -o <span class="o">{</span>dirname<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>_contigs
</pre></div>
</div>
<pre class="ansi-block literal-block">Welcome to GEM-indexer build 1.423 (beta) - (2013/04/01 01:02:13 GMT)
 (c) 2008-2013 Paolo Ribeca &lt;paolo.ribeca&#64;gmail.com&gt;
 (c) 2010-2013 Santiago Marco Sola &lt;santiagomsola&#64;gmail.com&gt;
 (c) 2010-2013 Leonor Frias Moya &lt;leonor.frias&#64;gmail.com&gt;
For the terms of use, run the program with the option --show-license.
************************************************************************
* WARNING: this is a beta version, provided for testing purposes only; *
*          check for updates at &lt;http://gemlibrary.sourceforge.net&gt;.   *
************************************************************************
Creating sequence and location files... done.
Computing DNA BWT (likely to take long)... done.
Generating index (likely to take long)... done.
Cleaning up... done.</pre>
<p>The path to the index file will be:
<strong>``{dirname}/{species}-{assembly}/{species}_contigs.gem``</strong></p>
</div>
</div>
<div class="section" id="compute-mappability-values-needed-for-bias-specific-normalizations">
<h2>Compute mappability values needed for bias specific normalizations<a class="headerlink" href="#compute-mappability-values-needed-for-bias-specific-normalizations" title="Permalink to this headline">¶</a></h2>
<p>In this case we can use the FASTA of the genome whithout contigs and
follow these step:</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>! gem-indexer -i {dirname}/{species}-{assembly}/{species}-{assembly}.fa \
   -o {dirname}/{species}-{assembly}/{species}-{assembly} -T 8

! gem-mappability -I {dirname}/{species}-{assembly}/{species}-{assembly}.gem -l 50 \
   -o {dirname}/{species}-{assembly}/{species}-{assembly}.50mer -T 8

! gem-2-wig -I {dirname}/{species}-{assembly}/{species}-{assembly}.gem \
   -i {dirname}/{species}-{assembly}/{species}-{assembly}.50mer.mappability \
   -o {dirname}/{species}-{assembly}/{species}-{assembly}.50mer

! wigToBigWig {dirname}/{species}-{assembly}/{species}-{assembly}.50mer.wig \
   {dirname}/{species}-{assembly}/{species}-{assembly}.50mer.sizes \
   {dirname}/{species}-{assembly}/{species}-{assembly}.50mer.bw

! bigWigToBedGraph {dirname}/{species}-{assembly}/{species}-{assembly}.50mer.bw  \
   {dirname}/{species}-{assembly}/{species}-{assembly}.50mer.bedGraph
</pre></div>
</div>
<pre class="ansi-block literal-block">Welcome to GEM-indexer build 1.423 (beta) - (2013/04/01 01:02:13 GMT)
 (c) 2008-2013 Paolo Ribeca &lt;paolo.ribeca&#64;gmail.com&gt;
 (c) 2010-2013 Santiago Marco Sola &lt;santiagomsola&#64;gmail.com&gt;
 (c) 2010-2013 Leonor Frias Moya &lt;leonor.frias&#64;gmail.com&gt;
For the terms of use, run the program with the option --show-license.
************************************************************************
* WARNING: this is a beta version, provided for testing purposes only; *
*          check for updates at &lt;http://gemlibrary.sourceforge.net&gt;.   *
************************************************************************
Creating sequence and location files... done.
Computing DNA BWT (likely to take long)... done.
Generating index (likely to take long)... done.
Cleaning up... done.
Welcome to GEM-mappability build 1.315 (beta) - (2013/03/29 02:59:40 GMT)
 (c) 2008-2013 Paolo Ribeca &lt;paolo.ribeca&#64;gmail.com&gt;
 (c) 2010-2013 Santiago Marco Sola &lt;santiagomsola&#64;gmail.com&gt;
 (c) 2010-2013 Leonor Frias Moya &lt;leonor.frias&#64;gmail.com&gt;
For the terms of use, run the program with the option --show-license.
************************************************************************
* WARNING: this is a beta version, provided for testing purposes only; *
*          check for updates at &lt;http://gemlibrary.sourceforge.net&gt;.   *
************************************************************************
Sat Jan 12 13:01:12 2019 -- Loading index (likely to take long)... done.
Sat Jan 12 13:01:23 2019 -- Starting (2647560565 positions to go)...
Sat Jan 12 20:34:17 2019 -- Writing frequencies to disk...
Sat Jan 12 20:34:27 2019 -- ...done.
Welcome to GEM-2-wig build 1.423 (beta) - (2013/04/01 01:02:13 GMT)
 (c) 2008-2013 Paolo Ribeca &lt;paolo.ribeca&#64;gmail.com&gt;
 (c) 2010-2013 Santiago Marco Sola &lt;santiagomsola&#64;gmail.com&gt;
 (c) 2010-2013 Leonor Frias Moya &lt;leonor.frias&#64;gmail.com&gt;
For the terms of use, run the program with the option --show-license.
************************************************************************
* WARNING: this is a beta version, provided for testing purposes only; *
*          check for updates at &lt;http://gemlibrary.sourceforge.net&gt;.   *
************************************************************************
Sat Jan 12 20:34:28 2019 -- Loading index (likely to take long)... done.
Sat Jan 12 20:34:41 2019 -- Inverting locations... done.
genome/Mus_musculus-GRCm38.p6/Mus_musculus-GRCm38.p6.50mer.bw is not a bpt b-plus tree index file</pre>
</div>
<div class="section" id="cleanup">
<h2>Cleanup<a class="headerlink" href="#cleanup" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span> rm -f <span class="o">{</span>dirname<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>.50mer.mappability
<span class="o">!</span> rm -f <span class="o">{</span>dirname<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>.50mer.wig
<span class="o">!</span> rm -f <span class="o">{</span>dirname<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>.50mer.bw
<span class="o">!</span> rm -f <span class="o">{</span>dirname<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>.50mer.sizes
<span class="o">!</span> rm -f <span class="o">{</span>dirname<span class="o">}</span>/<span class="o">{</span>species<span class="o">}</span>-<span class="o">{</span>assembly<span class="o">}</span>/*.log
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_1-Retrieve_published_HiC_datasets.html">Retrieve HiC dataset from NCBI</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Preparation of the reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3-Hi-C_quality_check.html">Hi-C quality check</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html">Iterative vs fragment-based mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html">Load mapped read-ends and generate a matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html#quality-check-of-the-hi-c-experiment">Quality check of the Hi-C experiment</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_6-Filtering_mapped_reads.html">Filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_7-Bin_filtering_and_normalization.html">Bin filtering and normalization</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_1-Retrieve_published_HiC_datasets.html"
                        title="previous chapter">Retrieve HiC dataset from NCBI</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_3-Hi-C_quality_check.html"
                        title="next chapter">Hi-C quality check</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_3-Hi-C_quality_check.html" title="Hi-C quality check"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_1-Retrieve_published_HiC_datasets.html" title="Retrieve HiC dataset from NCBI"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Feb 07, 2020.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>