<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Alignment of TAD boundaries &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Tutorials" href="../tutorial.html" />
    <link rel="next" title="TAD clustering" href="tutorial_3_clustering.html" />
    <link rel="prev" title="Getting started" href="tutorial_1_general.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_3_clustering.html" title="TAD clustering"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="../tutorial.html" accesskey="U">Tutorials</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="alignment-of-tad-boundaries">
<h1><a class="toc-backref" href="#id2">Alignment of TAD boundaries</a><a class="headerlink" href="#alignment-of-tad-boundaries" title="Permalink to this headline">¬∂</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#alignment-of-tad-boundaries" id="id2">Alignment of TAD boundaries</a><ul>
<li><a class="reference internal" href="#aligning-boundaries" id="id3">Aligning boundaries</a><ul>
<li><a class="reference internal" href="#check-alignment-consistency-through-randomization" id="id4">Check alignment consistency through randomization</a></li>
</ul>
</li>
<li><a class="reference internal" href="#alignment-objects" id="id5">Alignment objects</a><ul>
<li><a class="reference internal" href="#visualization" id="id6">Visualization</a></li>
<li><a class="reference internal" href="#the-get-column-function" id="id7">The get_column function</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<p>TADbit allows to use the information from different Hi-C experiments and to put it together in order to
decide whether some TAD boundaries are conserved or not.</p>
<p>Following with the example in the previous section (<a class="reference internal" href="tutorial_1_general.html#getting-start"><em>Getting started</em></a>), we will load one extra experiment
(from the same works of <a class="reference internal" href="../biblio.html#lieberman-aiden2009" id="id1">[Lieberman-Aiden2009]</a>):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="kn">import</span> <span class="n">Chromosome</span>

<span class="c"># initiate a chromosome object that will store all Hi-C data and analysis</span>
<span class="n">my_chrom</span> <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s">&#39;My fisrt chromosome&#39;</span><span class="p">,</span> <span class="n">centromere_search</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="c"># load Hi-C data</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s">&#39;First Hi-C experiment&#39;</span><span class="p">,</span> <span class="n">hic_data</span><span class="o">=</span><span class="s">&quot;../../scripts/sample_data/HIC_k562_chr19_chr19_100000_obs.txt&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s">&#39;Second Hi-C experiment&#39;</span><span class="p">,</span> <span class="n">hic_data</span><span class="o">=</span><span class="s">&quot;../../scripts/sample_data/HIC_gm06690_chr19_chr19_100000_obs.txt&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>

<span class="c"># run core tadbit function to find TADs, on each experiment</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">find_tad</span><span class="p">(</span><span class="s">&#39;First Hi-C experiment&#39;</span><span class="p">)</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">find_tad</span><span class="p">(</span><span class="s">&#39;Second Hi-C experiment&#39;</span><span class="p">)</span>

<span class="k">print</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<div class="highlight-python"><pre>[Experiment First Hi-C experiment (resolution: 100Kb, TADs: 43, Hi-C rows: 639, normalized: None), Experiment Second Hi-C experiment (resolution: 100Kb, TADs: 31, Hi-C rows: 639, normalized: None)]</pre>
</div>
<p>We now have loaded two Hi-C experiments, both at 100 Kb resolution, and have predicted the location of TADs in each of them (42 TADs detected in the first experiment and 31 in the second).</p>
<div class="section" id="aligning-boundaries">
<h2><a class="toc-backref" href="#id3">Aligning boundaries</a><a class="headerlink" href="#aligning-boundaries" title="Permalink to this headline">¬∂</a></h2>
<p>To align TAD boundaries several algorithms have been implemented
(see <a class="reference internal" href="../reference/reference_chromosome.html#pytadbit.chromosome.Chromosome.align_experiments" title="pytadbit.chromosome.Chromosome.align_experiments"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.chromosome.Chromosome.align_experiments()</span></tt></a>); our recommendation, however, is to use
the default &#8220;reciprocal&#8221; method (<a class="reference internal" href="../reference/reference_aligner.html#pytadbit.boundary_aligner.reciprocally.reciprocal" title="pytadbit.boundary_aligner.reciprocally.reciprocal"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.boundary_aligner.reciprocally.reciprocal()</span></tt></a>).</p>
<p><em>Note: If the align_experiments function is run with no argument, by default all the loaded experiments will be aligned.</em></p>
<p>Continuing with the example, the two loaded experiments are aligned as follow:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">align_experiments</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;First Hi-C experiment&quot;</span><span class="p">,</span> <span class="s">&quot;Second Hi-C experiment&quot;</span><span class="p">])</span>

<span class="k">print</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">alignment</span>
</pre></div>
</div>
<div class="highlight-python"><pre>{('First Hi-C experiment', 'Second Hi-C experiment'): Alignment of boundaries (length: 56, number of experiments: 2)}</pre>
</div>
<p>All the alignments done between the experiments belonging to the same chromosome are stored under the
alignment dictionary attached to the Chromosome object. Each alignment is an object itself
(see <a class="reference internal" href="../reference/reference_boundary_alignment.html#pytadbit.alignment.Alignment" title="pytadbit.alignment.Alignment"><tt class="xref py py-class docutils literal"><span class="pre">pytadbit.alignment.Alignment</span></tt></a>)</p>
<div class="section" id="check-alignment-consistency-through-randomization">
<h3><a class="toc-backref" href="#id4">Check alignment consistency through randomization</a><a class="headerlink" href="#check-alignment-consistency-through-randomization" title="Permalink to this headline">¬∂</a></h3>
<p>In order to check that the alignment makes sense and that it does not correspond to a random association of boundaries, the ‚Äúrandomize‚Äù parameter can be set to True when aligning:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">score</span><span class="p">,</span> <span class="n">pval</span> <span class="o">=</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">align_experiments</span><span class="p">(</span><span class="n">randomize</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">rnd_method</span><span class="o">=</span><span class="s">&quot;interpolate&quot;</span><span class="p">,</span>
                                         <span class="n">rnd_num</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>

<span class="k">print</span> <span class="s">&#39;score:&#39;</span><span class="p">,</span> <span class="n">score</span>
<span class="k">print</span> <span class="s">&#39;p-value:&#39;</span><span class="p">,</span> <span class="n">pval</span>
</pre></div>
</div>
<div class="highlight-python"><pre>score: 0.267857142857
p-value: 0.0</pre>
</div>
</div>
</div>
<div class="section" id="alignment-objects">
<h2><a class="toc-backref" href="#id5">Alignment objects</a><a class="headerlink" href="#alignment-objects" title="Permalink to this headline">¬∂</a></h2>
<div class="section" id="visualization">
<h3><a class="toc-backref" href="#id6">Visualization</a><a class="headerlink" href="#visualization" title="Permalink to this headline">¬∂</a></h3>
<p>The first function to call to check the quality of the generated alignments is the
<a class="reference internal" href="../reference/reference_boundary_alignment.html#pytadbit.alignment.Alignment.write_alignment" title="pytadbit.alignment.Alignment.write_alignment"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.alignment.Alignment.write_alignment()</span></tt></a>:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ali</span> <span class="o">=</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">alignment</span><span class="p">[(</span><span class="s">&#39;First Hi-C experiment&#39;</span><span class="p">,</span> <span class="s">&#39;Second Hi-C experiment&#39;</span><span class="p">)]</span>

<span class="k">print</span> <span class="n">ali</span>
</pre></div>
</div>
<div class="highlight-python"><pre>Alignment shown in 100 Kb (2 experiments) (scores: [34m0[m [34m1[m [34m2[m [36m3[m [0m4[m [1m5[m [33m6[m [33m7[m [35m8[m [35m9[m [31m10[m)
 First Hi-C experiment:|     [34m6[m|    [1m13[m| ---- | ---- |    [33m32[m| ---- |    [1m46[m| ---- |    [36m59[m|    [33m70[m|    [34m78[m| ---- | ---- |   [35m104[m|   [0m109[m|   [0m115[m|   [34m125[m| ---- |   [34m132[m|   [36m137[m|   [35m145[m|   [33m164[m|   [1m184[m|   [34m189[m|   [35m195[m|   [0m245[m|   [36m324[m|   [36m330[m|   [0m348[m|   [1m356[m|   [33m378[m|   [1m384[m| ---- |   [0m400[m| ---- |   [35m413[m| ---- |   [0m435[m|   [1m447[m|   [0m453[m|   [34m458[m|   [0m472[m|   [1m478[m|   [35m486[m|   [36m498[m|   [1m506[m| ---- |   [36m524[m|   [35m531[m|   [1m554[m|   [1m563[m| ---- |   [33m594[m|   [33m609[m| ---- |   [31m639[m
Second Hi-C experiment:|     [1m5[m|    [33m12[m|    [33m18[m|    [34m27[m| ---- |    [1m42[m|    [34m47[m|    [33m57[m| ---- | ---- |    [31m79[m|    [1m86[m|    [31m98[m| ---- | ---- |   [35m115[m| ---- |   [1m127[m| ---- | ---- | ---- | ---- | ---- | ---- |   [31m195[m|   [0m246[m| ---- | ---- | ---- | ---- |   [0m378[m| ---- |   [33m397[m| ---- |   [33m402[m|   [1m413[m|   [0m431[m| ---- | ---- | ---- | ---- | ---- |   [35m478[m|   [35m486[m|   [1m498[m| ---- |   [33m510[m| ---- |   [0m531[m|   [33m554[m|   [1m563[m|   [33m569[m|   [33m593[m|   [31m609[m|   [0m624[m|   [31m639[m</pre>
</div>
<p>The different colors, corresponding to the TADbit confidence in detecting the boundaries, show how conserved the boundaries are between (in this case) cell types.</p>
<p>Alignment can also be viewed using matplotlib:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ali</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/tutorial_2_boundary_alignment_17_0.png" src="../_images/tutorial_2_boundary_alignment_17_0.png" />
</div>
<div class="section" id="the-get-column-function">
<h3><a class="toc-backref" href="#id7">The get_column function</a><a class="headerlink" href="#the-get-column-function" title="Permalink to this headline">¬∂</a></h3>
<p>The <a class="reference internal" href="../reference/reference_boundary_alignment.html#pytadbit.alignment.Alignment.get_column" title="pytadbit.alignment.Alignment.get_column"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.alignment.Alignment.get_column()</span></tt></a> function allows to select specific columns of an alignment.</p>
<p>To select, for example, the third column of an alignment:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ali</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><pre>[(2, [&gt;-&lt;, &gt;1700&lt;])]</pre>
</div>
<p>The first element of the tuple is the column index, while the two values of the second element of the tuple
are the TADs associated to the aligned boundaries in that column. Note that TAD objects are represented
between the &#8216;&gt;&#8217; and &#8216;&lt;&#8217; symbols (see: <a class="reference internal" href="../reference/reference_boundary_alignment.html#pytadbit.alignment.TAD" title="pytadbit.alignment.TAD"><tt class="xref py py-class docutils literal"><span class="pre">pytadbit.alignment.TAD</span></tt></a>).</p>
<p>The <a class="reference internal" href="../reference/reference_boundary_alignment.html#pytadbit.alignment.Alignment.get_column" title="pytadbit.alignment.Alignment.get_column"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.alignment.Alignment.get_column()</span></tt></a> function can also take as an argument a function, in
order to select a column (or several) depending on a specific condition. For example, to select all the
boundaries with a score higher than 7:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">cond1</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;score&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">7</span>
</pre></div>
</div>
<p>and to the get the selected columns:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ali</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="n">cond1</span><span class="o">=</span><span class="n">cond1</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><pre>[(24, [&gt;19400&lt;, &gt;19400&lt;]), (43, [&gt;48500&lt;, &gt;48500&lt;]), (55, [&gt;63800&lt;, &gt;63800&lt;])]</pre>
</div>
<p>resulting, in the selection of these 3 columns.</p>
<p>To add a second condition, e.g. to select only the columns after the 50th column of the alignment:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">cond2</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;pos&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">50</span>
<span class="n">ali</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="n">cond1</span><span class="o">=</span><span class="n">cond1</span><span class="p">,</span> <span class="n">cond2</span><span class="o">=</span><span class="n">cond2</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><pre>[(55, [&gt;63800&lt;, &gt;63800&lt;])]</pre>
</div>
<p>Finally, to be more flexible, this conditions can be applied to only a given number of experiments (in this example of a pairwise alignment, it does not make a lot of sense):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ali</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="n">cond1</span><span class="o">=</span><span class="n">cond1</span><span class="p">,</span> <span class="n">cond2</span><span class="o">=</span><span class="n">cond2</span><span class="p">,</span> <span class="n">min_num</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><pre>[(53, [&gt;60800&lt;, &gt;60800&lt;]), (55, [&gt;63800&lt;, &gt;63800&lt;])]</pre>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads">Identification and Analysis of TADs</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_1_general.html">Getting started</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_1_general.html#find-topologically-associating-domains">Find Topologically Associating Domains</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="">Alignment of TAD boundaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3_clustering.html">TAD clustering</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_1_general.html"
                        title="previous chapter">Getting started</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_3_clustering.html"
                        title="next chapter">TAD clustering</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_3_clustering.html" title="TAD clustering"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             >previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="../tutorial.html" >Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright .
      Last updated on Jan 10, 2014.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit"><img
            style="position: fixed; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>