<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Mapping paired-end reads from NGS experiment (with GEM) &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Tutorials" href="../tutorial.html" />
    <link rel="next" title="Getting started" href="tutorial_1_general.html" />
    <link rel="prev" title="Tutorials" href="../tutorial.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="mapping-paired-end-reads-from-ngs-experiment-with-gem">
<h1>Mapping paired-end reads from NGS experiment (with GEM)<a class="headerlink" href="#mapping-paired-end-reads-from-ngs-experiment-with-gem" title="Permalink to this headline">¶</a></h1>
<div class="section" id="download-and-index-reference-genome">
<h2>Download and index reference genome<a class="headerlink" href="#download-and-index-reference-genome" title="Permalink to this headline">¶</a></h2>
<p>Download last human reference genome from ucsc:
<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/">http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/</a> It&#8217;s better to
download the file that contains each chromosome separately (like this we
could only keep those which we are interested in): hg38.chromFa.tar.gz
Uncompress the genome</p>
<div class="highlight-python"><div class="highlight"><pre>tar -xzf hg38.chromFa.tar.gz
</pre></div>
</div>
<p>Check the header of the fasta files (it&#8217;s important to know how
chromosomes are called in these files):</p>
<div class="highlight-python"><div class="highlight"><pre>head chr1.fa
</pre></div>
</div>
<p>Create a single file with all chromosomes together:</p>
<div class="highlight-python"><div class="highlight"><pre>cat chr1.fa chr2.fa chr3.fa chr4.fa chr5.fa chr6.fa chr7.fa chr8.fa chr9.fa chr10.fa chr11.fa chr12.fa chr13.fa chr14.fa  chr15.fa  chr16.fa  chr17.fa  chr18.fa  chr19.fa chr20.fa  chr21.fa  chr22.fa chrX.fa  chrY.fa chrM.fa &gt; hg38.fa
</pre></div>
</div>
<p>Index reference genome:</p>
<div class="highlight-python"><div class="highlight"><pre>gemtools index -i hg38.fa -t 8
</pre></div>
</div>
<p>Creates a file hg38.gem that we will be used for the mapping.</p>
</div>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<div class="section" id="retrieving-hi-c-experiment-sra-format">
<h3>Retrieving Hi-C experiment SRA format<a class="headerlink" href="#retrieving-hi-c-experiment-sra-format" title="Permalink to this headline">¶</a></h3>
<p>Along this tutorial we are going to use the dataset from [Dixon2012]_
that can be downloaded at:
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE35156">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE35156</a></p>
<p>For this part of the tutorial we are going to work with the experiments
conducted on <em>Homo sapiens</em> IMR90 and ESC (replicate 1 in both cases).</p>
<p>Each of this replica consists of severa SRA files:</p>
<div class="code python highlight-python"><div class="highlight"><pre>! ls ~/db/FASTQs/hsap/dixon_2012/*/*/*.sra
</pre></div>
</div>
<pre class="ansi-block literal-block">
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116344/SRR400260/SRR400260.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116344/SRR400261/SRR400261.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116344/SRR400262/SRR400262.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116344/SRR400263/SRR400263.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116345/SRR400264/SRR400264.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116345/SRR400265/SRR400265.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116345/SRR400266/SRR400266.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116345/SRR400267/SRR400267.sra
/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116345/SRR400268/SRR400268.sra
</pre>
</div>
<div class="section" id="first-quality-check-on-the-data">
<h3>First quality check on the data<a class="headerlink" href="#first-quality-check-on-the-data" title="Permalink to this headline">¶</a></h3>
<p>First we need to convert SRA files in FASTQ format, as TADbit does not
support reading directly from SRA. For this purpose we may want to use
the SRA-toolkit from the NCBI (<a class="reference external" href="https://github.com/ncbi/sratoolkit">https://github.com/ncbi/sratoolkit</a>). We
finally shall endup with a single FASTQ file per experimental replica,
with commands like (can last for hours):</p>
<p><code class="docutils literal"><span class="pre">fastq-dump</span> <span class="pre">-A</span> <span class="pre">SRX116345</span> <span class="pre">--gzip</span> <span class="pre">SRX116345/*/*.sra</span></code>
<code class="docutils literal"><span class="pre">fastq-dump</span> <span class="pre">-A</span> <span class="pre">SRX116344</span> <span class="pre">--gzip</span> <span class="pre">SRX116344/*/*.sra</span></code></p>
<p>the compressed FASTQ files should be:</p>
<div class="code python highlight-python"><div class="highlight"><pre>! ls ~/db/FASTQs/hsap/dixon_2012/
</pre></div>
</div>
<pre class="ansi-block literal-block">
SRX116344  SRX116345
</pre>
<p>Usig these files directly we can infer the quality of the Hi-C
experiment with TADbit:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.utils.fastq_utils</span> <span class="kn">import</span> <span class="n">quality_plot</span>

<span class="n">quality_plot</span><span class="p">(</span><span class="s">&#39;/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116344.fastq.gz&#39;</span><span class="p">,</span>
             <span class="n">nreads</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">r_enz</span><span class="o">=</span><span class="s">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">paired</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_12_0.png" src="../_images/tutorial_0_mapping_12_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.utils.fastq_utils</span> <span class="kn">import</span> <span class="n">quality_plot</span>

<span class="n">quality_plot</span><span class="p">(</span><span class="s">&#39;/home/fransua/db/FASTQs/hsap/dixon_2012/SRX116345.fastq.gz&#39;</span><span class="p">,</span>
             <span class="n">nreads</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">r_enz</span><span class="o">=</span><span class="s">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">paired</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_13_0.png" src="../_images/tutorial_0_mapping_13_0.png" />
</div>
<div class="section" id="iterative-mapping">
<h3>Iterative mapping<a class="headerlink" href="#iterative-mapping" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.mapper</span> <span class="kn">import</span> <span class="n">iterative_mapping</span>

<span class="n">name</span> <span class="o">=</span> <span class="s">&#39;SRX116344_200&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s">&#39;Mapping Read 1&#39;</span>

<span class="n">sams1</span> <span class="o">=</span> <span class="n">iterative_mapping</span><span class="p">(</span>
            <span class="n">gem_index_path</span>       <span class="o">=</span> <span class="s">&#39;ref_genome/hg38.gem&#39;</span><span class="p">,</span>
            <span class="n">fastq_path</span>           <span class="o">=</span> <span class="s">&#39;/scratch/db/FASTQs/hsap/dixon_2012/</span><span class="si">%s</span><span class="s">.fastq&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">out_sam_path</span>         <span class="o">=</span> <span class="s">&#39;/scratch/mapped_reads/</span><span class="si">%s</span><span class="s">_r1.sam&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">temp_dir</span>             <span class="o">=</span> <span class="s">&#39;/scratch/tmp_dir_</span><span class="si">%s</span><span class="s">/&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">range_start</span>          <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">,</span> <span class="mi">1</span> <span class="p">],</span> <span class="c"># starts with a flag sequence</span>
            <span class="n">range_stop</span>           <span class="o">=</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">35</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">55</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">65</span><span class="p">,</span> <span class="mi">70</span><span class="p">,</span> <span class="mi">75</span><span class="p">],</span>
            <span class="n">nthreads</span>             <span class="o">=</span> <span class="mi">8</span><span class="p">,</span>  <span class="c"># number of CPUs to use</span>
            <span class="n">max_reads_per_chunk</span>  <span class="o">=</span> <span class="mi">40000000</span><span class="p">,</span> <span class="c"># this is for computers with not too much RAM</span>
            <span class="n">single_end</span>           <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="k">print</span> <span class="s">&#39;  created </span><span class="si">%s</span><span class="s"> SAM files:&#39;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">sams1</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping Read 1
Split input file /scratch/db/FASTQs/hsap/dixon_2012/SRX116344_200.fastq into chunks
2 chunks obtained
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
  created 24 SAM files:
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">Mapping Read 2&#39;</span>

<span class="n">sams2</span> <span class="o">=</span> <span class="n">iterative_mapping</span><span class="p">(</span>
            <span class="n">gem_index_path</span>       <span class="o">=</span> <span class="s">&#39;ref_genome/hg38.gem&#39;</span><span class="p">,</span>
            <span class="n">fastq_path</span>           <span class="o">=</span> <span class="s">&#39;/scratch/db/FASTQs/hsap/dixon_2012/</span><span class="si">%s</span><span class="s">.fastq&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">out_sam_path</span>         <span class="o">=</span> <span class="s">&#39;/scratch/mapped_reads/</span><span class="si">%s</span><span class="s">_r2.sam&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">temp_dir</span>             <span class="o">=</span> <span class="s">&#39;/scratch/tmp_dir_</span><span class="si">%s</span><span class="s">/&#39;</span> <span class="o">%</span> <span class="n">name</span><span class="p">,</span>
            <span class="n">range_start</span>          <span class="o">=</span> <span class="p">[</span><span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">,</span> <span class="mi">102</span><span class="p">],</span>
            <span class="n">range_stop</span>           <span class="o">=</span> <span class="p">[</span><span class="mi">121</span><span class="p">,</span> <span class="mi">126</span><span class="p">,</span> <span class="mi">131</span><span class="p">,</span> <span class="mi">136</span><span class="p">,</span> <span class="mi">141</span><span class="p">,</span> <span class="mi">146</span><span class="p">,</span> <span class="mi">151</span><span class="p">,</span> <span class="mi">156</span><span class="p">,</span> <span class="mi">161</span><span class="p">,</span> <span class="mi">166</span><span class="p">,</span> <span class="mi">171</span><span class="p">,</span> <span class="mi">176</span><span class="p">],</span>
            <span class="n">nthreads</span>             <span class="o">=</span> <span class="mi">8</span><span class="p">,</span>
            <span class="n">max_reads_per_chunk</span>  <span class="o">=</span> <span class="mi">40000000</span><span class="p">,</span>
            <span class="n">single_end</span>           <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>

<span class="k">print</span> <span class="s">&#39;  created </span><span class="si">%s</span><span class="s"> SAM files:&#39;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">sams2</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping Read 2
Split input file /scratch/db/FASTQs/hsap/dixon_2012/SRX116344_200.fastq into chunks
2 chunks obtained
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1
Run iterative_mapping recursively on /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
Remove the chunks: /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.1 /scratch/tmp_dir_SRX116344_200/SRX116344_200.fastq.2
  created 24 SAM files:
</pre>
</div>
<div class="section" id="map-parsing">
<h3>Map parsing<a class="headerlink" href="#map-parsing" title="Permalink to this headline">¶</a></h3>
<p>In case we lost the sams1/sams2 lists of paths we can &#8220;reload&#8221; them like
this:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">os</span>

<span class="n">sams1</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">sams2</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">sam</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="s">&#39;/scratch/mapped_reads/&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="s">&#39;_r1.sam&#39;</span> <span class="ow">in</span> <span class="n">sam</span><span class="p">:</span>
        <span class="n">sams1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;/scratch/mapped_reads/&#39;</span> <span class="o">+</span> <span class="n">sam</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sams2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;/scratch/mapped_reads/&#39;</span> <span class="o">+</span> <span class="n">sam</span><span class="p">)</span>
<span class="k">print</span> <span class="n">sams1</span>
<span class="k">print</span> <span class="n">sams2</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
['/scratch/mapped_reads/SRX116344_200_r1.sam.1.3:1-30', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.6:1-45', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.7:1-50', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.8:1-55', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.12:1-75', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.12:1-75', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.6:1-45', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.1:1-20', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.1:1-20', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.9:1-60', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.4:1-35', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.10:1-65', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.4:1-35', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.9:1-60', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.5:1-40', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.2:1-25', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.8:1-55', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.7:1-50', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.11:1-70', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.2:1-25', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.3:1-30', '/scratch/mapped_reads/SRX116344_200_r1.sam.1.5:1-40', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.11:1-70', '/scratch/mapped_reads/SRX116344_200_r1.sam.2.10:1-65']
['/scratch/mapped_reads/SRX116344_200_r2.sam.1.11:102-171', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.2:102-126', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.9:102-161', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.12:102-176', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.5:102-141', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.2:102-126', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.8:102-156', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.10:102-166', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.5:102-141', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.9:102-161', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.7:102-151', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.6:102-146', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.4:102-136', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.1:102-121', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.11:102-171', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.4:102-136', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.7:102-151', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.6:102-146', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.3:102-131', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.8:102-156', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.10:102-166', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.3:102-131', '/scratch/mapped_reads/SRX116344_200_r2.sam.2.1:102-121', '/scratch/mapped_reads/SRX116344_200_r2.sam.1.12:102-176']
</pre>
<div class="section" id="mapping-analysis">
<h4>Mapping analysis<a class="headerlink" href="#mapping-analysis" title="Permalink to this headline">¶</a></h4>
<p>We collect mapped reads at all window sizes into a single file (a single
file for read1, and a single file for read2). These 2 files also contain
the placement of the restriction enzyme sites in the genome.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.parsers.sam_parser</span>    <span class="kn">import</span> <span class="n">parse_sam</span>
<span class="kn">from</span> <span class="nn">pytadbit.parsers.genome_parser</span> <span class="kn">import</span> <span class="n">parse_fasta</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c"># loads the genome</span>
<span class="n">genome_seq</span> <span class="o">=</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="s">&#39;/scratch/db/index_files/Homo_sapiens-79/Homo_sapiens.fa&#39;</span><span class="p">,</span>
                         <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;chr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s">&#39;X&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="s">&#39;MT&#39;</span><span class="p">]])</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Parsing 1 as chr1
Parsing 2 as chr2
Parsing 3 as chr3
Parsing 4 as chr4
Parsing 5 as chr5
Parsing 6 as chr6
Parsing 7 as chr7
Parsing 8 as chr8
Parsing 9 as chr9
Parsing 10 as chr10
Parsing 11 as chr11
Parsing 12 as chr12
Parsing 13 as chr13
Parsing 14 as chr14
Parsing 15 as chr15
Parsing 16 as chr16
Parsing 17 as chr17
Parsing 18 as chr18
Parsing 19 as chr19
Parsing 20 as chr20
Parsing 21 as chr21
Parsing 22 as chr22
Parsing X as chrX
Parsing Y as chrY
Parsing MT as chrMT
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">reads1</span> <span class="o">=</span> <span class="s">&#39;/scratch/results/</span><span class="si">%s</span><span class="s">_r1_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span> <span class="c"># new file with info of each &quot;read1&quot; and its placement with respect to RE sites</span>
<span class="n">reads2</span> <span class="o">=</span> <span class="s">&#39;/scratch/results/</span><span class="si">%s</span><span class="s">_r2_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span> <span class="c"># new file with info of each &quot;read2&quot; and its placement with respect to RE sites</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s">&#39;Parse SAM files...&#39;</span>
<span class="n">parse_sam</span><span class="p">(</span><span class="n">sams1</span><span class="p">,</span> <span class="n">sams2</span><span class="p">,</span> <span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">genome_seq</span><span class="p">,</span> <span class="s">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Parse SAM files...
Searching and mapping RE sites to the reference genome
Found 859855 RE sites
Loading read1
Loading read2
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.3:1-30
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.6:1-45
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.7:1-50
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.8:1-55
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.12:1-75
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.12:1-75
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.6:1-45
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.1:1-20
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.2:1-25
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.9:1-60
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.10:1-65
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.1:1-20
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.4:1-35
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.4:1-35
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.9:1-60
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.2:1-25
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.3:1-30
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.11:1-70
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.8:1-55
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.11:1-70
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.5:1-40
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.7:1-50
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.2.10:1-65
loading GEM file: /scratch/mapped_reads/SRX116344_200_r1.sam.1.5:1-40
Sorting reads
Removing temporary files...loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.11:102-171
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.2:102-126
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.9:102-161
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.12:102-176
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.5:102-141
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.2:102-126
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.10:102-166
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.8:102-156
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.11:102-171
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.6:102-146
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.5:102-141
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.9:102-161
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.7:102-151
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.1:102-121
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.4:102-136
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.6:102-146
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.4:102-136
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.3:102-131
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.7:102-151
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.8:102-156
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.3:102-131
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.10:102-166
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.1.12:102-176
loading GEM file: /scratch/mapped_reads/SRX116344_200_r2.sam.2.1:102-121
</pre>
</div>
</div>
<div class="section" id="plot-iterative-mapping">
<h3>Plot iterative mapping<a class="headerlink" href="#plot-iterative-mapping" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">plot_iterative_mapping</span>

<span class="n">lengths</span> <span class="o">=</span> <span class="n">plot_iterative_mapping</span><span class="p">(</span><span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">total_reads</span><span class="o">=</span><span class="mi">51493359</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_28_0.png" src="../_images/tutorial_0_mapping_28_0.png" />
</div>
<div class="section" id="merging-mapped-read1-and-read2">
<h3>Merging mapped &#8220;read1&#8221; and &#8220;read2&#8221;<a class="headerlink" href="#merging-mapped-read1-and-read2" title="Permalink to this headline">¶</a></h3>
<p>We create a new file that will contain only the reads mapped in both
ends (&#8220;read1&#8221; and &#8220;read2&#8221; uniquely mapped)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.mapper</span> <span class="kn">import</span> <span class="n">get_intersection</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">reads</span>  <span class="o">=</span> <span class="s">&#39;/scratch/results/</span><span class="si">%s</span><span class="s">_both_map.tsv&#39;</span> <span class="o">%</span> <span class="n">name</span>

<span class="n">get_intersection</span><span class="p">(</span><span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">reads</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Found 41123290 pair of reads mapping uniquely
</pre>
<div class="section" id="stats-on-intersection">
<h4>Stats on intersection<a class="headerlink" href="#stats-on-intersection" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">insert_sizes</span>

<span class="n">insert_sizes</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_34_0.png" src="../_images/tutorial_0_mapping_34_0.png" />
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="">Mapping paired-end reads from NGS experiment (with GEM)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads">Identification and Analysis of TADs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../tutorial.html"
                        title="previous chapter">Tutorials</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_1_general.html"
                        title="next chapter">Getting started</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             >next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright .
      Last updated on May 09, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>