<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Hi-C quality check &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Iterative vs fragment-based mapping" href="tutorial_4-Mapping.html" />
    <link rel="prev" title="Preparation of the reference genome" href="tutorial_2-Preparation_of_the_reference_genome.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_4-Mapping.html" title="Iterative vs fragment-based mapping"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_2-Preparation_of_the_reference_genome.html" title="Preparation of the reference genome"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="hi-c-quality-check">
<h1>Hi-C quality check<a class="headerlink" href="#hi-c-quality-check" title="Permalink to this headline">¶</a></h1>
<p>The file is organized in 4 lines per read: 1. starting with <code class="docutils literal notranslate"><span class="pre">&#64;</span></code>, the
header of the DNA sequence with the read id (plus optional fields) 2.
the DNA sequence 3. starting with <code class="docutils literal notranslate"><span class="pre">+</span></code>, the header of the sequence
quality (this line could be either a repetition of first line or empty)
4. the sequence quality (it is provided as PHRED score and it is not
human readable. Check
<a class="reference external" href="https://en.wikipedia.org/wiki/Phred_quality_score">https://en.wikipedia.org/wiki/Phred_quality_score</a> for more details)</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">bash</span>

<span class="n">dsrc</span> <span class="n">d</span> <span class="o">-</span><span class="n">s</span> <span class="n">FASTQs</span><span class="o">/</span><span class="n">mouse_B_rep1_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">dsrc</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="mi">8</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
&#64;SRR5344921.1
ACCACTGATGGCACCACTGTTGTACTTTAGGATCGATCACTCCTTATACGCACATAGGAAACACTGACTAGAGGG
+
&#64;B&#64;B&#64;&gt;1&gt;=;1&#64;EG0B0;1=E1:&gt;&gt;1111&lt;?=1E9=0&gt;&lt;C:&lt;&gt;11=11?1=ECDC1&gt;1=&#64;FD1&lt;:1?C#######
&#64;SRR5344921.2
GGGCTTTTTTTGGCTGGGAGACTATTTATAACTGCTTCTATTTCTTTAGGGGATATGGGACTGTTTAGAAGGTCA
+
BCCBBFGGGG&#64;GEGGGGDGEBGEF&#64;&lt;1FGGEFGGECGG&#64;FEGGGEG&gt;&#64;1=09/BFGGEG111C&lt;F1&gt;11BGF1=1
</pre>
<p>Count the number of lines in the file (4 times the number of reads)</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">bash</span>

<span class="n">dsrc</span> <span class="n">d</span> <span class="o">-</span><span class="n">s</span> <span class="n">FASTQs</span><span class="o">/</span><span class="n">mouse_B_rep1_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">dsrc</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
400000000
</pre>
<p>There are 400 M lines in the file, which means 100M reads in total.</p>
<div class="section" id="quality-check-before-mapping">
<h2>Quality check before mapping<a class="headerlink" href="#quality-check-before-mapping" title="Permalink to this headline">¶</a></h2>
<div class="section" id="check-restriction-enzyme-used">
<h3>Check restriction-enzyme used<a class="headerlink" href="#check-restriction-enzyme-used" title="Permalink to this headline">¶</a></h3>
<p>Most important to analyze Hi-C dataset is the restriction enzyme used in
the experiment. TADbit provides a simple function to check for it:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.mapping.restriction_enzymes</span> <span class="k">import</span> <span class="n">identify_re</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pat</span><span class="p">,</span> <span class="n">enz</span><span class="p">,</span> <span class="n">pv</span> <span class="o">=</span> <span class="n">identify_re</span><span class="p">(</span><span class="s1">&#39;FASTQs/mouse_B_rep1_1.fastq.dsrc&#39;</span><span class="p">)</span>

<span class="nb">print</span> <span class="s1">&#39;- Most probable pattern: </span><span class="si">%s</span><span class="s1">, matching enzymes: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">pat</span><span class="p">,</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">enz</span><span class="p">))</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
- Most probable pattern: GATC, matching enzymes: BssMI,DpnII,Sau3AI,MboI,Kzo9I,Bsp143I,NdeII,BstMBI,BfuCI
</pre>
</div>
<div class="section" id="plot-phred-score-and-ligation-digestion-sites">
<h3>Plot PHRED score and ligation/digestion sites<a class="headerlink" href="#plot-phred-score-and-ligation-digestion-sites" title="Permalink to this headline">¶</a></h3>
<p>In order to quickly assess the quality of the HiC experiment (before
mapping), and given that we know the restriction enzyme used, we can
check the proportion of reads with ligation sites as well as the number
of reads starting by a cut-site.</p>
<p>These numbers will give us a first hint on the <strong>efficiencies of two
critical steps in the HiC experiment, the digestion and the ligation.</strong></p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.utils.fastq_utils</span> <span class="k">import</span> <span class="n">quality_plot</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">r_enz</span> <span class="o">=</span> <span class="s1">&#39;MboI&#39;</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cell</span> <span class="o">=</span> <span class="s1">&#39;B&#39;</span>
<span class="n">repl</span> <span class="o">=</span> <span class="s1">&#39;rep1&#39;</span>
</pre></div>
</div>
<p>The plot on the top represents the typical per nucleotide quality
profile of NGS reads, with, in addition, the proportion of <code class="docutils literal notranslate"><span class="pre">N</span></code> found
at each position.</p>
<p>The second plot, is specific to Hi-C experiments. Given a restriction
enzyme the function searches for the presence of ligation sites and of
undigested restriction enzyme sites. Depending on the enzyme used the
function can differentiate between dangling-ends and undigested sites.</p>
<p>From these proportions some quality statistics can be inferred before
mapping: - The <strong>PHRED score</strong> and the <strong>number of unidentified
nucleotides</strong> (Ns) in the read sequence, which are routinely computed to
address the quality of high-throughput sequencing experiments. - The
numbers of <strong>undigested and unligated RE sites per-nucleotide</strong> along
the read to assess the quality of the Hi-C experiment. - The overall
<strong>percentage of digested sites</strong>, which relates directly to the RE
efficiency. - The <strong>percentage of non-ligated digested</strong>
(dangling-ends), which relates to the ligation efficiency. - The
<strong>percentage of read-ends with a ligation site</strong>, which is negatively
correlated with the percentage of dangling-ends.</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">quality_plot</span><span class="p">(</span><span class="s1">&#39;FASTQs/mouse_</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">_1.fastq.dsrc&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">repl</span><span class="p">),</span> <span class="n">r_enz</span><span class="o">=</span><span class="n">r_enz</span><span class="p">,</span> <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_3-Hi-C_quality_check_17_0.png" src="../_images/tutorial_3-Hi-C_quality_check_17_0.png" />
<pre class="ansi-block literal-block">
({'MboI': 3.8757}, {('MboI', 'MboI'): 20.5069})
</pre>
<p><em>Note: this plot is **compatible with the use of multiple restriction
enzymes*</em> (which is why the ligation site is labeled as MboI-MboI)*</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_1-Retrieve_published_HiC_datasets.html">Retrieve HiC dataset from NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_2-Preparation_of_the_reference_genome.html">Preparation of the reference genome</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Hi-C quality check</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html">Iterative vs fragment-based mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_4-Mapping.html#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html">Load mapped read-ends and generate a matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_5-Parsing_mapped_reads.html#quality-check-of-the-hi-c-experiment">Quality check of the Hi-C experiment</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_6-Filtering_mapped_reads.html">Filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_7-Bin_filtering_and_normalization.html">Bin filtering and normalization</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_2-Preparation_of_the_reference_genome.html"
                        title="previous chapter">Preparation of the reference genome</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_4-Mapping.html"
                        title="next chapter">Iterative vs fragment-based mapping</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_4-Mapping.html" title="Iterative vs fragment-based mapping"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_2-Preparation_of_the_reference_genome.html" title="Preparation of the reference genome"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Jun 23, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>