<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Getting started &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Tutorials" href="../tutorial.html" />
    <link rel="next" title="Alignment of TAD boundaries" href="tutorial_2_boundary_alignment.html" />
    <link rel="prev" title="Tutorials" href="../tutorial.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_2_boundary_alignment.html" title="Alignment of TAD boundaries"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="getting-started">
<span id="getting-start"></span><h1>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h1>
<div class="section" id="hi-c-data-format">
<h2>Hi-C data format<a class="headerlink" href="#hi-c-data-format" title="Permalink to this headline">¶</a></h2>
<p>Hi-C data are usually represented as symmetric matrices in a tab
separated file, as in the example below:</p>
<div class="highlight-python"><div class="highlight"><pre>chrT_001      chrT_002        chrT_003        chrT_004        chrT_005        chrT_006
chrT_001      629     164     88      105     10      35
chrT_002      86      612     175     110     40      29
chrT_003      159     216     437     105     43      73
chrT_004      100     111     146     278     70      42
chrT_005      16      36      26      79      243     39
chrT_006      19      50      53      42      37      224
</pre></div>
</div>
<p>TADbit allows to load most of this kind of matrices. A Hi-C matrix is
loaded as this <em>(note: this example loads human chromosome 19 from</em>
[Lieberman-Aiden2009]_ <em>results.)</em>:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="kn">import</span> <span class="n">Chromosome</span>

<span class="c1"># initiate a chromosome object that will store all Hi-C data and analysis</span>
<span class="n">my_chrom</span> <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="n">centromere_search</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                      <span class="n">species</span><span class="o">=</span><span class="s1">&#39;Homo sapiens&#39;</span><span class="p">,</span> <span class="n">assembly</span><span class="o">=</span><span class="s1">&#39;NCBI36&#39;</span><span class="p">)</span>

<span class="c1"># load Hi-C data</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s1">&#39;k562&#39;</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="s1">&#39;wild type&#39;</span><span class="p">,</span> <span class="n">exp_type</span><span class="o">=</span><span class="s1">&#39;Hi-C&#39;</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="s1">&#39;k562&#39;</span><span class="p">,</span>
                        <span class="n">project</span><span class="o">=</span><span class="s1">&#39;TADbit tutorial&#39;</span><span class="p">,</span>
                        <span class="n">hic_data</span><span class="o">=</span><span class="s2">&quot;../../scripts/sample_data/HIC_k562_chr19_chr19_100000_obs.txt&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s1">&#39;gm06690&#39;</span><span class="p">,</span>  <span class="n">cell_type</span><span class="o">=</span><span class="s1">&#39;cancer&#39;</span><span class="p">,</span> <span class="n">exp_type</span><span class="o">=</span><span class="s1">&#39;Hi-C&#39;</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="s1">&#39;gm06690&#39;</span><span class="p">,</span>
                        <span class="n">project</span><span class="o">=</span><span class="s1">&#39;TADbit tutorial&#39;</span><span class="p">,</span>
                        <span class="n">hic_data</span><span class="o">=</span><span class="s2">&quot;../../scripts/sample_data/HIC_gm06690_chr19_chr19_100000_obs.txt&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
/usr/local/lib/python2.7/dist-packages/pytadbit/parsers/hic_parser.py:93: UserWarning: WARNING: non integer values
  warn('WARNING: non integer values')
</pre>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">TADbit assumes that Hi-C data matrices starts at chromosome position 1. If your matrix do not represents the full chromosome length, fill the missing columns and rows with zeros.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">by default TADbit will not search for centromeres, however, as, in this example, we know that we are working with chromosome 19, we can ask TADbit search for it as explained in <a class="reference internal" href="#forbidden-regions"><span>Forbidden regions and centromeres</span></a></p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">TADbit parser is also able to read compressed data directly (gzip files).</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Many optional descriptive arguments can be passed to the chromosome and to the experiments. This arguments may be used by TADbit when generating the output.</p>
</div>
<div class="section" id="unrecognized-data-format">
<h3>Unrecognized data format<a class="headerlink" href="#unrecognized-data-format" title="Permalink to this headline">¶</a></h3>
<p>In some cases users may found data that are not readable by TADbit. In
such case, a parser function can be written and passed as an argument to
the <code class="docutils literal"><span class="pre">add_experiment</span></code> function:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">read_strange_file</span><span class="p">(</span><span class="n">f_handle</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    reads from file handler (already openned)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">nums</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f_handle</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="c1"># modify the following line to fit your parsing needings</span>
        <span class="c1">##</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="c1">##</span>
        <span class="c1"># feed the &quot;num&quot; list with the list of values you parsed</span>
        <span class="n">nums</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">values</span><span class="p">])</span>
    <span class="k">return</span> <span class="nb">tuple</span><span class="p">([</span><span class="n">nums</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nums</span><span class="p">))</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nums</span><span class="p">))]),</span> <span class="nb">len</span><span class="p">(</span><span class="n">nums</span><span class="p">)</span>
</pre></div>
</div>
<p>And call it as follow:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">other_chrom</span> <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;An other chromosome&#39;</span><span class="p">)</span>
<span class="n">other_chrom</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s1">&#39;First Hi-C experiment&#39;</span><span class="p">,</span> <span class="n">hic_data</span><span class="o">=</span><span class="s1">&#39;../../src/test/data/hESC_chr19-rep1.txt&#39;</span><span class="p">,</span>
                           <span class="n">parser</span><span class="o">=</span><span class="n">read_strange_file</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">20000</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="experiment-objects">
<h2>Experiment objects<a class="headerlink" href="#experiment-objects" title="Permalink to this headline">¶</a></h2>
<p>Experiments, when loaded, are stored in a special kind of list attached
to chromosome objects:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
[Experiment k562 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: None),
 Experiment gm06690 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: None)]
</pre>
<p>A specific Experiment can be accessed either by its name or by its
position in :class:<code class="docutils literal"><span class="pre">pytadbit.chromosome.ExperimentList</span></code> :</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s2">&quot;k562&quot;</span><span class="p">]</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
True
</pre>
<p>Each Experiment is an independent object with a list of associated functions
(see <a class="reference internal" href="../reference/reference_experiment.html#pytadbit.experiment.Experiment" title="pytadbit.experiment.Experiment"><code class="xref py py-class docutils literal"><span class="pre">pytadbit.experiment.Experiment</span></code></a>).</p>
<div class="section" id="basic-manipulation-of-experiments">
<span id="exp-operations"></span><h3>Basic manipulation of Experiments<a class="headerlink" href="#basic-manipulation-of-experiments" title="Permalink to this headline">¶</a></h3>
<p>Two Hi-C experiments can be summed up easily, resulting in a new Hi-c
experiment contatining the sum of the interaction counts of the summed
experiments:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">exp</span> <span class="o">=</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s2">&quot;k562&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s2">&quot;gm06690&quot;</span><span class="p">]</span>
<span class="k">print</span> <span class="n">exp</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Experiment k562+gm06690:
   resolution        : 100Kb
   TADs              : None
   Hi-C rows         : 639
   normalized        : None
   identifier        : k562+gm06690
   cell type         : wild type+cancer
   restriction enzyme: UNKNOWN
   project           : TADbit tutorial
</pre>
<pre class="ansi-block literal-block">
WARNING: experiments should be normalized before being summed
</pre>
<p><em>Note the last warning asking to normalize experiments before summing
them. Indeed normalizing the sum of different experiment should result
in bayesing the final result towards the experiment with more raw Hi-C
interaction counts.</em></p>
<p>The resulting experiment (which default name is the concatenation of the
summed experiments) can be added to the experiments of a given
chromosome.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="n">exp</span><span class="p">)</span>
<span class="k">print</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
[Experiment k562 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: None), Experiment gm06690 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: None), Experiment k562+gm06690 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: None)]
</pre>
</div>
<div class="section" id="hi-c-matric-visualization">
<h3>Hi-C matric visualization<a class="headerlink" href="#hi-c-matric-visualization" title="Permalink to this headline">¶</a></h3>
<p>To quickly view how does the interaction matrix look like, experiment objects have the <a class="reference internal" href="../reference/reference_experiment.html#pytadbit.experiment.Experiment.view" title="pytadbit.experiment.Experiment.view"><code class="xref py py-func docutils literal"><span class="pre">pytadbit.experiment.Experiment.view()</span></code></a> function</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">exp</span><span class="o">.</span><span class="n">view</span><span class="p">()</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
/usr/local/lib/python2.7/dist-packages/pytadbit/experiment.py:1090: RuntimeWarning: divide by zero encountered in log2
  img = axe.imshow(fun(matrix), origin='lower', vmin=vmin, vmax=vmax,
</pre>
<img alt="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_27_1.png" src="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_27_1.png" />
<pre class="ansi-block literal-block">
&lt;matplotlib.image.AxesImage at 0x7f6fa312e7d0&gt;
</pre>
<p>This plot shows the log2 interaction counts, resulting from the given
Hi-C experiment, and it can be drawn from individual experiments or from
the chromosome object.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">visualize</span><span class="p">([(</span><span class="s1">&#39;k562&#39;</span><span class="p">,</span> <span class="s1">&#39;gm06690&#39;</span><span class="p">),</span> <span class="s1">&#39;k562+gm06690&#39;</span><span class="p">])</span>
</pre></div>
</div>
<img alt="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_29_0.png" src="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_29_0.png" />
<p>Note how we pass the list of experiments to show to the <a class="reference internal" href="../reference/reference_chromosome.html#pytadbit.chromosome.Chromosome.visualize" title="pytadbit.chromosome.Chromosome.visualize"><code class="xref py py-func docutils literal"><span class="pre">pytadbit.chromosome.Chromosome.visualize()</span></code></a> we ask to view 3 experiments, the first two being grouped in a single matrix. This view is useful to compare experiments, and do not suppose a lose of information as Hi-C matrices are symetric.</p>
</div>
</div>
<div class="section" id="find-topologically-associating-domains">
<span id="run-tadbit"></span><h2>Find Topologically Associating Domains<a class="headerlink" href="#find-topologically-associating-domains" title="Permalink to this headline">¶</a></h2>
<p>Once an experiment has been loaded, the location of Topologically Associating Domains (TADs) can be estimated as:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">find_tad</span><span class="p">(</span><span class="s1">&#39;k562&#39;</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">my_chrom</span><span class="o">.</span><span class="n">find_tad</span><span class="p">(</span><span class="s1">&#39;gm06690&#39;</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<p><a class="reference internal" href="../reference/reference_chromosome.html#pytadbit.chromosome.Chromosome.find_tad" title="pytadbit.chromosome.Chromosome.find_tad"><code class="xref py py-func docutils literal"><span class="pre">pytadbit.chromosome.Chromosome.find_tad()</span></code></a> is called from our Chromosome object but is applied to a
specific experiment. Therefore, TADs found by TADBbit will be associated to this specific experiment.
They can be accessed as following:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">exp</span> <span class="o">=</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s2">&quot;k562&quot;</span><span class="p">]</span>
<span class="n">exp</span><span class="o">.</span><span class="n">tads</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
{1: {'brk': 5.0, 'end': 5.0, 'score': 2.0, 'start': 0.0},
 2: {'brk': 5.0, 'end': 5.0, 'score': 2.0, 'start': 0.0},
 3: {'brk': 11.0, 'end': 11.0, 'score': 2.0, 'start': 6.0},
 4: {'brk': 31.0, 'end': 31.0, 'score': 7.0, 'start': 12.0},
 5: {'brk': 45.0, 'end': 45.0, 'score': 5.0, 'start': 32.0},
 6: {'brk': 56.0, 'end': 56.0, 'score': 5.0, 'start': 46.0},
 7: {'brk': 69.0, 'end': 69.0, 'score': 7.0, 'start': 57.0},
 8: {'brk': 82.0, 'end': 82.0, 'score': 3.0, 'start': 70.0},
 9: {'brk': 103.0, 'end': 103.0, 'score': 4.0, 'start': 83.0},
 10: {'brk': 108.0, 'end': 108.0, 'score': 7.0, 'start': 104.0},
 11: {'brk': 128.0, 'end': 128.0, 'score': 6.0, 'start': 109.0},
 12: {'brk': 183.0, 'end': 183.0, 'score': 6.0, 'start': 129.0},
 13: {'brk': 194.0, 'end': 194.0, 'score': 4.0, 'start': 184.0},
 14: {'brk': 236.0, 'end': 236.0, 'score': 4.0, 'start': 195.0},
 15: {'brk': 244.0, 'end': 244.0, 'score': 5.0, 'start': 237.0},
 16: {'brk': 329.0, 'end': 329.0, 'score': 4.0, 'start': 324.0},
 17: {'brk': 347.0, 'end': 347.0, 'score': 4.0, 'start': 330.0},
 18: {'brk': 352.0, 'end': 352.0, 'score': 2.0, 'start': 348.0},
 19: {'brk': 377.0, 'end': 377.0, 'score': 8.0, 'start': 353.0},
 20: {'brk': 383.0, 'end': 383.0, 'score': 3.0, 'start': 378.0},
 21: {'brk': 399.0, 'end': 399.0, 'score': 6.0, 'start': 384.0},
 22: {'brk': 412.0, 'end': 412.0, 'score': 9.0, 'start': 400.0},
 23: {'brk': 432.0, 'end': 432.0, 'score': 3.0, 'start': 413.0},
 24: {'brk': 445.0, 'end': 445.0, 'score': 6.0, 'start': 433.0},
 25: {'brk': 471.0, 'end': 471.0, 'score': 4.0, 'start': 446.0},
 26: {'brk': 477.0, 'end': 477.0, 'score': 5.0, 'start': 472.0},
 27: {'brk': 485.0, 'end': 485.0, 'score': 7.0, 'start': 478.0},
 28: {'brk': 500.0, 'end': 500.0, 'score': 3.0, 'start': 486.0},
 29: {'brk': 505.0, 'end': 505.0, 'score': 7.0, 'start': 501.0},
 30: {'brk': 521.0, 'end': 521.0, 'score': 2.0, 'start': 506.0},
 31: {'brk': 530.0, 'end': 530.0, 'score': 5.0, 'start': 522.0},
 32: {'brk': 553.0, 'end': 553.0, 'score': 4.0, 'start': 531.0},
 33: {'brk': 562.0, 'end': 562.0, 'score': 6.0, 'start': 554.0},
 34: {'brk': 569.0, 'end': 569.0, 'score': 5.0, 'start': 563.0},
 35: {'brk': 593.0, 'end': 593.0, 'score': 6.0, 'start': 570.0},
 36: {'brk': 608.0, 'end': 608.0, 'score': 7.0, 'start': 594.0},
 37: {'brk': 638.0, 'end': 638.0, 'score': 10.0, 'start': 609.0}}
</pre>
<p>The &#8220;tads&#8221; variable returned in this example is a dictionary of TADs,
each of each is in turn a new dictionary containing information about
the start and end positions of a TAD.</p>
<p>&#8220;start&#8221; and &#8220;end&#8221; values correspond respectively to the start and end
positions of the given TAD in the chromosome (note that this numbers
have to be multiplied by the resolution of the experiment,
&#8220;exp.resolution&#8221;). The &#8220;brk&#8221; key corresponds to the value of &#8220;end&#8221;, all
&#8220;brk&#8221; together corresponds to all TAD&#8217;s boundaries.</p>
<div class="section" id="tads-to-text">
<h3>TADs to text<a class="headerlink" href="#tads-to-text" title="Permalink to this headline">¶</a></h3>
<p>TADs can also be seen or saved to a file using this write function:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">exp</span><span class="o">.</span><span class="n">write_tad_borders</span><span class="p">()</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
#      start   end score
2        1.0   6.0   2.0
3        1.0   6.0   2.0
4        7.0  12.0   2.0
5       13.0  32.0   7.0
6       33.0  46.0   5.0
7       47.0  57.0   5.0
8       58.0  70.0   7.0
9       71.0  83.0   3.0
10      84.0 104.0   4.0
11     105.0 109.0   7.0
12     110.0 129.0   6.0
13     130.0 184.0   6.0
14     185.0 195.0   4.0
15     196.0 237.0   4.0
16     238.0 245.0   5.0
17     325.0 330.0   4.0
18     331.0 348.0   4.0
19     349.0 353.0   2.0
20     354.0 378.0   8.0
21     379.0 384.0   3.0
22     385.0 400.0   6.0
23     401.0 413.0   9.0
24     414.0 433.0   3.0
25     434.0 446.0   6.0
26     447.0 472.0   4.0
27     473.0 478.0   5.0
28     479.0 486.0   7.0
29     487.0 501.0   3.0
30     502.0 506.0   7.0
31     507.0 522.0   2.0
32     523.0 531.0   5.0
33     532.0 554.0   4.0
34     555.0 563.0   6.0
35     564.0 570.0   5.0
36     571.0 594.0   6.0
37     595.0 609.0   7.0
38     610.0 639.0  10.0
</pre>
<p>Another way to view TADs is using the matrix visualization:</p>
</div>
<div class="section" id="tads-in-interaction-matrices">
<h3>TADs in interaction matrices<a class="headerlink" href="#tads-in-interaction-matrices" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_43_0.png" src="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_43_0.png" />
<p>In this case we can also put them side by side, view a given region, and
use nrmalized data instead of the log2 of raw data:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">visualize</span><span class="p">([(</span><span class="s1">&#39;k562&#39;</span><span class="p">,</span> <span class="s1">&#39;gm06690&#39;</span><span class="p">)],</span> <span class="n">paint_tads</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="p">(</span><span class="mi">490</span><span class="p">,</span><span class="mi">620</span><span class="p">),</span> <span class="n">normalized</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
/usr/local/lib/python2.7/dist-packages/pytadbit/experiment.py:1090: RuntimeWarning: invalid value encountered in log2
  img = axe.imshow(fun(matrix), origin='lower', vmin=vmin, vmax=vmax,
</pre>
<img alt="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_45_1.png" src="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_45_1.png" />
<p><em>Note that the width of the line is proportional to the score of the TAD
border.</em></p>
</div>
<div class="section" id="tads-in-density-plots">
<span id="density-plot"></span><h3>TADs in density plots<a class="headerlink" href="#tads-in-density-plots" title="Permalink to this headline">¶</a></h3>
<p>Finally TAD bourders can be seen using the density plot summary:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">tad_density_plot</span><span class="p">(</span><span class="s1">&#39;k562&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_50_0.png" src="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_50_0.png" />
<p>In this plot, each grey-filled-arc represents a TAD. The eight of the
TAD is proportional to the relative amount of interactions in this TAD.
If this relative amount of interactions is higher than 1 (dark grey
TADs) the number of interactions inside the TAD is higher than expected
accordin to its size (a black horizontal line, highlights this
theoretical threshold).</p>
<p>TAD borders in this plot are marked by colored triangles. The color
code, from blue to red, represents how confident is TADbit about the
identification of the border.</p>
</div>
</div>
<div class="section" id="finding-tads-in-related-hi-c-experiments">
<h2>Finding TADs in related Hi-C experiments<a class="headerlink" href="#finding-tads-in-related-hi-c-experiments" title="Permalink to this headline">¶</a></h2>
<p>TADbit also allows to search for TADs in a chromosome using the information of several Hi-C experiments. To do this experiments do not need to be summed up (like in <a class="reference internal" href="#exp-operations"><span>Basic manipulation of Experiments</span></a>), as the find_tad function has a batch_mode:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">find_tad</span><span class="p">([</span><span class="s1">&#39;k562&#39;</span><span class="p">,</span> <span class="s1">&#39;gm06690&#39;</span><span class="p">],</span> <span class="n">batch_mode</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="k">print</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
[Experiment k562 (resolution: 100Kb, TADs: 37, Hi-C rows: 639, normalized: visibility), Experiment gm06690 (resolution: 100Kb, TADs: 34, Hi-C rows: 639, normalized: visibility), Experiment k562+gm06690 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: None), Experiment batch_gm06690_k562 (resolution: 100Kb, TADs: 35, Hi-C rows: 639, normalized: visibility)]
</pre>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In this case a new experiment is also created (e.g. &#8220;batch_gm06690_k562&#8221;), it contains the TADs detected when taking into account the two experiments &#8220;gm06690&#8221; and &#8220;k562&#8221;.</p>
</div>
</div>
<div class="section" id="forbidden-regions-and-centromeres">
<span id="forbidden-regions"></span><h2>Forbidden regions and centromeres<a class="headerlink" href="#forbidden-regions-and-centromeres" title="Permalink to this headline">¶</a></h2>
<p>By default TADbit does not put limitation in sizes of TADs, owever it may make sense to set a maximum TAD size of 5 Mb in humans <a class="reference internal" href="../biblio.html#dekker2013" id="id1">[Dekker2013]</a>. In this case we can set this cutoff when first defining the chromosome (see the option in <code class="xref py py-class docutils literal"><span class="pre">pytadbit.Chromosome</span></code>), or just use the function <code class="xref py py-class docutils literal"><span class="pre">pytadbit.Chromosome.set_max_tad_size</span></code>:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">set_max_tad_size</span><span class="p">(</span><span class="mi">3000000</span><span class="p">)</span>
</pre></div>
</div>
<p>Once TADs are detected by the core <a class="reference internal" href="../reference/reference_tadbit.html#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><code class="xref py py-func docutils literal"><span class="pre">pytadbit.tadbit.tadbit()</span></code></a> function, TADbit checks that they are not larger than a given value (5 Mb here). If a TAD is larger than this value, it will be marked with a
<strong>negative score</strong>, and will be automatically excluded from the main TADbit functions.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="s1">&#39;k562&#39;</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_61_0.png" src="home/fransua/Box/tadbits/tadbit/doc/notebooks/tutorial_1_general_61_0.png" />
<p>Another optional inspection performed by TADbit is the presence of
centromeric regions. TADbit assumes that the larger gap found in a Hi-C
matrix corresponds to the centromere. In the case of this example we set
the argument &#8220;search_centromere=True&#8221;, but the default is False, as it
may give unexpected results in telocentric chromosomes.</p>
<p>The search for centromere is updated and refined each time a new
experiment is linked to a given Chromosome. Typically, TADs calculated
by the :func:<code class="docutils literal"><span class="pre">pytadbit.tadbit.tadbit</span></code> function include centromeric
regions; if a centromere is found, TADbit will split the TAD containing
it into two TADs, one ending before the centromere and one starting
after.</p>
</div>
<div class="section" id="saving-and-restoring-data">
<h2>Saving and restoring data<a class="headerlink" href="#saving-and-restoring-data" title="Permalink to this headline">¶</a></h2>
<p>In order to avoid having to calculate TAD positions each time, TADbit allows to save and load Chromosome
objects, with all the associated experiments. To save a Chromosome object:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">my_chrom</span><span class="o">.</span><span class="n">save_chromosome</span><span class="p">(</span><span class="s2">&quot;some_path.tdb&quot;</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>And to load it:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="kn">import</span> <span class="n">load_chromosome</span>

<span class="n">my_chrom</span> <span class="o">=</span> <span class="n">load_chromosome</span><span class="p">(</span><span class="s2">&quot;some_path.tdb&quot;</span><span class="p">)</span>

<span class="k">print</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
[Experiment k562 (resolution: 100Kb, TADs: 37, Hi-C rows: 639, normalized: None), Experiment gm06690 (resolution: 100Kb, TADs: 34, Hi-C rows: 639, normalized: None), Experiment k562+gm06690 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: None), Experiment batch_gm06690_k562 (resolution: 100Kb, TADs: 35, Hi-C rows: 639, normalized: None)]
</pre>
<p><em>Note: while information about TADs can be saved, in order to save disk space, raw Hi-C data are not stored in this way but can be loaded again for each experiment:</em></p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expr</span> <span class="o">=</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s2">&quot;k562&quot;</span><span class="p">]</span>

<span class="n">expr</span><span class="o">.</span><span class="n">load_hic_data</span><span class="p">(</span><span class="s2">&quot;../../scripts/sample_data/HIC_k562_chr19_chr19_100000_obs.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads">Identification and Analysis of TADs</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="">Getting started</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_2_boundary_alignment.html">Alignment of TAD boundaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_3_clustering.html">TAD clustering</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../tutorial.html"
                        title="previous chapter">Tutorials</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_2_boundary_alignment.html"
                        title="next chapter">Alignment of TAD boundaries</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_2_boundary_alignment.html" title="Alignment of TAD boundaries"
             >next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright .
      Last updated on Apr 12, 2016.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>