<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Compartments and TADs detection &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Comparison of Hi-C experiments" href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" />
    <link rel="prev" title="Bin filtering and normalization" href="tutorial_7-Bin_filtering_and_normalization.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" title="Comparison of Hi-C experiments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_7-Bin_filtering_and_normalization.html" title="Bin filtering and normalization"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="compartments-and-tads-detection">
<h1>Compartments and TADs detection<a class="headerlink" href="#compartments-and-tads-detection" title="Permalink to this headline">¶</a></h1>
<p>Here, we present the analysis to detect the compartments in Mouse B and
iPS cells. In this example, we will use the GC-content (guanine-cytosine
content) to identify which bins belong to the A or B compartments. The
percentage of bases that are either guanine or cytosine on a DNA strand
correlates directly with gene density and is a good measure to identify
open and close chromatine.</p>
<p><strong>*Note</strong>: Compartments are detected on the full genome matrix.*</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.parsers.hic_parser</span> <span class="k">import</span> <span class="n">load_hic_data_from_bam</span>
<span class="kn">from</span> <span class="nn">pytadbit.parsers.genome_parser</span> <span class="k">import</span> <span class="n">get_gc_content</span><span class="p">,</span> <span class="n">parse_fasta</span>
<span class="kn">from</span> <span class="nn">cPickle</span> <span class="k">import</span> <span class="n">load</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reso</span> <span class="o">=</span> <span class="mi">200000</span>
<span class="n">base_path</span> <span class="o">=</span> <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">.bam&#39;</span>
<span class="n">bias_path</span> <span class="o">=</span> <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/04_normalizing/biases_</span><span class="si">{0}</span><span class="s1">_both_</span><span class="si">{1}</span><span class="s1">kb.biases&#39;</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rich_in_A</span> <span class="o">=</span> <span class="n">get_gc_content</span><span class="p">(</span><span class="n">parse_fasta</span><span class="p">(</span><span class="s1">&#39;genome/Mus_musculus-GRCm38.p6/Mus_musculus-GRCm38.p6.fa&#39;</span><span class="p">),</span>
                           <span class="n">by_chrom</span><span class="o">=</span><span class="kc">True</span> <span class="p">,</span><span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Loading cached genome
</pre>
<div class="section" id="compartments">
<h2>Compartments<a class="headerlink" href="#compartments" title="Permalink to this headline">¶</a></h2>
<div class="section" id="mouse-b-cells">
<h3>Mouse B cells<a class="headerlink" href="#mouse-b-cells" title="Permalink to this headline">¶</a></h3>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cell</span>   <span class="o">=</span> <span class="s1">&#39;mouse_B&#39;</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">),</span>
                                  <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                  <span class="n">biases</span><span class="o">=</span><span class="n">bias_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">reso</span> <span class="o">/</span> <span class="mi">1000</span><span class="p">),</span>
                                  <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
(Matrix size 13641x13641)                                                    [2018-10-11 11:30:12]

- Parsing BAM (122 chunks)                                                   [2018-10-11 11:30:12]
   .......... .......... .......... .......... ..........     50/122
   .......... .......... .......... .......... ..........    100/122
   .......... .......... ..                                  122/122

- Getting matrices                                                           [2018-10-11 11:33:14]
   .......... .......... .......... .......... ..........     50/122
   .......... .......... .......... .......... ..........    100/122
   .......... .......... ..                                  122/122
</pre>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span>! mkdir -p results/fragment/$cell\_both/05_segmenting
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span> <span class="o">=</span> <span class="s1">&#39;chr3&#39;</span>
<span class="n">corr</span> <span class="o">=</span> <span class="n">hic_data</span><span class="o">.</span><span class="n">find_compartments</span><span class="p">(</span><span class="n">show_compartment_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">crms</span><span class="o">=</span><span class="p">[</span><span class="n">crm</span><span class="p">],</span> <span class="n">vmin</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">rich_in_A</span><span class="o">=</span><span class="n">rich_in_A</span><span class="p">,</span>
        <span class="n">savedata</span><span class="o">=</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/05_segmenting/compartments_</span><span class="si">{1}</span><span class="s1">_</span><span class="si">{2}</span><span class="s1">.tsv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">crm</span><span class="p">,</span> <span class="n">reso</span><span class="p">),</span>
        <span class="n">savedir</span><span class="o">=</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/05_segmenting/eigenvectors_</span><span class="si">{1}</span><span class="s1">_</span><span class="si">{2}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">crm</span><span class="p">,</span> <span class="n">reso</span><span class="p">))</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_10_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_10_0.png" />
</div>
<div class="section" id="mouse-ips-cells">
<h3>Mouse iPS cells<a class="headerlink" href="#mouse-ips-cells" title="Permalink to this headline">¶</a></h3>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cell</span>   <span class="o">=</span> <span class="s1">&#39;mouse_PSC&#39;</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">),</span>
                                  <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                  <span class="n">biases</span><span class="o">=</span><span class="n">bias_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">reso</span> <span class="o">/</span> <span class="mi">1000</span><span class="p">),</span>
                                  <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
(Matrix size 13641x13641)                                                    [2018-10-11 11:43:44]

- Parsing BAM (122 chunks)                                                   [2018-10-11 11:43:45]
   .......... .......... .......... .......... ..........     50/122
   .......... .......... .......... .......... ..........    100/122
   .......... .......... ..                                  122/122

- Getting matrices                                                           [2018-10-11 11:48:20]
   .......... .......... .......... .......... ..........     50/122
   .......... .......... .......... .......... ..........    100/122
   .......... .......... ..                                  122/122
</pre>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span>! mkdir -p results/fragment/$cell\_both/05_segmenting
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span> <span class="o">=</span> <span class="s1">&#39;chr3&#39;</span>
<span class="n">corr</span> <span class="o">=</span> <span class="n">hic_data</span><span class="o">.</span><span class="n">find_compartments</span><span class="p">(</span><span class="n">show_compartment_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">crms</span><span class="o">=</span><span class="p">[</span><span class="n">crm</span><span class="p">],</span> <span class="n">vmin</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">rich_in_A</span><span class="o">=</span><span class="n">rich_in_A</span><span class="p">,</span>
        <span class="n">savedata</span><span class="o">=</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/05_segmenting/compartments_</span><span class="si">{1}</span><span class="s1">_</span><span class="si">{2}</span><span class="s1">.tsv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">crm</span><span class="p">,</span> <span class="n">reso</span><span class="p">),</span>
        <span class="n">savedir</span><span class="o">=</span><span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/05_segmenting/eigenvectors_</span><span class="si">{1}</span><span class="s1">_</span><span class="si">{2}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">crm</span><span class="p">,</span> <span class="n">reso</span><span class="p">))</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_15_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_15_0.png" />
</div>
<div class="section" id="compare">
<h3>Compare<a class="headerlink" href="#compare" title="Permalink to this headline">¶</a></h3>
<p>The assignments of the compartments for the two cell types are stored in
two different files:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span>! head -n 20 results/fragment/mouse_B_both/05_segmenting/compartments_chr3_200000.tsv
</pre></div>
</div>
<pre class="ansi-block literal-block">
## CHR chr3 Eigenvector: 1
#   start   end     rich in A       type
chr3        16      44      0.40    B
chr3        45      52      0.50    A
chr3        53      72      0.40    B
chr3        73      76      0.53    A
chr3        77      96      0.39    B
chr3        97      98      0.82    A
chr3        99      100     0.82    B
chr3        101     101     nan     A
chr3        102     108     0.46    B
chr3        109     111     0.62    A
chr3        112     135     0.40    B
chr3        136     139     0.59    A
chr3        140     153     0.44    B
chr3        154     156     0.64    A
chr3        157     162     0.51    B
chr3        163     164     0.87    A
chr3        165     178     0.46    B
chr3        179     181     0.61    A
</pre>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span>! head -n 20 results/fragment/mouse_PSC_both/05_segmenting/compartments_chr3_200000.tsv
</pre></div>
</div>
<pre class="ansi-block literal-block">
## CHR chr3 Eigenvector: 1
#   start   end     rich in A       type
chr3        16      42      0.40    B
chr3        43      52      0.49    A
chr3        53      75      0.40    B
chr3        76      76      nan     A
chr3        77      96      0.39    B
chr3        97      97      nan     A
chr3        98      109     0.43    B
chr3        110     111     0.84    A
chr3        112     135     0.40    B
chr3        136     145     0.48    A
chr3        146     152     0.47    B
chr3        153     156     0.57    A
chr3        157     160     0.56    B
chr3        161     165     0.54    A
chr3        166     169     0.54    B
chr3        170     183     0.46    A
chr3        184     184     nan     B
chr3        188     189     0.84    A
</pre>
<p>In another folder, we also saved the coordinates of each computed
eigenvector:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span>! ls -lh results/fragment/mouse_B_both/05_segmenting/eigenvectors_chr3_200000
</pre></div>
</div>
<pre class="ansi-block literal-block">
total 52K
-rw-r--r-- 1 systemd-network users 49K oct 11 11:43 chr3_EigVect1.tsv
</pre>
<p><strong>*Note</strong>: In this file the first line shows the eigenvector index with
it’s corresponding eigenvalue (the first column should always be the one
you selected, even if it is not the first eigenvector).*</p>
<p>Then there are the coordinates of the eigenvectors. In the first column,
the coordinates correspond to the assignment of the A and B
compartments, positive values for A compartments and negative values for
B compartments.</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span>! head -n 20 results/fragment/mouse_B_both/05_segmenting/eigenvectors_chr3_200000/chr3_EigVect1.tsv
</pre></div>
</div>
<pre class="ansi-block literal-block">
# EV_1 (199.3674)   EV_2 (38.9365)  EV_3 (22.7067)
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
nan nan     nan
-0.047058774338678624       -0.0008416946845418646  0.041851430073603285
-0.04570124930041133        -0.00491013140668968    0.038322756862219455
-0.045914264537434225       -0.002039448345766831   0.03596787271950662
-0.04485768885458887        -0.003962351416821032   0.0423027589274437
</pre>
<p>Load eigenvectors coordinates from files:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results/fragment/mouse_B_both/05_segmenting/eigenvectors_chr3_200000/chr3_EigVect1.tsv&#39;</span><span class="p">)</span>

<span class="n">header</span> <span class="o">=</span> <span class="n">fh</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>

<span class="n">ev1_B</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>
    <span class="n">evc1</span><span class="p">,</span> <span class="n">evc2</span><span class="p">,</span> <span class="n">evc3</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="n">ev1_B</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">evc1</span><span class="p">))</span>

<span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results/fragment/mouse_PSC_both/05_segmenting/eigenvectors_chr3_200000/chr3_EigVect1.tsv&#39;</span><span class="p">)</span>

<span class="n">header</span> <span class="o">=</span> <span class="n">fh</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>

<span class="n">ev1_PSC</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>
    <span class="n">evc1</span><span class="p">,</span> <span class="n">evc2</span><span class="p">,</span> <span class="n">evc3</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="n">ev1_PSC</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">evc1</span><span class="p">))</span>

<span class="n">diff</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ev1_B</span><span class="p">)):</span>
    <span class="n">diff</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ev1_B</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">ev1_PSC</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="section" id="spot-changes-in-activity">
<h3>Spot changes in activity<a class="headerlink" href="#spot-changes-in-activity" title="Permalink to this headline">¶</a></h3>
<p>Plot the difference between each eigenvector along chromosome</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mf">0.07</span><span class="p">,</span> <span class="s1">&#39;More active in B cell&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)),</span> <span class="n">diff</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">where</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="o">&gt;</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">diff</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.09</span><span class="p">,</span> <span class="s1">&#39;More active in PSC cell&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)),</span> <span class="n">diff</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">where</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="o">&lt;</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">diff</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;difference of EV&#39;</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Genomic coordinate (</span><span class="si">%s</span><span class="s1"> kb)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">reso</span> <span class="o">/</span> <span class="mi">1000</span><span class="p">))</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_29_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_29_0.png" />
</div>
<div class="section" id="correlate-eigenvectors">
<h3>Correlate eigenvectors<a class="headerlink" href="#correlate-eigenvectors" title="Permalink to this headline">¶</a></h3>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ev1_B</span><span class="p">)):</span>
    <span class="k">if</span> <span class="n">ev1_B</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">ev1_PSC</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ev1_B</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ev1_PSC</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s1">&#39;ro&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">ev1_B</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">ev1_PSC</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ev1_B</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ev1_PSC</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s1">&#39;bo&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ev1_B</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ev1_PSC</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;B cell EV1&#39;</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;PSC cell EV1&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_31_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_31_0.png" />
</div>
</div>
<div class="section" id="tads">
<h2>TADs<a class="headerlink" href="#tads" title="Permalink to this headline">¶</a></h2>
<p>Now, we move to the TADs detection. In this notebook we will detect TAD
borders at 100kbp resolution.</p>
<!-- The comparison of TAD borders at high resolutions becomes difficult because the border positions are not as sharply defined as at lower resolutions.

An example of consistency between TAD borders is shown in the following graph. TAD borders are called at 1 kb resolution (insulation score-based method). We assume that there should be a high ratio of conservation between the 4 replicates and as we see that's true for resolutions lower than approximately 100kbp. If our bin size is for example 50kbp, we only reach the same ratio of consistency if we consider TAD borders found 2 bins away as being the same border in the different replicates.

<img src="images/TAD_calling_resolution.png"> --><div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="k">import</span> <span class="n">Chromosome</span>
<span class="kn">from</span> <span class="nn">pytadbit.parsers.hic_parser</span> <span class="k">import</span> <span class="n">load_hic_data_from_bam</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">base_path</span> <span class="o">=</span> <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">.bam&#39;</span>
<span class="n">bias_path</span> <span class="o">=</span> <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/04_normalizing/biases_</span><span class="si">{0}</span><span class="s1">_both_</span><span class="si">{1}</span><span class="s1">kb.biases&#39;</span>

<span class="n">reso</span> <span class="o">=</span> <span class="mi">100000</span>

<span class="n">cel1</span> <span class="o">=</span> <span class="s1">&#39;mouse_B&#39;</span>
<span class="n">cel2</span> <span class="o">=</span> <span class="s1">&#39;mouse_PSC&#39;</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data1</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel1</span><span class="p">),</span>
                                   <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                   <span class="n">region</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span>
                                   <span class="n">biases</span><span class="o">=</span><span class="n">bias_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel1</span><span class="p">,</span> <span class="n">reso</span> <span class="o">/</span> <span class="mi">1000</span><span class="p">),</span>
                                   <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">hic_data2</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel2</span><span class="p">),</span>
                                   <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                   <span class="n">region</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span>
                                   <span class="n">biases</span><span class="o">=</span><span class="n">bias_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel2</span><span class="p">,</span> <span class="n">reso</span> <span class="o">/</span> <span class="mi">1000</span><span class="p">),</span>
                                   <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
(Matrix size 1601x1601)                                                      [2018-10-17 00:42:22]

- Parsing BAM (101 chunks)                                                   [2018-10-17 00:42:22]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101

- Getting matrices                                                           [2018-10-17 00:42:31]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101


(Matrix size 1601x1601)                                                      [2018-10-17 00:42:42]

- Parsing BAM (101 chunks)                                                   [2018-10-17 00:42:42]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101

- Getting matrices                                                           [2018-10-17 00:42:51]
   .......... .......... .......... .......... ..........     50/101
   .......... .......... .......... .......... ..........    100/101
   .                                                         101/101
</pre>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span> <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="n">chrname</span><span class="p">)</span>
<span class="n">crm</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span>
                   <span class="n">hic_data</span><span class="o">=</span><span class="p">[</span><span class="n">hic_data1</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">)],</span>
                   <span class="n">norm_data</span><span class="o">=</span><span class="p">[</span><span class="n">hic_data1</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span><span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">)],</span>
                   <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">)</span>
<span class="n">crm</span><span class="o">.</span><span class="n">add_experiment</span><span class="p">(</span><span class="s1">&#39;mouse_PSC&#39;</span><span class="p">,</span>
                   <span class="n">hic_data</span><span class="o">=</span><span class="p">[</span><span class="n">hic_data2</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">)],</span>
                   <span class="n">norm_data</span><span class="o">=</span><span class="p">[</span><span class="n">hic_data2</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span><span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">)],</span>
                   <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">)</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">visualize</span><span class="p">([(</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">)])</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_39_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_39_0.png" />
<div class="section" id="tad-caller-algorithms">
<h3>TAD caller algorithms<a class="headerlink" href="#tad-caller-algorithms" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="tadbit">
<h3>TADbit<a class="headerlink" href="#tadbit" title="Permalink to this headline">¶</a></h3>
<p>TADbit is the original TAD caller algorithm TADbit is a breakpoint
detection algorithm that returns the optimal segmentation of the
chromosome under BIC-penalized likelihood. The model assumes that counts
have a Poisson distribution and that the expected value of the counts
decreases like a power-law with the linear distance on the chromosome.</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">find_tad</span><span class="p">([</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">],</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">visualize</span><span class="p">([(</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">)],</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_44_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_44_0.png" />
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">visualize</span><span class="p">([(</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">)],</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">360</span><span class="p">))</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_45_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_45_0.png" />
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">B</span> <span class="o">=</span> <span class="n">crm</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">]</span>
<span class="n">PSC</span> <span class="o">=</span> <span class="n">crm</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s1">&#39;mouse_PSC&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
[Experiment mouse_B (resolution: 100 kb, TADs: 96, Hi-C rows: 1601, normalized: visibility),
 Experiment mouse_PSC (resolution: 100 kb, TADs: 118, Hi-C rows: 1601, normalized: visibility)]
</pre>
</div>
<div class="section" id="topdom">
<h3>TopDom<a class="headerlink" href="#topdom" title="Permalink to this headline">¶</a></h3>
<p>TopDom identifies TAD borders based on the assumption that contact
frequencies between regions upstream and downstream of a border are
lower than those between two regions within a TAD. The algorithm only
depends on a single parameter corresponding to the window size. The
algorithm provides a measure (from 0 to 10) of confidence on the
accuracy of the border detection
(<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/26704975">https://www.ncbi.nlm.nih.gov/pubmed/26704975</a>).</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">find_tad</span><span class="p">([</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">],</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">use_topdom</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">visualize</span><span class="p">([(</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">)],</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_51_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_51_0.png" />
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">visualize</span><span class="p">([(</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">)],</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">paint_tads</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">360</span><span class="p">))</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_52_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_52_0.png" />
</div>
<div class="section" id="insulation-score">
<h3>Insulation score<a class="headerlink" href="#insulation-score" title="Permalink to this headline">¶</a></h3>
<p>Insulation score (Crane et al. 2015 <a class="reference external" href="https://doi.org/10.1038/nature14450">https://doi.org/10.1038/nature14450</a>)
can be used to build an insulation profile of the genome and, with a
simple transformation, to identify TAD borders.</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.tadbit</span> <span class="k">import</span> <span class="n">insulation_score</span><span class="p">,</span> <span class="n">insulation_to_borders</span>
</pre></div>
</div>
<p>First we need to normalize the matrices by visibility and by decay:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">hic_data1</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">()</span>
<span class="n">hic_data1</span><span class="o">.</span><span class="n">normalize_expected</span><span class="p">()</span>

<span class="n">hic_data2</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">()</span>
<span class="n">hic_data2</span><span class="o">.</span><span class="n">normalize_expected</span><span class="p">()</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
iterative correction
  - copying matrix
  - computing biases
rescaling to factor 1
  - getting the sum of the matrix
    =&gt; 1553.816
  - rescaling biases
iterative correction
  - copying matrix
  - computing biases
rescaling to factor 1
  - getting the sum of the matrix
    =&gt; 1570.990
  - rescaling biases
</pre>
<p>The two important parameter to define are the window size, the distance
from the diagonal and the delta. - the square size should be 500 kb as
close as possible from the diagonal - the delta is to look for increases
in insulation around a given bin. Should be around 100 kb (in out case
we define it as 200 kb as we are working at 100 kb resolution, and
working with only one bin is a bit to little)</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">insc1</span><span class="p">,</span> <span class="n">delta1</span> <span class="o">=</span> <span class="n">insulation_score</span><span class="p">(</span><span class="n">hic_data1</span><span class="p">,</span> <span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">insc2</span><span class="p">,</span> <span class="n">delta2</span> <span class="o">=</span> <span class="n">insulation_score</span><span class="p">(</span><span class="n">hic_data2</span><span class="p">,</span> <span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
- computing insulation in band 1-4
- computing insulation in band 1-4
</pre>
<p>Once defined the insulation score and the values of delta can be used to
search for borders.</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">borders1</span> <span class="o">=</span> <span class="n">insulation_to_borders</span><span class="p">(</span><span class="n">insc1</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">delta1</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">min_strength</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">borders2</span> <span class="o">=</span> <span class="n">insulation_to_borders</span><span class="p">(</span><span class="n">insc2</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">delta2</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">min_strength</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
<p>Currently the representation is not available in TADbit as for the other
methods, but we can easily plot it:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;B cell&#39;</span><span class="p">)</span>
<span class="n">l1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">insc1</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">float</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">insc1</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]))],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Insulation score&#39;</span><span class="p">)</span>
<span class="n">l2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">delta1</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">float</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">insc1</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]))],</span>
         <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Delta value&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">borders1</span><span class="p">:</span>
    <span class="n">l3</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">b</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mf">2.3</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;darkgreen&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">c</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Border strength&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">2.5</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">360</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">l1</span> <span class="o">+</span> <span class="n">l2</span> <span class="o">+</span> <span class="n">l3</span><span class="p">,</span> <span class="p">[</span><span class="n">l</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">l1</span> <span class="o">+</span> <span class="n">l2</span> <span class="o">+</span> <span class="n">l3</span><span class="p">],</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;PSC&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">insc2</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">float</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">insc2</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]))])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">delta2</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">float</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">insc2</span><span class="p">[(</span><span class="n">wsize</span><span class="p">)]))],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="k">for</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">borders2</span><span class="p">:</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">b</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mf">2.3</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;darkgreen&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">c</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="o">-</span><span class="mf">2.5</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">360</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_65_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_65_0.png" />
</div>
</div>
<div class="section" id="comparison-of-tad-borders">
<h2>Comparison of TAD borders<a class="headerlink" href="#comparison-of-tad-borders" title="Permalink to this headline">¶</a></h2>
<p>The TAD borders can be aligned, using a simple reciprocal best hit
strategy:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ali</span> <span class="o">=</span> <span class="n">crm</span><span class="o">.</span><span class="n">align_experiments</span><span class="p">([</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">],</span> <span class="n">max_dist</span><span class="o">=</span><span class="n">reso</span><span class="p">)</span>
</pre></div>
</div>
<p>In the plots below, each arc represents a TAD. Between two consecutive
arcs the triangle mark the border. This triangle is colored depending on
the confidence of the TAD border call.</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ali</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">ymax</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_70_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_70_0.png" />
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ali</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">focus</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">350</span><span class="p">),</span> <span class="n">ymax</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_71_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_71_0.png" />
<div class="section" id="statistical-significance-of-the-tad-borders-alignments">
<h3>Statistical significance of the TAD borders alignments<a class="headerlink" href="#statistical-significance-of-the-tad-borders-alignments" title="Permalink to this headline">¶</a></h3>
<p>In order to asses how well two experiments align, or how conserved are
the TAD borders between two experiments, we can compare the overlap
between our experiments with the overlap of simulated random
distributions of TAD borders.</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ali</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="n">crm</span><span class="o">.</span><span class="n">align_experiments</span><span class="p">([</span><span class="s1">&#39;mouse_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_PSC&#39;</span><span class="p">],</span> <span class="n">randomize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span> <span class="n">ali</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Alignment shown in 100 Kb (2 experiments) (scores: <span class="ansi-blue">0</span> <span class="ansi-blue">1</span> <span class="ansi-blue">2</span> <span class="ansi-cyan">3</span> 4 <span class="ansi-bold">5</span> <span class="ansi-yellow">6</span> <span class="ansi-yellow">7</span> <span class="ansi-magenta">8</span> <span class="ansi-magenta">9</span> <span class="ansi-red">10</span>)
  mouse_B:|    <span class="ansi-red">30</span>|    <span class="ansi-red">73</span>|    78|    <span class="ansi-red">85</span>|    <span class="ansi-red">90</span>|   <span class="ansi-red">104</span>| ---- |   <span class="ansi-magenta">141</span>| ---- |   146|   153|   158|   <span class="ansi-red">163</span>|   <span class="ansi-magenta">179</span>|   <span class="ansi-magenta">192</span>|   197|   <span class="ansi-red">205</span>| ---- | ---- |   <span class="ansi-red">227</span>| ---- |   <span class="ansi-red">270</span>| ---- |   <span class="ansi-red">290</span>| ---- |   <span class="ansi-red">305</span>|   <span class="ansi-red">312</span>|   <span class="ansi-yellow">317</span>|   <span class="ansi-red">323</span>|   <span class="ansi-red">328</span>| ---- | ---- |   <span class="ansi-red">356</span>| ---- |   <span class="ansi-magenta">368</span>|   <span class="ansi-magenta">375</span>|   <span class="ansi-red">382</span>|   <span class="ansi-red">389</span>|   <span class="ansi-bold">403</span>| ---- |   <span class="ansi-yellow">408</span>|   <span class="ansi-red">418</span>| ---- | ---- | ---- |   <span class="ansi-red">512</span>|   <span class="ansi-red">535</span>|   <span class="ansi-red">548</span>|   <span class="ansi-red">553</span>| ---- |   <span class="ansi-red">576</span>| ---- |   <span class="ansi-red">590</span>|   <span class="ansi-red">608</span>|   <span class="ansi-red">625</span>|   <span class="ansi-yellow">653</span>|   <span class="ansi-yellow">660</span>| ---- |   <span class="ansi-red">685</span>| ---- |   <span class="ansi-red">699</span>| ---- |   <span class="ansi-red">752</span>|   <span class="ansi-red">759</span>|   <span class="ansi-red">789</span>|   <span class="ansi-red">796</span>| ---- | ---- |   <span class="ansi-red">830</span>|   <span class="ansi-red">837</span>| ---- |   <span class="ansi-red">870</span>|   <span class="ansi-red">877</span>|   <span class="ansi-red">883</span>| ---- |   <span class="ansi-red">901</span>|   <span class="ansi-red">908</span>| ---- |   <span class="ansi-red">943</span>|   <span class="ansi-red">950</span>|   <span class="ansi-red">967</span>| ---- |   <span class="ansi-red">976</span>|   <span class="ansi-red">986</span>|  <span class="ansi-red">1000</span>| ---- | ---- |  <span class="ansi-red">1020</span>|  <span class="ansi-red">1027</span>| ---- |  <span class="ansi-red">1034</span>| ---- | ---- |  <span class="ansi-red">1047</span>| ---- | ---- | ---- |  <span class="ansi-red">1075</span>|  <span class="ansi-red">1082</span>|  <span class="ansi-magenta">1093</span>|  <span class="ansi-cyan">1135</span>|  <span class="ansi-red">1155</span>|  <span class="ansi-red">1169</span>| ---- |  <span class="ansi-red">1211</span>|  <span class="ansi-red">1235</span>| ---- | ---- | ---- |  <span class="ansi-red">1275</span>|  <span class="ansi-red">1280</span>|  <span class="ansi-red">1292</span>|  <span class="ansi-red">1297</span>| ---- |  1302|  <span class="ansi-red">1307</span>|  <span class="ansi-bold">1313</span>|  <span class="ansi-red">1326</span>| ---- | ---- | ---- |  <span class="ansi-red">1352</span>| ---- | ---- |  <span class="ansi-red">1382</span>| ---- | ---- | ---- |  1423| ---- |  1428| ---- |  <span class="ansi-yellow">1445</span>|  <span class="ansi-magenta">1452</span>| ---- |  <span class="ansi-red">1459</span>| ---- |  <span class="ansi-red">1470</span>| ---- | ---- |  <span class="ansi-bold">1517</span>|  <span class="ansi-bold">1522</span>| ---- |  <span class="ansi-red">1540</span>| ---- | ---- | ---- |  <span class="ansi-magenta">1579</span>| ---- |  <span class="ansi-bold">1585</span>|  <span class="ansi-cyan">1590</span>|  <span class="ansi-yellow">1596</span>|  <span class="ansi-red">1601</span>
mouse_PSC:|    <span class="ansi-red">30</span>|    <span class="ansi-red">73</span>| ---- |    <span class="ansi-red">85</span>| ---- |   <span class="ansi-red">104</span>|   <span class="ansi-red">139</span>| ---- |   <span class="ansi-magenta">144</span>| ---- | ---- | ---- |   <span class="ansi-red">162</span>|   <span class="ansi-red">179</span>|   <span class="ansi-red">192</span>| ---- |   <span class="ansi-red">205</span>|   <span class="ansi-yellow">214</span>|   <span class="ansi-red">223</span>| ---- |   <span class="ansi-red">257</span>|   <span class="ansi-red">270</span>|   <span class="ansi-red">288</span>| ---- |   298|   <span class="ansi-red">305</span>|   <span class="ansi-red">311</span>|   <span class="ansi-red">317</span>|   <span class="ansi-magenta">324</span>| ---- |   <span class="ansi-magenta">331</span>|   <span class="ansi-red">338</span>|   <span class="ansi-red">356</span>|   <span class="ansi-red">366</span>| ---- |   <span class="ansi-bold">375</span>|   <span class="ansi-yellow">382</span>|   <span class="ansi-yellow">388</span>| ---- |   <span class="ansi-red">405</span>| ---- |   <span class="ansi-red">418</span>|   <span class="ansi-red">454</span>|   <span class="ansi-red">472</span>|   <span class="ansi-red">489</span>|   <span class="ansi-red">512</span>|   <span class="ansi-red">535</span>|   <span class="ansi-red">547</span>|   <span class="ansi-red">552</span>|   <span class="ansi-red">572</span>| ---- |   <span class="ansi-red">585</span>|   <span class="ansi-red">591</span>| ---- |   <span class="ansi-red">625</span>|   <span class="ansi-red">653</span>|   <span class="ansi-red">660</span>|   <span class="ansi-red">674</span>| ---- |   <span class="ansi-red">690</span>|   <span class="ansi-red">700</span>|   <span class="ansi-red">731</span>|   <span class="ansi-red">752</span>|   <span class="ansi-red">759</span>|   <span class="ansi-red">789</span>|   <span class="ansi-magenta">796</span>|   <span class="ansi-magenta">801</span>|   <span class="ansi-red">820</span>|   <span class="ansi-red">831</span>|   <span class="ansi-red">838</span>|   <span class="ansi-red">862</span>|   <span class="ansi-red">870</span>|   <span class="ansi-magenta">877</span>|   <span class="ansi-yellow">883</span>|   891|   <span class="ansi-red">900</span>|   <span class="ansi-red">907</span>|   <span class="ansi-red">936</span>|   <span class="ansi-red">943</span>|   <span class="ansi-red">949</span>| ---- |   <span class="ansi-red">969</span>|   <span class="ansi-red">976</span>|   <span class="ansi-yellow">985</span>| ---- |  <span class="ansi-magenta">1007</span>|  <span class="ansi-yellow">1013</span>|  <span class="ansi-red">1019</span>| ---- |  <span class="ansi-red">1030</span>| ---- |  <span class="ansi-yellow">1038</span>|  <span class="ansi-red">1045</span>| ---- |  <span class="ansi-red">1051</span>|  <span class="ansi-red">1060</span>|  <span class="ansi-red">1072</span>| ---- | ---- |  <span class="ansi-blue">1092</span>|  <span class="ansi-red">1136</span>|  <span class="ansi-red">1156</span>|  <span class="ansi-red">1169</span>|  <span class="ansi-red">1175</span>|  <span class="ansi-red">1212</span>|  <span class="ansi-red">1235</span>|  <span class="ansi-yellow">1243</span>|  1258|  <span class="ansi-magenta">1271</span>| ---- |  <span class="ansi-red">1280</span>|  <span class="ansi-red">1292</span>| ---- |  <span class="ansi-bold">1300</span>| ---- |  <span class="ansi-red">1307</span>|  <span class="ansi-bold">1313</span>|  <span class="ansi-red">1327</span>|  <span class="ansi-red">1333</span>|  <span class="ansi-red">1338</span>|  <span class="ansi-yellow">1347</span>|  <span class="ansi-red">1352</span>|  <span class="ansi-red">1358</span>|  <span class="ansi-red">1378</span>| ---- |  <span class="ansi-red">1385</span>|  <span class="ansi-red">1392</span>|  <span class="ansi-red">1419</span>| ---- |  <span class="ansi-magenta">1425</span>| ---- |  <span class="ansi-red">1430</span>|  <span class="ansi-red">1444</span>| ---- |  <span class="ansi-red">1454</span>| ---- |  <span class="ansi-red">1463</span>|  <span class="ansi-red">1470</span>|  <span class="ansi-yellow">1488</span>|  <span class="ansi-yellow">1496</span>|  <span class="ansi-red">1516</span>|  <span class="ansi-red">1521</span>|  <span class="ansi-red">1528</span>|  <span class="ansi-red">1539</span>|  <span class="ansi-red">1546</span>|  <span class="ansi-red">1567</span>|  <span class="ansi-magenta">1576</span>| ---- |  <span class="ansi-red">1581</span>| ---- | ---- | ---- |  <span class="ansi-red">1601</span>
</pre>
<p>This analysis returns an alignment score between 0 and 1 (0: no match,
1: all borders aligned), a p-value (usually equal zero), the proportion
of borders conserved in the second experiment, and the proportion of
borders conserved in the first experiment:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stats</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
(0.3202614379084967, 0.0, 0.7604166666666666, 0.8050847457627118)
</pre>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span> <span class="s1">&#39;Alignment score: </span><span class="si">%.3f</span><span class="s1">, p-value: </span><span class="si">%.4f</span><span class="se">\n</span><span class="s1">  proportion of borders of T0 found in T60: </span><span class="si">%.3f</span><span class="s1">, of T60 in T0 </span><span class="si">%.3f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">stats</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Alignment score: 0.320, p-value: 0.0000
  proportion of borders of T0 found in T60: 0.760, of T60 in T0 0.805
</pre>
</div>
<div class="section" id="save-chromosome-object-with-tad-definition">
<h3>Save Chromosome object (with TAD definition)<a class="headerlink" href="#save-chromosome-object-with-tad-definition" title="Permalink to this headline">¶</a></h3>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span><span class="o">.</span><span class="n">save_chromosome</span><span class="p">(</span><span class="s1">&#39;results/fragment/chr3.tdb&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="extra-choosing-the-best-resolution-to-call-tad-borders">
<h3>Extra: choosing the best resolution to call TAD borders<a class="headerlink" href="#extra-choosing-the-best-resolution-to-call-tad-borders" title="Permalink to this headline">¶</a></h3>
<p>TopDom and the methodology based on insulation score are relatively fast
computationally and allow to call TAD borders at very high resolution.
<strong>However</strong> being able to call TAD borders at high resolution does not
mean that we should do it. <strong>As the resolution increases, so does the
noise</strong>.</p>
<p>In order to assess which is the best resolution in order to call TAD
borders, a good strategy is to test the consistency of several
resolutions.</p>
<p>In the example below, we use the methodology based on the insulation
score as it is the fastest (tadbit strategy is almost not usable bellow
20 kb). We compare the number of TAD borders that are shared (plus-minus
one bin) between the two replicates (iPS and B cells).</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.parsers.hic_parser</span> <span class="k">import</span> <span class="n">load_hic_data_from_bam</span>
<span class="kn">from</span> <span class="nn">pytadbit.tadbit</span> <span class="k">import</span> <span class="n">insulation_score</span><span class="p">,</span> <span class="n">insulation_to_borders</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">base_path</span> <span class="o">=</span> <span class="s1">&#39;results/fragment/</span><span class="si">{0}</span><span class="s1">_both/03_filtering/valid_reads12_</span><span class="si">{0}</span><span class="s1">.bam&#39;</span>
<span class="n">cel1</span> <span class="o">=</span> <span class="s1">&#39;mouse_B&#39;</span>
<span class="n">cel2</span> <span class="o">=</span> <span class="s1">&#39;mouse_PSC&#39;</span>
<span class="n">borders</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">resos</span> <span class="o">=</span> <span class="p">[</span><span class="mi">5000</span><span class="p">,</span> <span class="mi">10000</span><span class="p">,</span> <span class="mi">20000</span><span class="p">,</span> <span class="mi">30000</span><span class="p">,</span> <span class="mi">40000</span><span class="p">,</span> <span class="mi">50000</span><span class="p">,</span> <span class="mi">60000</span><span class="p">,</span> <span class="mi">80000</span><span class="p">,</span> <span class="mi">100000</span><span class="p">,</span> <span class="mi">150000</span><span class="p">,</span> <span class="mi">200000</span><span class="p">]</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;chr</span><span class="si">%d</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">cs</span> <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">20</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;chrX&#39;</span><span class="p">]:</span>
    <span class="n">borders</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">reso</span> <span class="ow">in</span> <span class="n">resos</span><span class="p">:</span>
        <span class="n">hic_data1</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel1</span><span class="p">),</span>
                                           <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                           <span class="n">region</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
                                           <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">hic_data2</span> <span class="o">=</span> <span class="n">load_hic_data_from_bam</span><span class="p">(</span><span class="n">base_path</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cel2</span><span class="p">),</span>
                                           <span class="n">resolution</span><span class="o">=</span><span class="n">reso</span><span class="p">,</span>
                                           <span class="n">region</span><span class="o">=</span><span class="n">c</span><span class="p">,</span>
                                           <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">hic_data1</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">(</span><span class="n">silent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">hic_data1</span><span class="o">.</span><span class="n">normalize_expected</span><span class="p">()</span>
        <span class="n">hic_data2</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">(</span><span class="n">silent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">hic_data2</span><span class="o">.</span><span class="n">normalize_expected</span><span class="p">()</span>
        <span class="n">wsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">500000</span> <span class="o">/</span> <span class="n">reso</span><span class="p">))</span>
        <span class="n">delta</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100000</span> <span class="o">/</span> <span class="n">reso</span><span class="p">)</span>
        <span class="n">insc1</span><span class="p">,</span> <span class="n">delta1</span> <span class="o">=</span> <span class="n">insulation_score</span><span class="p">(</span><span class="n">hic_data1</span><span class="p">,</span> <span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="n">delta</span><span class="p">)</span>
        <span class="n">insc2</span><span class="p">,</span> <span class="n">delta2</span> <span class="o">=</span> <span class="n">insulation_score</span><span class="p">(</span><span class="n">hic_data2</span><span class="p">,</span> <span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="n">delta</span><span class="p">)</span>
        <span class="n">borders1</span> <span class="o">=</span> <span class="n">insulation_to_borders</span><span class="p">(</span><span class="n">insc1</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">delta1</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">min_strength</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
        <span class="n">borders2</span> <span class="o">=</span> <span class="n">insulation_to_borders</span><span class="p">(</span><span class="n">insc2</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">delta2</span><span class="p">[</span><span class="n">wsize</span><span class="p">],</span> <span class="n">min_strength</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
        <span class="n">borders</span><span class="p">[</span><span class="n">c</span><span class="p">][</span><span class="n">reso</span><span class="p">]</span> <span class="o">=</span> <span class="n">borders1</span><span class="p">,</span> <span class="n">borders2</span>
</pre></div>
</div>
<p>We align TAD borders and keep the proportion of borders aligned between
cell types</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.alignment</span> <span class="k">import</span> <span class="n">align</span>
</pre></div>
</div>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scores</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">props1</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">props2</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">borders</span><span class="p">:</span>
    <span class="n">scores</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">props1</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">props2</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">reso</span> <span class="ow">in</span> <span class="n">resos</span><span class="p">:</span>
        <span class="n">ali</span> <span class="o">=</span> <span class="n">align</span><span class="p">([[</span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">b</span><span class="p">]</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">borders</span><span class="p">[</span><span class="n">c</span><span class="p">][</span><span class="n">reso</span><span class="p">]],</span> <span class="n">max_dist</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">props1</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ali</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">props2</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ali</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
        <span class="n">scores</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ali</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
<p>We plot the results, as boxplots to group the values of each
chromosomes:</p>
<div class="code ipython2 highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">bp</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">boxplot</span><span class="p">([</span><span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span><span class="o">+</span> <span class="n">y</span><span class="p">,</span>
                 <span class="p">[(</span><span class="n">props1</span><span class="p">[</span><span class="n">c</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">props2</span><span class="p">[</span><span class="n">c</span><span class="p">][</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">props1</span><span class="p">])</span>
                  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">resos</span><span class="p">))],</span> <span class="n">positions</span><span class="o">=</span><span class="n">resos</span><span class="p">,</span> <span class="n">widths</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get_xdata</span><span class="p">())</span> <span class="o">/</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">bp</span><span class="p">[</span><span class="s1">&#39;medians&#39;</span><span class="p">]],</span>
         <span class="p">[</span><span class="mi">0</span><span class="n">m</span><span class="o">.</span><span class="n">get_ydata</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">bp</span><span class="p">[</span><span class="s1">&#39;medians&#39;</span><span class="p">]],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">205000</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">resos</span><span class="p">,</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">reso</span> <span class="o">/</span> <span class="mi">1000</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;kb&#39;</span>  <span class="k">for</span> <span class="n">reso</span> <span class="ow">in</span> <span class="n">resos</span><span class="p">],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Proportion of aligned borders</span><span class="se">\n</span><span class="s1">PSC vs B cell (all chromosomes)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Resolution at which TAD borders are called&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_90_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_90_0.png" />
<p>In the case of these experiments it seems that looking for TAD borders
bellow 50 kb is dangerous, as very few are reproducible. Calling TAD
borders at 100 kb resolution might be here a good idea.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Compartments and TADs detection</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_9-Compare_and_merge_Hi-C_experiments.html">Comparison of Hi-C experiments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_7-Bin_filtering_and_normalization.html"
                        title="previous chapter">Bin filtering and normalization</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html"
                        title="next chapter">Comparison of Hi-C experiments</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" title="Comparison of Hi-C experiments"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_7-Bin_filtering_and_normalization.html" title="Bin filtering and normalization"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Jun 21, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>