<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Mapping paired-end reads from NGS experiment (with GEM) &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Tutorials" href="../tutorial.html" />
    <link rel="next" title="Getting started" href="tutorial_1_general.html" />
    <link rel="prev" title="Tutorials" href="../tutorial.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="mapping-paired-end-reads-from-ngs-experiment-with-gem">
<h1>Mapping paired-end reads from NGS experiment (with GEM)<a class="headerlink" href="#mapping-paired-end-reads-from-ngs-experiment-with-gem" title="Permalink to this headline">¶</a></h1>
<div class="section" id="download-and-index-reference-genome">
<h2>Download and index reference genome<a class="headerlink" href="#download-and-index-reference-genome" title="Permalink to this headline">¶</a></h2>
<p>Download last human reference genome from ucsc:
<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/">http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/</a> It&#8217;s better to
download the file that contains each chromosome separately (like this we
could only keep those which we are interested in): hg38.chromFa.tar.gz
Uncompress the genome</p>
<div class="highlight-python"><div class="highlight"><pre>tar -xzf hg38.chromFa.tar.gz
</pre></div>
</div>
<p>Check the header of the fasta files (it&#8217;s important to know how
chromosomes are called in these files):</p>
<div class="highlight-python"><div class="highlight"><pre>head chr1.fa
</pre></div>
</div>
<p>Create a single file with all chromosomes together:</p>
<div class="highlight-python"><div class="highlight"><pre>cat chr1.fa chr2.fa chr3.fa chr4.fa chr5.fa chr6.fa chr7.fa chr8.fa chr9.fa chr10.fa chr11.fa chr12.fa chr13.fa chr14.fa  chr15.fa  chr16.fa  chr17.fa  chr18.fa  chr19.fa chr20.fa  chr21.fa  chr22.fa chrX.fa  chrY.fa chrM.fa &gt; hg38.fa
</pre></div>
</div>
<p>Index reference genome:</p>
<div class="highlight-python"><div class="highlight"><pre>gemtools index -i hg38.fa -t 8
</pre></div>
</div>
<p>Creates a file hg38.gem that we will be used for the mapping.</p>
</div>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<div class="section" id="retrieving-hi-c-experiment-sra-format">
<h3>Retrieving Hi-C experiment SRA format<a class="headerlink" href="#retrieving-hi-c-experiment-sra-format" title="Permalink to this headline">¶</a></h3>
<p>Along this tutorial we are going to use the dataset from [Rao21014]_
that can be downloaded at:
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63525">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63525</a></p>
<p>For this part of the tutorial we are going to work with the experiments
conducted in <em>Homo sapiens</em> GM12878 cell line with three diferent
restriction enzymes.</p>
<p>First we need to convert SRA files in FASTQ format, as TADbit does not
support reading directly from SRA. For this purpose we may want to use
the SRA-toolkit from the NCBI (<a class="reference external" href="https://github.com/ncbi/sratoolkit">https://github.com/ncbi/sratoolkit</a>). We
finally shall endup with a single FASTQ file per experimental replica,
with commands like (can last for hours):</p>
<ul>
<li><p class="first">HindIII dilution Hi-C replica:</p>
<div class="highlight-python"><div class="highlight"><pre>fastq-dump -A SRR1658625 -DQ &#39;+&#39; --defline-seq &#39;@$ac.$si&#39; --split-files -X 50000000 -O HiC035
</pre></div>
</div>
</li>
<li><p class="first">NcoI dilution Hi-C replica:</p>
<div class="highlight-python"><div class="highlight"><pre>fastq-dump -A SRR1658632 -DQ &#39;+&#39; --defline-seq &#39;@$ac.$si&#39; --split-files -X 50000000 -O HiC036
</pre></div>
</div>
</li>
<li><p class="first">MboI <em>in situ</em> Hi-C replica:</p>
<div class="highlight-python"><div class="highlight"><pre>fastq-dump -A SRR1658572 -DQ &#39;+&#39; --defline-seq &#39;@$ac.$si&#39; -X 50000000 -O HiC003
</pre></div>
</div>
</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the parameter used here for <cite>fastq-dump</cite> are for generating clean FASTQ files, <cite>-DQ &#8216;+&#8217;</cite> tells to put a single plus sign as header of the quality input, <cite>&#8211;defline-seq &#8216;&#64;$ac.$si&#8217;</cite> reduces the information in the headers to the accession number and the read id, <cite>&#8211;split-files</cite> is to separate read1 and read2 in diferent files (note that it is only used in the two first examples in order to show different cases), finally <cite>-X 50000000</cite> is to download only 50 Milion reads.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">FASTQ files can be used compressed in order to save space. TADbit recognize [DSRC](<a class="reference external" href="https://github.com/lrog/dsrc">https://github.com/lrog/dsrc</a>), gunzip, bunzip2 and zip compression formats</p>
</div>
</div>
<div class="section" id="first-quality-check-on-the-data">
<h3>First quality check on the data<a class="headerlink" href="#first-quality-check-on-the-data" title="Permalink to this headline">¶</a></h3>
<p>the compressed FASTQ files should be:</p>
<div class="code python highlight-python"><div class="highlight"><pre>! wc -l HiC036/SRR1658632_1.fastq
</pre></div>
</div>
<pre class="ansi-block literal-block">
10354635 HiC036/SRR1658632_1.fastq
</pre>
<p>Usig these files directly we can infer the quality of the Hi-C
experiment with TADbit:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.utils.fastq_utils</span> <span class="kn">import</span> <span class="n">quality_plot</span>

<span class="n">quality_plot</span><span class="p">(</span><span class="s1">&#39;HiC003/SRR1658572.fastq&#39;</span><span class="p">,</span> <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;MboI&#39;</span><span class="p">,</span> <span class="n">paired</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_13_0.png" src="../_images/tutorial_0_mapping_13_0.png" />
<pre class="ansi-block literal-block">
(10.9159, 38.5393)
</pre>
<p>The plot on the top represents the typical per nucleotide quality
profile of NGS reads, with the addition of the proportion of &#8220;N&#8221; found
at each position. This plot is done over the first milion reads, as this
is usually enough to asses the general quality of the experiment.</p>
<p>The second plot, is specific to Hi-C experiments. Given a restriction
enzyme the function searches for the presence of ligation sites and of
undigested restriction enzyme sites. Depending on the enzyme used the
function can differentiate between dangling-ends and undigested sites.</p>
<p>From this proportion some qulaity statistic can be inferred before the
mapping: - The percentage of digested sites wich is the ratio of
digested over undigested sites found over the reads analyzed - The
percentage os dangling-ends which is the number of time a digested site
is found at the begining of a read - The percentage of ligation sites
which is the number of times a ligation site is found in the processed
reads. This number has to be transfrom to reflect the reality as it
corresponds only to the sequenced part of the DNA fragment. For example,
in this case, if the mean fragment size is 400 nucleotides and we found
that 11% of 1 million reads present 1 ligation site, we should expect a
proportional number of ligation sites in the insert (unsequenced space
between the ends of the reads). Thus, in this case we would have a final
number 22% of the reads with at least one ligation site.</p>
<p>For the other replicates used in this tutorial, the enzymes (DpnII and
HindIII) allow to differentiate between dangling-ends and undigested
sites, as their cut site is inside the recognition pattern (<em>note that
in the bellow plot we are now only seing one of the ends of the reads as
we used the ``&#8211;split-file`` we have the one end per file</em>):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">quality_plot</span><span class="p">(</span><span class="s1">&#39;HiC035/SRR1658625_1.fastq&#39;</span><span class="p">,</span> <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;HindIII&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_15_0.png" src="../_images/tutorial_0_mapping_15_0.png" />
<pre class="ansi-block literal-block">
(3.1608, 11.3747)
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">quality_plot</span><span class="p">(</span><span class="s1">&#39;HiC036/SRR1658632_1.fastq&#39;</span><span class="p">,</span> <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;NcoI&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_16_0.png" src="../_images/tutorial_0_mapping_16_0.png" />
<pre class="ansi-block literal-block">
(1.2954, 17.1003)
</pre>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the function returns two numbers that can be used to further analyze the quality of the reads, e.g. when, after mapping we can confirm the mean size of the insert. These numbers are the proportion of dangling-ends and the proportion of ligation sites.</p>
</div>
</div>
<div class="section" id="id1">
<h3>Mapping<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>TADbit implements a flexible mapping that can be set to mimmic either
the <strong>iterative mapping</strong> as proposed by [Imakaev2012]_ , or other
mapping strategies based on the restriction enzyme fragments we call
them <strong>fragment-based mapping</strong>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">pytadbit</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.mapper</span> <span class="kn">import</span> <span class="n">full_mapping</span>
</pre></div>
</div>
<div class="section" id="iterative-mapping">
<h4>Iterative mapping<a class="headerlink" href="#iterative-mapping" title="Permalink to this headline">¶</a></h4>
<p>The bellow example would be following the iterative mapping strategy,
defining the windows over which to do the iterations:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;Mapping the first end of the read</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="n">mapped_r1</span> <span class="o">=</span> <span class="n">full_mapping</span><span class="p">(</span><span class="s1">&#39;/scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/SRR1658572.fastq&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/01_it-mapped_read1&#39;</span><span class="p">,</span>
                         <span class="n">windows</span><span class="o">=</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="mi">25</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">50</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">75</span><span class="p">),(</span><span class="mi">1</span><span class="p">,</span><span class="mi">100</span><span class="p">)),</span>
                         <span class="n">frag_map</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">nthreads</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">clean</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                         <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;/scratch/test/rao2014/tmp_HiC003&#39;</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping the first end of the read

Preparing FASTQ file
  - conversion to MAP format
  - trimming reads 1-25
Mapping reads in window 1-25...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_NbfVlq
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_NbfVlq -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_NbfVlq_full_1-25
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_NbfVlq
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_NbfVlq_full_1-25.map
Preparing MAP file
  - trimming reads 1-50
   x removing original input /scratch/test/rao2014/tmp_HiC003/SRR1658572_NbfVlq_filt_1-25.map
Mapping reads in window 1-50...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_HxWwM8
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_HxWwM8 -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_HxWwM8_full_1-50
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_HxWwM8
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_HxWwM8_full_1-50.map
Preparing MAP file
  - trimming reads 1-75
   x removing original input /scratch/test/rao2014/tmp_HiC003/SRR1658572_HxWwM8_filt_1-50.map
Mapping reads in window 1-75...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_nUa3eh
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_nUa3eh -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_nUa3eh_full_1-75
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_nUa3eh
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_nUa3eh_full_1-75.map
Preparing MAP file
  - trimming reads 1-100
   x removing original input /scratch/test/rao2014/tmp_HiC003/SRR1658572_nUa3eh_filt_1-75.map
Mapping reads in window 1-100...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_IvYOT3
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_IvYOT3 -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_IvYOT3_full_1-100
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_IvYOT3
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_IvYOT3_full_1-100.map
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;Mapping the first end of the read</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="n">mapped_r2</span> <span class="o">=</span> <span class="n">full_mapping</span><span class="p">(</span><span class="s1">&#39;/scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/SRR1658572.fastq&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/01_it-mapped_read2&#39;</span><span class="p">,</span>
                         <span class="n">windows</span><span class="o">=</span><span class="p">((</span><span class="mi">101</span><span class="p">,</span><span class="mi">125</span><span class="p">),</span> <span class="p">(</span><span class="mi">101</span><span class="p">,</span><span class="mi">150</span><span class="p">),</span> <span class="p">(</span><span class="mi">101</span><span class="p">,</span><span class="mi">175</span><span class="p">),(</span><span class="mi">101</span><span class="p">,</span><span class="mi">200</span><span class="p">)),</span>
                         <span class="n">frag_map</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">nthreads</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">clean</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                         <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;/scratch/test/rao2014/tmp_HiC003&#39;</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping the first end of the read

Preparing FASTQ file
  - conversion to MAP format
  - trimming reads 101-125
Mapping reads in window 101-125...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_L3K0wQ
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_L3K0wQ -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_L3K0wQ_full_101-125
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_L3K0wQ
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_L3K0wQ_full_101-125.map
Preparing MAP file
  - trimming reads 101-150
   x removing original input /scratch/test/rao2014/tmp_HiC003/SRR1658572_L3K0wQ_filt_101-125.map
Mapping reads in window 101-150...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_0AfFmK
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_0AfFmK -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_0AfFmK_full_101-150
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_0AfFmK
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_0AfFmK_full_101-150.map
Preparing MAP file
  - trimming reads 101-175
   x removing original input /scratch/test/rao2014/tmp_HiC003/SRR1658572_0AfFmK_filt_101-150.map
Mapping reads in window 101-175...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_pYEcnx
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_pYEcnx -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_pYEcnx_full_101-175
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_pYEcnx
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_pYEcnx_full_101-175.map
Preparing MAP file
  - trimming reads 101-200
   x removing original input /scratch/test/rao2014/tmp_HiC003/SRR1658572_pYEcnx_filt_101-175.map
Mapping reads in window 101-200...
TO GEM /scratch/test/rao2014/tmp_HiC003/SRR1658572_lEDzwc
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i /scratch/test/rao2014/tmp_HiC003/SRR1658572_lEDzwc -o /scratch/test/rao2014/tmp_HiC003/SRR1658572_lEDzwc_full_101-200
Parsing result...
   x removing GEM input /scratch/test/rao2014/tmp_HiC003/SRR1658572_lEDzwc
   x removing map /scratch/test/rao2014/tmp_HiC003/SRR1658572_lEDzwc_full_101-200.map
</pre>
</div>
<div class="section" id="fragment-based-mapping">
<h4>Fragment-based mapping<a class="headerlink" href="#fragment-based-mapping" title="Permalink to this headline">¶</a></h4>
<p>And here bellow an alternative mapping, fragment-based, in this case the
restriction enzyme (RE) name is needed. And the windows parameter is
only used for defining which part of read should be used.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;Mapping the first end of the read</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="n">mapped_r1</span> <span class="o">=</span> <span class="n">full_mapping</span><span class="p">(</span><span class="s1">&#39;/scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/SRR1658572.fastq&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/01_mapped_read1&#39;</span><span class="p">,</span>
                         <span class="n">windows</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">100</span><span class="p">),</span>
                         <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;MboI&#39;</span><span class="p">,</span> <span class="n">frag_map</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">nthreads</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">clean</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;tmp&#39;</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping the first end of the read

Preparing FASTQ file
  - conversion to MAP format
  - trimming reads 1-100
Mapping reads in window 1-100...
TO GEM ../nbpictures//tmp/SRR1658572_eQYIHW
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i ../nbpictures//tmp/SRR1658572_eQYIHW -o ../nbpictures//tmp/SRR1658572_eQYIHW_full_1-100
Parsing result...
   x removing GEM input ../nbpictures//tmp/SRR1658572_eQYIHW
   x removing map ../nbpictures//tmp/SRR1658572_eQYIHW_full_1-100.map
  - splitting into restriction enzyme (RE) fragments using ligation sites
  - ligation sites are replaced by RE sites to match the reference genome
    * enzyme: MboI, ligation site: GATCGATC, RE site: GATC
Preparing MAP file
  - trimming reads 1-100
Mapping fragments of remaining reads...
TO GEM ../nbpictures//tmp/SRR1658572_Sj_OYP
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i ../nbpictures//tmp/SRR1658572_Sj_OYP -o ../nbpictures//tmp/SRR1658572_Sj_OYP_frag_1-100
Parsing result...
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">Mapping the second end of the read</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="n">mapped_r2</span> <span class="o">=</span> <span class="n">full_mapping</span><span class="p">(</span><span class="s1">&#39;/scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/SRR1658572.fastq&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;/scratch/test/rao2014/HiC003/01_mapped_read2&#39;</span><span class="p">,</span>
                         <span class="n">windows</span><span class="o">=</span><span class="p">(</span><span class="mi">101</span><span class="p">,</span><span class="mi">200</span><span class="p">),</span>
                         <span class="n">r_enz</span><span class="o">=</span><span class="s1">&#39;MboI&#39;</span><span class="p">,</span> <span class="n">frag_map</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">nthreads</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">clean</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;tmp&#39;</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Mapping the second end of the read

Preparing FASTQ file
  - conversion to MAP format
  - trimming reads 101-200
Mapping reads in window 101-200...
TO GEM ../nbpictures//tmp/SRR1658572_3OpPDr
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i ../nbpictures//tmp/SRR1658572_3OpPDr -o ../nbpictures//tmp/SRR1658572_3OpPDr_full_101-200
Parsing result...
   x removing GEM input ../nbpictures//tmp/SRR1658572_3OpPDr
   x removing map ../nbpictures//tmp/SRR1658572_3OpPDr_full_101-200.map
  - splitting into restriction enzyme (RE) fragments using ligation sites
  - ligation sites are replaced by RE sites to match the reference genome
    * enzyme: MboI, ligation site: GATCGATC, RE site: GATC
Preparing MAP file
  - trimming reads 101-200
Mapping fragments of remaining reads...
TO GEM ../nbpictures//tmp/SRR1658572_IgIT9e
/usr/local/bin/gem-mapper -I /scratch/db/Genomes/index_files/Homo_sapiens-79/Homo_sapiens.gem -q offset-33 -m 0.04 -s 0 --allow-incomplete-strata 0.00 --granularity 10000 --max-decoded-matches 1 --min-decoded-strata 0 --min-insert-size 0 --max-insert-size 0 --min-matched-bases 0.8 --gem-quality-threshold 26 --max-big-indel-length 15 --mismatch-alphabet ACGT -E 0.30 --max-extendable-matches 20 --max-extensions-per-match 1 -e 0.04 -T 8 -i ../nbpictures//tmp/SRR1658572_IgIT9e -o ../nbpictures//tmp/SRR1658572_IgIT9e_frag_101-200
Parsing result...
</pre>
<p>The Fragment based mapping works in 2 steps: - First read ends are
aligned entirely, assuming that no ligation occured in them. - Second,
for the read ends that were not mapped, the function searches for a
ligation site (in the case of MboI this would correspond to <code class="docutils literal"><span class="pre">GATCGATC</span></code>
and in the case of HindIII to <code class="docutils literal"><span class="pre">AAGCTAGCTT</span></code>). The read-end is splitted
accordingly replacing the ligation site by two RE sites:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">read</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">part</span><span class="o">-</span><span class="n">one</span><span class="o">---</span><span class="n">AAGCTAGCTT</span><span class="o">----</span><span class="n">read</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">part</span><span class="o">-</span><span class="n">two</span>
</pre></div>
</div>
<p>will be splitted in:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">read</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">part</span><span class="o">-</span><span class="n">one</span><span class="o">---</span><span class="n">AAGCTT</span>
</pre></div>
</div>
<p>and</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">AAGCTT</span><span class="o">----</span><span class="n">read</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">part</span><span class="o">-</span><span class="n">two</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">if no ligation site is found step two is repeated using digested RE site as split point (<cite>AAGCT</cite> in the case of HindIII). This in order to be protected against that sequencing errors. When this path is followed the digested RE site is removed, but not replaced.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">both mapping strategies can be mixed up, for example defining the windows as in the iterative mapping, and also gives a RE name and setting <cite>frag_map=True</cite>.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">for the following part of the tutorial we are going to continue using the result of the fragment based mapping, as it contains the most complex results.</p>
</div>
</div>
</div>
<div class="section" id="map-parsing">
<h3>Map parsing<a class="headerlink" href="#map-parsing" title="Permalink to this headline">¶</a></h3>
<p>In case we lost the lists of paths we can &#8220;reload&#8221; them like this:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">os</span>

<span class="n">mapped_r1</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">mapped_r2</span> <span class="o">=</span> <span class="p">[]</span>

<span class="n">r1_dir</span> <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/01_mapped_read1&#39;</span>
<span class="n">r2_dir</span> <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/01_mapped_read2&#39;</span>

<span class="k">for</span> <span class="n">mapped</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">r1_dir</span><span class="p">):</span>
    <span class="n">mapped_r1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">r1_dir</span><span class="p">,</span> <span class="n">mapped</span><span class="p">))</span>
<span class="k">for</span> <span class="n">mapped</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">r2_dir</span><span class="p">):</span>
    <span class="n">mapped_r2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">r2_dir</span><span class="p">,</span> <span class="n">mapped</span><span class="p">))</span>

<span class="k">print</span> <span class="s1">&#39;Output files of the maping of the first end of the reads:</span><span class="se">\n</span><span class="s1"> - &#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1"> - &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mapped_r1</span><span class="p">)</span>
<span class="k">print</span> <span class="s1">&#39;Output files of the maping of the second end of the reads:</span><span class="se">\n</span><span class="s1"> - &#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1"> - &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mapped_r2</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Output files of the maping of the first end of the reads:
 - /scratch/test/rao2014/HiC003/01_mapped_read1/SRR1658572_frag_1-100.map
 - /scratch/test/rao2014/HiC003/01_mapped_read1/SRR1658572_full_1-100.map
Output files of the maping of the second end of the reads:
 - /scratch/test/rao2014/HiC003/01_mapped_read2/SRR1658572_frag_101-200.map
 - /scratch/test/rao2014/HiC003/01_mapped_read2/SRR1658572_full_101-200.map
</pre>
<p>We collect mapped reads at all window sizes into a single file (a single
file for read1, and a single file for read2). These 2 files also contain
the placement of the restriction enzyme sites in the genome.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.parsers.map_parser</span>    <span class="kn">import</span> <span class="n">parse_map</span>
<span class="kn">from</span> <span class="nn">pytadbit.parsers.genome_parser</span> <span class="kn">import</span> <span class="n">parse_fasta</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c1"># loads the genome</span>
<span class="n">genome_seq</span> <span class="o">=</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="s1">&#39;/scratch/db/Genomes/genomes/Homo_sapiens-79/Homo_sapiens.fa&#39;</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Parsing 1
Parsing 2
Parsing 3
Parsing 4
Parsing 5
Parsing 6
Parsing 7
Parsing 8
Parsing 9
Parsing 10
Parsing 11
Parsing 12
Parsing 13
Parsing 14
Parsing 15
Parsing 16
Parsing 17
Parsing 18
Parsing 19
Parsing 20
Parsing 21
Parsing 22
Parsing X
Parsing Y
Parsing MT
</pre>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<ul class="last simple">
<li>Make sure that your renaming corresponds to the chromosomes in the files (same order)</li>
<li>The chromosome names should be the same as the one used to generate the index file used by the mapper. Otherwise you will endup with no read mapped.</li>
</ul>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">TADbit comes with a list of file utilities like <cite>mkdir</cite> to ease the handling of files and folders.</p>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.utils.file_handling</span> <span class="kn">import</span> <span class="n">mkdir</span>

<span class="n">mkdir</span><span class="p">(</span><span class="s1">&#39;/scratch/test/rao2014/HiC003/02_parsed_reads/&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c1"># new file with info of each &quot;read1&quot; and its placement with respect to RE sites</span>
<span class="n">reads1</span> <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/02_parsed_reads/read1.tsv&#39;</span>
<span class="c1"># new file with info of each &quot;read2&quot; and its placement with respect to RE sites</span>
<span class="n">reads2</span> <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/02_parsed_reads/read2.tsv&#39;</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Parsing MAP files is a slow process because reads have to be placed in between the closest RE sites, and they also need to be sorted in order to ease the computation of the intersection.</p>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="s1">&#39;Parse MAP files...&#39;</span>
<span class="n">parse_map</span><span class="p">(</span><span class="n">mapped_r1</span><span class="p">,</span> <span class="n">mapped_r2</span><span class="p">,</span> <span class="n">out_file1</span><span class="o">=</span><span class="n">reads1</span><span class="p">,</span> <span class="n">out_file2</span><span class="o">=</span><span class="n">reads2</span><span class="p">,</span> <span class="n">genome_seq</span><span class="o">=</span><span class="n">genome_seq</span><span class="p">,</span>
          <span class="n">re_name</span><span class="o">=</span><span class="s1">&#39;MboI&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">ncpus</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Parse MAP files...
Searching and mapping RE sites to the reference genome
Found 7191117 RE sites
Loading read1
loading file: /scratch/test/rao2014/HiC003/01_mapped_read1/SRR1658572_frag_1-100.map
loading file: /scratch/test/rao2014/HiC003/01_mapped_read1/SRR1658572_full_1-100.map
Merge sort.....................................................
Getting Multiple contacts
Loading read2
loading file: /scratch/test/rao2014/HiC003/01_mapped_read2/SRR1658572_frag_101-200.map
loading file: /scratch/test/rao2014/HiC003/01_mapped_read2/SRR1658572_full_101-200.map
Merge sort...................................................
Getting Multiple contacts
</pre>
<pre class="ansi-block literal-block">
({0: {1: 14363191, 2: 38080970}, 1: {1: 13755441, 2: 37571439}},
 {0: 5361490, 1: 4879802})
</pre>
<p>The output files generated here are BED-like files, with a header with
the size of the chromosomes used, followed by the list of reads mapped.</p>
<p>The column for each read are:</p>
<ul class="simple">
<li>Read Id</li>
<li>Chromosome name</li>
<li>genomic position</li>
<li>strand (1: positive strand, 0: negative strand)</li>
<li>mapped read length (this length can be extended when the fragment is
found to be spanning over the insert paired-end read)</li>
<li>upstream RE site position</li>
<li>downstream RE site position</li>
</ul>
</div>
<div class="section" id="mapping-analysis">
<h3>Mapping analysis<a class="headerlink" href="#mapping-analysis" title="Permalink to this headline">¶</a></h3>
<div class="section" id="plot-mapping-efficiency">
<h4>Plot mapping efficiency<a class="headerlink" href="#plot-mapping-efficiency" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">plot_iterative_mapping</span>

<span class="n">lengths</span> <span class="o">=</span> <span class="n">plot_iterative_mapping</span><span class="p">(</span><span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">total_reads</span><span class="o">=</span><span class="mi">50000000</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_49_0.png" src="../_images/tutorial_0_mapping_49_0.png" />
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">in some cases the mapping efficiency might be above 100% as fragment based mapping split the original reads into fragments and map them independently.</p>
</div>
</div>
</div>
<div class="section" id="merging-mapped-read1-and-read2">
<h3>Merging mapped &#8220;read1&#8221; and &#8220;read2&#8221;<a class="headerlink" href="#merging-mapped-read1-and-read2" title="Permalink to this headline">¶</a></h3>
<p>We create a new file that will contain only the reads mapped in both
ends (&#8220;read1&#8221; and &#8220;read2&#8221; uniquely mapped)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping</span> <span class="kn">import</span> <span class="n">get_intersection</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">reads</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/02_parsed_reads/both_map.tsv&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">get_intersection</span><span class="p">(</span><span class="n">reads1</span><span class="p">,</span> <span class="n">reads2</span><span class="p">,</span> <span class="n">reads</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Getting intersection of reads 1 and reads 2:

  .......... .......... .......... .......... ........
Found 44674766 pair of reads mapping uniquely
Sorting easch temporary file by genomic coordinate
    1025/1025 sorted files
Removing temporary files...
</pre>
<pre class="ansi-block literal-block">
(44674766, {2: 7044121, 3: 20910, 4: 51})
</pre>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the numbers returned by the function corresponds to the total number of mapped reads, and the number of time multiple contact where found (in this case ~7M contacts to occur between 3 fragments at the same time, ~20k between 4 and 51 between 5).</p>
</div>
<div class="section" id="descriptive-statistics">
<h4>Descriptive statistics<a class="headerlink" href="#descriptive-statistics" title="Permalink to this headline">¶</a></h4>
<p>In order to confirm the size of the inserts fed to the sequencer, we can
look at the distribution of genomic distances between the mapped read1
and read2 of dangling-ends. From this analysis we can extract the
maximum insert size, that is an important value to classify reads during
the filtering process.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">insert_sizes</span>

<span class="n">insert_sizes</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_59_0.png" src="../_images/tutorial_0_mapping_59_0.png" />
<pre class="ansi-block literal-block">
[355.0, 610.0]
</pre>
</div>
</div>
<div class="section" id="simple-descriptive-stats">
<h3>Simple descriptive stats<a class="headerlink" href="#simple-descriptive-stats" title="Permalink to this headline">¶</a></h3>
<div class="section" id="how-the-count-in-interaction-falls-as-the-genomic-distance-is-larger">
<h4>How the count in interaction falls as the genomic distance is larger<a class="headerlink" href="#how-the-count-in-interaction-falls-as-the-genomic-distance-is-larger" title="Permalink to this headline">¶</a></h4>
<p>Here we want to see how the interaction between to two genomic region
decays as the distance between these two loci is larger. Theexpectation
is that at distances between 700 kb and 10 Mb the decay in logarithm
scale is -1.</p>
<p>In the example below are represented the interactions in between genomic
regions that, each, spans over 10 kb (resolution parameter).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">plot_distance_vs_interactions</span>

<span class="n">plot_distance_vs_interactions</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">max_diff</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
((-1.0699233238000596, 15.456024589131042, -0.9992506172369725),
 (-1.1471902596970704, 15.335536453854143, -0.99565698091932198),
 (-2.4979074477418139, 23.753684595554244, -0.92419616437542085))
</pre>
<img alt="../_images/tutorial_0_mapping_63_1.png" src="../_images/tutorial_0_mapping_63_1.png" />
</div>
<div class="section" id="genomic-coverage-of-our-reads">
<h4>Genomic coverage of our reads<a class="headerlink" href="#genomic-coverage-of-our-reads" title="Permalink to this headline">¶</a></h4>
<p>In the plot above we want to see the distribution of the reads mapped in
the genome (regardless of their interactions). Here, te expecation is to
see a minimum number of reads mapping in all positions of the genome
with falls around centromeres and telomeres.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">plot_genomic_distribution</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_67_0.png" src="../_images/tutorial_0_mapping_67_0.png" />
<p><em>The picks in the plot correspond to PCR artifact that we will remove in
the filtering step (see bellow)</em></p>
<p>This plot can be zoomed in the y axis in order to avoid depending on
these artifacts. The plot can also be generated only for a given number
of chromosomes</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5000</span><span class="p">),</span>
                          <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">],</span> <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_70_0.png" src="../_images/tutorial_0_mapping_70_0.png" />
</div>
<div class="section" id="interaction-matrix">
<h4>Interaction matrix<a class="headerlink" href="#interaction-matrix" title="Permalink to this headline">¶</a></h4>
<p>The plot above is probablythe most informative, in order to infer the
qualtity of an Hi-C experiment. This plot represents the matrix of
interaction, the distribution of these interaction as an histogram or as
a function of genomic distance. Some statistics on the specificity of
these interaction, like the cis-to-trans ratio (expected to be at least
between 40 and 60%), and the 3 first eigen vectors of the matrix
highlighting the principal structural features of the matrix (in
non-normalized matrices eigen-vectors are not very informative however).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_74_0.png" src="../_images/tutorial_0_mapping_74_0.png" />
</div>
</div>
</div>
<div class="section" id="filter-reads">
<h2>Filter reads<a class="headerlink" href="#filter-reads" title="Permalink to this headline">¶</a></h2>
<p>In order to remove interactions between reads that are experimental
artifacts, or just uninoformative, a series of adjustable filters can be
applied:</p>
<ol class="arabic simple">
<li>self-circle : reads are comming from a single RE fragment and point
to the outside (—-&lt;===—===&gt;—)</li>
<li>dangling-end : reads are comming from a single RE fragment and point
to the inside (—-===&gt;—&lt;===—)</li>
<li>error : reads are comming from a single RE fragment and point in the
same direction</li>
<li>extra dangling-end : reads are comming from different RE fragment
but are close enough (&lt; max_molecule length) and point to the
inside. Maximum molecule length parameter can be set at 1.5 or 2
times the 99.9 percentile returned by insert_size function above.</li>
<li>too close from RES : semi-dangling-end filter, start position of one
of the read is too close (5 bp by default) from RE cutting site
(with 4 base-pair-cutter enzyme it can be set to 4nt). This filter
is in general not taken into account in <em>in situ</em> Hi-C experiments,
and with 4bp cutters as the ligation may happen only one side of the
DNA fragments.</li>
<li>too short : remove reads comming from small restriction less than
100 bp (default) because they are comparable to the read length, and
are thus probably artifacts.</li>
<li>too large : remove reads comming from large restriction fragments
(default: 100 Kb, P &lt; 10-5 to occur in a randomized genome) as they
likely represent poorly assembled or repeated regions</li>
<li>over-represented : reads coming from the top 0.5% most frequently
detected restriction fragments, they may be prone to PCR artifacts
or represent fragile regions of the genome or genome assembly errors</li>
<li>duplicated : the combination of the start positions (and direction)
of the reads is repeated -&gt; PCR artifact (only keep one copy)</li>
<li>random breaks : start position of one of the read is too far (more
than min_dist_to_re) from RE cutting site. Non-canonical enzyme
activity or random physical breakage of the chromatin. The
min_dist_to_re parameter can be set as the 99.9 percentile of the
distribution of insert sizes, as this parameter is important it is
recommended to try different combinations and observe how it may
affect the quality of the resulting interaction matrix</li>
</ol>
<p>The function <code class="docutils literal"><span class="pre">filter_reads</span></code> works in parallel (4 threads), and creates
one file per filter (10 files, which path are an extension of the input
file containing the reads).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.filter</span> <span class="kn">import</span> <span class="n">filter_reads</span>

<span class="n">masked</span> <span class="o">=</span> <span class="n">filter_reads</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">max_molecule_length</span><span class="o">=</span><span class="mi">610</span><span class="p">,</span> <span class="n">min_dist_to_re</span><span class="o">=</span><span class="mi">915</span><span class="p">,</span>
                      <span class="n">over_represented</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">max_frag_size</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span>
                      <span class="n">min_frag_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">re_proximity</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Filtered reads (and percentage of total):

     Mapped both                :     58868016 (100.00%)
  -----------------------------------------------------
   1- self-circle               :        19763 (  0.03%)
   2- dangling-end              :      1617270 (  2.75%)
   3- error                     :         7907 (  0.01%)
   4- extra dangling-end        :     11123319 ( 18.90%)
   5- too close from RES        :     15365731 ( 26.10%)
   6- too short                 :      6742667 ( 11.45%)
   7- too large                 :          917 (  0.00%)
   8- over-represented          :      2111005 (  3.59%)
   9- duplicated                :       274430 (  0.47%)
  10- random breaks             :        36700 (  0.06%)
</pre>
<p>Previous function creates one file per filter. Each containing the list
of IDs of the reads falling into the corresponding filter. In order to
apply filter, the function <code class="docutils literal"><span class="pre">apply_filter</span></code> will create a new file
without the reads contained in the files. By default all filters are
applied.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.filter</span> <span class="kn">import</span> <span class="n">apply_filter</span>
<span class="n">filt_reads</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/filtered_map.tsv&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">apply_filter</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">filt_reads</span><span class="p">,</span> <span class="n">masked</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
saving to file 33541378 reads without .
</pre>
<pre class="ansi-block literal-block">
33541378
</pre>
<p>As mentioned, it is usually a good idea not to consider the filter
number 5: &#8220;too close form REs&#8221;. To do so, we can just use the
<code class="docutils literal"><span class="pre">filters</span></code> parameter as (the less conservative approach would be to use
only the filters 1, 2, 3, 9 and 10):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">apply_filter</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">filt_reads</span><span class="p">,</span> <span class="n">masked</span><span class="p">,</span> <span class="n">filters</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
saving to file 40026192 reads without .
</pre>
<pre class="ansi-block literal-block">
40026192
</pre>
<p>Filters can also be applied in a &#8220;reverse&#8221; way in order to select only
&#8220;bad&#8221; reads.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">sc_de</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/self_circles_and_dangling-ends.tsv&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">apply_filter</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">sc_de</span><span class="p">,</span> <span class="n">masked</span><span class="p">,</span> <span class="n">filters</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
saving to file 1637033 reads with .
</pre>
<pre class="ansi-block literal-block">
1637033
</pre>
<p>This can be used for example to analyze the distribution of
dangling-ends and self-circle along the genome</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">sc_de</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">],</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">300</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_88_0.png" src="../_images/tutorial_0_mapping_88_0.png" />
<p>Once filtered the peaks previously seen should disapeear:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">],</span>
                          <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">6000</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_90_0.png" src="../_images/tutorial_0_mapping_90_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">plot_genomic_distribution</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span> <span class="n">first_read</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">chr_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">],</span>
                          <span class="n">nreads</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">6000</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_91_0.png" src="../_images/tutorial_0_mapping_91_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_92_0.png" src="../_images/tutorial_0_mapping_92_0.png" />
<p>These maps can be zoomed to a given region, like first chromosome:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">250</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_94_0.png" src="../_images/tutorial_0_mapping_94_0.png" />
<p>Same as above, calling the focus using directly chromosome name and
using a smaller resolution (100 kb):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">focus</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_96_0.png" src="../_images/tutorial_0_mapping_96_0.png" />
</div>
<div class="section" id="filtering-and-normalization">
<h2>Filtering and normalization<a class="headerlink" href="#filtering-and-normalization" title="Permalink to this headline">¶</a></h2>
<div class="section" id="removal-of-columns-having-to-few-data">
<h3>Removal of columns having to few data<a class="headerlink" href="#removal-of-columns-having-to-few-data" title="Permalink to this headline">¶</a></h3>
<p>Depending on the normalization method, the presence of columns with high
proportion of zeros can prevent to converge into a satisfactory result.</p>
<p>For this part of the processing of the data we will start to work on
full matrices. This step is critical in the sense that <strong>we have to
decide at which resolution we are going to analyze the data</strong>. For this
tutorial we will use a resolution of 1 Mb.</p>
<p><em>Note</em> : as all previous steps ended in the generation of a single file,
we just need to load the name of the saved file with the filtered reads:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">filt_reads</span>  <span class="o">=</span> <span class="s1">&#39;/scratch/test/rao2014/HiC003/filtered_map.tsv&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit</span> <span class="kn">import</span> <span class="n">load_hic_data_from_reads</span>

<span class="n">hic_data</span> <span class="o">=</span> <span class="n">load_hic_data_from_reads</span><span class="p">(</span><span class="n">filt_reads</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000000</span><span class="p">)</span>
</pre></div>
</div>
<p>We can visualize the matrix using the same function as before, with the
file of reads:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.mapping.analyze</span> <span class="kn">import</span> <span class="n">hic_map</span>

<span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_104_0.png" src="../_images/tutorial_0_mapping_104_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_data</span><span class="o">.</span><span class="n">filter_columns</span><span class="p">(</span><span class="n">draw_hist</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
WARNING: removing columns having more than 2327 zeroes:
   123   124   125   126   127   128   129   130   131   132   133   134   135   136   137   138   139   140   141   142
   143   342   343   584   585   742   882   929   930   931   932   953  1124  1294  1295  1296  1297  1440  1441  1541
  1585  1586  1587  1588  1589  1590  1591  1592  1593  1594  1595  1596  1597  1598  1599  1600  1601  1602  1720  1721
  1722  1866  1867  1868  1950  1986  1987  2085  2086  2087  2088  2089  2090  2091  2092  2093  2094  2095  2096  2097
  2098  2099  2100  2101  2102  2200  2201  2202  2203  2204  2205  2206  2207  2208  2209  2210  2211  2212  2213  2214
  2215  2216  2217  2218  2307  2308  2309  2310  2311  2312  2313  2314  2315  2316  2317  2318  2319  2320  2321  2322
  2323  2324  2325  2326  2327  2446  2447  2448  2449  2450  2451  2452  2453  2454  2455  2524  2525  2526  2601  2602
  2603  2604  2605  2691  2692  2752  2790  2791  2792  2793  2794  2795  2796  2797  2801  2802  2837  2838  2839  2840
  2841  2842  2843  2844  2845  2846  2850  2851  2946  2947  2948  2949  3044  3045  3046  3047  3048  3049  3050  3051
  3052  3053  3054  3055  3057  3058  3059  3060  3061  3062  3063  3064  3065  3066  3067  3068  3069  3070  3071  3072
  3073  3074  3075  3076  3077  3078  3079  3080  3081  3082  3083  3084  3085  3086  3087  3088  3089  3090  3091  3092
  3093  3094  3095  3096  3097  3098  3099  3100  3101  3102  3103
/home/fransua/.miniconda2/lib/python2.7/site-packages/numpy/core/numeric.py:474: ComplexWarning: Casting complex values to real discards the imaginary part
  return array(a, dtype, copy=False, order=order)
</pre>
<img alt="../_images/tutorial_0_mapping_105_1.png" src="../_images/tutorial_0_mapping_105_1.png" />
<pre class="ansi-block literal-block">
Found 237 of 3103 columnswith poor signal
</pre>
<pre class="ansi-block literal-block">
/home/fransua/.miniconda2/lib/python2.7/site-packages/pytadbit/utils/hic_filtering.py:149: ComplexWarning: Casting complex values to real discards the imaginary part
  round(root, 3), ' '.join(

WARNING: removing columns having less than 2886.461 counts:
   123   124   125   126   127   128   129   130   131   132   133   134   135   136   137   138   139   140   141   142
   143   145   342   343   584   585   742   882   929   930   931   932   953  1124  1294  1295  1296  1297  1440  1441
  1541  1585  1586  1587  1588  1589  1590  1591  1592  1593  1594  1595  1596  1597  1598  1599  1600  1601  1602  1603
  1608  1609  1720  1721  1722  1866  1867  1868  1950  1986  1987  2085  2086  2087  2088  2089  2090  2091  2092  2093
  2094  2095  2096  2097  2098  2099  2100  2101  2102  2200  2201  2202  2203  2204  2205  2206  2207  2208  2209  2210
  2211  2212  2213  2214  2215  2216  2217  2218  2307  2308  2309  2310  2311  2312  2313  2314  2315  2316  2317  2318
  2319  2320  2321  2322  2323  2324  2325  2326  2327  2329  2446  2447  2448  2449  2450  2451  2452  2453  2454  2455
  2524  2525  2526  2601  2602  2603  2604  2605  2691  2692  2752  2790  2791  2792  2793  2794  2795  2796  2797  2801
  2802  2837  2838  2839  2840  2841  2842  2843  2844  2845  2846  2850  2851  2852  2946  2947  2948  2949  3044  3045
  3046  3047  3048  3049  3050  3051  3052  3053  3054  3055  3057  3058  3059  3060  3061  3062  3063  3064  3065  3066
  3067  3068  3069  3070  3071  3072  3073  3074  3075  3076  3077  3078  3079  3080  3081  3082  3083  3084  3085  3086
  3087  3088  3089  3090  3091  3092  3093  3094  3095  3096  3097  3098  3099  3100  3101  3102  3103
</pre>
<p>Filtered columns (to high count of zeroes, or to low mean value) will be
skipped in most of analysis available and are now shaded in the matrix
representation:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_107_0.png" src="../_images/tutorial_0_mapping_107_0.png" />
</div>
<div class="section" id="normalization">
<h3>Normalization<a class="headerlink" href="#normalization" title="Permalink to this headline">¶</a></h3>
<p>TADbit implements ICE normalization strategy [Imakaev2012]_ which
basically consists constructing a new in dividing each cell</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_data</span><span class="o">.</span><span class="n">normalize_hic</span><span class="p">(</span><span class="n">iterations</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">max_dev</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
iterative correction
  - copying matrix
  - computing baises
          2836.000       27884.987       44592.000    0   0.89830
         20089.022       30027.383       85246.499    1   1.83896
         15842.132       31075.219       40561.598    2   0.49020
         25400.491       31543.690       47025.449    3   0.49080
         24496.468       31783.468       37210.830    4   0.22927
         28213.858       31908.415       37511.228    5   0.17559
         28468.121       31976.815       35032.068    6   0.10973
         29843.664       32015.253       34841.364    7   0.08827
rescaling to factor 1
</pre>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For a faster normalization like the one used in <a class="reference internal" href="../biblio.html#rao2014" id="id2">[Rao2014]</a> just set the iterations parameter to 0</p>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">hic_map</span><span class="p">(</span><span class="n">hic_data</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_0_mapping_112_0.png" src="../_images/tutorial_0_mapping_112_0.png" />
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="">Mapping paired-end reads from NGS experiment (with GEM)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../tutorial.html"
                        title="previous chapter">Tutorials</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_1_general.html"
                        title="next chapter">Getting started</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_1_general.html" title="Getting started"
             >next</a> |</li>
        <li class="right" >
          <a href="../tutorial.html" title="Tutorials"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright .
      Last updated on Apr 25, 2016.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>