<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Compartments and TADs detection &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Comparison of Hi-C experiments" href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" />
    <link rel="prev" title="Bin filtering and normalization" href="tutorial_7-Bin_filtering_and_normalization.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" title="Comparison of Hi-C experiments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_7-Bin_filtering_and_normalization.html" title="Bin filtering and normalization"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="compartments-and-tads-detection">
<h1>Compartments and TADs detection<a class="headerlink" href="#compartments-and-tads-detection" title="Permalink to this headline">¶</a></h1>
<p>Here, we present the analysis to detect the compartments in Mouse B and
iPS cells. In this example, we will use the GC-content (guanine-cytosine
content) to identify which bins belong to the A or B compartments. The
percentage of bases that are either guanine or cytosine on a DNA strand
correlates directly with gene density and is a good measure to identify
open and close chromatine.</p>
<p><strong>*Note</strong>: Compartments are detected on the full genome matrix.*</p>
<div class="section" id="compartments">
<h2>Compartments<a class="headerlink" href="#compartments" title="Permalink to this headline">¶</a></h2>
<div class="section" id="mouse-b-cells">
<h3>Mouse B cells<a class="headerlink" href="#mouse-b-cells" title="Permalink to this headline">¶</a></h3>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_10_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_10_0.png" />
</div>
<div class="section" id="mouse-ips-cells">
<h3>Mouse iPS cells<a class="headerlink" href="#mouse-ips-cells" title="Permalink to this headline">¶</a></h3>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_15_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_15_0.png" />
</div>
<div class="section" id="compare">
<h3>Compare<a class="headerlink" href="#compare" title="Permalink to this headline">¶</a></h3>
<p>The assignments of the compartments for the two cell types are stored in
two different files:</p>
<p>In another folder, we also saved the coordinates of each computed
eigenvector:</p>
<p><strong>*Note</strong>: In this file the first line shows the eigenvector index with
it’s corresponding eigenvalue (the first column should always be the one
you selected, even if it is not the first eigenvector).*</p>
<p>Then there are the coordinates of the eigenvectors. In the first column,
the coordinates correspond to the assignment of the A and B
compartments, positive values for A compartments and negative values for
B compartments.</p>
<p>Load eigenvectors coordinates from files:</p>
</div>
<div class="section" id="spot-changes-in-activity">
<h3>Spot changes in activity<a class="headerlink" href="#spot-changes-in-activity" title="Permalink to this headline">¶</a></h3>
<p>Plot the difference between each eigenvector along chromosome</p>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_29_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_29_0.png" />
</div>
<div class="section" id="correlate-eigenvectors">
<h3>Correlate eigenvectors<a class="headerlink" href="#correlate-eigenvectors" title="Permalink to this headline">¶</a></h3>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_31_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_31_0.png" />
</div>
</div>
</div>
<div class="section" id="tads">
<h1>TADs<a class="headerlink" href="#tads" title="Permalink to this headline">¶</a></h1>
<p>Now, we move to the TADs detection. In this notebook we will detect TAD
borders at 100kbp resolution.</p>
<!-- The comparison of TAD borders at high resolutions becomes difficult because the border positions are not as sharply defined as at lower resolutions.

An example of consistency between TAD borders is shown in the following graph. TAD borders are called at 1 kb resolution (insulation score-based method). We assume that there should be a high ratio of conservation between the 4 replicates and as we see that's true for resolutions lower than approximately 100kbp. If our bin size is for example 50kbp, we only reach the same ratio of consistency if we consider TAD borders found 2 bins away as being the same border in the different replicates.

<img src="images/TAD_calling_resolution.png"> --><img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_39_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_39_0.png" />
<p>TADbit is the original TAD caller algorithm TADbit is a breakpoint
detection algorithm that returns the optimal segmentation of the
chromosome under BIC-penalized likelihood. The model assumes that counts
have a Poisson distribution and that the expected value of the counts
decreases like a power-law with the linear distance on the chromosome.</p>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_44_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_44_0.png" />
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_45_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_45_0.png" />
<p>TopDom identifies TAD borders based on the assumption that contact
frequencies between regions upstream and downstream of a border are
lower than those between two regions within a TAD. The algorithm only
depends on a single parameter corresponding to the window size. The
algorithm provides a measure (from 0 to 10) of confidence on the
accuracy of the border detection
(<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/26704975">https://www.ncbi.nlm.nih.gov/pubmed/26704975</a>).</p>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_51_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_51_0.png" />
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_52_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_52_0.png" />
<p>Insulation score (Crane et al. 2015 <a class="reference external" href="https://doi.org/10.1038/nature14450">https://doi.org/10.1038/nature14450</a>)
can be used to build an insulation profile of the genome and, with a
simple transformation, to identify TAD borders.</p>
<p>First we need to normalize the matrices by visibility and by decay:</p>
<p>The two important parameter to define are the window size, the distance
from the diagonal and the delta. - the square size should be 500 kb as
close as possible from the diagonal - the delta is to look for increases
in insulation around a given bin. Should be around 100 kb (in out case
we define it as 200 kb as we are working at 100 kb resolution, and
working with only one bin is a bit to little)</p>
<p>Once defined the insulation score and the values of delta can be used to
search for borders.</p>
<p>Currently the representation is not available in TADbit as for the other
methods, but we can easily plot it:</p>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_65_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_65_0.png" />
<div class="section" id="comparison-of-tad-borders">
<h2>Comparison of TAD borders<a class="headerlink" href="#comparison-of-tad-borders" title="Permalink to this headline">¶</a></h2>
<p>The TAD borders can be aligned, using a simple reciprocal best hit
strategy:</p>
<p>In the plots below, each arc represents a TAD. Between two consecutive
arcs the triangle mark the border. This triangle is colored depending on
the confidence of the TAD border call.</p>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_70_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_70_0.png" />
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_71_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_71_0.png" />
<div class="section" id="statistical-significance-of-the-tad-borders-alignments">
<h3>Statistical significance of the TAD borders alignments<a class="headerlink" href="#statistical-significance-of-the-tad-borders-alignments" title="Permalink to this headline">¶</a></h3>
<p>In order to asses how well two experiments align, or how conserved are
the TAD borders between two experiments, we can compare the overlap
between our experiments with the overlap of simulated random
distributions of TAD borders.</p>
<p>This analysis returns an alignment score between 0 and 1 (0: no match,
1: all borders aligned), a p-value (usually equal zero), the proportion
of borders conserved in the second experiment, and the proportion of
borders conserved in the first experiment:</p>
</div>
<div class="section" id="save-chromosome-object-with-tad-definition">
<h3>Save Chromosome object (with TAD definition)<a class="headerlink" href="#save-chromosome-object-with-tad-definition" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="extra-choosing-the-best-resolution-to-call-tad-borders">
<h3>Extra: choosing the best resolution to call TAD borders<a class="headerlink" href="#extra-choosing-the-best-resolution-to-call-tad-borders" title="Permalink to this headline">¶</a></h3>
<p>TopDom and the methodology based on insulation score are relatively fast
computationally and allow to call TAD borders at very high resolution.
<strong>However</strong> being able to call TAD borders at high resolution does not
mean that we should do it. <strong>As the resolution increases, so does the
noise</strong>.</p>
<p>In order to assess which is the best resolution in order to call TAD
borders, a good strategy is to test the consistency of several
resolutions.</p>
<p>In the example below, we use the methodology based on the insulation
score as it is the fastest (tadbit strategy is almost not usable bellow
20 kb). We compare the number of TAD borders that are shared (plus-minus
one bin) between the two replicates (iPS and B cells).</p>
<p>We align TAD borders and keep the proportion of borders aligned between
cell types</p>
<p>We plot the results, as boxplots to group the values of each
chromosomes:</p>
<img alt="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_90_0.png" src="tutorial/../nbpictures//tutorial_8-Compartments_and_TADs_detection_90_0.png" />
<p>In the case of these experiments it seems that looking for TAD borders
bellow 50 kb is dangerous, as very few are reproducible. Calling TAD
borders at 100 kb resolution might be here a good idea.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux under conda</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html#software">Software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html#installing-tadbit-on-gnu-linux">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Compartments and TADs detection</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tads">TADs</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_9-Compare_and_merge_Hi-C_experiments.html">Comparison of Hi-C experiments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_7-Bin_filtering_and_normalization.html"
                        title="previous chapter">Bin filtering and normalization</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html"
                        title="next chapter">Comparison of Hi-C experiments</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_9-Compare_and_merge_Hi-C_experiments.html" title="Comparison of Hi-C experiments"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_7-Bin_filtering_and_normalization.html" title="Bin filtering and normalization"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Jun 21, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>