<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Parameter optimization for IMP. &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Tutorials" href="../tutorial.html" />
    <link rel="next" title="Three-dimensional modeling of chromatine structure" href="tutorial_6_modelling_and_analysis.html" />
    <link rel="prev" title="Hi-C data processing" href="tutorial_4_data_normalization.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial_6_modelling_and_analysis.html" title="Three-dimensional modeling of chromatine structure"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_4_data_normalization.html" title="Hi-C data processing"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Tutorials</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="parameter-optimization-for-imp">
<h1>Parameter optimization for IMP.<a class="headerlink" href="#parameter-optimization-for-imp" title="Permalink to this headline">¶</a></h1>
<p><a href="#id1"><span class="problematic" id="id2">*</span></a>Recover data from previous section by loading the previously saved chromosome and its Hi-C data.*Import the necessary libraries.
.. code:: python</p>
<blockquote>
<div># Libraries
from pytadbit import load_chromosome # to load chromosomes
from pytadbit.imp.impoptimizer import IMPoptimizer</div></blockquote>
<p>First, load the chromosome from previous tutorial (<span class="xref std std-ref">run_tadbit</span>).
.. code:: python</p>
<blockquote>
<div># Load the chromosome
my_chrom = load_chromosome(&#8216;some_path.tdb&#8217;)</div></blockquote>
<p>Next, load Hi-C data for each experiment (Hi-C data is not saved inside chromosome objects because of their size):
.. code:: python</p>
<blockquote>
<div><p># Loop over experiments in chromosome and load Hi-C data.
res = 100000</p>
<dl class="docutils">
<dt>for exp in my_chrom.experiments:</dt>
<dd><dl class="first docutils">
<dt>try:</dt>
<dd><dl class="first docutils">
<dt>exp.load_hic_data(&#8216;../../scripts/sample_data/HIC_{0}_{1}_{1}_{2}_obs.txt&#8217;.format(</dt>
<dd>exp.name, my_chrom.name, res))</dd>
</dl>
<p class="last">exp.normalize_hic(factor=1)</p>
</dd>
<dt>except IOError:</dt>
<dd>print &#8216;file not found for experiment: &#8216; + exp.name
continue</dd>
</dl>
<p class="last">print exp</p>
</dd>
</dl>
</div></blockquote>
<pre class="ansi-block literal-block">
Experiment k562:
   resolution        : 100Kb
   TADs              : 37
   Hi-C rows         : 639
   normalized        : visibility_factor:1
   identifier        : k562
   cell type         : wild type
   restriction enzyme: UNKNOWN
   project           : TADbit tutorial

Experiment gm06690:
   resolution        : 100Kb
   TADs              : 34
   Hi-C rows         : 639
   normalized        : visibility_factor:1
   identifier        : gm06690
   cell type         : cancer
   restriction enzyme: UNKNOWN
   project           : TADbit tutorial

file not found for experiment: k562+gm06690
file not found for experiment: batch_gm06690_k562
</pre>
<pre class="ansi-block literal-block">
/usr/local/lib/python2.7/dist-packages/pytadbit/parsers/hic_parser.py:93: UserWarning: WARNING: non integer values
  warn('WARNING: non integer values')
</pre>
<p>The log indicates that experiment &#8220;k562+gm06690&#8221; had no file. Such experiment was built ad-hoc in our previous tutorial and needs to be created again by summing the Hi-C matrices from the individual experiments.
.. code:: python</p>
<blockquote>
<div>my_chrom.experiments[&#8216;k562+gm06690&#8217;] = my_chrom.experiments[&#8216;k562&#8217;] + my_chrom.experiments[&#8216;gm06690&#8217;]</div></blockquote>
<p>We are now going to use the experiment &#8220;gm06690&#8221; for the modelling. The
first thing we need to do is to filter columns with low counts:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">exp</span> <span class="o">=</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span><span class="p">[</span><span class="s">&#39;gm06690&#39;</span><span class="p">]</span>
<span class="n">exp</span><span class="o">.</span><span class="n">filter_columns</span><span class="p">(</span><span class="n">draw_hist</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="k">print</span> <span class="n">my_chrom</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
/usr/lib/python2.7/dist-packages/numpy/core/numeric.py:460: ComplexWarning: Casting complex values to real discards the imaginary part
  return array(a, dtype, copy=False, order=order)
</pre>
<img alt="../_images/tutorial_5_parameter_optimization_11_1.png" src="../_images/tutorial_5_parameter_optimization_11_1.png" />
<pre class="ansi-block literal-block">
[Experiment k562 (resolution: 100Kb, TADs: 37, Hi-C rows: 639, normalized: visibility_factor:1), Experiment gm06690 (resolution: 100Kb, TADs: 34, Hi-C rows: 639, normalized: visibility_factor:1), Experiment k562+gm06690 (resolution: 100Kb, TADs: None, Hi-C rows: 639, normalized: visibility_factor:2), Experiment batch_gm06690_k562 (resolution: 100Kb, TADs: 35, Hi-C rows: 639, normalized: None)]
</pre>
<pre class="ansi-block literal-block">
/usr/local/lib/python2.7/dist-packages/pytadbit/utils/hic_filtering.py:136: ComplexWarning: Casting complex values to real discards the imaginary part
  round(root, 3), ' '.join(

WARNING: removing columns having less than 67.485 counts:
   246   247   248   249   250   251   252   253   254   255   256   257   258   259   260   261   262   263   264   265
   266   267   268   269   270   271   272   273   274   275   276   277   278   279   280   281   282   283   284   285
   286   287   288   289   290   291   292   293   294   295   296   297   298   299   300   301   302   303   304   305
   306   307   308   309   310   311   312   313   314   315   316   317   318   319   320   321   322   323   324   639
</pre>
<div class="section" id="optimization-of-imp-3d-modeling-parameters">
<h2>Optimization of IMP 3D modeling parameters<a class="headerlink" href="#optimization-of-imp-3d-modeling-parameters" title="Permalink to this headline">¶</a></h2>
<p>In the previous tutorial we found a specific TAD (region 406 to 448) that seemed quite conserved accross different cell types.</p>
<p>Next, we will optimize the three IMP parameters for this TAD. The IMP parameters to optimize are maximal distance between two non-interacting particles (maxdist), Upper-bound Z-score (upfreq) and Lower-bound Z-score (lowfreq). For details see Bau &amp; Marti-Renom. METHODS <a class="reference internal" href="../biblio.html#bau2012" id="id3">[Baù2012]</a>.
.. code:: python</p>
<blockquote>
<div>optimizer = IMPoptimizer(exp, 100, 200, n_models=50, n_keep=25, cutoff=2000)</div></blockquote>
<pre class="ansi-block literal-block">
Preparing to optimize 101 particles
</pre>
<p><code class="docutils literal"><span class="pre">cutoff</span></code> value corresponds to the distance limit, in nanometers, to consider if two particles of a model are interacting or not. A wise choice corresponds to two time the resolution times the scale factor (0.01), which in this case is <span class="math">\(cutoff = 100Kb \times 2 \times scale = 100000 \times 2 \times 0.01 = 2000\)</span></p>
<p><em>Note: Usually the number of models to generate and to keep, should be respectively 500 and 100.</em>
.. code:: python</p>
<blockquote>
<div><p># Optimize parameters. Be aware that this step is CPU intensive. If you want to se the progress, set verbose=True.
optimizer.run_grid_search(n_cpus=8, lowfreq_range=(-1, 0, 0.2), upfreq_range=(0.2, 0.8, 0.2),</p>
<blockquote>
<div>maxdist_range=(2000, 3500, 500), verbose=True)</div></blockquote>
</div></blockquote>
<pre class="ansi-block literal-block">
 1  0.2 -1 2000 0.01 0.767098689596
 2  0.2 -0.8 2000 0.01 0.767098689596
 3  0.2 -0.6 2000 0.01 0.764430378531
 4  0.2 -0.4 2000 0.01 0.769877194832
 5  0.2 -0.2 2000 0.01 0.780250808861
 6  0.2 0 2000 0.01 0.777301796587
 7  0.4 -1 2000 0.01 0.772393224867
 8  0.4 -0.8 2000 0.01 0.772393224867
 9  0.4 -0.6 2000 0.01 0.772958040825
10  0.4 -0.4 2000 0.01 0.772804035198
11  0.4 -0.2 2000 0.01 0.778396325547
12  0.4 0 2000 0.01 0.778557679165
13  0.6 -1 2000 0.01 0.780712289443
14  0.6 -0.8 2000 0.01 0.780712289443
15  0.6 -0.6 2000 0.01 0.782092378436
16  0.6 -0.4 2000 0.01 0.783871795733
17  0.6 -0.2 2000 0.01 0.784238133139
18  0.6 0 2000 0.01 0.78502636527
19  0.8 -1 2000 0.01 0.787091209619
20  0.8 -0.8 2000 0.01 0.787091209619
21  0.8 -0.6 2000 0.01 0.786614740119
22  0.8 -0.4 2000 0.01 0.792475503423
23  0.8 -0.2 2000 0.01 0.790170250267
24  0.8 0 2000 0.01 0.788238926824
25  0.2 -1 2500 0.01 0.750871626725
26  0.2 -0.8 2500 0.01 0.750871626725
27  0.2 -0.6 2500 0.01 0.751733987004
28  0.2 -0.4 2500 0.01 0.7582618866
29  0.2 -0.2 2500 0.01 0.771518216885
30  0.2 0 2500 0.01 0.773504325552
31  0.4 -1 2500 0.01 0.756774760421
32  0.4 -0.8 2500 0.01 0.756774760421
33  0.4 -0.6 2500 0.01 0.759833450921
34  0.4 -0.4 2500 0.01 0.765192308125
35  0.4 -0.2 2500 0.01 0.772765218419
36  0.4 0 2500 0.01 0.774522497614
37  0.6 -1 2500 0.01 0.780682090473
38  0.6 -0.8 2500 0.01 0.780682090473
39  0.6 -0.6 2500 0.01 0.781857090923
40  0.6 -0.4 2500 0.01 0.783029516904
41  0.6 -0.2 2500 0.01 0.790986026532
42  0.6 0 2500 0.01 0.789542526866
43  0.8 -1 2500 0.01 0.786372973446
44  0.8 -0.8 2500 0.01 0.786372973446
45  0.8 -0.6 2500 0.01 0.787207607051
46  0.8 -0.4 2500 0.01 0.788810781589
47  0.8 -0.2 2500 0.01 0.795164267021
48  0.8 0 2500 0.01 0.795552255769
49  0.2 -1 3000 0.01 0.736950317918
50  0.2 -0.8 3000 0.01 0.736950317918
51  0.2 -0.6 3000 0.01 0.740476733274
52  0.2 -0.4 3000 0.01 0.753609629213
53  0.2 -0.2 3000 0.01 0.764380672057
54  0.2 0 3000 0.01 0.771119903648
55  0.4 -1 3000 0.01 0.758070700552
56  0.4 -0.8 3000 0.01 0.758070700552
57  0.4 -0.6 3000 0.01 0.752965160316
58  0.4 -0.4 3000 0.01 0.761098324731
59  0.4 -0.2 3000 0.01 0.774111941221
60  0.4 0 3000 0.01 0.775126798787
61  0.6 -1 3000 0.01 0.772880007549
62  0.6 -0.8 3000 0.01 0.772880007549
63  0.6 -0.6 3000 0.01 0.771506004428
64  0.6 -0.4 3000 0.01 0.775450414
65  0.6 -0.2 3000 0.01 0.786066844799
66  0.6 0 3000 0.01 0.791412592662
67  0.8 -1 3000 0.01 0.786054130073
68  0.8 -0.8 3000 0.01 0.786054130073
69  0.8 -0.6 3000 0.01 0.78560321088
70  0.8 -0.4 3000 0.01 0.784784920366
71  0.8 -0.2 3000 0.01 0.793887281274
72  0.8 0 3000 0.01 0.796732555769
73  0.2 -1 3500 0.01 0.739942301447
74  0.2 -0.8 3500 0.01 0.739942301447
75  0.2 -0.6 3500 0.01 0.74297913941
76  0.2 -0.4 3500 0.01 0.751257211108
77  0.2 -0.2 3500 0.01 0.772119574201
78  0.2 0 3500 0.01 0.772849566162
79  0.4 -1 3500 0.01 0.745002332126
80  0.4 -0.8 3500 0.01 0.745002332126
81  0.4 -0.6 3500 0.01 0.746930311926
82  0.4 -0.4 3500 0.01 0.751131192823
83  0.4 -0.2 3500 0.01 0.768242207472
84  0.4 0 3500 0.01 0.765639010622
85  0.6 -1 3500 0.01 0.757275221091
86  0.6 -0.8 3500 0.01 0.757275221091
87  0.6 -0.6 3500 0.01 0.759466814558
88  0.6 -0.4 3500 0.01 0.757620114436
89  0.6 -0.2 3500 0.01 0.77102095279
90  0.6 0 3500 0.01 0.773079965496
91  0.8 -1 3500 0.01 0.766944515046
92  0.8 -0.8 3500 0.01 0.766944515046
93  0.8 -0.6 3500 0.01 0.769696862072
94  0.8 -0.4 3500 0.01 0.777119886412
95  0.8 -0.2 3500 0.01 0.778091343506
96  0.8 0 3500 0.01 0.778697968972
</pre>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The above warning is given when a small matrix is loaded. TADbit has a filtering function that is applied to all Hi-C matrices with the aim of removing entire rows with very low counts. Those rows/colums are treated then for modeling as &#8220;missing-data&#8221; points. This flitering function can only be applied for relatively large matrices.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">By default TADbit does not store the models generated during the optimization, however, in case they are needed, the option savedata may allow to store them.</p>
</div>
<div class="section" id="optimizing-from-experiment">
<h3>Optimizing from Experiment<a class="headerlink" href="#optimizing-from-experiment" title="Permalink to this headline">¶</a></h3>
<p>The exact same as above can be done from Experiment objects directly:
.. code:: python</p>
<blockquote>
<div>#optimizer = exp.optimal_imp_parameters(100, 200, n_cpus=8, n_models=50, n_keep=25, cutoff=1000,
#                                       lowfreq_range=(-1, 0, 0.2), upfreq_range=(0.2, 0.8, 0.2),
#                                       maxdist_range=(2000, 3500, 500),
#                                       verbose=False)</div></blockquote>
</div>
</div>
<div class="section" id="visualize-the-results">
<h2>Visualize the results<a class="headerlink" href="#visualize-the-results" title="Permalink to this headline">¶</a></h2>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer</span><span class="o">.</span><span class="n">write_result</span><span class="p">(</span><span class="s">&#39;results.log&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c"># Visualize the results of the optimization.</span>
<span class="n">optimizer</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/tutorial_5_parameter_optimization_23_0.png" src="../_images/tutorial_5_parameter_optimization_23_0.png" />
<p>We can also ask to mark on the plot the best N combination of parameters with the &#8220;show_best&#8221; parameter.
.. code:: python</p>
<blockquote>
<div># Visualize the results of the optimization and mark the best 10 parameter sets
optimizer.plot_2d(show_best=20)</div></blockquote>
<img alt="../_images/tutorial_5_parameter_optimization_25_0.png" src="../_images/tutorial_5_parameter_optimization_25_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">axes_range</span> <span class="o">=</span> <span class="p">[[</span><span class="nb">float</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">optimizer</span><span class="o">.</span><span class="n">scale_range</span><span class="p">],</span>
                    <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">optimizer</span><span class="o">.</span><span class="n">maxdist_range</span><span class="p">],</span>
                    <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">optimizer</span><span class="o">.</span><span class="n">upfreq_range</span><span class="p">],</span>
                    <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">optimizer</span><span class="o">.</span><span class="n">lowfreq_range</span><span class="p">]]</span>

<span class="k">print</span> <span class="n">axes_range</span>
<span class="p">[</span><span class="nb">round</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">axes_range</span><span class="p">[</span><span class="mi">3</span><span class="p">]]</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">optimizer</span><span class="o">.</span><span class="n">_result_to_array</span><span class="p">()</span>
<span class="n">wax</span> <span class="o">=</span> <span class="p">[</span><span class="nb">round</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">axes_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<span class="n">zax</span> <span class="o">=</span> <span class="p">[</span><span class="nb">round</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">axes_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<span class="n">xax</span> <span class="o">=</span> <span class="p">[</span><span class="nb">round</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">axes_range</span><span class="p">[</span><span class="mi">3</span><span class="p">]]</span>
<span class="n">yax</span> <span class="o">=</span> <span class="p">[</span><span class="nb">round</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">axes_range</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span>
<span class="n">sort_result</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([(</span><span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">l</span><span class="p">],</span> <span class="n">wax</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">zax</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">xax</span><span class="p">[</span><span class="n">l</span><span class="p">],</span> <span class="n">yax</span><span class="p">[</span><span class="n">k</span><span class="p">])</span>
                              <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">wax</span><span class="p">))</span>
                              <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">zax</span><span class="p">))</span>
                              <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">yax</span><span class="p">))</span>
                              <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xax</span><span class="p">))</span>
                              <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">l</span><span class="p">])</span>
                              <span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                             <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="k">print</span> <span class="n">sort_result</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
[[0.01], [2000.0, 2500.0, 3000.0, 3500.0], [0.2, 0.4, 0.6, 0.8], [-1.0, -0.8, -0.6, -0.4, -0.2, 0.0]]
(0.79673255576921265, 0.01, 3000.0, 0.0, 0.8)
</pre>
<p>One can also visualize the parameter optimization according to ne of the three optimization parameters.
.. code:: python</p>
<blockquote>
<div># Visualize the results of the optimization based on the lowfreq parameter.
optimizer.plot_2d(axes=(&#8216;upfreq&#8217;, &#8216;lowfreq&#8217;, &#8216;maxdist&#8217;, &#8216;scale&#8217;),show_best=10)</div></blockquote>
<img alt="../_images/tutorial_5_parameter_optimization_28_0.png" src="../_images/tutorial_5_parameter_optimization_28_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">skip</span><span class="o">=</span><span class="p">{</span><span class="s">&quot;scale&quot;</span><span class="p">:</span><span class="mf">0.01</span><span class="p">},</span> <span class="n">show_best</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_5_parameter_optimization_29_0.png" src="../_images/tutorial_5_parameter_optimization_29_0.png" />
<p>TADbit also provides the possibility to view it all together in a 3D plot (note that, while here its a static image, inside matplotlib GUI you would be able to turn around and zoom):
.. code:: python</p>
<blockquote>
<div># Visualize the results of the optimization using a 3D representation with the three optimization parameters in the axis.
optimizer.plot_3d(axes=(&#8216;scale&#8217;, &#8216;maxdist&#8217;, &#8216;upfreq&#8217;, &#8216;lowfreq&#8217;))</div></blockquote>
<img alt="../_images/tutorial_5_parameter_optimization_31_0.png" src="../_images/tutorial_5_parameter_optimization_31_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer</span><span class="o">.</span><span class="n">run_grid_search</span><span class="p">(</span><span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">lowfreq_range</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">1.</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">),</span> <span class="n">upfreq_range</span><span class="o">=</span><span class="p">(</span><span class="mf">0.6</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
                          <span class="n">scale_range</span><span class="o">=</span><span class="p">[</span><span class="mf">0.01</span><span class="p">],</span> <span class="n">maxdist_range</span><span class="o">=</span><span class="p">[</span><span class="mi">2000</span><span class="p">,</span><span class="mi">2250</span><span class="p">,</span><span class="mi">2500</span><span class="p">,</span><span class="mi">2750</span><span class="p">,</span><span class="mi">3000</span><span class="p">],</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/tutorial_5_parameter_optimization_33_0.png" src="../_images/tutorial_5_parameter_optimization_33_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">show_best</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_5_parameter_optimization_34_0.png" src="../_images/tutorial_5_parameter_optimization_34_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer</span><span class="o">.</span><span class="n">write_result</span><span class="p">(</span><span class="s">&#39;results.log&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer2</span> <span class="o">=</span> <span class="n">IMPoptimizer</span><span class="p">(</span><span class="n">exp</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="n">n_models</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_keep</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
Preparing to optimize 101 particles
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer2</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s">&#39;results.log&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer2</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">keys</span><span class="p">()[</span><span class="mi">105</span><span class="p">]</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">
('0.01', '3500', '0.8', '-0.2')
</pre>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">optimizer2</span><span class="o">.</span><span class="n">plot_2d</span><span class="p">(</span><span class="n">show_best</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/tutorial_5_parameter_optimization_39_0.png" src="../_images/tutorial_5_parameter_optimization_39_0.png" />
</div>
<div class="section" id="retrieve-best-parameters">
<h2>Retrieve best parameters<a class="headerlink" href="#retrieve-best-parameters" title="Permalink to this headline">¶</a></h2>
<p>Once done, best results can be returned as a dictionary to be used for modeling (see next section of the tutorial)
.. code:: python</p>
<blockquote>
<div><p>config = optimizer.get_best_parameters_dict(reference=&#8217;gm cell from Job Dekker 2009&#8217;)</p>
<p>print config</p>
</div></blockquote>
<pre class="ansi-block literal-block">
{'maxdist': 2750.0, 'upfreq': 0.8, 'kforce': 5, 'reference': 'gm cell from Job Dekker 2009', 'lowfreq': 0.0, 'scale': 0.01}
</pre>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#identification-and-analysis-of-tads">Identification and Analysis of TADs</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#three-dimentional-modelling">Three-dimentional modelling</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tutorial_4_data_normalization.html">Hi-C data processing</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="">Parameter optimization for IMP.</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_6_modelling_and_analysis.html">Three-dimensional modeling of chromatine structure</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial_7_single_model_analysis.html">Structural analysis of a chromatin model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_4_data_normalization.html"
                        title="previous chapter">Hi-C data processing</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_6_modelling_and_analysis.html"
                        title="next chapter">Three-dimensional modeling of chromatine structure</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial_6_modelling_and_analysis.html" title="Three-dimensional modeling of chromatine structure"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_4_data_normalization.html" title="Hi-C data processing"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright .
      Last updated on May 08, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>