<!doctype html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Installing TADbit on GNU/Linux &#8212; TADbit 1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/forkme_nature.css" />
    <link rel="stylesheet" type="text/css" href="_static/black-on-white.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="shortcut icon" href="_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorials" href="tutorial.html" />
    <link rel="prev" title="Welcome to TADbit’s documentation!" href="index.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial.html" title="Tutorials"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to TADbit’s documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">TADbit 1.0 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Installing TADbit on GNU/Linux</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="installing-tadbit-on-gnu-linux">
<h1>Installing TADbit on GNU/Linux<a class="headerlink" href="#installing-tadbit-on-gnu-linux" title="Permalink to this headline">¶</a></h1>
<p><strong>TADbit requires python2 &gt;= 2.6 or python3 &gt;= 3.6 as well as several
dependencies that are listed below.</strong></p>
<section id="dependencies">
<h2>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h2>
<section id="conda">
<h3>Conda<a class="headerlink" href="#conda" title="Permalink to this headline">¶</a></h3>
<p>Conda (<a class="reference external" href="http://conda.pydata.org/docs/index.html">http://conda.pydata.org/docs/index.html</a>) is a package manager,
mainly hosting python programs, that is very useful when no root access
is available and the softwares have complicated dependencies.</p>
<p>To install it (<strong>in case you don’t already have it</strong>) just download the
installer from <a class="reference external" href="http://conda.pydata.org/miniconda.html">http://conda.pydata.org/miniconda.html</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

wget -nv https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh -O miniconda.sh
</pre></div>
</div>
<pre class="ansi-block literal-block">2018-09-28 00:39:07 URL:https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh [41787735/41787735] -&gt; &quot;miniconda.sh&quot; [1]</pre>
<p>And run it with all the default options. The installer will create a
<code class="docutils literal notranslate"><span class="pre">miniconda2</span></code> folder in your home directory where all the programs that
you need will be stored (including python).</p>
<p>Alternatively you can also use this oneliner:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

bash miniconda.sh -b -p <span class="nv">$HOME</span>/miniconda2
</pre></div>
</div>
<pre class="ansi-block literal-block">PREFIX=/home/fransua/miniconda2
installing: python-2.7.15-h1571d57_0 ...
installing: ca-certificates-2018.03.07-0 ...
installing: conda-env-2.6.0-1 ...
installing: libgcc-ng-8.2.0-hdf63c60_1 ...
installing: libstdcxx-ng-8.2.0-hdf63c60_1 ...
installing: libffi-3.2.1-hd88cf55_4 ...
installing: ncurses-6.1-hf484d3e_0 ...
installing: openssl-1.0.2p-h14c3975_0 ...
installing: yaml-0.1.7-had09818_2 ...
installing: zlib-1.2.11-ha838bed_2 ...
installing: libedit-3.1.20170329-h6b74fdf_2 ...
installing: readline-7.0-h7b6447c_5 ...
installing: tk-8.6.8-hbc83047_0 ...
installing: sqlite-3.24.0-h84994c4_0 ...
installing: asn1crypto-0.24.0-py27_0 ...
installing: certifi-2018.8.24-py27_1 ...
installing: chardet-3.0.4-py27_1 ...
installing: enum34-1.1.6-py27_1 ...
installing: futures-3.2.0-py27_0 ...
installing: idna-2.7-py27_0 ...
installing: ipaddress-1.0.22-py27_0 ...
installing: pycosat-0.6.3-py27h14c3975_0 ...
installing: pycparser-2.18-py27_1 ...
installing: pysocks-1.6.8-py27_0 ...
installing: ruamel_yaml-0.15.46-py27h14c3975_0 ...
installing: six-1.11.0-py27_1 ...
installing: cffi-1.11.5-py27he75722e_1 ...
installing: setuptools-40.2.0-py27_0 ...
installing: cryptography-2.3.1-py27hc365091_0 ...
installing: wheel-0.31.1-py27_0 ...
installing: pip-10.0.1-py27_0 ...
installing: pyopenssl-18.0.0-py27_0 ...
installing: urllib3-1.23-py27_0 ...
installing: requests-2.19.1-py27_0 ...
installing: conda-4.5.11-py27_0 ...
installation finished.</pre>
<pre class="ansi-block literal-block">Python 2.7.15 :: Anaconda, Inc.</pre>
</section>
<section id="python-libraries">
<h3>Python libraries<a class="headerlink" href="#python-libraries" title="Permalink to this headline">¶</a></h3>
<p><strong>Required:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">python</span><span class="o">-</span><span class="n">scipy</span>
<span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">python</span><span class="o">-</span><span class="n">numpy</span>
</pre></div>
</div>
<p>Optional packages (but <strong>highly</strong> recommended):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">python</span><span class="o">-</span><span class="n">matplotlib</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Alternative install</strong>, you can install <em>python-setuptools</em></p>
</div>
<p>and use easy_install to get these packages
(e.g. “<code class="docutils literal notranslate"><span class="pre">easy_install</span> <span class="pre">scipy</span></code>”).</p>
<p>With conda you can install most of the needed dependencies:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

<span class="c1">## required</span>
conda config --add channels bioconda
conda config --add channels conda-forge
conda install -y -q -c bioconda mcl
conda install -y -q future
conda install -y -q h5py
conda install -y -q samtools
conda install -y -q pysam
conda install -y -q matplotlib-base
conda install -y -q scipy

<span class="c1">## optional</span>
conda install -y -q jupyter                                    <span class="c1"># this notebook :)</span>
conda install -y -q -c bioconda sra-tools                      <span class="c1"># to download raw data from released experiment</span>
</pre></div>
</div>
</section>
<section id="imp-3d-modeling">
<h3>IMP - 3D modeling<a class="headerlink" href="#imp-3d-modeling" title="Permalink to this headline">¶</a></h3>
<section id="from-the-repositories">
<h4>From the repositories<a class="headerlink" href="#from-the-repositories" title="Permalink to this headline">¶</a></h4>
<p>Since version 2.5 IMP is available in several repositories, like Ubuntu</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sudo</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">imp</span>
</pre></div>
</div>
<p>or in <a class="reference external" href="http://conda.pydata.org/docs/intro.html">anaconda</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">conda</span><span class="o">.</span><span class="n">anaconda</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">salilab</span> <span class="n">imp</span>
</pre></div>
</div>
<p>These options may be easier than the source compilation.</p>
</section>
<section id="from-source">
<h4>From source<a class="headerlink" href="#from-source" title="Permalink to this headline">¶</a></h4>
<p>Check <a class="reference external" href="https://integrativemodeling.org/download-linux.html">https://integrativemodeling.org/download-linux.html</a></p>
</section>
</section>
<section id="mcl-clustering">
<h3>MCL - clustering<a class="headerlink" href="#mcl-clustering" title="Permalink to this headline">¶</a></h3>
<p>MCL is the program used for clustering the 3D models generated by IMP.
It can be downloaded from <a class="reference external" href="http://micans.org/mcl/">http://micans.org/mcl/</a>; on Debian/Ubuntu
machines it can be automatically installed with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sudo</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">mcl</span>
</pre></div>
</div>
<p>or in anaconda <a class="reference external" href="http://conda.pydata.org/docs/intro.html">http://conda.pydata.org/docs/intro.html</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">y</span> <span class="o">-</span><span class="n">q</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">mcl</span>
</pre></div>
</div>
<p><em>Note: if the MCL executable is not found by TADbit, an alternative
clustering method will be used. Nevertheless we strongly recommend to
use MCL.</em></p>
</section>
</section>
<section id="gem-mapper">
<h2>GEM Mapper<a class="headerlink" href="#gem-mapper" title="Permalink to this headline">¶</a></h2>
<p>The default mapper in TADbit is GEM, but bowtie2 and hisat2 are also
supported.</p>
<p>GEM version 3, bowtie2 and hisat2 are available in bioconda.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

conda -y -q -c bioconda gem3-mapper
conda -y -q -c bioconda bowtie2
conda -y -q -c bioconda hisat2
</pre></div>
</div>
<p>If you prefer the old good GEM version 2, go to the download page:
<a class="reference external" href="https://sourceforge.net/projects/gemlibrary/files/gem-library/Binary%20pre-release%202/">https://sourceforge.net/projects/gemlibrary/files/gem-library/Binary%20pre-release%202/</a>
and download the <code class="docutils literal notranslate"><span class="pre">i3</span></code> version (the other version is for older
computers, and you usually won’t have to use it).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

wget -nv -O GEM.tbz2 https://sourceforge.net/projects/gemlibrary/files/gem-library/Binary%20pre-release%203/GEM-binaries-Linux-x86_64-core_i3-20130406-045632.tbz2/download
</pre></div>
</div>
<pre class="ansi-block literal-block">2019-01-15 14:12:09 URL:https://netcologne.dl.sourceforge.net/project/gemlibrary/gem-library/Binary%20pre-release%203/GEM-binaries-Linux-x86_64-core_i3-20130406-045632.tbz2 [33847526/33847526] -&gt; &quot;GEM.tbz2&quot; [1]</pre>
<p>Uncompress the archive:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

tar -xjvf GEM.tbz2
</pre></div>
</div>
<pre class="ansi-block literal-block">GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer_bwt-dna
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/transcriptome-2-genome
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-mappability
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer_generate
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-mappability-retriever
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/External/
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/External/gemtools
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/External/LICENSE
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gemtools
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-rna-mapper
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-retriever
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-2-sam
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer_fasta2meta+cont
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gtf-2-junctions
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-2-wig
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-2-gem
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-info
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/compute-transcriptome
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/LICENSE
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/splits-2-junctions
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-mapper
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-indexer.man
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-mapper.man
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-2-sam.man
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-2-gem.man
GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-mappability.man</pre>
<p>And copy the needed binaries to somewhere in your PATH, like:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

rm -f GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/LICENCE
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

cp GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/* ~/miniconda2/bin/
</pre></div>
</div>
<p>Cleanup</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

rm -rf GEM-binaries-Linux-x86_64-core_i3-20121106-022124
rm -f GEM.tbz2
</pre></div>
</div>
</section>
<section id="dryhic-for-oned-normalization">
<h2>DryHiC for oneD normalization<a class="headerlink" href="#dryhic-for-oned-normalization" title="Permalink to this headline">¶</a></h2>
<p>Install dryhic from: <a class="reference external" href="https://github.com/qenvio/dryhic">https://github.com/qenvio/dryhic</a></p>
<p>From an R console type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s2">&quot;devtools&quot;</span><span class="p">)</span>

<span class="n">devtools</span><span class="p">::</span><span class="n">install_github</span><span class="p">(</span><span class="s2">&quot;qenvio/dryhic&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Or</strong> execute this cell:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

R -e <span class="s1">&#39;</span>
<span class="s1">install.packages(&quot;devtools&quot;, repos=&quot;http://cran.us.r-project.org&quot;); devtools::install_github(&quot;qenvio/dryhic&quot;)&#39;</span>
</pre></div>
</div>
<pre class="ansi-block literal-block">R version 3.4.2 (2017-09-28) -- &quot;Short Summer&quot;
Copyright (C) 2017 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

&gt; install.packages(&quot;devtools&quot;, repos=&quot;http://cran.us.r-project.org&quot;); devtools::install_github(&quot;qenvio/dryhic&quot;, force=TRUE)
gfortran   -fpic  -g -O2 -fstack-protector-strong  -c fwdb.f -o fwdb.o
gfortran   -fpic  -g -O2 -fstack-protector-strong  -c viterbi.f -o viterbi.o
gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o dryhic.so fwdb.o viterbi.o -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR
&gt;
&gt;</pre>
<pre class="ansi-block literal-block">Installing package into ‘/home/fransua/R/x86_64-pc-linux-gnu-library/3.4’
(as ‘lib’ is unspecified)
trying URL 'http://cran.us.r-project.org/src/contrib/devtools_1.13.6.tar.gz'
Content type 'application/x-gzip' length 486446 bytes (475 KB)
==================================================
downloaded 475 KB

* installing *source* package ‘devtools’ ...
** package ‘devtools’ successfully unpacked and MD5 sums checked
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
* DONE (devtools)

The downloaded source packages are in
    ‘/tmp/Rtmp1j1AUG/downloaded_packages’
Downloading GitHub repo qenvio/dryhic&#64;master
from URL https://api.github.com/repos/qenvio/dryhic/zipball/master
Installing dryhic
'/usr/lib/R/bin/R' --no-site-file --no-environ --no-save --no-restore --quiet  \
  CMD INSTALL '/tmp/Rtmp1j1AUG/devtools6acf4c30139e/qenvio-dryhic-764e0f0'  \
  --library='/home/fransua/R/x86_64-pc-linux-gnu-library/3.4' --install-tests

* installing *source* package ‘dryhic’ ...
** libs
installing to /home/fransua/R/x86_64-pc-linux-gnu-library/3.4/dryhic/libs
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (dryhic)</pre>
</section>
<section id="dsrc-fastq-compressor">
<h2>DSRC FASTQ compressor<a class="headerlink" href="#dsrc-fastq-compressor" title="Permalink to this headline">¶</a></h2>
<p>DSRC is a FASTQ compressor, it’s not needed, but we use it as the size
of the files is significantly smaller than using gunzip (&gt;30%), and,
more importantly, the access to them can be parallelized, and is much
faster than any other alternative.</p>
<p>It can be downloaded from <a class="reference external" href="https://github.com/lrog/dsrc">https://github.com/lrog/dsrc</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

wget -nv http://sun.aei.polsl.pl/dsrc/download/2.0rc/dsrc
</pre></div>
</div>
<pre class="ansi-block literal-block">2018-09-28 00:46:48 URL:http://sun.aei.polsl.pl/dsrc/download/2.0rc/dsrc [1761768/1761768] -&gt; &quot;dsrc&quot; [1]</pre>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

chmod +x dsrc
</pre></div>
</div>
<p>And copy to somewhere in your PATH, like:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

mv dsrc ~/miniconda2/bin/
</pre></div>
</div>
<section id="chimera-visualization">
<h3>Chimera - visualization<a class="headerlink" href="#chimera-visualization" title="Permalink to this headline">¶</a></h3>
<p>Chimera is a program used for visualization and analysis of molecular
structures. It is used in TADbit to visualize the generated 3D models.
Chimera is available at: <a class="reference external" href="http://www.cgl.ucsf.edu/chimera/">http://www.cgl.ucsf.edu/chimera/</a></p>
<p><em>This software is only needed for the visualization of 3D models from
inside TADbit.</em></p>
</section>
<section id="liftover">
<h3>LiftOver<a class="headerlink" href="#liftover" title="Permalink to this headline">¶</a></h3>
<p>TADbit provides a wrapper for the LiftOver tool [Fujita2011]_ (download
it from: <a class="reference external" href="http://hgdownload.cse.ucsc.edu/admin/exe/">http://hgdownload.cse.ucsc.edu/admin/exe/</a> ).</p>
<p>This can be used to ease the conversion of genomic TAD coordinates
(e.g.: to align human TADs with mouse TADs).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A ‘chain’ file may also be downloaded. For example, to convert coordinates to hg19, the chain file needed may be found at: <a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/</a></p>
</div>
</section>
</section>
<section id="tadbit">
<h2>TADbit<a class="headerlink" href="#tadbit" title="Permalink to this headline">¶</a></h2>
<p>Once all the needed library/software have been installed, TADbit can be
downloaded, unpacked and installed as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="mi">3</span><span class="n">DGenomes</span><span class="o">/</span><span class="n">tadbit</span><span class="o">/</span><span class="n">archive</span><span class="o">/</span><span class="n">master</span><span class="o">.</span><span class="n">zip</span> <span class="o">-</span><span class="n">O</span> <span class="n">tadbit</span><span class="o">.</span><span class="n">zip</span>
<span class="n">unzip</span> <span class="n">tadbit</span><span class="o">.</span><span class="n">zip</span>
<span class="n">cd</span> <span class="n">tadbit</span><span class="o">-</span><span class="n">master</span>
<span class="n">sudo</span> <span class="n">python</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">install</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>IMP not found problem
If you are under <strong>debian/Ubuntu machines</strong>, and you have
followed the quick tutorial proposed here to install IMP,
you may encounter some warning message stating that IMP is
not installed. This is because root user is not using the
PYTHONPATH declared in your bashrc file. You can perfectly
ignore the warning, or just proceed like this:</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>sudo PYTHONPATH=$PYTHONPATH python setup.py install
</pre></div>
</div>
<p>Finally, run the test script to check that the installation completed
successfully.</p>
<p>To do so, move to the test directory and run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">test</span>
<span class="n">python</span> <span class="n">test_all</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<section id="conda-builds">
<h3>Conda builds<a class="headerlink" href="#conda-builds" title="Permalink to this headline">¶</a></h3>
<p>Alternatively we regularly build a conda package in bioconda. The
packages come without IMP and with gem v3 as the default mapper</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

conda install -c bioconda tadbit
</pre></div>
</div>
<p>If you want to install TADbit with IMP in conda this line will create a
environment with them:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

conda create -n tadbit <span class="nv">python</span><span class="o">=</span><span class="m">3</span>.7 r-base<span class="o">=</span><span class="m">3</span>.6.1 r-essentials<span class="o">=</span><span class="m">3</span>.6 r-devtools imp <span class="nv">samtools</span><span class="o">=</span><span class="m">1</span>.12 <span class="nv">jupyter_client</span><span class="o">=</span><span class="m">6</span>.1 tadbit -c conda-forge -c salilab -c bioconda
</pre></div>
</div>
<p>If you wish then to have the last version of TADbit from github remove
the TADbit conda package leaving the dependencies:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash

conda remove tadbit --force
</pre></div>
</div>
<p>And install TADbit from source as explained before.</p>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installing TADbit on GNU/Linux</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#dependencies">Dependencies</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#conda">Conda</a></li>
<li class="toctree-l3"><a class="reference internal" href="#python-libraries">Python libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#imp-3d-modeling">IMP - 3D modeling</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mcl-clustering">MCL - clustering</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#gem-mapper">GEM Mapper</a></li>
<li class="toctree-l2"><a class="reference internal" href="#dryhic-for-oned-normalization">DryHiC for oneD normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#dsrc-fastq-compressor">DSRC FASTQ compressor</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#chimera-visualization">Chimera - visualization</a></li>
<li class="toctree-l3"><a class="reference internal" href="#liftover">LiftOver</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tadbit">TADbit</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#conda-builds">Conda builds</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to TADbit’s documentation!</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial.html"
                        title="next chapter">Tutorials</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial.html" title="Tutorials"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to TADbit’s documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">TADbit 1.0 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Installing TADbit on GNU/Linux</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Sep 26, 2021.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.2.0.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>