<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Tutorials &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Retrieve HiC dataset from NCBI" href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html" />
    <link rel="prev" title="Installing TADbit on GNU/Linux under conda" href="install.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html" title="Retrieve HiC dataset from NCBI"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installing TADbit on GNU/Linux under conda"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Tadbit 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tutorials">
<h1>Tutorials<a class="headerlink" href="#tutorials" title="Permalink to this headline">¶</a></h1>
<div class="section" id="from-fastq-files-to-interaction-matrices">
<h2>From FASTQ files to interaction matrices<a class="headerlink" href="#from-fastq-files-to-interaction-matrices" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html">Retrieve HiC dataset from NCBI</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html#compression">Compression</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html#cleanup">Cleanup</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html#references">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html">Search for the reference genome</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#mus-musculus-s-reference-genome-sequence"><em>Mus musculus</em>’s reference genome sequence</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#download-from-the-ncbi">Download from the NCBI</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#list-of-chromosomes-contigs">List of chromosomes/contigs</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#sequences-of-each-chromosome-contig">Sequences of each chromosome/contig</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#concatenate-all-contigs-chromosomes-into-single-files">Concatenate all contigs/chromosomes into single files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#creation-of-an-index-file-for-gem-mapper">Creation of an index file for GEM mapper</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#compute-mappability-values-needed-for-bias-specific-normalizations">Compute mappability values needed for bias specific normalizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#cleanup">Cleanup</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_3-Hi-C_quality_check.html">FASTQ format</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_3-Hi-C_quality_check.html#quality-check-before-mapping">Quality check before mapping</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_3-Hi-C_quality_check.html#check-restriction-enzyme-used">Check restriction-enzyme used</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_3-Hi-C_quality_check.html#plot-phred-score-and-ligation-digestion-sites">Plot PHRED score and ligation/digestion sites</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html">Iterative vs fragment-based mapping</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#advantages-of-iterative-mapping">Advantages of iterative mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#advantages-of-fragment-based-mapping">Advantages of fragment-based mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#mapping">Mapping</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#iterative-mapping">Iterative mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#fragment-based-mapping">Fragment-based mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#references">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html">Load mapped read-ends and generate a matrix</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#extract-uniquely-mapped-reads">Extract uniquely mapped reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#keep-only-uniquely-mapped-reads-pairs">Keep only uniquely mapped reads pairs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#quality-check-of-the-hi-c-experiment">Quality check of the Hi-C experiment</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#interaction-count-vs-genomic-separation">Interaction count vs. genomic separation!</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#coverage-per-bin">Coverage per bin</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#first-hi-c-map">First Hi-C map</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#insert-size">Insert size</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#distance-between-mapped-read-ends-classified-by-their-orientation">Distance between mapped read-ends classified by their orientation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#references">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_6-Filtering_mapped_reads.html">Filtering</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_6-Filtering_mapped_reads.html#filter-out-dangling-ends-and-self-circles">Filter out dangling ends and self-circles</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_6-Filtering_mapped_reads.html#apply-filters-on-the-data">Apply filters on the data</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_6-Filtering_mapped_reads.html#save-to-bam">Save to BAM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html">The HiC_data object</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#filter-bins-with-low-interaction-counts">Filter bins with low interaction counts</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#normalization-algorithms">Normalization algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#iterative-correction-and-eigenvector-decomposition-ice">Iterative Correction and Eigenvector decomposition (ICE)</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#vanilla-coverage-normalization">Vanilla coverage normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#square-root-vanilla-coverage-sqrt-normalization">Square root vanilla coverage (SQRT) normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#oned-normalization">OneD normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#comparison">Comparison</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#other-normalizations">Other normalizations</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#best-normalization">Best normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#references">References</a></li>
</ul>
</div>
</div>
<div class="section" id="identification-and-analysis-of-tads-and-compartments">
<h2>Identification and Analysis of TADs and Compartments<a class="headerlink" href="#identification-and-analysis-of-tads-and-compartments" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html">Compartments</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#mouse-b-cells">Mouse B cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#mouse-ips-cells">Mouse iPS cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#compare">Compare</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#tads">TADs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#tad-caller-algorithms">TAD caller algorithms</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#comparison-of-tad-borders">Comparison of TAD borders</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#save-chromosome-object-with-tad-definition">Save Chromosome object (with TAD definition)</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#extra-choosing-the-best-resolution-to-call-tad-borders">Extra: choosing the best resolution to call TAD borders</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html">Comparison of Hi-C experiments</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html#load-previous-data">Load previous data</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html#comparison-between-replicates">Comparison between replicates</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html#comparison-between-cell-types">Comparison between cell types</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html#merge-hi-c-experiments">Merge Hi-C experiments</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html#normalizing-merged-data">Normalizing merged data</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="three-dimentional-modelling">
<h2>Three-dimentional modelling<a class="headerlink" href="#three-dimentional-modelling" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_10-Modeling_parameters_optimization.html">Experimental data assessment and model parameters optimisation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_10-Modeling_parameters_optimization.html#data-preparation">Data preparation</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_10-Modeling_parameters_optimization.html#data-modellability-assessment-via-mmp-score">Data modellability assessment via MMP score</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_10-Modeling_parameters_optimization.html#data-transformation-and-scoring-function">Data Transformation and scoring function</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_10-Modeling_parameters_optimization.html#optimization-of-parameters">Optimization of parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial/tutorial_11-3D_Models_production_and_analysis.html">Modeling a region of the chromatin</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_11-3D_Models_production_and_analysis.html#load-previous-data">Load previous data</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_11-3D_Models_production_and_analysis.html#modeling">Modeling</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_11-3D_Models_production_and_analysis.html#structural-models">Structural Models</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial/tutorial_11-3D_Models_production_and_analysis.html#model-analysis">Model analysis</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installing TADbit on GNU/Linux under conda</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html#software">Software</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html#installing-tadbit-on-gnu-linux">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#from-fastq-files-to-interaction-matrices">From FASTQ files to interaction matrices</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html">Retrieve HiC dataset from NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html">Search for the reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#download-from-the-ncbi">Download from the NCBI</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_2-Preparation_of_the_reference_genome.html#creation-of-an-index-file-for-gem-mapper">Creation of an index file for GEM mapper</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_3-Hi-C_quality_check.html">FASTQ format</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_3-Hi-C_quality_check.html#quality-check-before-mapping">Quality check before mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html">Iterative vs fragment-based mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_4-Mapping.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html">Load mapped read-ends and generate a matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#quality-check-of-the-hi-c-experiment">Quality check of the Hi-C experiment</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_5-Parsing_mapped_reads.html#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_6-Filtering_mapped_reads.html">Filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html">The HiC_data object</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#filter-bins-with-low-interaction-counts">Filter bins with low interaction counts</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#normalization-algorithms">Normalization algorithms</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_7-Bin_filtering_and_normalization.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#identification-and-analysis-of-tads-and-compartments">Identification and Analysis of TADs and Compartments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html">Compartments</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_8-Compartments_and_TADs_detection.html#tads">TADs</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html">Comparison of Hi-C experiments</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_9-Compare_and_merge_Hi-C_experiments.html#merge-hi-c-experiments">Merge Hi-C experiments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#three-dimentional-modelling">Three-dimentional modelling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_10-Modeling_parameters_optimization.html">Experimental data assessment and model parameters optimisation</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorial/tutorial_11-3D_Models_production_and_analysis.html">Modeling a region of the chromatin</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tools.html">TADbit tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference/index.html">Reference Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="install.html"
                        title="previous chapter">Installing TADbit on GNU/Linux under conda</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html"
                        title="next chapter">Retrieve HiC dataset from NCBI</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="tutorial/tutorial_1-Retrieve_published_HiC_datasets.html" title="Retrieve HiC dataset from NCBI"
             >next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installing TADbit on GNU/Linux under conda"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Tadbit 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Jun 21, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>