<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>From FASTQ to matrix &#8212; Tadbit 0.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="HiC_data class" href="reference_hic_data.html" />
    <link rel="prev" title="Reference Guide" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_hic_data.html" title="HiC_data class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Reference Guide"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Reference Guide</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="from-fastq-to-matrix">
<h1>From FASTQ to matrix<a class="headerlink" href="#from-fastq-to-matrix" title="Permalink to this headline">¶</a></h1>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pytadbit.mapping.mapper.get_intersection">
<code class="descclassname">pytadbit.mapping.mapper.</code><code class="descname">get_intersection</code><span class="sig-paren">(</span><em>fname1</em>, <em>fname2</em>, <em>out_path</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/mapper.html#get_intersection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.mapper.get_intersection" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.full_mapper.full_mapping">
<code class="descclassname">pytadbit.mapping.full_mapper.</code><code class="descname">full_mapping</code><span class="sig-paren">(</span><em>gem_index_path</em>, <em>fastq_path</em>, <em>out_map_dir</em>, <em>r_enz=None</em>, <em>frag_map=True</em>, <em>min_seq_len=15</em>, <em>windows=None</em>, <em>add_site=True</em>, <em>clean=False</em>, <em>get_nread=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/full_mapper.html#full_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.full_mapper.full_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Do the mapping</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gem_index_path</strong> &#8211; path to index file created from a reference genome
using gem-index tool</li>
<li><strong>fastq_path</strong> &#8211; PATH to fastq file, either compressed or not.</li>
<li><strong>out_map_dir</strong> &#8211; path to a directory where to store mapped reads in MAP
format .</li>
<li><strong>r_enz</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; name of the restriction enzyme used in the experiment e.g.
HindIII. This is optional if frag_map option is False</li>
<li><strong>frag_map</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; two step mapper, first full length is mapped, then
remaining, unmapped reads, are divided into restriction-enzyme fragments
andeach is mapped.</li>
<li><strong>add_site</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; when splitting the sequence by ligated sites found,
removes the ligation site, and put back the original RE site.</li>
<li><strong>min_seq_len</strong> (<em>15</em>) &#8211; minimum size of a fragment to map</li>
<li><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; <p>tuple of ranges for begining and end of the
mapping. This parameter allows to do classical iterative mapping, e.g.</p>
<blockquote>
<div>windows=((1,25),(1,30),(1,35),(1,40),(1,45),(1,50))</div></blockquote>
<dl class="docutils">
<dt>A unique window can also be passed, for trimming, like this:</dt>
<dd>windows=((1,101),)</dd>
</dl>
</li>
<li><strong>clean</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; remove intermedite files created in temp_dir</li>
<li><strong>nthreads</strong> (<em>4</em>) &#8211; number of threads to use for mapping (number of CPUs)</li>
<li><strong>max_edit_distance</strong> (<em>0.04</em>) &#8211; The maximum number of edit operations allowed
while verifying candidate matches by dynamic programming.</li>
<li><strong>mismatches</strong> (<em>0.04</em>) &#8211; The maximum number of nucleotide substitutions
allowed while mapping each k-mer. It is always guaranteed that, however
other options are chosen, all the matches up to the specified number of
substitutions will be found by the program.</li>
<li><strong>temp_dir</strong> (<em>/tmp</em>) &#8211; important to change. Intermediate FASTQ files will be
written there.</li>
<li><strong>get_nreads</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; returns a list of lists where each element contains
a path and the number of reads processed</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a list of paths to generated outfiles. To be passed to 
<code class="xref py py-func docutils literal"><span class="pre">pytadbit.parsers.map_parser.parse_map()</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="quality-check-and-plotting">
<h2>Quality check and plotting<a class="headerlink" href="#quality-check-and-plotting" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pytadbit.mapping.analyze.load_hic_data_from_reads">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">load_hic_data_from_reads</code><span class="sig-paren">(</span><em>fnam</em>, <em>resolution</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/parsers/hic_parser.html#load_hic_data_from_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.load_hic_data_from_reads" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fnam</strong> &#8211; tsv file with reads1 and reads2</li>
<li><strong>resolution</strong> &#8211; the resolution of the experiment (size of a bin in
bases)</li>
<li><strong>genome_seq</strong> &#8211; a dictionary containing the genomic sequence by
chromosome</li>
<li><strong>get_sections</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; for very very high resolution, when the column
index does not fit in memory</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.hic_map">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">hic_map</code><span class="sig-paren">(</span><em>data</em>, <em>resolution=None</em>, <em>normalized=False</em>, <em>masked=None</em>, <em>by_chrom=False</em>, <em>savefig=None</em>, <em>show=False</em>, <em>savedata=None</em>, <em>focus=None</em>, <em>clim=None</em>, <em>cmap='jet'</em>, <em>pdf=False</em>, <em>decay=True</em>, <em>perc=10</em>, <em>name=None</em>, <em>decay_resolution=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#hic_map"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.hic_map" title="Permalink to this definition">¶</a></dt>
<dd><p>function to retrieve data from HiC-data object. Data can be stored as
a square matrix, or drawn using matplotlib</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>data</strong> &#8211; can be either a path to a file with pre-processed reads
(filtered or not), or a Hi-C-data object</li>
<li><strong>resolution</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; at which to bin the data (try having a dense matrix
with &lt; 10% of cells with zero interaction counts). Note: not necessary
if a hic_data object is passed as &#8216;data&#8217;.</li>
<li><strong>normalized</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; used normalized data, based on precalculated biases</li>
<li><strong>masked</strong> &#8211; a list of columns to be removed. Usually because to few
interactions</li>
<li><strong>by_chrom</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; <p>data can be stored in a partitioned way. This
parameter can take the values of:</p>
<blockquote>
<div><ul>
<li>&#8216;intra&#8217;: one output per each chromosome will be created</li>
<li><dl class="first docutils">
<dt>&#8216;inter&#8217;: one output per each possible pair of chromosome will be</dt>
<dd>created</dd>
</dl>
</li>
<li>&#8216;all&#8217;  : both of the above outputs</li>
</ul>
</div></blockquote>
</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path where to store the output images. Note that, if
the by_chrom option is used, then savefig will be the name of the
directory containing the output files.</li>
<li><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path where to store the output matrices. Note that, if
the by_chrom option is used, then savefig will be the name of the
directory containing the output files.</li>
<li><strong>focus</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; can be either two number (i.e.: (1, 100)) specifying the
start and end position of the sub-matrix to display (start and end, along
the diagonal of the original matrix); or directly a chromosome name; or
two chromosome names (i.e.: focus=(&#8216;chr2, chrX&#8217;)), in order to store the
data corresponding to inter chromosomal interactions between these two
chromosomes</li>
<li><strong>decay</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; plot the correlation between genomic distance and
interactions (usually a decay).</li>
<li><strong>force_image</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; force to generate an image even if resolution is
crazy...</li>
<li><strong>clim</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; cutoff for the upper and lower bound in the coloring scale
of the heatmap</li>
<li><strong>pdf</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; when using the bny_chrom option, to specify the format of
the stored images</li>
<li><strong>cmap</strong> (<em>Reds</em>) &#8211; color map to be used for the heatmap</li>
<li><strong>decay_resolution</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; chromatin fragment size to consider when
calculating decay of the number of interactions with genomic distance.
Default is equal to resolution of the matrix.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.correlate_matrices">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">correlate_matrices</code><span class="sig-paren">(</span><em>hic_data1</em>, <em>hic_data2</em>, <em>max_dist=10</em>, <em>intra=False</em>, <em>axe=None</em>, <em>savefig=None</em>, <em>show=False</em>, <em>savedata=None</em>, <em>normalized=False</em>, <em>remove_bad_columns=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#correlate_matrices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.correlate_matrices" title="Permalink to this definition">¶</a></dt>
<dd><p>Compare the iteractions of two Hi-C matrices at a given distance,
with spearman rank correlation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>hic_data1</strong> &#8211; Hi-C-data object</li>
<li><strong>hic_data2</strong> &#8211; Hi-C-data object</li>
<li><strong>resolution</strong> (<em>1</em>) &#8211; to be used for scaling the plot</li>
<li><strong>max_dist</strong> (<em>10</em>) &#8211; maximum distance from diagonal (e.g. 10 mean we will
not look further than 10 times the resolution)</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to save the plot</li>
<li><strong>intra</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; only takes into account intra-chromosomal contacts</li>
<li><strong>show</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; displays the plot</li>
<li><strong>normalized</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; use normalized data</li>
<li><strong>remove_bads</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; computes the union of bad columns between samples
and exclude them from the comparison</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of correlations and list of genomic distances</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.eig_correlate_matrices">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">eig_correlate_matrices</code><span class="sig-paren">(</span><em>hic_data1</em>, <em>hic_data2</em>, <em>nvect=6</em>, <em>normalized=False</em>, <em>savefig=None</em>, <em>show=False</em>, <em>savedata=None</em>, <em>remove_bad_columns=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#eig_correlate_matrices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.eig_correlate_matrices" title="Permalink to this definition">¶</a></dt>
<dd><p>Compare the iteractions of two Hi-C matrices using their 6 first
eigenvectors, with pearson correlation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>hic_data1</strong> &#8211; Hi-C-data object</li>
<li><strong>hic_data2</strong> &#8211; Hi-C-data object</li>
<li><strong>nvect</strong> (<em>6</em>) &#8211; number of eigenvectors to compare</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to save the plot</li>
<li><strong>show</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; displays the plot</li>
<li><strong>normalized</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; use normalized data</li>
<li><strong>remove_bads</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; computes the union of bad columns between samples
and exclude them from the comparison</li>
<li><strong>kwargs</strong> &#8211; any argument to pass to matplotlib imshow function</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">matrix of correlations</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.plot_distance_vs_interactions">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">plot_distance_vs_interactions</code><span class="sig-paren">(</span><em>data</em>, <em>min_diff=1</em>, <em>max_diff=1000</em>, <em>show=False</em>, <em>genome_seq=None</em>, <em>resolution=None</em>, <em>axe=None</em>, <em>savefig=None</em>, <em>normalized=False</em>, <em>plot_each_cell=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#plot_distance_vs_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.plot_distance_vs_interactions" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> &#8211; input file name, or HiC_data object or list of lists</li>
<li><strong>min_diff</strong> (<em>10</em>) &#8211; lower limit (in number of bins)</li>
<li><strong>max_diff</strong> (<em>1000</em>) &#8211; upper limit (in number of bins) to look for</li>
<li><strong>resolution</strong> (<em>100</em>) &#8211; group reads that are closer than this resolution
parameter</li>
<li><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a matplotlib.axes.Axes object to define the plot
appearance</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Param_hash False plot_each_cell:</th></tr>
<tr class="field-even field"><td>&#160;</td><td class="field-body"><p class="first">if false, only the mean distances by bin
will be represented, otherwise each pair of interactions will be plotted.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">slope, intercept and R square of each of the 3 correlations</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.plot_iterative_mapping">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">plot_iterative_mapping</code><span class="sig-paren">(</span><em>fnam1</em>, <em>fnam2</em>, <em>total_reads=None</em>, <em>axe=None</em>, <em>savefig=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#plot_iterative_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.plot_iterative_mapping" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fnam</strong> &#8211; input file name</li>
<li><strong>total_reads</strong> &#8211; total number of reads in the initial FASTQ file</li>
<li><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a matplotlib.axes.Axes object to define the plot
appearance</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dictionary with the number of reads per mapped length</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.plot_genomic_distribution">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">plot_genomic_distribution</code><span class="sig-paren">(</span><em>fnam</em>, <em>first_read=True</em>, <em>resolution=10000</em>, <em>ylim=None</em>, <em>yscale=None</em>, <em>savefig=None</em>, <em>show=False</em>, <em>savedata=None</em>, <em>chr_names=None</em>, <em>nreads=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#plot_genomic_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.plot_genomic_distribution" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fnam</strong> &#8211; input file name</li>
<li><strong>first_read</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; uses first read.</li>
<li><strong>resolution</strong> (<em>100</em>) &#8211; group reads that are closer than this resolution
parameter</li>
<li><strong>ylim</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a tuple of lower and upper bound for the y axis</li>
<li><strong>yscale</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; if set_bad to &#8220;log&#8221; values will be represented in log2 
scale</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
<li><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path where to store the output read counts per bin.</li>
<li><strong>chr_names</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; can pass a list of chromosome names in case only some
them the need to be plotted (this option may last even more than default)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.insert_sizes">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">insert_sizes</code><span class="sig-paren">(</span><em>fnam</em>, <em>savefig=None</em>, <em>nreads=None</em>, <em>max_size=99.9</em>, <em>axe=None</em>, <em>show=False</em>, <em>xlog=False</em>, <em>stats=('median'</em>, <em>'perc_max')</em>, <em>too_large=10000</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#insert_sizes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.insert_sizes" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the distribution of dangling-ends lengths
:param fnam: input file name
:param None savefig: path where to store the output images.
:param 99.9 max_size: top percentage of distances to consider, within the</p>
<blockquote>
<div>top 0.01% are usually found very long outliers.</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>xlog</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; represent x axis in logarithmic scale</li>
<li><strong>'perc_max'</strong><strong>) </strong><strong>stats</strong> (<em></em><em>(</em><em>'median'</em><em>,</em><em></em>) &#8211; <p>returns this set of values calculated from the
distribution of insert/fragment sizes. Possible values are:</p>
<blockquote>
<div><ul>
<li>&#8216;median&#8217; median of the distribution</li>
<li>&#8216;perc_max&#8217; percentil defined by the other parameter &#8216;max_size&#8217;</li>
<li><dl class="first docutils">
<dt>&#8216;first_deacay&#8217; starting from the median of the distribution to the</dt>
<dd>first window where 10 consecutive insert sizes are counted less than
a given value (this given value is equal to the sum of all
sizes divided by 100 000)</dd>
</dl>
</li>
<li>&#8216;MAD&#8217; Double Median Adjusted Deviation</li>
</ul>
</div></blockquote>
</li>
<li><strong>too_large</strong> (<em>10000</em>) &#8211; upper bound limit for fragment size to consider</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the median value and the percentile inputed as max_size.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.plot_strand_bias_by_distance">
<code class="descclassname">pytadbit.mapping.analyze.</code><code class="descname">plot_strand_bias_by_distance</code><span class="sig-paren">(</span><em>fnam</em>, <em>nreads=None</em>, <em>half_step=20</em>, <em>half_len=2000</em>, <em>full_step=500</em>, <em>full_len=50000</em>, <em>savefig=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#plot_strand_bias_by_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.plot_strand_bias_by_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Classify reads into for categories depending on the strand on which each end 
is mapped, and plots the proportion of each of these categories in function 
of the genomic distance between them.
The four categories are:</p>
<blockquote>
<div><ul class="simple">
<li>Both read-ends mapped in the forward strand</li>
<li>Both read-ends mapped in the reverse strand</li>
<li>First read-end in the forward strand1, second in the reverse</li>
<li>First read-end in the reverse strand1, second in the forward</li>
</ul>
</div></blockquote>
<p>Note: First/second read-ends are according to their genomic coordinates.</p>
<p>The plot is divided in two halves, in order to use different zooms for 
read-ends mapped very close, and read-ends further (by default the first 
half goes from a distance of 0 to 2 kb, and the second from 2 kb to 50 kb).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fnam</strong> &#8211; input file name with the intersection of the two read-ends mapped</li>
<li><strong>nreads</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; number of reads to process (default: all reads)</li>
<li><strong>half_len</strong> (<em>2000</em>) &#8211; limit in the X axis of the first plot</li>
<li><strong>half_step</strong> (<em>20</em>) &#8211; to bin distances between read-ends in the first plot</li>
<li><strong>full_len</strong> (<em>2000</em>) &#8211; limit in the X axis of the second plot</li>
<li><strong>full_step</strong> (<em>20</em>) &#8211; to bin distances between read-ends in the second plot</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path where to store the output images.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.fastq_utils.quality_plot">
<code class="descclassname">pytadbit.utils.fastq_utils.</code><code class="descname">quality_plot</code><span class="sig-paren">(</span><em>fnam</em>, <em>r_enz=None</em>, <em>nreads=None</em>, <em>axe=None</em>, <em>savefig=None</em>, <em>paired=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/utils/fastq_utils.html#quality_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.fastq_utils.quality_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the sequencing quality of a given FASTQ file. If a restrinction enzyme
(RE) name is provided, can also represent the distribution of digested and
undigested RE sites and estimate an expected proportion of dangling-ends.</p>
<p>Proportion of dangling-ends is inferred by counting the number of times a
dangling-end site, is found at the beginning of any of the reads (divided by
the number of reads).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fnam</strong> &#8211; path to FASTQ file</li>
<li><strong>nreads</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; max number of reads to read, not necesary to read all</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
<li><strong>paired</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; is input FASTQ contains both ends</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the percentage of dangling-ends (sensu stricto) and the percentage of
reads with at least a ligation site.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="filtering">
<h2>Filtering<a class="headerlink" href="#filtering" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pytadbit.mapping.filter.filter_reads">
<code class="descclassname">pytadbit.mapping.filter.</code><code class="descname">filter_reads</code><span class="sig-paren">(</span><em>fnam</em>, <em>output=None</em>, <em>max_molecule_length=500</em>, <em>over_represented=0.005</em>, <em>max_frag_size=100000</em>, <em>min_frag_size=100</em>, <em>re_proximity=5</em>, <em>verbose=True</em>, <em>savedata=None</em>, <em>min_dist_to_re=750</em>, <em>fast=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/filter.html#filter_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.filter.filter_reads" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter mapped pair of reads in order to remove experimental artifacts (e.g.
dangling-ends, self-circle, PCR artifacts...)</p>
<dl class="docutils">
<dt>Applied filters are:    </dt>
<dd><dl class="first last docutils">
<dt>1- self-circle <span class="classifier-delimiter">:</span> <span class="classifier">reads are coming from a single RE fragment and</span></dt>
<dd>point to the outside (&#8212;-&lt;===&#8212;===&gt;&#8212;)</dd>
<dt>2- dangling-end <span class="classifier-delimiter">:</span> <span class="classifier">reads are coming from a single RE fragment and</span></dt>
<dd>point to the inside (&#8212;-===&gt;&#8212;&lt;===&#8212;)</dd>
<dt>3- error <span class="classifier-delimiter">:</span> <span class="classifier">reads are coming from a single RE fragment and</span></dt>
<dd>point in the same direction</dd>
<dt>4- extra dangling-end <span class="classifier-delimiter">:</span> <span class="classifier">reads are coming from different RE fragment but</span></dt>
<dd>are close enough (&lt; max_molecule length) and point to the inside</dd>
<dt>5- too close from RES <span class="classifier-delimiter">:</span> <span class="classifier">semi-dangling-end filter, start position of one</span></dt>
<dd>of the read is too close (5 bp by default) from RE cutting site. This
filter is skipped in case read is involved in multi-contact. This
filter may be too conservative for 4bp cutter REs.</dd>
<dt>6- too short <span class="classifier-delimiter">:</span> <span class="classifier">remove reads coming from small restriction less</span></dt>
<dd>than 100 bp (default) because they are comparable to the read length</dd>
<dt>7- too large <span class="classifier-delimiter">:</span> <span class="classifier">remove reads coming from large restriction</span></dt>
<dd>fragments (default: 100 Kb, P &lt; 10-5 to occur in a randomized genome)
as they likely represent poorly assembled or repeated regions</dd>
<dt>8- over-represented <span class="classifier-delimiter">:</span> <span class="classifier">reads coming from the top 0.5% most frequently</span></dt>
<dd>detected restriction fragments, they may be prone to PCR artifacts or
represent fragile regions of the genome or genome assembly errors</dd>
<dt>9- duplicated <span class="classifier-delimiter">:</span> <span class="classifier">the combination of the start positions of the</span></dt>
<dd>reads is repeated -&gt; PCR artifact (only keep one copy)</dd>
<dt>10- random breaks <span class="classifier-delimiter">:</span> <span class="classifier">start position of one of the read is too far (</span></dt>
<dd>more than min_dist_to_re) from RE cutting site. Non-canonical
enzyme activity or random physical breakage of the chromatin.</dd>
</dl>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fnam</strong> &#8211; path to file containing the pair of reads in tsv format, file
generated by <a class="reference internal" href="#pytadbit.mapping.mapper.get_intersection" title="pytadbit.mapping.mapper.get_intersection"><code class="xref py py-func docutils literal"><span class="pre">pytadbit.mapping.mapper.get_intersection()</span></code></a></li>
<li><strong>output</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; PATH where to write files containing IDs of filtered
reads. Uses fnam by default.</li>
<li><strong>max_molecule_length</strong> (<em>500</em>) &#8211; facing reads that are within
max_molecule_length, will be classified as &#8216;extra dangling-ends&#8217;</li>
<li><strong>over_represented</strong> (<em>0.005</em>) &#8211; to remove the very top fragment containing
more reads</li>
<li><strong>max_frag_size</strong> (<em>100000</em>) &#8211; maximum fragment size allowed (fragments should
not span over several bins)</li>
<li><strong>min_frag_size</strong> (<em>100</em>) &#8211; remove fragment that are too short (shorter than
the sequenced read length)</li>
<li><strong>re_proximity</strong> (<em>5</em>) &#8211; should be adjusted according to RE site, to filter
semi-dangling-ends</li>
<li><strong>min_dist_to_re</strong> (<em>750</em>) &#8211; minimum distance the start of a read should be
from a RE site (usually 1.5 times the insert size). Applied in filter 10</li>
<li><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; PATH where to write the number of reads retained by
each filter</li>
<li><strong>fast</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; parallel version, requires 4 CPUs and more RAM memory</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">dicitonary with, as keys, the kind of filter applied, and as values
a set of read IDs to be removed</p>
</td>
</tr>
</tbody>
</table>
<p><em>Note: Filtering is not exclusive, one read can be filtered several times.</em></p>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.filter.apply_filter">
<code class="descclassname">pytadbit.mapping.filter.</code><code class="descname">apply_filter</code><span class="sig-paren">(</span><em>fnam</em>, <em>outfile</em>, <em>masked</em>, <em>filters=None</em>, <em>reverse=False</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/filter.html#apply_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.filter.apply_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new file with reads filtered</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fnam</strong> &#8211; input file path, where non-filtered read are stored</li>
<li><strong>outfile</strong> &#8211; output file path, where filtered read will be stored</li>
<li><strong>masked</strong> &#8211; dictionary given by the
<a class="reference internal" href="#pytadbit.mapping.filter.filter_reads" title="pytadbit.mapping.filter.filter_reads"><code class="xref py py-func docutils literal"><span class="pre">pytadbit.mapping.filter.filter_reads()</span></code></a></li>
<li><strong>filters</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; list of numbers corresponding to the filters we want
to apply (numbers correspond to the keys in the masked dictionary)</li>
<li><strong>reverse</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; if set, the resulting outfile will only contain the
reads filtered, not the valid pairs.</li>
<li><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/2/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; </li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">number of reads kept</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.restriction_enzymes.repaired">
<code class="descclassname">pytadbit.mapping.restriction_enzymes.</code><code class="descname">repaired</code><span class="sig-paren">(</span><em>r_enz</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/restriction_enzymes.html#repaired"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.restriction_enzymes.repaired" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the resulting sequence after reparation of two digested and repaired
ends, marking dangling ends.</p>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.restriction_enzymes.map_re_sites">
<code class="descclassname">pytadbit.mapping.restriction_enzymes.</code><code class="descname">map_re_sites</code><span class="sig-paren">(</span><em>enzyme_name</em>, <em>genome_seq</em>, <em>frag_chunk=100000</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/restriction_enzymes.html#map_re_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.restriction_enzymes.map_re_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>map all restriction enzyme (RE) sites of a given enzyme in a genome.
Position of a RE site is defined as the genomic coordinate of the first
nucleotide after the first cut (genomic coordinate starts at 1).
In the case of HindIII the genomic coordinate is this one:</p>
<dl class="docutils">
<dt>123456 789...</dt>
<dd><div class="first line-block">
<div class="line"><br /></div>
</div>
<p class="last">v</p>
</dd>
</dl>
<p>&#8212;&#8211;A|AGCT T&#8212;&#8212;&#8212;&#8212;&#8211;
&#8212;&#8211;T TCGA|A&#8212;&#8212;&#8212;&#8212;&#8211;</p>
<p>In this example the coordinate of the RE site would be 7.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>enzyme_name</strong> &#8211; name of the enzyme to map (upper/lower case are
important)</li>
<li><strong>genome_seq</strong> &#8211; a dictionary containing the genomic sequence by
chromosome</li>
<li><strong>frag_chunk</strong> (<em>100000</em>) &#8211; in order to optimize the search for nearby RE
sites, each chromosome is splitted into chunks.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.restriction_enzymes.religated">
<code class="descclassname">pytadbit.mapping.restriction_enzymes.</code><code class="descname">religated</code><span class="sig-paren">(</span><em>r_enz</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/mapping/restriction_enzymes.html#religated"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.restriction_enzymes.religated" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the resulting sequence after religation of two digested and repaired
ends.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">From FASTQ to matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="#quality-check-and-plotting">Quality check and plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filtering">Filtering</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="reference_hic_data.html">HiC_data class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_optimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_structuralmodel.html">StructuralModel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Reference Guide</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_hic_data.html"
                        title="next chapter">HiC_data class</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_hic_data.html" title="HiC_data class"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Reference Guide"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.2 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Reference Guide</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Apr 20, 2017.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>