<!doctype html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Alignment class &#8212; TADbit 1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/forkme_nature.css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="IMPoptimizer class" href="reference_modelling_impoptimizer.html" />
    <link rel="prev" title="Boundary Aligner" href="reference_aligner.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_modelling_impoptimizer.html" title="IMPoptimizer class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_aligner.html" title="Boundary Aligner"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">TADbit 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Reference Guide</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Alignment class</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="alignment-class">
<h1>Alignment class<a class="headerlink" href="#alignment-class" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="pytadbit.alignment.Alignment">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">pytadbit.alignment.</span></span><span class="sig-name descname"><span class="pre">Alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">experiments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">consensus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#Alignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment of TAD borders</p>
<dl class="py method">
<dt class="sig sig-object py" id="pytadbit.alignment.Alignment.draw">
<span class="sig-name descname"><span class="pre">draw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">focus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extras</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ymax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ali_colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('grey',)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">savefig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shape</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ellipse'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#Alignment.draw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.Alignment.draw" title="Permalink to this definition">¶</a></dt>
<dd><p>Draw alignments as a plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>focus</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – can pass a tuple (bin_start, bin_stop) to display the
alignment between these genomic bins</p></li>
<li><p><strong>extras</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – list of coordinates (genomic bin) where to draw a
red cross</p></li>
<li><p><strong>ymax</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – limit the y axis up to a given value</p></li>
<li><p><strong>)</strong> (<em>(</em><em>'grey'</em><em>,</em>) – successive colors for alignment</p></li>
<li><p><strong>normalized</strong> (<em>True</em>) – normalized Hi-C count are plotted instead of raw
data.</p></li>
<li><p><strong>shape</strong> (<em>'ellipse'</em>) – which kind of shape to use as schematic
representation of TADs. Implemented: ‘ellipse’, ‘rectangle’,
‘triangle’</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pytadbit.alignment.Alignment.get_column">
<span class="sig-name descname"><span class="pre">get_column</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cond1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#Alignment.get_column"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.Alignment.get_column" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of column responding to a given characteristic.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cond1</strong> – can be either a column number or a function to search for
a given condition that may fulfill boundaries.</p></li>
<li><p><strong>cond2</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – a second function to search for a given condition
that may fulfill boundaries.</p></li>
<li><p><strong>min_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – Number of boundaries in a given column that have to
pass the defined conditions (all boundaries may pass them by default).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of column, each column represented by a tuple which
first element is the column number in the alignment and second
element is the list of boundaries in this column.</p>
</dd>
</dl>
<p><strong>Example</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ali</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="s1">&#39;ALL&#39;</span><span class="p">,</span> <span class="n">crm</span><span class="o">.</span><span class="n">get_alignment</span><span class="p">(</span><span class="s1">&#39;ALL&#39;</span><span class="p">),</span>
                <span class="n">crm</span><span class="o">.</span><span class="n">experiments</span><span class="p">,</span> <span class="n">score</span><span class="p">)</span>
<span class="n">ali</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="c1"># [(3, [&gt;55&lt;, ,&gt;56&lt;, &gt;55&lt;, &gt;58&lt;])]</span>

<span class="c1"># now we want boundaries with high scores</span>
<span class="n">cond1</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">5</span>
<span class="c1"># and we want boundaries to be detected in Experiments exp1 and exp2</span>
<span class="n">cond2</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;exp&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;exp1&#39;</span><span class="p">,</span> <span class="s1">&#39;exp2&#39;</span><span class="p">]</span>

<span class="n">ali</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="n">cond1</span><span class="o">=</span><span class="n">cond1</span><span class="p">,</span> <span class="n">cond2</span><span class="o">=</span><span class="n">cond2</span><span class="p">,</span> <span class="n">min_num</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="c1"># [(33, [&gt;268&lt;, &gt;192&lt;, &gt;-&lt;, &gt;-&lt;]),</span>
<span class="c1">#  (46, [&gt;324&lt;, &gt;323&lt;, &gt;335&lt;, &gt;329&lt;]),</span>
<span class="c1">#  (51, [&gt;348&lt;, &gt;335&lt;, &gt;357&lt;, &gt;347&lt;]),</span>
<span class="c1">#  (56, [&gt;374&lt;, &gt;358&lt;, &gt;383&lt;, &gt;-&lt;]),</span>
<span class="c1">#  (64, [&gt;397&lt;, &gt;396&lt;, &gt;407&lt;, &gt;-&lt;]),</span>
<span class="c1">#  (77, [&gt;444&lt;, &gt;442&lt;, &gt;456&lt;, &gt;-&lt;])]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pytadbit.alignment.Alignment.itercolumns">
<span class="sig-name descname"><span class="pre">itercolumns</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#Alignment.itercolumns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.Alignment.itercolumns" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over columns in the alignment</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pytadbit.alignment.Alignment.iteritems">
<span class="sig-name descname"><span class="pre">iteritems</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#Alignment.iteritems"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.Alignment.iteritems" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over experiment names and aligned boundaries</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pytadbit.alignment.Alignment.itervalues">
<span class="sig-name descname"><span class="pre">itervalues</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#Alignment.itervalues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.Alignment.itervalues" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over experiment names and aligned boundaries</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pytadbit.alignment.Alignment.write_alignment">
<span class="sig-name descname"><span class="pre">write_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xpers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">string</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ftype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ansi'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#Alignment.write_alignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.Alignment.write_alignment" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Print alignment of TAD boundaries between different experiments.</dt><dd><p>Alignments are displayed with colors according to the TADbit
confidence score for each boundary.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>names</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – if None print all experiments</p></li>
<li><p><strong>xpers</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – if None print all experiments</p></li>
<li><p><strong>string</strong> (<em>False</em>) – return string instead of printing</p></li>
<li><p><strong>ftype</strong> (<em>ansi</em>) – display colors in ‘ansi’ or ‘html’ format</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pytadbit.alignment.TAD">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">pytadbit.alignment.</span></span><span class="sig-name descname"><span class="pre">TAD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#TAD"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.TAD" title="Permalink to this definition">¶</a></dt>
<dd><p>Specific class of TADs, used only within Alignment objects.
It is directly inheriting from python dict.
a TAD these keys:</p>
<blockquote>
<div><ul class="simple">
<li><p>‘start’: position of the TAD</p></li>
<li><p>‘end’: position of the TAD</p></li>
<li><p>‘score’: of the prediction of boundary</p></li>
<li><p>‘brk’: same as ‘end’</p></li>
<li><p>‘pos’: in the alignment (column number)</p></li>
<li><p>‘exp’: Experiment this TAD belongs to</p></li>
<li><p>‘index’: of this TAD within all TADs in the Experiment</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pytadbit.alignment.randomization_test">
<span class="sig-prename descclassname"><span class="pre">pytadbit.alignment.</span></span><span class="sig-name descname"><span class="pre">randomization_test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xpers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rnd_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'interpolate'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'reciprocal'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#randomization_test"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.randomization_test" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the probability that original alignment is better than an
alignment of randomized boundaries.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tads</strong> – original TADs of each experiment to align</p></li>
<li><p><strong>distr</strong> – the function to interpolate TAD lengths from probability</p></li>
<li><p><strong>score</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – just to print it when verbose</p></li>
<li><p><strong>num</strong> (<em>1000</em>) – number of random alignment to generate for comparison</p></li>
<li><p><strong>verbose</strong> (<em>False</em>) – to print something nice</p></li>
<li><p><strong>method</strong> (<em>interpolate</em>) – how to generate random tads (alternative is
‘shuffle’). ‘interpolate’ will calculate the distribution of TAD lengths,
and generate a random set of TADs according to this distribution (see
<a class="reference internal" href="#pytadbit.alignment.generate_rnd_tads" title="pytadbit.alignment.generate_rnd_tads"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.alignment.generate_rnd_tads()</span></code></a>). In contrast, the ‘shuffle’
method uses directly the set of observed TADs and shuffle them (see
<a class="reference internal" href="#pytadbit.alignment.generate_shuffle_tads" title="pytadbit.alignment.generate_shuffle_tads"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.alignment.generate_shuffle_tads()</span></code></a>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pytadbit.alignment.generate_rnd_tads">
<span class="sig-prename descclassname"><span class="pre">pytadbit.alignment.</span></span><span class="sig-name descname"><span class="pre">generate_rnd_tads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_len</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#generate_rnd_tads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.generate_rnd_tads" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates random TADs over a chromosome of a given size according to a given
distribution of lengths of TADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome_len</strong> – length of the chromosome</p></li>
<li><p><strong>distr</strong> – function that returns a TAD length depending on a p value</p></li>
<li><p><strong>bin_size</strong> – size of the bin of the Hi-C experiment</p></li>
<li><p><strong>start</strong> (<em>0</em>) – starting position in the chromosome</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of TADs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pytadbit.alignment.generate_shuffle_tads">
<span class="sig-prename descclassname"><span class="pre">pytadbit.alignment.</span></span><span class="sig-name descname"><span class="pre">generate_shuffle_tads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tads</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/alignment.html#generate_shuffle_tads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pytadbit.alignment.generate_shuffle_tads" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a shuffle version of a given list of TADs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tads</strong> – list of TADs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of shuffled TADs</p>
</dd>
</dl>
</dd></dl>

</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_hic_data.html">HiC_data class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impoptimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impmodel.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_aligner.html"
                        title="previous chapter">Boundary Aligner</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_modelling_impoptimizer.html"
                        title="next chapter">IMPoptimizer class</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_modelling_impoptimizer.html" title="IMPoptimizer class"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_aligner.html" title="Boundary Aligner"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">TADbit 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Reference Guide</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Alignment class</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Sep 25, 2021.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.2.0.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>