<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>TAD detection functions &#8212; TADbit 1.0 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="TAD clustering" href="reference_clustering.html" />
    <link rel="prev" title="HiC_data class" href="reference_hic_data.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_clustering.html" title="TAD clustering"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_hic_data.html" title="HiC_data class"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">TADbit 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Reference Guide</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tad-detection-functions">
<h1>TAD detection functions<a class="headerlink" href="#tad-detection-functions" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="pytadbit.tadbit.tadbit">
<code class="sig-prename descclassname">pytadbit.tadbit.</code><code class="sig-name descname">tadbit</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">remove=None</em>, <em class="sig-param">n_cpus=1</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">max_tad_size='max'</em>, <em class="sig-param">no_heuristic=0</em>, <em class="sig-param">use_topdom=False</em>, <em class="sig-param">topdom_window=5</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/tadbit.html#tadbit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.tadbit.tadbit" title="Permalink to this definition">¶</a></dt>
<dd><p>The TADbit algorithm works on raw chromosome interaction count data.
The normalization is neither necessary nor recommended,
since the data is assumed to be discrete counts.</p>
<p>TADbit is a breakpoint detection algorithm that returns the optimal
segmentation of the chromosome under BIC-penalized likelihood. The
model assumes that counts have a Poisson distribution and that the
expected value of the counts decreases like a power-law with the
linear distance on the chromosome. This expected value of the counts
at position (i,j) is corrected by the counts at diagonal positions
(i,i) and (j,j). This normalizes for different restriction enzyme
site densities and ‘mappability’ of the reads in case a bin contains
repeated regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> – a square matrix of interaction counts in the HI-C data or a list
of such matrices for replicated experiments. The counts must be evenly
sampled and not normalized. x might be either a list of list, a path to
a file or a file handler</p></li>
<li><p><strong>norm</strong> (<em>'visibility'</em>) – kind of normalization to use. Choose between
‘visibility’ of ‘Imakaev’</p></li>
<li><p><strong>remove</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – a python list of lists of booleans mapping positively
columns to remove (if None only columns with a 0 in the diagonal will be
removed)</p></li>
<li><p><strong>n_cpus</strong> (<em>1</em>) – The number of CPUs to allocate to TADbit. If
n_cpus=’max’ the total number of CPUs will be used</p></li>
<li><p><strong>max_tad_size</strong> (<em>auto</em>) – an integer defining maximum size of TAD. Default
(auto or max) defines it as the number of rows/columns</p></li>
<li><p><strong>no_heuristic</strong> (<em>False</em>) – whether to use or not some heuristics</p></li>
<li><p><strong>use_topdom</strong> (<em>False</em>) – whether to use TopDom algorithm to find tads or not (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/26704975">http://www.ncbi.nlm.nih.gov/pubmed/26704975</a>, <a class="reference external" href="http://zhoulab.usc.edu/TopDom/">http://zhoulab.usc.edu/TopDom/</a>)</p></li>
<li><p><strong>topdom_window</strong> (<em>5</em>) – the window size for topdom algorithm</p></li>
<li><p><strong>get_weights</strong> (<em>False</em>) – either to return the weights corresponding to the
Hi-C count (weights are a normalization dependent of the count of each
columns)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the <code class="xref py py-func docutils literal notranslate"><span class="pre">list()</span></code> of topologically associated domains’
boundaries, and the corresponding list associated log likelihoods.
If no weights are given, it may also return calculated weights.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pytadbit.tadbit.batch_tadbit">
<code class="sig-prename descclassname">pytadbit.tadbit.</code><code class="sig-name descname">batch_tadbit</code><span class="sig-paren">(</span><em class="sig-param">directory</em>, <em class="sig-param">parser=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/tadbit.html#batch_tadbit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.tadbit.batch_tadbit" title="Permalink to this definition">¶</a></dt>
<dd><p>Use tadbit on directories of data files.
All files in the specified directory will be considered data file. The
presence of non data files will cause the function to either crash or
produce aberrant results.</p>
<p>Each file has to contain the data for a single unit/chromosome. The
files can be separated in sub-directories corresponding to single
experiments or any other organization. Data files that should be
considered replicates have to start with the same characters, until
the character sep. For instance, all replicates of the unit
‘chr1’ should start with ‘chr1_’, using the default value of sep.</p>
<p>The data files are read through read.delim. You can pass options
to read.delim through the list read_options. For instance
if the files have no header, use read_options=list(header=FALSE) and if
they also have row names, read_options=list(header=FALSE, row.names=1).</p>
<p>Other arguments such as max_size, n_CPU and verbose are passed to
<a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><code class="xref py py-func docutils literal notranslate"><span class="pre">tadbit()</span></code></a>.</p>
<p>NOTE: only used externally, not from Chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory</strong> – the directory containing the data files</p></li>
<li><p><strong>kwargs</strong> – arguments passed to <a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><code class="xref py py-func docutils literal notranslate"><span class="pre">tadbit()</span></code></a> function</p></li>
<li><p><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – a parser function that takes file name as input and
returns a tuple representing the matrix of data. Tuple is a
concatenation of column1 + column2 + column3 + …</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <code class="xref py py-func docutils literal notranslate"><span class="pre">list()</span></code> where each element has the name of the
unit/chromosome, and is the output of <a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><code class="xref py py-func docutils literal notranslate"><span class="pre">tadbit()</span></code></a> run on the
corresponding files assumed to be replicates</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pytadbit.tadbit.print_result_r">
<code class="sig-prename descclassname">pytadbit.tadbit.</code><code class="sig-name descname">print_result_r</code><span class="sig-paren">(</span><em class="sig-param">result</em>, <em class="sig-param">write=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/tadbit.html#print_result_r"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.tadbit.print_result_r" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a table summarizing the TADs found by tadbit. This function outputs
something similar to the R function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>result</strong> – the <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a> that returns <a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><code class="xref py py-func docutils literal notranslate"><span class="pre">tadbit()</span></code></a></p></li>
<li><p><strong>write</strong> (<em>True</em>) – print table. If False, returns the string</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>if write is False, returns a string corresponding to the table of
results</p>
</dd>
</dl>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_hic_data.html">HiC_data class</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impoptimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impmodel.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_hic_data.html"
                        title="previous chapter">HiC_data class</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_clustering.html"
                        title="next chapter">TAD clustering</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_clustering.html" title="TAD clustering"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_hic_data.html" title="HiC_data class"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">TADbit 1.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Reference Guide</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Oct 21, 2020.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>