<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Experiment class &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Reference Guide" href="index.html" />
    <link rel="next" title="TAD detection functions" href="reference_tadbit.html" />
    <link rel="prev" title="Chromosome class" href="reference_chromosome.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_tadbit.html" title="TAD detection functions"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_chromosome.html" title="Chromosome class"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="experiment-class">
<h1>Experiment class<a class="headerlink" href="#experiment-class" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="pytadbit.experiment.Experiment">
<em class="property">class </em><tt class="descclassname">pytadbit.experiment.</tt><tt class="descname">Experiment</tt><big>(</big><em>name</em>, <em>resolution</em>, <em>hic_data=None</em>, <em>norm_data=None</em>, <em>tad_def=None</em>, <em>parser=None</em>, <em>no_warn=False</em>, <em>weights=None</em>, <em>conditions=None</em>, <em>identifier=None</em>, <em>cell_type=None</em>, <em>enzyme=None</em>, <em>exp_type='Hi-C'</em>, <em>**kw_descr</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Hi-C experiment.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> &#8211; name of the experiment</li>
<li><strong>resolution</strong> &#8211; the resolution of the experiment (size of a bin in
bases)</li>
<li><strong>identifier</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; some identifier relative to the Hi-C data</li>
<li><strong>cell_type</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; cell type on which the experiment was done</li>
<li><strong>enzyme</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; restriction enzyme used in  the Hi-C experiment</li>
<li><strong>exp_type</strong> (<em>Hi-C</em>) &#8211; name of the experiment used (currently only Hi-C is
supported)</li>
<li><strong>hic_data</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; whether a file or a list of lists corresponding to
the Hi-C data</li>
<li><strong>tad_def</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a file or a dict with precomputed TADs for this
experiment</li>
<li><strong>parser</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; <p>a parser function that returns a tuple of lists 
representing the data matrix and the length of a row/column. With
the file example.tsv:</p>
<div class="highlight-python"><div class="highlight"><pre>chrT_001       chrT_002        chrT_003        chrT_004
chrT_001       629     164     88      105
chrT_002       164     612     175     110
chrT_003       88      175     437     100
chrT_004       105     110     100     278
</pre></div>
</div>
<p>the output of parser(&#8216;example.tsv&#8217;) would be be:
<tt class="docutils literal"><span class="pre">[([629,</span> <span class="pre">164,</span> <span class="pre">88,</span> <span class="pre">105,</span> <span class="pre">164,</span> <span class="pre">612,</span> <span class="pre">175,</span> <span class="pre">110,</span> <span class="pre">88,</span> <span class="pre">175,</span> <span class="pre">437,</span> <span class="pre">100,</span> <span class="pre">105,</span>
<span class="pre">110,</span> <span class="pre">100,</span> <span class="pre">278]),</span> <span class="pre">4]</span></tt></p>
</li>
<li><strong>conditions</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; <tt class="xref py py-func docutils literal"><span class="pre">list()</span></tt> of experimental conditions, e.g. 
the cell type, the enzyme... (i.e.: [&#8216;HindIII&#8217;, &#8216;cortex&#8217;, &#8216;treatment&#8217;]).
This parameter may be used to compare the effect of this conditions on
the TADs</li>
<li><strong>filter_columns</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; filter the columns with unexpectedly high 
content of low values</li>
<li><strong>kw_descr</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; <p>any other argument passed would be stored as
complementary descriptive field. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">exp</span>  <span class="o">=</span> <span class="n">Experiment</span><span class="p">(</span><span class="s">&#39;k562_rep2&#39;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span>
                  <span class="n">identifier</span><span class="o">=</span><span class="s">&#39;SRX015263&#39;</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="s">&#39;K562&#39;</span><span class="p">,</span>
                  <span class="n">enzyme</span><span class="o">=</span><span class="s">&#39;HindIII&#39;</span><span class="p">,</span> <span class="n">cylce</span><span class="o">=</span><span class="s">&#39;synchronized&#39;</span><span class="p">)</span>
<span class="k">print</span> <span class="n">exp</span>

<span class="c"># Experiment k562_rep2:</span>
<span class="c">#    resolution        : 100Kb</span>
<span class="c">#    TADs              : None</span>
<span class="c">#    Hi-C rows         : None</span>
<span class="c">#    normalized        : None</span>
<span class="c">#    identifier        : SRX015263</span>
<span class="c">#    cell type         : K562</span>
<span class="c">#    restriction enzyme: HindIII</span>
<span class="c">#    cylce             : synchronized</span>
</pre></div>
</div>
<p><em>note that these fields may appear in the header of generated out files</em></p>
</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>TODO: doc conditions
TODO: normalization</p>
<dl class="method">
<dt id="pytadbit.experiment.Experiment.filter_columns">
<tt class="descname">filter_columns</tt><big>(</big><em>silent=False</em>, <em>draw_hist=False</em>, <em>savefig=None</em>, <em>diagonal=True</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.filter_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.filter_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Call filtering function, to remove artefactual columns in a given Hi-C
matrix. This function will detect columns with very low interaction
counts; columns passing through a cell with no interaction in the
diagonal; and columns with NaN values (in this case NaN will be replaced
by zero in the original Hi-C data matrix). Filtered out columns will be
stored in the dictionary Experiment._zeros.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>silent</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; does not warn for removed columns</li>
<li><strong>draw_hist</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; shows the distribution of mean values by column
the polynomial fit, and the cut applied.</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
<li><strong>diagonal</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; remove row/columns with zero in the diagonal</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.get_hic_matrix">
<tt class="descname">get_hic_matrix</tt><big>(</big><em>focus=None</em>, <em>diagonal=True</em>, <em>normalized=False</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.get_hic_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.get_hic_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Hi-C matrix.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>focus</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; if a tuple is passed (start, end), wil return a Hi-C
matrix starting at start, and ending at end (all inclusive).</li>
<li><strong>diagonal</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; replace the values in the diagonal by one. Used
for the filtering in order to smooth the distribution of mean values</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Para False normalized:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body"><p class="first">returns normalized data instead of raw Hi-C</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of lists representing the Hi-C data matrix of the
current experiment</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.get_hic_zscores">
<tt class="descname">get_hic_zscores</tt><big>(</big><em>normalized=True</em>, <em>zscored=True</em>, <em>remove_zeros=True</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.get_hic_zscores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.get_hic_zscores" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize the Hi-C raw data. The result will be stored into
the private Experiment._zscore list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>normalized</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; whether to normalize the result using the
weights (see <a class="reference internal" href="#pytadbit.experiment.Experiment.normalize_hic" title="pytadbit.experiment.Experiment.normalize_hic"><tt class="xref py py-func docutils literal"><span class="pre">normalize_hic()</span></tt></a>)</li>
<li><strong>zscored</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; calculate the z-score of the data</li>
<li><strong>remove_zeros</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; remove null interactions. Dangerous, null
interaction are informative.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.load_hic_data">
<tt class="descname">load_hic_data</tt><big>(</big><em>hic_data</em>, <em>parser=None</em>, <em>wanted_resolution=None</em>, <em>data_resolution=None</em>, <em>silent=False</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.load_hic_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.load_hic_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a Hi-C experiment to the Chromosome object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>hic_data</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; whether a file or a list of lists corresponding to
the Hi-C data</li>
<li><strong>name</strong> &#8211; name of the experiment</li>
<li><strong>force</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; overwrite the experiments loaded under the same 
name</li>
<li><strong>parser</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; <p>a parser function that returns a tuple of lists
representing the data matrix and the length of a row/column. 
With the file example.tsv:</p>
<div class="highlight-python"><div class="highlight"><pre>chrT_001   chrT_002        chrT_003        chrT_004
chrT_001   629     164     88      105
chrT_002   86      612     175     110
chrT_003   159     216     437     105
chrT_004   100     111     146     278
</pre></div>
</div>
<p>the output of parser(&#8216;example.tsv&#8217;) would be:
<tt class="docutils literal"><span class="pre">[([629,</span> <span class="pre">86,</span> <span class="pre">159,</span> <span class="pre">100,</span> <span class="pre">164,</span> <span class="pre">612,</span> <span class="pre">216,</span> <span class="pre">111,</span> <span class="pre">88,</span> <span class="pre">175,</span> <span class="pre">437,</span> <span class="pre">146,</span> <span class="pre">105,</span>
<span class="pre">110,</span> <span class="pre">105,</span> <span class="pre">278]),</span> <span class="pre">4]</span></tt></p>
</li>
<li><strong>resolution</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; resolution of the experiment in the file; it
will be adjusted to the resolution of the experiment. By default the
file is expected to contain a Hi-C experiment with the same resolution
as the <tt class="xref py py-class docutils literal"><span class="pre">pytadbit.Experiment</span></tt> created, and no change is made</li>
<li><strong>filter_columns</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; filter the columns with unexpectedly high 
content of low values</li>
<li><strong>silent</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; does not warn for removed columns</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.load_norm_data">
<tt class="descname">load_norm_data</tt><big>(</big><em>norm_data</em>, <em>parser=None</em>, <em>resolution=None</em>, <em>normalization='visibility'</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.load_norm_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.load_norm_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a normalized Hi-C experiment to the Chromosome object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>norm_data</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; whether a file or a list of lists corresponding to
the normalized Hi-C data</li>
<li><strong>name</strong> &#8211; name of the experiment</li>
<li><strong>force</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; overwrite the experiments loaded under the same 
name</li>
<li><strong>parser</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; <p>a parser function that returns a tuple of lists
representing the data matrix and the length of a row/column. 
With the file example.tsv:</p>
<div class="highlight-python"><div class="highlight"><pre>chrT_001   chrT_002        chrT_003        chrT_004
chrT_001   12.5    164     8.8     0.5
chrT_002   8.6     61.2    1.5     1.1
chrT_003   15.9    21.6    3.7     0.5
chrT_004   0.0     1.1     1.6     2.8
</pre></div>
</div>
</li>
<li><strong>resolution</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; resolution of the experiment in the file; it
will be adjusted to the resolution of the experiment. By default the
file is expected to contain a Hi-C experiment with the same resolution
as the <tt class="xref py py-class docutils literal"><span class="pre">pytadbit.Experiment</span></tt> created, and no change is made</li>
<li><strong>filter_columns</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; filter the columns with unexpectedly high 
content of low values</li>
<li><strong>silent</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; does not warn for removed columns</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.load_tad_def">
<tt class="descname">load_tad_def</tt><big>(</big><em>tad_def</em>, <em>weights=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.load_tad_def"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.load_tad_def" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div>Add the Topologically Associated Domains definition detection to Slice</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>tad_def</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a file or a dict with precomputed TADs for this
experiment</li>
<li><strong>name</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; name of the experiment, if None f_name will be used</li>
<li><strong>weights</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; Store information about the weights, corresponding
to the normalization of the Hi-C data (see tadbit function
documentation)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.model_region">
<tt class="descname">model_region</tt><big>(</big><em>start=1</em>, <em>end=None</em>, <em>n_models=5000</em>, <em>n_keep=1000</em>, <em>n_cpus=1</em>, <em>verbose=0</em>, <em>keep_all=False</em>, <em>close_bins=1</em>, <em>outfile=None</em>, <em>config={'maxdist': 600</em>, <em>'upfreq': 0.3</em>, <em>'reference': 'victor corces dataset 2013'</em>, <em>'kforce': 5</em>, <em>'lowfreq': -0.7</em>, <em>'scale': 0.01}</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.model_region"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.model_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates of three-dimentional models using IMP, for a given segment of
chromosome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>start</strong> (<em>1</em>) &#8211; first bin to model (bin number)</li>
<li><strong>end</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; last bin to model (bin number). By default goes to the
last bin.</li>
<li><strong>n_models</strong> (<em>5000</em>) &#8211; number of modes to generate</li>
<li><strong>n_keep</strong> (<em>1000</em>) &#8211; number of models used in the final analysis 
(usually the top 20% of the generated models). The models are ranked
according to their objective function value (the lower the better)</li>
<li><strong>keep_all</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; whether or not to keep the discarded models (if
True, models will be stored under tructuralModels.bad_models)</li>
<li><strong>close_bins</strong> (<em>1</em>) &#8211; number of particles away (i.e. the bin number
difference) a particle pair must be in order to be considered as
neighbors (e.g. 1 means consecutive particles)</li>
<li><strong>n_cpus</strong> &#8211; number of CPUs to use</li>
<li><strong>verbose</strong> (<em>0</em>) &#8211; the information printed can be: nothing (0), the
objective function value the selected models (1), the objective
function value of all the models (2), all the modeling 
information (3)</li>
<li><strong>config</strong> (<em>CONFIG[&#8216;dmel_01&#8217;]</em>) &#8211; <p>a dictionary containing the standard
parameters used to generate the models. The dictionary should
contain the keys kforce, maxdist, upfreq and lowfreq.
Examples can be seen by doing:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">pytadbit.imp.CONFIG</span> <span class="kn">import</span> <span class="n">CONFIG</span>
</pre></div>
</div>
<p>where CONFIG is a dictionarry of dictionnaries to be passed to this
function:</p>
<div class="highlight-python"><div class="highlight"><pre>CONFIG = {
 &#39;dmel_01&#39;: {
     # use these paramaters with the Hi-C data from:
     &#39;reference&#39; : &#39;victor corces dataset 2013&#39;,

     # Force applied to the restraints inferred to neighbor particles
     &#39;kforce&#39;    : 5,

     # Maximum experimental contact distance
     &#39;maxdist&#39;   : 600, # OPTIMIZATION: 500-1200

     # Minimum and maximum thresholds used to decide which experimental values have to be
     # included in the computation of restraints. Z-score values bigger than upfreq
     # and less that lowfreq will be include, whereas all the others will be rejected
     &#39;upfreq&#39;    : 0.3, # OPTIMIZATION: min/max Z-score

     &#39;lowfreq&#39;   : -0.7 # OPTIMIZATION: min/max Z-score

     # How much space (radius in nm) ocupies a nucleotide
     &#39;scale&#39;     : 0.005
     }
 }
</pre></div>
</div>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a <a class="reference internal" href="reference_imp_structuralmodels.html#pytadbit.imp.structuralmodels.StructuralModels" title="pytadbit.imp.structuralmodels.StructuralModels"><tt class="xref py py-class docutils literal"><span class="pre">pytadbit.imp.structuralmodels.StructuralModels</span></tt></a> object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.normalize_hic">
<tt class="descname">normalize_hic</tt><big>(</big><em>factor=1</em>, <em>iterations=0</em>, <em>max_dev=0.1</em>, <em>silent=False</em>, <em>rowsums=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.normalize_hic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.normalize_hic" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize the Hi-C data. This normalization step does the same of
the <a class="reference internal" href="reference_tadbit.html#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.tadbit.tadbit()</span></tt></a> function (default parameters),</p>
<p>It fills the Experiment.norm variable with the Hi-C values divided by
the calculated weight.</p>
<p>The weight of a given cell in column i and row j corresponds to the
square root of the product of the sum of column i by the sum of row
j.</p>
<p>normalization is done according to this formula:</p>
<div class="math">
\[weight_{(I,J)} = \frac{\sum^N_{j=0}\sum^N_{i=0}(matrix(i,j))}
                       {\sum^N_{i=0}(matrix(i,J)) \times \sum^N_{j=0}(matrix(I,j))}\]</div>
<p>with N being the number or rows/columns of the Hi-C matrix in both
cases.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>factor</strong> (<em>1</em>) &#8211; final mean number of normalized interactions wanted
per cell</li>
<li><strong>silent</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; does not warn when overwriting weights</li>
<li><strong>rowsums</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; input a list of rowsums calculated elsewhere</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.optimal_imp_parameters">
<tt class="descname">optimal_imp_parameters</tt><big>(</big><em>start=1, end=None, n_models=500, n_keep=100, n_cpus=1, upfreq_range=(0, 1, 0.1), close_bins=1, lowfreq_range=(-1, 0, 0.1), scale_range=[0.01], maxdist_range=(400, 1400, 100), dcutoff_range=[2], outfile=None, verbose=True, corr='spearman', off_diag=1, savedata=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.optimal_imp_parameters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.optimal_imp_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the optimal set of parameters to be used for the 3D modeling in
IMP.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>start</strong> (<em>1</em>) &#8211; first bin to model (bin number)</li>
<li><strong>end</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; last bin to model (bin number). By default goes to the
last bin.</li>
<li><strong>n_models</strong> (<em>500</em>) &#8211; number of modes to generate</li>
<li><strong>n_keep</strong> (<em>100</em>) &#8211; number of models used in the final analysis (usually
the top 20% of the generated models). The models are ranked
according to their objective function value (the lower the better)</li>
<li><strong>close_bins</strong> (<em>1</em>) &#8211; number of particles away (i.e. the bin number
difference) a particle pair must be in order to be considered as
neighbors (e.g. 1 means consecutive particles)</li>
<li><strong>n_cpus</strong> &#8211; number of CPUs to use</li>
<li><strong>verbose</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; if set to True, information about the distance,
force and Z-score between particles will be printed</li>
<li><strong>lowfreq_range</strong> (<em>(-1,0,0.1)</em>) &#8211; range of lowfreq values to be 
optimized. The last value of the input tuple is the incremental step
for the lowfreq values</li>
<li><strong>upfreq_range</strong> (<em>(0,1,0.1,0.1)</em>) &#8211; range of upfreq values to be
optimized. The last value of the input tuple is the incremental step
for the upfreq values</li>
<li><strong>maxdist_range</strong> (<em>(400,1400,100)</em>) &#8211; upper and lower bounds used to
search for the optimal maximum experimental distance. The last value
of the input tuple is the incremental step for maxdist values</li>
<li><strong>scale_range</strong> (<em>[0.01]</em>) &#8211; upper and lower bounds used to search for
the optimal scale parameter (nm per nucleotide). The last value of
the input tuple is the incremental step for scale parameter values</li>
<li><strong>dcutoff_range</strong> (<em>[2]</em>) &#8211; upper and lower bounds used to search for
the optimal distance cutoff parameter (distance, in number of beads,
from which to consider 2 beads as being close). The last value of the
input tuple is the incremental step for scale parameter values</li>
<li><strong>cutoff</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; distance cutoff (nm) to define whether two particles
are in contact or not, default is 2 times resolution, times scale.</li>
<li><strong>verbose</strong> &#8211; print the results to the standard output</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<dl class="last docutils">
<dt>Each of the <em>_range</em> parameters accept tuples in the form</dt>
<dd><p class="first"><em>(start, end, step)</em>, or a list with the list of values to test.</p>
<dl class="last docutils">
<dt>E.g.:</dt>
<dd><ul class="first last simple">
<li>scale_range=[0.001, 0.005, 0.006] will test these three values.</li>
<li>scale_range=(0.001, 0.005, 0.001) will test the values 0.001,
0.002, 0.003, 0.004 and 0.005</li>
</ul>
</dd>
</dl>
</dd>
</dl>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a tuple containing:<ul class="simple">
<li>a 3D numpy array with the values of the correlations calculated</li>
<li>the range of scale used</li>
<li>the range of maxdist used</li>
<li>the range of upfreq used</li>
<li>the range of lowfreq used</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.print_hic_matrix">
<tt class="descname">print_hic_matrix</tt><big>(</big><em>print_it=True</em>, <em>normalized=False</em>, <em>zeros=False</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.print_hic_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.print_hic_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Hi-C matrix as string</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>print_it</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; Otherwise, returns the string</li>
<li><strong>normalized</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; returns normalized data, instead of raw Hi-C</li>
<li><strong>zeros</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; take into account filtered columns</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of lists representing the Hi-C data matrix of the
current experiment</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.set_resolution">
<tt class="descname">set_resolution</tt><big>(</big><em>resolution</em>, <em>keep_original=True</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.set_resolution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.set_resolution" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new value for the resolution. Copy the original data into
Experiment._ori_hic and replace the Experiment.hic_data
with the data corresponding to new data 
(<tt class="xref py py-func docutils literal"><span class="pre">pytadbit.Chromosome.compare_condition()</span></tt>).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>resolution</strong> &#8211; an integer representing the resolution. This number
must be a multiple of the original resolution, and higher than it</li>
<li><strong>keep_original</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; either to keep or not the original data</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.view">
<tt class="descname">view</tt><big>(</big><em>tad=None</em>, <em>focus=None</em>, <em>paint_tads=False</em>, <em>axe=None</em>, <em>show=True</em>, <em>logarithm=True</em>, <em>normalized=False</em>, <em>relative=True</em>, <em>decorate=True</em>, <em>savefig=None</em>, <em>where='both'</em>, <em>clim=None</em>, <em>cmap='jet'</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.view"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.view" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize the matrix of Hi-C interactions</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>tad</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; <p>a given TAD in the form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">{</span><span class="s">&#39;start&#39;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span>
 <span class="s">&#39;end&#39;</span>  <span class="p">:</span> <span class="n">end</span><span class="p">,</span>
 <span class="s">&#39;brk&#39;</span>  <span class="p">:</span> <span class="n">end</span><span class="p">,</span>
 <span class="s">&#39;score&#39;</span><span class="p">:</span> <span class="n">score</span><span class="p">}</span>
</pre></div>
</div>
<p><strong>Alternatively</strong> a list of the TADs can be passed (all the TADs
between the first and last one passed will be showed. Thus, passing
more than two TADs might be superfluous)</p>
</li>
<li><strong>focus</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a tuple with the start and end positions of the 
region to visualize</li>
<li><strong>paint_tads</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; draw a box around the TADs defined for this
experiment</li>
<li><strong>axe</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; an axe object from matplotlib can be passed in order
to customize the picture</li>
<li><strong>show</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; either to pop-up matplotlib image or not</li>
<li><strong>logarithm</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; show the logarithm values</li>
<li><strong>normalized</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; show the normalized data (weights might have
been calculated previously). <em>Note: white rows/columns may appear in
the matrix displayed; these rows correspond to filtered rows (see</em>
<a class="reference internal" href="reference_utils.html#pytadbit.utils.hic_filtering.hic_filtering_for_modelling" title="pytadbit.utils.hic_filtering.hic_filtering_for_modelling"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.utils.hic_filtering.hic_filtering_for_modelling()</span></tt></a> <em>)</em></li>
<li><strong>relative</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; color scale is relative to the whole matrix of
data, not only to the region displayed</li>
<li><strong>decorate</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; draws color bar, title and axes labels</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
<li><strong>clim</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; tuple with minimum and maximum value range for color
scale. I.e. clim=(-4, 10)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.write_interaction_pairs">
<tt class="descname">write_interaction_pairs</tt><big>(</big><em>fname</em>, <em>normalized=True</em>, <em>zscored=True</em>, <em>diagonal=False</em>, <em>cutoff=None</em>, <em>header=False</em>, <em>true_position=False</em>, <em>uniq=True</em>, <em>remove_zeros=False</em>, <em>focus=None</em>, <em>format='tsv'</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.write_interaction_pairs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.write_interaction_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a tab separated file with all the pairwise interactions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fname</strong> &#8211; file name where to write the  pairwise interactions</li>
<li><strong>zscored</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; computes the z-score of the log10(data)</li>
<li><strong>normalized</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; use the weights to normalize the data</li>
<li><strong>cutoff</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; if defined, only the zscores above the cutoff will
be writen to the file</li>
<li><strong>uniq</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; only writes one representent per interacting pair</li>
<li><strong>true_position</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; if, true writes genomic coordinates,
otherwise, genomic bin.</li>
<li><strong>focus</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; writes interactions between the start and stop bin
passed to this parameter.</li>
<li><strong>format</strong> (<em>&#8216;tsv&#8217;</em>) &#8211; in which to write the file, can be tab separated
(tsv) or JSON (json)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.experiment.Experiment.write_tad_borders">
<tt class="descname">write_tad_borders</tt><big>(</big><em>savedata=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/experiment.html#Experiment.write_tad_borders"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.experiment.Experiment.write_tad_borders" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a table summarizing the TADs found by tadbit. This function outputs
something similar to the R function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>savedata</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the density data
generated (1 column per step + 1 for particle number). If None, print
a table.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_optimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_model.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_chromosome.html"
                        title="previous chapter">Chromosome class</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_tadbit.html"
                        title="next chapter">TAD detection functions</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_tadbit.html" title="TAD detection functions"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_chromosome.html" title="Chromosome class"
             >previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" >Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright .
      Last updated on Feb 10, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>