<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>From FASTQ to matrix &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Reference Guide" href="index.html" />
    <link rel="next" title="Chromosome class" href="reference_chromosome.html" />
    <link rel="prev" title="Reference Guide" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_chromosome.html" title="Chromosome class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Reference Guide"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="from-fastq-to-matrix">
<h1>From FASTQ to matrix<a class="headerlink" href="#from-fastq-to-matrix" title="Permalink to this headline">¶</a></h1>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pytadbit.mapping.mapper.get_intersection">
<tt class="descclassname">pytadbit.mapping.mapper.</tt><tt class="descname">get_intersection</tt><big>(</big><em>fname1</em>, <em>fname2</em>, <em>out_path</em>, <em>verbose=False</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/mapper.html#get_intersection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.mapper.get_intersection" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Merges the two files corresponding to each reads sides. Reads found in both</dt>
<dd>files are merged and written in an output file.</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fname1</strong> &#8211; path to a tab separated file generated by the function
<tt class="xref py py-func docutils literal"><span class="pre">pytadbit.parsers.sam_parser.parse_sam()</span></tt></li>
<li><strong>fname2</strong> &#8211; path to a tab separated file generated by the function
<tt class="xref py py-func docutils literal"><span class="pre">pytadbit.parsers.sam_parser.parse_sam()</span></tt></li>
<li><strong>out_path</strong> &#8211; path to an outfile. It will written in a similar format as
the inputs</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.mapper.iterative_mapping">
<tt class="descclassname">pytadbit.mapping.mapper.</tt><tt class="descname">iterative_mapping</tt><big>(</big><em>gem_index_path</em>, <em>fastq_path</em>, <em>out_sam_path</em>, <em>range_start</em>, <em>range_stop</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/mapper.html#iterative_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.mapper.iterative_mapping" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fastq_path</strong> &#8211; 152 bases first 76 from one end, next 76 from the other
end. Both to be read from left to right.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="quality-check-and-plotting">
<h2>Quality check and plotting<a class="headerlink" href="#quality-check-and-plotting" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pytadbit.mapping.analyze.hic_map">
<tt class="descclassname">pytadbit.mapping.analyze.</tt><tt class="descname">hic_map</tt><big>(</big><em>data</em>, <em>resolution</em>, <em>biases=None</em>, <em>masked=None</em>, <em>by_chrom=False</em>, <em>savefig=None</em>, <em>show=False</em>, <em>savedata=None</em>, <em>focus=None</em>, <em>clim=None</em>, <em>cmap='Reds'</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#hic_map"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.hic_map" title="Permalink to this definition">¶</a></dt>
<dd><p>function to retrieve data from HiC-data object. Data can be stored as
a square matrix, or drawn using matplotlib</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>data</strong> &#8211; can be either a path to a file with pre-processed reads
(filtered or not), or a Hi-C-data object</li>
<li><strong>resolution</strong> &#8211; at which to bin the data (try having a dense matrix
with &lt; 10% of cells with zero interaction counts).</li>
<li><strong>biases</strong> &#8211; a list of biases, one per column, each cell (i, j) will be
divided by the product: BixBj</li>
<li><strong>masked</strong> &#8211; a list of columns to be removed. Usually because to few
interactions</li>
<li><strong>by_chrom</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; <p>data can be stored in a partitioned way. This
parameter can take the values of:</p>
<blockquote>
<div><ul>
<li>&#8216;intra&#8217;: one output per each chromosome will be created</li>
<li><dl class="first docutils">
<dt>&#8216;inter&#8217;: one output per each possible pair of chromosome will be</dt>
<dd>created</dd>
</dl>
</li>
<li>&#8216;all&#8217;  : both of the above outputs</li>
</ul>
</div></blockquote>
</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path where to store the output images. Note that, if
the by_chrom option is used, then savefig will be the name of the
directory containing the output files.</li>
<li><strong>savedata</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path where to store the output matrices. Note that, if
the by_chrom option is used, then savefig will be the name of the
directory containing the output files.</li>
<li><strong>focus</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; can be either two number (i.e.: (1, 100)) specifying the
start and end position of the sub-matrix to display (start and end, along
the diagonal of the original matrix); or directly a chromosome name; or
two chromosome names (i.e.: focus=(&#8216;chr2, chrX&#8217;)), in order to store the
data corresponding to inter chromosomal interactions between these two
chromosomes</li>
<li><strong>clim</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; cutoff for the upper and lower bound in the coloring scale
of the heatmap</li>
<li><strong>cmap</strong> (<em>Reds</em>) &#8211; color map to be used for the heatmap</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.plot_distance_vs_interactions">
<tt class="descclassname">pytadbit.mapping.analyze.</tt><tt class="descname">plot_distance_vs_interactions</tt><big>(</big><em>fnam</em>, <em>min_diff=100</em>, <em>max_diff=1000000</em>, <em>resolution=100</em>, <em>axe=None</em>, <em>savefig=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#plot_distance_vs_interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.plot_distance_vs_interactions" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fnam</strong> &#8211; input file name</li>
<li><strong>min_diff</strong> (<em>100</em>) &#8211; lower limit kn genomic distance (usually equal to read
length)</li>
<li><strong>max_diff</strong> (<em>1000000</em>) &#8211; upper limit in genomic distance to look for</li>
<li><strong>resolution</strong> (<em>100</em>) &#8211; group reads that are closer than this resolution
parameter</li>
<li><strong>axe</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a matplotlib.axes.Axes object to define the plot
appearance</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.plot_iterative_mapping">
<tt class="descclassname">pytadbit.mapping.analyze.</tt><tt class="descname">plot_iterative_mapping</tt><big>(</big><em>fnam1</em>, <em>fnam2</em>, <em>total_reads=None</em>, <em>axe=None</em>, <em>savefig=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#plot_iterative_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.plot_iterative_mapping" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fnam</strong> &#8211; input file name</li>
<li><strong>total_reads</strong> &#8211; total number of reads in the initial FASTQ file</li>
<li><strong>axe</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a matplotlib.axes.Axes object to define the plot
appearance</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dictionary with the number of reads per mapped length</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.analyze.plot_genomic_distribution">
<tt class="descclassname">pytadbit.mapping.analyze.</tt><tt class="descname">plot_genomic_distribution</tt><big>(</big><em>fnam</em>, <em>first_read=True</em>, <em>resolution=10000</em>, <em>axe=None</em>, <em>ylim=None</em>, <em>savefig=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/analyze.html#plot_genomic_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.analyze.plot_genomic_distribution" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fnam</strong> &#8211; input file name</li>
<li><strong>first_read</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; map first read.</li>
<li><strong>resolution</strong> (<em>100</em>) &#8211; group reads that are closer than this resolution
parameter</li>
<li><strong>axe</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a matplotlib.axes.Axes object to define the plot
appearance</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.fastq_utils.quality_plot">
<tt class="descclassname">pytadbit.utils.fastq_utils.</tt><tt class="descname">quality_plot</tt><big>(</big><em>fnam</em>, <em>nreads=None</em>, <em>axe=None</em>, <em>savefig=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/fastq_utils.html#quality_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.fastq_utils.quality_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the qualities</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fnam</strong> &#8211; path to FASTQ file</li>
<li><strong>nreads</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; max number of reads to read, not necesary to read all</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="filtering">
<h2>Filtering<a class="headerlink" href="#filtering" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pytadbit.mapping.filter.filter_reads">
<tt class="descclassname">pytadbit.mapping.filter.</tt><tt class="descname">filter_reads</tt><big>(</big><em>fnam</em>, <em>max_molecule_length=500</em>, <em>over_represented=0.005</em>, <em>max_frag_size=100000</em>, <em>min_frag_size=100</em>, <em>re_proximity=5</em>, <em>verbose=True</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/filter.html#filter_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.filter.filter_reads" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Apply different filters on pair of reads:</dt>
<dd><dl class="first last docutils">
<dt>1- self-circle <span class="classifier-delimiter">:</span> <span class="classifier">reads are comming from a single RE fragment and</span></dt>
<dd>point to the outside (&#8212;-&lt;===&#8212;===&gt;&#8212;)</dd>
<dt>2- dangling-end <span class="classifier-delimiter">:</span> <span class="classifier">reads are comming from a single RE fragment and</span></dt>
<dd>point to the inside (&#8212;-===&gt;&#8212;&lt;===&#8212;)</dd>
<dt>3- extra dangling-end <span class="classifier-delimiter">:</span> <span class="classifier">reads are comming from different RE fragment but</span></dt>
<dd>are close enough (&lt; max_molecule length) and point to the inside</dd>
<dt>4- error <span class="classifier-delimiter">:</span> <span class="classifier">reads are comming from a single RE fragment and</span></dt>
<dd>point in the same direction</dd>
<dt>5- duplicated <span class="classifier-delimiter">:</span> <span class="classifier">the combination of the start positions of the</span></dt>
<dd>reads is repeated -&gt; PCR artifact (only keep one copy)</dd>
<dt>6- too close from RE <span class="classifier-delimiter">:</span> <span class="classifier">start position of one of the read is too close (</span></dt>
<dd>5 bp by default) from RE cutting site. Non-canonical enzyme activity
or random physical breakage of the chromatin.</dd>
<dt>7- too short <span class="classifier-delimiter">:</span> <span class="classifier">remove reads comming from small restriction less</span></dt>
<dd>than 100 bp (default) because they are comparable to the read length</dd>
<dt>8- too large <span class="classifier-delimiter">:</span> <span class="classifier">remove reads comming from large restriction</span></dt>
<dd>fragments (default: 100 Kb, P &lt; 10-5 to occur in a randomized genome)
as they likely represent poorly assembled or repeated regions</dd>
<dt>9- over-represented <span class="classifier-delimiter">:</span> <span class="classifier">reads coming from the top 0.5% most frequently</span></dt>
<dd>detected restriction fragments, they may be prone to PCR artifacts or
represent fragile regions of the genome or genome assembly errors</dd>
</dl>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fnam</strong> &#8211; path to file containing the pair of reads in tsv format, file
generated by <a class="reference internal" href="#pytadbit.mapping.mapper.get_intersection" title="pytadbit.mapping.mapper.get_intersection"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.mapping.mapper.get_intersection()</span></tt></a></li>
<li><strong>max_molecule_length</strong> (<em>500</em>) &#8211; </li>
<li><strong>over_represented</strong> (<em>0.005</em>) &#8211; </li>
<li><strong>max_frag_size</strong> (<em>100000</em>) &#8211; </li>
<li><strong>min_frag_size</strong> (<em>100</em>) &#8211; </li>
<li><strong>re_proximity</strong> (<em>5</em>) &#8211; </li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">dicitonary with, as keys, the kind of filter applied, and as values
a set of read IDs to be removed</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.mapping.filter.apply_filter">
<tt class="descclassname">pytadbit.mapping.filter.</tt><tt class="descname">apply_filter</tt><big>(</big><em>fnam</em>, <em>outfile</em>, <em>masked</em>, <em>filters=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/mapping/filter.html#apply_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.mapping.filter.apply_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new file with reads filtered</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>fnam</strong> &#8211; input file path, where non-filtered read are stored</li>
<li><strong>outfile</strong> &#8211; output file path, where filtered read will be stored</li>
<li><strong>masked</strong> &#8211; dictionary given by the
<a class="reference internal" href="#pytadbit.mapping.filter.filter_reads" title="pytadbit.mapping.filter.filter_reads"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.mapping.filter.filter_reads()</span></tt></a></li>
<li><strong>filters</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; list of numbers corresponding to the filters we want
to apply (numbers correspond to the keys in the masked dictionary)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="">From FASTQ to matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mapping">Mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="#quality-check-and-plotting">Quality check and plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filtering">Filtering</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_optimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_model.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Reference Guide</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_chromosome.html"
                        title="next chapter">Chromosome class</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_chromosome.html" title="Chromosome class"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Reference Guide"
             >previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" >Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright .
      Last updated on Feb 10, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>