<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Chromosome class &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Experiment class" href="reference_experiment.html" />
    <link rel="prev" title="From FASTQ to matrix" href="reference_mapping.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_experiment.html" title="Experiment class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_mapping.html" title="From FASTQ to matrix"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Reference Guide</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="chromosome-class">
<h1>Chromosome class<a class="headerlink" href="#chromosome-class" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="pytadbit.chromosome.Chromosome">
<em class="property">class </em><code class="sig-prename descclassname">pytadbit.chromosome.</code><code class="sig-name descname">Chromosome</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">species=None</em>, <em class="sig-param">assembly=None</em>, <em class="sig-param">experiment_resolutions=None</em>, <em class="sig-param">experiment_tads=None</em>, <em class="sig-param">experiment_hic_data=None</em>, <em class="sig-param">experiment_norm_data=None</em>, <em class="sig-param">experiment_names=None</em>, <em class="sig-param">max_tad_size=inf</em>, <em class="sig-param">chr_len=0</em>, <em class="sig-param">parser=None</em>, <em class="sig-param">centromere_search=False</em>, <em class="sig-param">silent=False</em>, <em class="sig-param">**kw_descr</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>A Chromosome object designed to deal with Topologically Associating Domains
predictions from different experiments, in different cell types for a given
chromosome of DNA, and to compare them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – name of the chromosome (might be a chromosome name for example)</p></li>
<li><p><strong>species</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – species name</p></li>
<li><p><strong>assembly</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – version number of the genomic assembly used</p></li>
<li><p><strong>resolutions</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – list of resolutions corresponding to a list of
experiments passed.</p></li>
<li><p><strong>experiment_hic_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <code class="xref py py-func docutils literal notranslate"><span class="pre">list()</span></code> of paths to files
containing the Hi-C count matrices corresponding to different experiments</p></li>
<li><p><strong>experiment_tads</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <code class="xref py py-func docutils literal notranslate"><span class="pre">list()</span></code> of paths to files
containing the definition of TADs corresponding to different experiments</p></li>
<li><p><strong>experiment_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <code class="xref py py-func docutils literal notranslate"><span class="pre">list()</span></code> of the names of each
experiment</p></li>
<li><p><strong>max_tad_size</strong> (<em>infinite</em>) – maximum TAD size allowed. TADs longer than
this value will not be considered, and size of the corresponding
chromosome size will be reduced accordingly</p></li>
<li><p><strong>chr_len</strong> (<em>0</em>) – size of the DNA chromosome in bp. By default it will be
inferred from the distribution of TADs</p></li>
<li><p><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>a parser function that returns a tuple of lists
representing the data matrix and the length of a row/column. With
the file example.tsv:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chrT_001</span>       <span class="n">chrT_002</span>        <span class="n">chrT_003</span>        <span class="n">chrT_004</span>
<span class="n">chrT_001</span>       <span class="mi">629</span>     <span class="mi">164</span>     <span class="mi">88</span>      <span class="mi">105</span>
<span class="n">chrT_002</span>       <span class="mi">164</span>     <span class="mi">612</span>     <span class="mi">175</span>     <span class="mi">110</span>
<span class="n">chrT_003</span>       <span class="mi">88</span>      <span class="mi">175</span>     <span class="mi">437</span>     <span class="mi">100</span>
<span class="n">chrT_004</span>       <span class="mi">105</span>     <span class="mi">110</span>     <span class="mi">100</span>     <span class="mi">278</span>
</pre></div>
</div>
<p>the output of parser(‘example.tsv’) would be be:
<code class="docutils literal notranslate"><span class="pre">[([629,</span> <span class="pre">164,</span> <span class="pre">88,</span> <span class="pre">105,</span> <span class="pre">164,</span> <span class="pre">612,</span> <span class="pre">175,</span> <span class="pre">110,</span> <span class="pre">88,</span> <span class="pre">175,</span> <span class="pre">437,</span> <span class="pre">100,</span> <span class="pre">105,</span>
<span class="pre">110,</span> <span class="pre">100,</span> <span class="pre">278]),</span> <span class="pre">4]</span></code></p>
</p></li>
<li><p><strong>kw_descr</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>any other argument passed would be stored as
complementary descriptive field. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">crm</span>  <span class="o">=</span> <span class="n">Chromosome</span><span class="p">(</span><span class="s1">&#39;19&#39;</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s1">&#39;mus musculus&#39;</span><span class="p">,</span>
                  <span class="n">subspecies</span><span class="o">=</span><span class="s1">&#39;musculus musculus&#39;</span><span class="p">,</span>
                  <span class="n">skin_color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
<span class="nb">print</span> <span class="n">crm</span>

<span class="c1"># Chromosome 19:</span>
<span class="c1">#    0  experiment loaded:</span>
<span class="c1">#    0  alignment loaded:</span>
<span class="c1">#    species         : mus musculus</span>
<span class="c1">#    assembly version: UNKNOWN</span>
<span class="c1">#    subspecies      : musculus musculus</span>
<span class="c1">#    skin_color      : black</span>
</pre></div>
</div>
<p><em>note that these fields may appear in the header of generated out files</em></p>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Chromosome object</p>
</dd>
</dl>
<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.add_experiment">
<code class="sig-name descname">add_experiment</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">resolution=None</em>, <em class="sig-param">tad_def=None</em>, <em class="sig-param">hic_data=None</em>, <em class="sig-param">norm_data=None</em>, <em class="sig-param">replace=False</em>, <em class="sig-param">parser=None</em>, <em class="sig-param">conditions=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.add_experiment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.add_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a Hi-C experiment to Chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – name of the experiment or of the Experiment object</p></li>
<li><p><strong>resolution</strong> – resolution of the experiment (needed if name is not
an Experiment object)</p></li>
<li><p><strong>hic_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – whether a file or a list of lists corresponding to
the Hi-C data</p></li>
<li><p><strong>tad_def</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a file or a dict with precomputed TADs for this
experiment</p></li>
<li><p><strong>replace</strong> (<em>False</em>) – overwrite the experiments loaded under the same
name</p></li>
<li><p><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>a parser function that returns a tuple of lists
representing the data matrix and the length of a row/column. With
a file example.tsv containing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chrT_001</span>   <span class="n">chrT_002</span>        <span class="n">chrT_003</span>        <span class="n">chrT_004</span>
<span class="n">chrT_001</span>   <span class="mi">629</span>     <span class="mi">164</span>     <span class="mi">88</span>      <span class="mi">105</span>
<span class="n">chrT_002</span>   <span class="mi">164</span>     <span class="mi">612</span>     <span class="mi">175</span>     <span class="mi">110</span>
<span class="n">chrT_003</span>   <span class="mi">88</span>      <span class="mi">175</span>     <span class="mi">437</span>     <span class="mi">100</span>
<span class="n">chrT_004</span>   <span class="mi">105</span>     <span class="mi">110</span>     <span class="mi">100</span>     <span class="mi">278</span>
</pre></div>
</div>
<p>the output of parser(‘example.tsv’) would be:
<code class="docutils literal notranslate"><span class="pre">[([629,</span> <span class="pre">164,</span> <span class="pre">88,</span> <span class="pre">105,</span> <span class="pre">164,</span> <span class="pre">612,</span> <span class="pre">175,</span> <span class="pre">110,</span> <span class="pre">88,</span> <span class="pre">175,</span> <span class="pre">437,</span> <span class="pre">100,</span> <span class="pre">105,</span>
<span class="pre">110,</span> <span class="pre">100,</span> <span class="pre">278]),</span> <span class="pre">4]</span></code></p>
</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.align_experiments">
<code class="sig-name descname">align_experiments</code><span class="sig-paren">(</span><em class="sig-param">names=None</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">randomize=False</em>, <em class="sig-param">rnd_method='interpolate'</em>, <em class="sig-param">rnd_num=1000</em>, <em class="sig-param">get_score=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.align_experiments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.align_experiments" title="Permalink to this definition">¶</a></dt>
<dd><p>Align the predicted boundaries of two different experiments. The
resulting alignment will be stored in the self.experiment list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>names</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – list of names of the experiments to align. If None,
align all</p></li>
<li><p><strong>experiment1</strong> – name of the first experiment to align</p></li>
<li><p><strong>experiment2</strong> – name of the second experiment to align</p></li>
<li><p><strong>penalty</strong> (<em>-0.1</em>) – penalty for inserting a gap in the alignment</p></li>
<li><p><strong>max_dist</strong> (<em>100000</em>) – maximum distance between two boundaries
allowing match (100Kb seems fair with HUMAN chromosomes)</p></li>
<li><p><strong>verbose</strong> (<em>False</em>) – if True, print some information about the
alignments</p></li>
<li><p><strong>randomize</strong> (<em>False</em>) – check the alignment quality by comparing
randomized boundaries over Chromosomes of the same size. This will
return a extra value, the p-value of accepting that the observed
alignment is not better than a random alignment</p></li>
<li><p><strong>get_score</strong> (<em>False</em>) – returns alignemnt object, alignment score and
percentage of identity from one side and from the other</p></li>
<li><p><strong>rnd_method</strong> (<em>interpolate</em>) – by default uses the interpolation of TAD
distribution. The alternative method is ‘shuffle’, where TADs are
simply shuffled</p></li>
<li><p><strong>rnd_num</strong> (<em>1000</em>) – number of randomizations to do</p></li>
<li><p><strong>method</strong> (<em>reciprocal</em>) – if global, Needleman-Wunsch is used to align
(see <a class="reference internal" href="reference_aligner.html#pytadbit.boundary_aligner.globally.needleman_wunsch" title="pytadbit.boundary_aligner.globally.needleman_wunsch"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.boundary_aligner.globally.needleman_wunsch()</span></code></a>);
if reciprocal, a method based on reciprocal closest boundaries is
used (see <a class="reference internal" href="reference_aligner.html#pytadbit.boundary_aligner.reciprocally.reciprocal" title="pytadbit.boundary_aligner.reciprocally.reciprocal"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.boundary_aligner.reciprocally.reciprocal()</span></code></a>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>an alignment object or, if the randomizattion was invoked,
an alignment object, and a list of statistics that are, the alignment
score, the probability that observed alignment performs better than
randoms, the proportion of borders from the first experiment found
aligned in the second experiment and the proportion of borders from
the second experiment found aligned in the first experiment.
Returned calues can be catched like this:</p>
<blockquote>
<div><p>ali = crm.align_experiments()</p>
</div></blockquote>
<p>or, with randomization test:</p>
<blockquote>
<div><p>ali, (score, pval, prop1, prop2) = crm.align_experiments(randomize=True)</p>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.find_tad">
<code class="sig-name descname">find_tad</code><span class="sig-paren">(</span><em class="sig-param">experiments</em>, <em class="sig-param">name=None</em>, <em class="sig-param">n_cpus=1</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">max_tad_size='max'</em>, <em class="sig-param">heuristic=True</em>, <em class="sig-param">batch_mode=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.find_tad"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.find_tad" title="Permalink to this definition">¶</a></dt>
<dd><p>Call the <a class="reference internal" href="reference_tadbit.html#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.tadbit.tadbit()</span></code></a> function to calculate the
position of Topologically Associated Domain boundaries</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>experiment</strong> – A square matrix of interaction counts of Hi-C
data or a list of such matrices for replicated experiments. The
counts must be evenly sampled and not normalized. ‘experiment’
can be either a list of lists, a path to a file or a file handler</p></li>
<li><p><strong>normalized</strong> (<em>True</em>) – if False simple normalization will be computed,
as well as a simple column filtering will be applied (remove columns
where value at the diagonal is null)</p></li>
<li><p><strong>n_cpus</strong> (<em>1</em>) – The number of CPUs to allocate to TADbit. If
n_cpus=’max’ the total number of CPUs will be used</p></li>
<li><p><strong>max_tad_size</strong> (<em>max</em>) – an integer defining the maximum size of a
TAD. Default (auto) defines it as the number of rows/columns</p></li>
<li><p><strong>heuristic</strong> (<em>True</em>) – whether to use or not some heuristics</p></li>
<li><p><strong>batch_mode</strong> (<em>False</em>) – if True, all the experiments will be
concatenated into one for the search of TADs. The resulting TADs
found are stored under the name ‘batch’ plus a concatenation of the
experiment names passed (e.g.: if experiments=[‘exp1’, ‘exp2’], the
name would be: ‘batch_exp1_exp2’).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.get_experiment">
<code class="sig-name descname">get_experiment</code><span class="sig-paren">(</span><em class="sig-param">name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.get_experiment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.get_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch an Experiment according to its name.
This can also be done directly with Chromosome.experiments[name].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name</strong> – name of the experiment to select</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.Experiment</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.get_tad_hic">
<code class="sig-name descname">get_tad_hic</code><span class="sig-paren">(</span><em class="sig-param">tad</em>, <em class="sig-param">x_name</em>, <em class="sig-param">normed=True</em>, <em class="sig-param">matrix_num=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.get_tad_hic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.get_tad_hic" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve the Hi-C data matrix corresponding to a given TAD.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tad</strong> – a given TAD (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>)</p></li>
<li><p><strong>x_name</strong> – name of the experiment</p></li>
<li><p><strong>normed</strong> (<em>True</em>) – if True, normalize the Hi-C data</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Hi-C data matrix for the given TAD</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.iter_tads">
<code class="sig-name descname">iter_tads</code><span class="sig-paren">(</span><em class="sig-param">x_name</em>, <em class="sig-param">normed=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.iter_tads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.iter_tads" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over the TADs corresponding to a given experiment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x_name</strong> – name of the experiment</p></li>
<li><p><strong>normed</strong> (<em>True</em>) – normalize Hi-C data returned</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p>Hi-C data corresponding to each TAD</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.save_chromosome">
<code class="sig-name descname">save_chromosome</code><span class="sig-paren">(</span><em class="sig-param">out_f</em>, <em class="sig-param">fast=True</em>, <em class="sig-param">divide=True</em>, <em class="sig-param">force=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.save_chromosome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.save_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a Chromosome object to a file (it uses <a class="reference external" href="https://docs.python.org/3/library/pickle.html#pickle.load" title="(in Python v3.8)"><code class="xref py py-func docutils literal notranslate"><span class="pre">pickle.load()</span></code></a> from
the <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle" title="(in Python v3.8)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pickle</span></code></a>). Once saved, the object can be loaded with
<a class="reference internal" href="#pytadbit.chromosome.load_chromosome" title="pytadbit.chromosome.load_chromosome"><code class="xref py py-func docutils literal notranslate"><span class="pre">load_chromosome()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>out_f</strong> – path to the file where to store the <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle" title="(in Python v3.8)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pickle</span></code></a>
object</p></li>
<li><p><strong>fast</strong> (<em>True</em>) – if True, skip Hi-C data and weights</p></li>
<li><p><strong>divide</strong> (<em>True</em>) – if True writes two pickles, one with what would
result by using the fast option, and the second with the Hi-C and
weights data. The second file name will be extended by ‘_hic’ (ie:
with out_f=’chromosome12.pik’ we would obtain chromosome12.pik and
chromosome12.pik_hic). When loaded <a class="reference internal" href="#pytadbit.chromosome.load_chromosome" title="pytadbit.chromosome.load_chromosome"><code class="xref py py-func docutils literal notranslate"><span class="pre">load_chromosome()</span></code></a> will
automatically search for both files</p></li>
<li><p><strong>force</strong> (<em>False</em>) – overwrite the existing file</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.set_max_tad_size">
<code class="sig-name descname">set_max_tad_size</code><span class="sig-paren">(</span><em class="sig-param">value</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.set_max_tad_size"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.set_max_tad_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Change the maximum size allowed for TADs. It also applies to the
computed experiments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>value</strong> – an int value (default is 5000000)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.tad_density_plot">
<code class="sig-name descname">tad_density_plot</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">focus=None</em>, <em class="sig-param">extras=None</em>, <em class="sig-param">normalized=True</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">shape='ellipse'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.tad_density_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.tad_density_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Draw an summary of the TAD found in a given experiment and their density
in terms of relative Hi-C interaction count.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – name of the experiment to visualize</p></li>
<li><p><strong>focus</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – can pass a tuple (bin_start, bin_stop) to display the
alignment between these genomic bins</p></li>
<li><p><strong>extras</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – list of coordinates (genomic bin) where to draw a
red cross</p></li>
<li><p><strong>ymax</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – limit the y axis up to a given value</p></li>
<li><p><strong>)</strong> (<em>(</em><em>'grey'</em><em>,</em>) – successive colors for alignment</p></li>
<li><p><strong>normalized</strong> (<em>True</em>) – normalized Hi-C count are plotted instead of raw
data.</p></li>
<li><p><strong>shape</strong> (<em>'ellipse'</em>) – which kind of shape to use as schematic
representation of TADs. Implemented: ‘ellipse’, ‘rectangle’,
‘triangle’</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.chromosome.Chromosome.visualize">
<code class="sig-name descname">visualize</code><span class="sig-paren">(</span><em class="sig-param">names=None</em>, <em class="sig-param">tad=None</em>, <em class="sig-param">focus=None</em>, <em class="sig-param">paint_tads=False</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">show=True</em>, <em class="sig-param">logarithm=True</em>, <em class="sig-param">normalized=False</em>, <em class="sig-param">relative=True</em>, <em class="sig-param">decorate=True</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">clim=None</em>, <em class="sig-param">scale=(8</em>, <em class="sig-param">6)</em>, <em class="sig-param">cmap='jet'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#Chromosome.visualize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.Chromosome.visualize" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize the matrix of Hi-C interactions of a given experiment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>names</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – name of the experiment to visualize, or list of
experiment names. If None, all experiments will be shown</p></li>
<li><p><strong>tad</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>a given TAD in the form:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span>
 <span class="s1">&#39;end&#39;</span>  <span class="p">:</span> <span class="n">end</span><span class="p">,</span>
 <span class="s1">&#39;brk&#39;</span>  <span class="p">:</span> <span class="n">end</span><span class="p">,</span>
 <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="n">score</span><span class="p">}</span>
</pre></div>
</div>
<p><strong>Alternatively</strong> a list of the TADs can be passed (all the TADs
between the first and last one passed will be showed. Thus, passing
more than two TADs might be superfluous)</p>
</p></li>
<li><p><strong>focus</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a tuple with the start and end positions of the
region to visualize</p></li>
<li><p><strong>paint_tads</strong> (<em>False</em>) – draw a box around the TADs defined for this
experiment</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – an axe object from matplotlib can be passed in order
to customize the picture</p></li>
<li><p><strong>show</strong> (<em>True</em>) – either to pop-up matplotlib image or not</p></li>
<li><p><strong>logarithm</strong> (<em>True</em>) – show the logarithm values</p></li>
<li><p><strong>normalized</strong> (<em>True</em>) – show the normalized data (weights might have
been calculated previously). <em>Note: white rows/columns may appear in
the matrix displayed; these rows correspond to filtered rows (see</em>
<a class="reference internal" href="reference_utils.html#pytadbit.utils.hic_filtering.hic_filtering_for_modelling" title="pytadbit.utils.hic_filtering.hic_filtering_for_modelling"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.hic_filtering.hic_filtering_for_modelling()</span></code></a> <em>)</em></p></li>
<li><p><strong>relative</strong> (<em>True</em>) – color scale is relative to the whole matrix of
data, not only to the region displayed</p></li>
<li><p><strong>decorate</strong> (<em>True</em>) – draws color bar, title and axes labels</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>clim</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – tuple with minimum and maximum value range for color
scale. I.e. clim=(-4, 10)</p></li>
<li><p><strong>cmap</strong> (<em>'jet'</em>) – color map from matplotlib. Can also be a
preconfigured cmap object.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<div class="section" id="load-chromosome">
<h2>Load chromosome<a class="headerlink" href="#load-chromosome" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pytadbit.chromosome.load_chromosome">
<code class="sig-prename descclassname">pytadbit.chromosome.</code><code class="sig-name descname">load_chromosome</code><span class="sig-paren">(</span><em class="sig-param">in_f</em>, <em class="sig-param">fast=2</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#load_chromosome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.load_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a Chromosome object from a file. A Chromosome object can be saved with
the <a class="reference internal" href="#pytadbit.chromosome.Chromosome.save_chromosome" title="pytadbit.chromosome.Chromosome.save_chromosome"><code class="xref py py-func docutils literal notranslate"><span class="pre">Chromosome.save_chromosome()</span></code></a> function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>in_f</strong> – path to a saved Chromosome object file</p></li>
<li><p><strong>fast</strong> (<em>2</em>) – if fast=2 do not load the Hi-C data (in the case that they
were saved in a separate file see <a class="reference internal" href="#pytadbit.chromosome.Chromosome.save_chromosome" title="pytadbit.chromosome.Chromosome.save_chromosome"><code class="xref py py-func docutils literal notranslate"><span class="pre">Chromosome.save_chromosome()</span></code></a>).
If fast is equal to 1, the weights will be skipped from load to save
memory. Finally if fast=0, both the weights and Hi-C data will be loaded</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a Chromosome object</p>
</dd>
</dl>
<p>TODO: remove first try/except type error… this is loading old experiments</p>
</dd></dl>

</div>
<div class="section" id="experimentlist-class">
<h2>ExperimentList class<a class="headerlink" href="#experimentlist-class" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pytadbit.chromosome.ExperimentList">
<em class="property">class </em><code class="sig-prename descclassname">pytadbit.chromosome.</code><code class="sig-name descname">ExperimentList</code><span class="sig-paren">(</span><em class="sig-param">thing</em>, <em class="sig-param">crm</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#ExperimentList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.ExperimentList" title="Permalink to this definition">¶</a></dt>
<dd><p>Inherited from python built in <code class="xref py py-func docutils literal notranslate"><span class="pre">list()</span></code>, modified for TADbit
<code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.Experiment</span></code>.</p>
<p>Mainly, <cite>getitem</cite>, <cite>setitem</cite>, and <cite>append</cite> were modified in order to
be able to search for experiments by index or by name, and to add
experiments simply using Chromosome.experiments.append(Experiment).</p>
<p>The whole ExperimentList object is linked to a Chromosome instance
(<code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.Chromosome</span></code>).</p>
<dl class="method">
<dt id="pytadbit.chromosome.ExperimentList.append">
<code class="sig-name descname">append</code><span class="sig-paren">(</span><em class="sig-param">exp</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/chromosome.html#ExperimentList.append"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.chromosome.ExperimentList.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append object to the end of the list.</p>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Chromosome class</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#load-chromosome">Load chromosome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#experimentlist-class">ExperimentList class</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impoptimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impmodel.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_mapping.html"
                        title="previous chapter">From FASTQ to matrix</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_experiment.html"
                        title="next chapter">Experiment class</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_experiment.html" title="Experiment class"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_mapping.html" title="From FASTQ to matrix"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Reference Guide</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Feb 07, 2020.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>