<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tadbit functions &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Reference Guide" href="index.html" />
    <link rel="next" title="TAD clustering" href="reference_clustering.html" />
    <link rel="prev" title="Experiment class" href="reference_experiment.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_clustering.html" title="TAD clustering"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_experiment.html" title="Experiment class"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="tadbit-functions">
<h1>Tadbit functions<a class="headerlink" href="#tadbit-functions" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="pytadbit.tadbit.tadbit">
<tt class="descclassname">pytadbit.tadbit.</tt><tt class="descname">tadbit</tt><big>(</big><em>x</em>, <em>n_cpus=1</em>, <em>verbose=True</em>, <em>max_tad_size='max'</em>, <em>no_heuristic=False</em>, <em>get_weights=False</em>, <em>use_visibility=False</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/tadbit.html#tadbit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.tadbit.tadbit" title="Permalink to this definition">¶</a></dt>
<dd><p>The TADBit algorithm works on raw chromosome interaction count data.
The normalization is neither necessary nor recommended,
since the data is assumed to be discrete counts.</p>
<p>TADBit is a breakpoint detection algorithm that returns the optimal
segmentation of the chromosome under BIC-penalized likelihood. The
model assumes that counts have a Poisson distribution and that the
expected value of the counts decreases like a power-law with the
linear distance on the chromosome. This expected value of the counts
at position (i,j) is corrected by the counts at diagonal positions
(i,i) and (j,j). This normalizes for different restriction enzynme
site densities and &#8216;mappability&#8217; of the reads in case a bin contains
repeated regions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>x</strong> &#8211; a square matrix of interaction counts in the HI-C data or a list
of such matrices for replicated experiments. The counts must be evenly
sampled and not normalized. x might be either a list of list, a path to
a file or a file handler</li>
<li><strong>n_cpus</strong> (<em>1</em>) &#8211; The number of CPUs to allocate to TADBit. If
n_cpus=&#8217;max&#8217; the total number of CPUs will be used</li>
<li><strong>max_tad_size</strong> (<em>auto</em>) &#8211; an integer defining maximum size of TAD. Default
(auto) defines it as the number of rows/columns</li>
<li><strong>no_heuristic</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; whether to use or not some heuristics</li>
<li><strong>get_weights</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; either to return the weights corresponding to the
Hi-C count (weights are a normalization dependent of the count of each
columns)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the <a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><tt class="xref py py-func docutils literal"><span class="pre">list()</span></tt></a> of topologically associated domains&#8217;
boundaries, and the corresponding list associated log likelihoods.
Depending on the value of the get_weights parameter, may also return
weights.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.tadbit.batch_tadbit">
<tt class="descclassname">pytadbit.tadbit.</tt><tt class="descname">batch_tadbit</tt><big>(</big><em>directory</em>, <em>parser=None</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/tadbit.html#batch_tadbit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.tadbit.batch_tadbit" title="Permalink to this definition">¶</a></dt>
<dd><p>Use tadbit on directories of data files.
All files in the specified directory will be considered data file. The
presence of non data files will cause the function to either crash or
produce aberrant results.</p>
<p>Each file has to contain the data for a single unit/chromosome. The
files can be separated in sub-directories corresponding to single
experiments or any other organization. Data files that should be
considered replicates have to start with the same characters, until
the character sep. For instance, all replicates of the unit
&#8216;chr1&#8217; should start with &#8216;chr1_&#8217;, using the default value of sep.</p>
<p>The data files are read through read.delim. You can pass options
to read.delim through the list read_options. For instance
if the files have no header, use read_options=list(header=FALSE) and if
they also have row names, read_options=list(header=FALSE, row.names=1).</p>
<p>Other arguments such as max_size, n_CPU and verbose are passed to
<a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><tt class="xref py py-func docutils literal"><span class="pre">tadbit()</span></tt></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>directory</strong> &#8211; the directory containing the data files</li>
<li><strong>kwargs</strong> &#8211; arguments passed to <a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><tt class="xref py py-func docutils literal"><span class="pre">tadbit()</span></tt></a> function</li>
<li><strong>parser</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a parser function that takes file name as input and
returns a tuple representing the matrix of data. Tuple is a
concatenation of column1 + column2 + column3 + ...</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A <a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><tt class="xref py py-func docutils literal"><span class="pre">list()</span></tt></a> where each element has the name of the
unit/chromosome, and is the output of <a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><tt class="xref py py-func docutils literal"><span class="pre">tadbit()</span></tt></a> run on the
corresponding files assumed to be replicates</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.tadbit.print_result_r">
<tt class="descclassname">pytadbit.tadbit.</tt><tt class="descname">print_result_r</tt><big>(</big><em>result</em>, <em>write=True</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/tadbit.html#print_result_r"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.tadbit.print_result_r" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a table summarizing the TADs found by tadbit. This function outputs
something similar to the R function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>result</strong> &#8211; the <a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><tt class="xref py py-class docutils literal"><span class="pre">dict</span></tt></a> that returns <a class="reference internal" href="#pytadbit.tadbit.tadbit" title="pytadbit.tadbit.tadbit"><tt class="xref py py-func docutils literal"><span class="pre">tadbit()</span></tt></a></li>
<li><strong>write</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; print table. If False, returns the string</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">if write is False, returns a string corresponding to the table of
results</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADBit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">Tadbit functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp.html">Generate a 3D models</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_experiment.html"
                        title="previous chapter">Experiment class</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_clustering.html"
                        title="next chapter">TAD clustering</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_clustering.html" title="TAD clustering"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_experiment.html" title="Experiment class"
             >previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" >Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright .
      Last updated on Nov 07, 2013.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>