<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>IMPoptimizer class &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="IMPmodel class" href="reference_modelling_impmodel.html" />
    <link rel="prev" title="Alignment class" href="reference_boundary_alignment.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_modelling_impmodel.html" title="IMPmodel class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_boundary_alignment.html" title="Alignment class"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Reference Guide</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="impoptimizer-class">
<h1>IMPoptimizer class<a class="headerlink" href="#impoptimizer-class" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer">
<em class="property">class </em><code class="descclassname">pytadbit.modelling.impoptimizer.</code><code class="descname">IMPoptimizer</code><span class="sig-paren">(</span><em>experiment</em>, <em>start</em>, <em>end</em>, <em>n_models=500</em>, <em>n_keep=100</em>, <em>close_bins=1</em>, <em>container=None</em>, <em>single_particle_restraints=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer" title="Permalink to this definition">¶</a></dt>
<dd><p>This class optimizes a set of parameters (scale, kbending, maxdist, lowfreq, and
upfreq) in order to maximize the correlation between the contact matrix computed on
the generted models (generated by IMP, or lammps) and the input contact map.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>experiment</strong> – an instance of the class pytadbit.experiment.Experiment</li>
<li><strong>start</strong> – first bin to model (bin number, inclusive and starting at 1)</li>
<li><strong>end</strong> – last bin to model (bin number)</li>
<li><strong>n_models</strong> (<em>5000</em>) – number of models to generate</li>
<li><strong>n_keep</strong> (<em>1000</em>) – number of models to use in the final analysis (usually
the top 20% of the generated models). The models are ranked according to
their objective function value (the lower the better)</li>
<li><strong>close_bins</strong> (<em>1</em>) – number of particles away (i.e. the bin number difference) a
particle pair must be in order to be considered as neighbors (e.g. 1 means
nearest neighbors particles)</li>
<li><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – distance cutoff (nm) to define whether two particles
are in contact or not in the models, default is 2.0 * resolution * scale.</li>
<li><strong>container</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – restrains particle to be within a given object.
NOTE: The container can only be a ‘cylinder’ of a given height closed
by hemispheres. This cylinder is defined by its radius, its height (if
height=0 the container is a sphere), and the strength (k) of the harmonic
force applied in the restraint. For example, to model the confinement
in E. coli (2 micrometers of length, and 0.5 micrometer of width),
container = [‘cylinder’, 250, 1500, 50] should be used, and
in a typical spherical mammalian nuclei (about 6 micrometers of diameter),
container = [‘cylinder’, 3000, 0, 50]</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.get_best_parameters_dict">
<code class="descname">get_best_parameters_dict</code><span class="sig-paren">(</span><em>reference=None</em>, <em>with_corr=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.get_best_parameters_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.get_best_parameters_dict" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – a description of the dataset optimized</li>
<li><strong>with_corr</strong> (<em>False</em>) – if True, returns also the correlation value</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dict that can be used for modelling, see config parameter in
<a class="reference internal" href="reference_experiment.html#pytadbit.experiment.Experiment.model_region" title="pytadbit.experiment.Experiment.model_region"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.experiment.Experiment.model_region()</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.load_from_file">
<code class="descname">load_from_file</code><span class="sig-paren">(</span><em>f_name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.load_from_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.load_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the optimized parameters from a file generated with the function:
pytadbit.modelling.impoptimizer.IMPoptimizer.write_result.
This function does not overwrite the parameters that were already
loaded or calculated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>f_name</strong> – file name with the absolute path</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.load_from_file_OLD">
<code class="descname">load_from_file_OLD</code><span class="sig-paren">(</span><em>f_name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.load_from_file_OLD"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.load_from_file_OLD" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the optimized parameters from a file generated with the function:
pytadbit.modelling.impoptimizer.IMPoptimizer.write_result.
This function does not overwrite the parameters that were already
loaded or calculated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>f_name</strong> – file name with the absolute path</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.load_grid_search">
<code class="descname">load_grid_search</code><span class="sig-paren">(</span><em>filenames</em>, <em>corr='spearman'</em>, <em>off_diag=1</em>, <em>verbose=True</em>, <em>n_cpus=1</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.load_grid_search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.load_grid_search" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads one file or a list of files containing pre-calculated Structural
Models (keep_models parameter used). And correlate each set of models
with real data. Useful to run different correlation on the same data
avoiding to re-calculate each time the models.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>filenames</strong> – either a path to a file or a list of paths.</li>
<li><strong>corr</strong> (<em>spearman</em>) – correlation coefficient to use</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>‘param 1 off_diag:
:param True verbose: print the results to the standard output</p>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.load_grid_search_OLD">
<code class="descname">load_grid_search_OLD</code><span class="sig-paren">(</span><em>filenames</em>, <em>corr='spearman'</em>, <em>off_diag=1</em>, <em>verbose=True</em>, <em>n_cpus=1</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.load_grid_search_OLD"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.load_grid_search_OLD" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads one file or a list of files containing pre-calculated Structural
Models (keep_models parameter used). And correlate each set of models
with real data. Useful to run different correlation on the same data
avoiding to re-calculate each time the models.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>filenames</strong> – either a path to a file or a list of paths.</li>
<li><strong>corr</strong> (<em>spearman</em>) – correlation coefficient to use</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>‘param 1 off_diag:
:param True verbose: print the results to the standard output</p>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.plot_2d">
<code class="descname">plot_2d</code><span class="sig-paren">(</span><em>axes=('scale'</em>, <em>'kbending'</em>, <em>'maxdist'</em>, <em>'lowfreq'</em>, <em>'upfreq')</em>, <em>show_best=0</em>, <em>skip=None</em>, <em>savefig=None</em>, <em>clim=None</em>, <em>cmap='inferno'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.plot_2d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.plot_2d" title="Permalink to this definition">¶</a></dt>
<dd><p>A grid of heatmaps representing the result of the optimization.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>axes</strong> (<em>'scale'</em><em>,</em><em>'kbending'</em><em>,</em><em>'maxdist'</em><em>,</em><em>'lowfreq'</em><em>,</em><em>'upfreq'</em>) – list of
axes to be represented in the plot. The order will define which
parameter will be placed on the x, y, z or w axe.</li>
<li><strong>show_best</strong> (<em>0</em>) – number of best correlation values to highlight in
the plot</li>
<li><strong>skip</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – if passed (as a dictionary), fix a given axe,
e.g.: {‘scale’: 0.001, ‘maxdist’: 500}</li>
<li><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
<li><strong>clim</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – color scale. If None, the max and min values of the input are used.</li>
<li><strong>cmap</strong> (<em>inferno</em>) – matplotlib colormap</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.run_grid_search">
<code class="descname">run_grid_search</code><span class="sig-paren">(</span><em>scale_range=0.01</em>, <em>kbending_range=0.0</em>, <em>maxdist_range=(400</em>, <em>1500</em>, <em>100)</em>, <em>lowfreq_range=(-1</em>, <em>0</em>, <em>0.1)</em>, <em>upfreq_range=(0</em>, <em>1</em>, <em>0.1)</em>, <em>dcutoff_range=2</em>, <em>corr='spearman'</em>, <em>off_diag=1</em>, <em>savedata=None</em>, <em>n_cpus=1</em>, <em>verbose=True</em>, <em>use_HiC=True</em>, <em>use_confining_environment=True</em>, <em>use_excluded_volume=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.run_grid_search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.run_grid_search" title="Permalink to this definition">¶</a></dt>
<dd><p>This function calculates the correlation between the models generated
by IMP and the input data for the four main IMP parameters (scale,
kbending, maxdist, lowfreq and upfreq) in the given ranges of values.
The range can be expressed as a list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>n_cpus</strong> – number of CPUs to use</li>
<li><strong>scale_range</strong> (<em>0.01</em>) – upper and lower bounds used to search for
the optimal scale parameter (unit nm per nucleotide). The last value of
the input tuple is the incremental step for scale parameter values</li>
<li><strong>kbending_range</strong> (<em>(</em><em>0</em><em>,</em><em>2.0</em><em>,</em><em>0.5</em><em>)</em>) – values of the bending rigidity
strength to enforce in the models</li>
<li><strong>maxdist_range</strong> (<em>(</em><em>400</em><em>,</em><em>1400</em><em>,</em><em>100</em><em>)</em>) – upper and lower bounds
used to search for the optimal maximum experimental distance.
The last value of the input tuple is the incremental step for maxdist
values</li>
<li><strong>lowfreq_range</strong> (<em>(</em><em>-1</em><em>,</em><em>0</em><em>,</em><em>0.1</em><em>)</em>) – range of lowfreq values to be
optimized. The last value of the input tuple is the incremental
step for the lowfreq values. To be precise “freq” refers to the
Z-score.</li>
<li><strong>upfreq_range</strong> (<em>(</em><em>0</em><em>,</em><em>1</em><em>,</em><em>0.1</em><em>)</em>) – range of upfreq values to be optimized.
The last value of the input tuple is the incremental step for the
upfreq values. To be precise “freq” refers to the Z-score.</li>
<li><strong>dcutoff_range</strong> (<em>2</em>) – upper and lower bounds used to search for
the optimal distance cutoff parameter (distance, in number of beads,
from which to consider 2 beads as being close). The last value of the
input tuple is the incremental step for scale parameter values.</li>
<li><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – concatenate all generated models into a dictionary
and save it into a file named by this argument</li>
<li><strong>verbose</strong> (<em>True</em>) – print the results to the standard output</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.impoptimizer.IMPoptimizer.write_result">
<code class="descname">write_result</code><span class="sig-paren">(</span><em>f_name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/impoptimizer.html#IMPoptimizer.write_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.impoptimizer.IMPoptimizer.write_result" title="Permalink to this definition">¶</a></dt>
<dd><p>This function writes a log file of all the values tested for each
parameter, and the resulting correlation value.</p>
<p>This file can be used to load or merge data a posteriori using
the function pytadbit.modelling.impoptimizer.IMPoptimizer.load_from_file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>f_name</strong> – file name with the absolute path</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impmodel.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_boundary_alignment.html"
                        title="previous chapter">Alignment class</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_modelling_impmodel.html"
                        title="next chapter">IMPmodel class</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_modelling_impmodel.html" title="IMPmodel class"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_boundary_alignment.html" title="Alignment class"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Reference Guide</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Jun 21, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>