<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Parsers &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Boundary Aligner" href="reference_aligner.html" />
    <link rel="prev" title="TAD clustering" href="reference_clustering.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_aligner.html" title="Boundary Aligner"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_clustering.html" title="TAD clustering"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Reference Guide</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="parsers">
<h1>Parsers<a class="headerlink" href="#parsers" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="pytadbit.parsers.hic_parser.read_matrix">
<code class="sig-prename descclassname">pytadbit.parsers.hic_parser.</code><code class="sig-name descname">read_matrix</code><span class="sig-paren">(</span><em class="sig-param">things</em>, <em class="sig-param">parser=None</em>, <em class="sig-param">hic=True</em>, <em class="sig-param">resolution=1</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/parsers/hic_parser.html#read_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.parsers.hic_parser.read_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Read and checks a matrix from a file (using
<code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.parser.hic_parser.autoreader()</span></code>) or a list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>things</strong> – might be either a file name, a file handler or a list of
list (all with same length)</p></li>
<li><p><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>a parser function that returns a tuple of lists
representing the data matrix,
with this file example.tsv:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chrT_001</span>    <span class="n">chrT_002</span>    <span class="n">chrT_003</span>    <span class="n">chrT_004</span>
<span class="n">chrT_001</span>    <span class="mi">629</span>    <span class="mi">164</span>    <span class="mi">88</span>    <span class="mi">105</span>
<span class="n">chrT_002</span>    <span class="mi">86</span>    <span class="mi">612</span>    <span class="mi">175</span>    <span class="mi">110</span>
<span class="n">chrT_003</span>    <span class="mi">159</span>    <span class="mi">216</span>    <span class="mi">437</span>    <span class="mi">105</span>
<span class="n">chrT_004</span>    <span class="mi">100</span>    <span class="mi">111</span>    <span class="mi">146</span>    <span class="mi">278</span>
</pre></div>
</div>
<p>the output of parser(‘example.tsv’) might be:
<code class="docutils literal notranslate"><span class="pre">([629,</span> <span class="pre">86,</span> <span class="pre">159,</span> <span class="pre">100,</span> <span class="pre">164,</span> <span class="pre">612,</span> <span class="pre">216,</span> <span class="pre">111,</span> <span class="pre">88,</span> <span class="pre">175,</span> <span class="pre">437,</span> <span class="pre">146,</span> <span class="pre">105,</span> <span class="pre">110,</span>
<span class="pre">105,</span> <span class="pre">278])</span></code></p>
</p></li>
<li><p><strong>resolution</strong> (<em>1</em>) – resolution of the matrix</p></li>
<li><p><strong>hic</strong> (<em>True</em>) – if False, TADbit assumes that files contains normalized
data</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the corresponding matrix concatenated into a huge list, also
returns number or rows</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pytadbit.parsers.hic_parser.load_hic_data_from_reads">
<code class="sig-prename descclassname">pytadbit.parsers.hic_parser.</code><code class="sig-name descname">load_hic_data_from_reads</code><span class="sig-paren">(</span><em class="sig-param">fnam</em>, <em class="sig-param">resolution</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/parsers/hic_parser.html#load_hic_data_from_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.parsers.hic_parser.load_hic_data_from_reads" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fnam</strong> – tsv file with reads1 and reads2</p></li>
<li><p><strong>resolution</strong> – the resolution of the experiment (size of a bin in
bases)</p></li>
<li><p><strong>genome_seq</strong> – a dictionary containing the genomic sequence by
chromosome</p></li>
<li><p><strong>get_sections</strong> (<em>False</em>) – for very very high resolution, when the column
index does not fit in memory</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pytadbit.parsers.genome_parser.parse_fasta">
<code class="sig-prename descclassname">pytadbit.parsers.genome_parser.</code><code class="sig-name descname">parse_fasta</code><span class="sig-paren">(</span><em class="sig-param">f_names</em>, <em class="sig-param">chr_names=None</em>, <em class="sig-param">chr_filter=None</em>, <em class="sig-param">chr_regexp=None</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">save_cache=True</em>, <em class="sig-param">reload_cache=False</em>, <em class="sig-param">only_length=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/parsers/genome_parser.html#parse_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.parsers.genome_parser.parse_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a list of fasta files, or just one fasta.</p>
<p>WARNING: The order is important</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f_names</strong> – list of pathes to files, or just a single path</p></li>
<li><p><strong>chr_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – pass list of chromosome names, or just one. If None
are passed, then chromosome names will be inferred from fasta headers</p></li>
<li><p><strong>chr_filter</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – use only chromosome in the input list</p></li>
<li><p><strong>chr_regexp</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – use only chromosome matching</p></li>
<li><p><strong>save_cache</strong> (<em>True</em>) – save a cached version of this file for faster
loadings (~4 times faster)</p></li>
<li><p><strong>reload_cache</strong> (<em>False</em>) – reload cached genome</p></li>
<li><p><strong>only_length</strong> (<em>False</em>) – returns dictionary with length of genome,not sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a sorted dictionary with chromosome names as keys, and sequences
as values (sequence in upper case)</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pytadbit.parsers.sam_parser.parse_sam">
<code class="sig-prename descclassname">pytadbit.parsers.sam_parser.</code><code class="sig-name descname">parse_sam</code><span class="sig-paren">(</span><em class="sig-param">f_names1</em>, <em class="sig-param">f_names2=None</em>, <em class="sig-param">out_file1=None</em>, <em class="sig-param">out_file2=None</em>, <em class="sig-param">genome_seq=None</em>, <em class="sig-param">re_name=None</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">clean=True</em>, <em class="sig-param">mapper=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/parsers/sam_parser.html#parse_sam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.parsers.sam_parser.parse_sam" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse sam/bam file using pysam tools.</p>
<dl class="simple">
<dt>Keep a summary of the results into 2 tab-separated files that will contain 6</dt><dd><p>columns: read ID, Chromosome, position, strand (either 0 or 1), mapped
sequence lebgth, position of the closest upstream RE site, position of
the closest downstream RE site</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f_names1</strong> – a list of path to sam/bam files corresponding to the
mapping of read1, can also  be just one file</p></li>
<li><p><strong>f_names1</strong> – a list of path to sam/bam files corresponding to the
mapping of read2, can also  be just one file</p></li>
<li><p><strong>out_file1</strong> – path to outfile tab separated format containing mapped
read1 information</p></li>
<li><p><strong>out_file1</strong> – path to outfile tab separated format containing mapped
read2 information</p></li>
<li><p><strong>genome_seq</strong> – a dictionary generated by <code class="xref py py-func docutils literal notranslate"><span class="pre">pyatdbit.parser.genome_parser.parse_fasta()</span></code>.
containing the genomic sequence</p></li>
<li><p><strong>re_name</strong> – name of the restriction enzyme used</p></li>
<li><p><strong>mapper</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – software used to map (supported are GEM and BOWTIE2).
Guessed from file by default.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pytadbit.parsers.tad_parser.parse_tads">
<code class="sig-prename descclassname">pytadbit.parsers.tad_parser.</code><code class="sig-name descname">parse_tads</code><span class="sig-paren">(</span><em class="sig-param">handler</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/parsers/tad_parser.html#parse_tads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.parsers.tad_parser.parse_tads" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a tab separated value file that contains the list of TADs of a given
experiment. This file might have been generated whith the
<code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.tadbit.print_result_R()</span></code> or with the R binding for tadbit</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>handler</strong> – path to file</p></li>
<li><p><strong>bin_size</strong> (<em>1</em>) – resolution of the experiment</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of TADs and list of weights, each TAD being a dict of type:</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">TAD_num</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span>
           <span class="s1">&#39;end&#39;</span>  <span class="p">:</span> <span class="n">end</span><span class="p">,</span>
           <span class="s1">&#39;brk&#39;</span>  <span class="p">:</span> <span class="n">end</span><span class="p">,</span>
           <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="n">score</span><span class="p">}}</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impoptimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impmodel.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_clustering.html"
                        title="previous chapter">TAD clustering</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_aligner.html"
                        title="next chapter">Boundary Aligner</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_aligner.html" title="Boundary Aligner"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_clustering.html" title="TAD clustering"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Reference Guide</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Feb 07, 2020.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>