<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>IMPoptimizer class &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/black-on-white.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Reference Guide" href="index.html" />
    <link rel="next" title="IMPmodel class" href="reference_imp_model.html" />
    <link rel="prev" title="Alignment class" href="reference_boundary_alignment.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_imp_model.html" title="IMPmodel class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_boundary_alignment.html" title="Alignment class"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="impoptimizer-class">
<h1>IMPoptimizer class<a class="headerlink" href="#impoptimizer-class" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer">
<em class="property">class </em><tt class="descclassname">pytadbit.imp.impoptimizer.</tt><tt class="descname">IMPoptimizer</tt><big>(</big><em>experiment</em>, <em>start</em>, <em>end</em>, <em>n_models=500</em>, <em>n_keep=100</em>, <em>close_bins=1</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer" title="Permalink to this definition">¶</a></dt>
<dd><p>This class optimizes a set of paramaters (scale, maxdist, lowfreq and
upfreq) in order to maximize the correlation between the models generated 
by IMP and the input data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>experiment</strong> &#8211; an instance of the class pytadbit.experiment.Experiment</li>
<li><strong>start</strong> &#8211; first bin to model (bin number)</li>
<li><strong>end</strong> &#8211; last bin to model (bin number)</li>
<li><strong>n_models</strong> (<em>5000</em>) &#8211; number of models to generate</li>
<li><strong>n_keep</strong> (<em>1000</em>) &#8211; number of models used in the final analysis (usually 
the top 20% of the generated models). The models are ranked according to
their objective function value (the lower the better)</li>
<li><strong>close_bins</strong> (<em>1</em>) &#8211; number of particles away (i.e. the bin number 
difference) a particle pair must be in order to be considered as 
neighbors (e.g. 1 means consecutive particles)</li>
<li><strong>cutoff</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; distance cutoff (nm) to define whether two particles
are in contact or not, default is 2 times resolution, times scale.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer.get_best_parameters_dict">
<tt class="descname">get_best_parameters_dict</tt><big>(</big><em>reference=None</em>, <em>with_corr=False</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer.get_best_parameters_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer.get_best_parameters_dict" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a description of the dataset optimized</li>
<li><strong>with_corr</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; if True, returns also the correlation value</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dict that can be used for modelling, see config parameter in
<a class="reference internal" href="reference_experiment.html#pytadbit.experiment.Experiment.model_region" title="pytadbit.experiment.Experiment.model_region"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.experiment.Experiment.model_region()</span></tt></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer.load_from_file">
<tt class="descname">load_from_file</tt><big>(</big><em>f_name</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer.load_from_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer.load_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the optimized parameters from a file generated with the function:
pytadbit.imp.impoptimizer.IMPoptimizer.write_result.
This function does not overwrite the parameters that were already 
loaded or calculated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>f_name</strong> &#8211; file name with the absolute path</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer.load_grid_search">
<tt class="descname">load_grid_search</tt><big>(</big><em>filenames</em>, <em>corr='spearman'</em>, <em>off_diag=1</em>, <em>verbose=True</em>, <em>n_cpus=1</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer.load_grid_search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer.load_grid_search" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads one file or a list of files containing pre-calculated Structural
Models (keep_models parameter used). And correlate each set of models
with real data. Usefull to run different correlation on the same data
avoiding to re-calculate each time the models.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>filenames</strong> &#8211; either a path to a file or a list of paths.</li>
<li><strong>corr</strong> (<em>spearman</em>) &#8211; correlation coefficient to use</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>&#8216;param 1 off_diag:
:param True verbose: print the results to the standard output</p>
</dd></dl>

<dl class="method">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer.plot_2d">
<tt class="descname">plot_2d</tt><big>(</big><em>axes=('scale'</em>, <em>'maxdist'</em>, <em>'upfreq'</em>, <em>'lowfreq')</em>, <em>show_best=0</em>, <em>skip=None</em>, <em>savefig=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer.plot_2d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer.plot_2d" title="Permalink to this definition">¶</a></dt>
<dd><p>A grid of heatmaps representing the result of the optimization.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>axes</strong> (<em>&#8216;scale&#8217;,&#8217;maxdist&#8217;,&#8217;upfreq&#8217;,&#8217;lowfreq&#8217;</em>) &#8211; list of axes to be
represented in the plot. The order will define which parameter will
be placed on the x, y, z or w axe.</li>
<li><strong>show_best</strong> (<em>0</em>) &#8211; number of best correlation values to highlight in 
the plot</li>
<li><strong>skip</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; if passed (as a dictionary), fix a given axe,
e.g.: {&#8216;scale&#8217;: 0.001, &#8216;maxdist&#8217;: 500}</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer.plot_3d">
<tt class="descname">plot_3d</tt><big>(</big><em>axes=('scale'</em>, <em>'maxdist'</em>, <em>'upfreq'</em>, <em>'lowfreq')</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer.plot_3d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer.plot_3d" title="Permalink to this definition">¶</a></dt>
<dd><p>A grid of heatmaps representing the result of the optimization.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>axes</strong> (<em>&#8216;scale&#8217;,&#8217;maxdist&#8217;,&#8217;upfreq&#8217;,&#8217;lowfreq&#8217;</em>) &#8211; tuple of axes to be
represented in the plot. The order will define which parameter will 
be placed on the x, y, z or w axe.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer.run_grid_search">
<tt class="descname">run_grid_search</tt><big>(</big><em>upfreq_range=(0</em>, <em>1</em>, <em>0.1)</em>, <em>lowfreq_range=(-1</em>, <em>0</em>, <em>0.1)</em>, <em>maxdist_range=(400</em>, <em>1500</em>, <em>100)</em>, <em>scale_range=0.01</em>, <em>dcutoff_range=2</em>, <em>corr='spearman'</em>, <em>off_diag=1</em>, <em>savedata=None</em>, <em>n_cpus=1</em>, <em>verbose=True</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer.run_grid_search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer.run_grid_search" title="Permalink to this definition">¶</a></dt>
<dd><p>This function calculates the correlation between the models generated 
by IMP and the input data for the four main IMP parameters (scale, 
maxdist, lowfreq and upfreq) in the given ranges of values.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>n_cpus</strong> &#8211; number of CPUs to use</li>
<li><strong>lowfreq_range</strong> (<em>(-1,0,0.1)</em>) &#8211; range of lowfreq values to be 
optimized. The last value of the input tuple is the incremental 
step for the lowfreq values</li>
<li><strong>upfreq_range</strong> (<em>(0,1,0.1)</em>) &#8211; range of upfreq values to be optimized.
The last value of the input tuple is the incremental step for the
upfreq values</li>
<li><strong>maxdist_range</strong> (<em>(400,1400,100)</em>) &#8211; upper and lower bounds
used to search for the optimal maximum experimental distance. The 
last value of the input tuple is the incremental step for maxdist 
values</li>
<li><strong>scale_range</strong> (<em>0.01</em>) &#8211; upper and lower bounds used to search for
the optimal scale parameter (nm per nucleotide). The last value of
the input tuple is the incremental step for scale parameter values</li>
<li><strong>dcutoff_range</strong> (<em>2</em>) &#8211; upper and lower bounds used to search for
the optimal distance cutoff parameter (distance, in number of beads,
from which to consider 2 beads as being close). The last value of the
input tuple is the incremental step for scale parameter values</li>
<li><strong>savedata</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; concatenate all generated models into a dictionary
and save it into a file named by this argument</li>
<li><strong>verbose</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; print the results to the standard output</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pytadbit.imp.impoptimizer.IMPoptimizer.write_result">
<tt class="descname">write_result</tt><big>(</big><em>f_name</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/imp/impoptimizer.html#IMPoptimizer.write_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.imp.impoptimizer.IMPoptimizer.write_result" title="Permalink to this definition">¶</a></dt>
<dd><p>This function writes a log file of all the values tested for each 
parameter, and the resulting correlation value.</p>
<p>This file can be used to load or merge data a posteriori using 
the function pytadbit.imp.impoptimizer.IMPoptimizer.load_from_file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>f_name</strong> &#8211; file name with the absolute path</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_model.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_boundary_alignment.html"
                        title="previous chapter">Alignment class</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_imp_model.html"
                        title="next chapter">IMPmodel class</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_imp_model.html" title="IMPmodel class"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_boundary_alignment.html" title="Alignment class"
             >previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" >Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright .
      Last updated on Mar 31, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>