<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>StructuralModels class &#8212; Tadbit 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/black-on-white.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Utilities" href="reference_utils.html" />
    <link rel="prev" title="IMPmodel class" href="reference_modelling_impmodel.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="reference_utils.html" title="Utilities"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_modelling_impmodel.html" title="IMPmodel class"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Reference Guide</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <dl class="function">
<dt id="pytadbit.modelling.imp_modelling.generate_3d_models">
<code class="sig-prename descclassname">pytadbit.modelling.imp_modelling.</code><code class="sig-name descname">generate_3d_models</code><span class="sig-paren">(</span><em class="sig-param">zscores</em>, <em class="sig-param">resolution</em>, <em class="sig-param">nloci</em>, <em class="sig-param">start=1</em>, <em class="sig-param">n_models=5000</em>, <em class="sig-param">n_keep=1000</em>, <em class="sig-param">close_bins=1</em>, <em class="sig-param">n_cpus=1</em>, <em class="sig-param">keep_all=False</em>, <em class="sig-param">verbose=0</em>, <em class="sig-param">outfile=None</em>, <em class="sig-param">config=None</em>, <em class="sig-param">values=None</em>, <em class="sig-param">experiment=None</em>, <em class="sig-param">coords=None</em>, <em class="sig-param">zeros=None</em>, <em class="sig-param">first=None</em>, <em class="sig-param">container=None</em>, <em class="sig-param">use_HiC=True</em>, <em class="sig-param">use_confining_environment=True</em>, <em class="sig-param">use_excluded_volume=True</em>, <em class="sig-param">single_particle_restraints=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/imp_modelling.html#generate_3d_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.imp_modelling.generate_3d_models" title="Permalink to this definition">¶</a></dt>
<dd><p>This function generates three-dimensional models starting from Hi-C data.
The final analysis will be performed on the n_keep top models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>zscores</strong> – the dictionary of the Z-score values calculated from the
Hi-C pairwise interactions</p></li>
<li><p><strong>resolution</strong> – number of nucleotides per Hi-C bin. This will be the
number of nucleotides in each model’s particle</p></li>
<li><p><strong>nloci</strong> – number of particles to model (may not all be present in
zscores)</p></li>
<li><p><strong>experiment</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – experiment from which to do the modelling (used only
for descriptive purpose)</p></li>
<li><p><strong>coords</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>a dictionary or a list of dictionaries like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="s1">&#39;crm&#39;</span>  <span class="p">:</span> <span class="s1">&#39;19&#39;</span><span class="p">,</span>
 <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="mi">14637</span><span class="p">,</span>
 <span class="s1">&#39;end&#39;</span>  <span class="p">:</span> <span class="mi">15689</span><span class="p">}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>n_models</strong> (<em>5000</em>) – number of models to generate</p></li>
<li><p><strong>n_keep</strong> (<em>1000</em>) – number of models used in the final analysis (usually
the top 20% of the generated models). The models are ranked according to
their objective function value (the lower the better)</p></li>
<li><p><strong>keep_all</strong> (<em>False</em>) – whether or not to keep the discarded models (if
True, models will be stored under StructuralModels.bad_models)</p></li>
<li><p><strong>close_bins</strong> (<em>1</em>) – number of particles away (i.e. the bin number
difference) a particle pair must be in order to be considered as
neighbors (e.g. 1 means consecutive particles)</p></li>
<li><p><strong>n_cpus</strong> – number of CPUs to use</p></li>
<li><p><strong>verbose</strong> (<em>False</em>) – if set to True, information about the distance, force
and Z-score between particles will be printed. If verbose is 0.5 than
constraints will be printed only for the first model calculated.</p></li>
<li><p><strong>values</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – the normalized Hi-C data in a list of lists (equivalent
to a square matrix)</p></li>
<li><p><strong>config</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>a dictionary containing the standard
parameters used to generate the models. The dictionary should contain
the keys kforce, lowrdist, maxdist, upfreq and lowfreq. Examples can be
seen by doing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pytadbit.modelling.CONFIG</span> <span class="kn">import</span> <span class="n">CONFIG</span>

<span class="n">where</span> <span class="n">CONFIG</span> <span class="ow">is</span> <span class="n">a</span> <span class="n">dictionary</span> <span class="n">of</span> <span class="n">dictionaries</span> <span class="n">to</span> <span class="n">be</span> <span class="n">passed</span> <span class="n">to</span> <span class="n">this</span> <span class="n">function</span><span class="p">:</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CONFIG</span> <span class="o">=</span> <span class="p">{</span>
 <span class="c1"># Paramaters for the Hi-C dataset from:</span>
 <span class="s1">&#39;reference&#39;</span> <span class="p">:</span> <span class="s1">&#39;victor corces dataset 2013&#39;</span><span class="p">,</span>

 <span class="c1"># Force applied to the restraints inferred to neighbor particles</span>
 <span class="s1">&#39;kforce&#39;</span>    <span class="p">:</span> <span class="mi">5</span><span class="p">,</span>

 <span class="c1"># Space occupied by a nucleotide (nm)</span>
 <span class="s1">&#39;scale&#39;</span>     <span class="p">:</span> <span class="mf">0.005</span>

 <span class="c1"># Strength of the bending interaction</span>
 <span class="s1">&#39;kbending&#39;</span>     <span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span> <span class="c1"># OPTIMIZATION:</span>

 <span class="c1"># Maximum experimental contact distance</span>
 <span class="s1">&#39;maxdist&#39;</span>   <span class="p">:</span> <span class="mi">600</span><span class="p">,</span> <span class="c1"># OPTIMIZATION: 500-1200</span>

 <span class="c1"># Minimum thresholds used to decide which experimental values have to be</span>
 <span class="c1"># included in the computation of restraints. Z-score values bigger than upfreq</span>
 <span class="c1"># and less that lowfreq will be include, whereas all the others will be rejected</span>
 <span class="s1">&#39;lowfreq&#39;</span>   <span class="p">:</span> <span class="o">-</span><span class="mf">0.7</span> <span class="c1"># OPTIMIZATION: min/max Z-score</span>

 <span class="c1"># Maximum threshold used to decide which experimental values have to be</span>
 <span class="c1"># included in the computation of restraints. Z-score values greater than upfreq</span>
 <span class="c1"># and less than lowfreq will be included, while all the others will be rejected</span>
 <span class="s1">&#39;upfreq&#39;</span>    <span class="p">:</span> <span class="mf">0.3</span> <span class="c1"># OPTIMIZATION: min/max Z-score</span>

 <span class="p">}</span>
</pre></div>
</div>
</p></li>
<li><p><strong>first</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – particle number at which model should start (0 should be
used inside TADbit)</p></li>
<li><p><strong>container</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – restrains particle to be within a given object. Can
only be a ‘cylinder’, which is, in fact a cylinder of a given height to
which are added hemispherical ends. This cylinder is defined by a radius,
its height (with a height of 0 the cylinder becomes a sphere) and the
force applied to the restraint. E.g. for modeling E. coli genome (2
micrometers length and 0.5 micrometer of width), these values could be
used: [‘cylinder’, 250, 1500, 50], and for a typical mammalian nuclei
(6 micrometers diameter): [‘cylinder’, 3000, 0, 50]</p></li>
</ul>
</dd>
</dl>
<p>:param True  use_HiC
:param True  use_confining_environment
:param True  use_excluded_volume
:param: None single_particle_restraints: a list containing restraints to single particles.</p>
<blockquote>
<div><dl class="simple">
<dt>Each restraint in the list is itself a list with the following information:</dt><dd><p>[bin, [position_x, position_y, position_z], type, kforce, radius]
bin: bin number of the particle to restraint
[position_x, position_y, position_z](nm): center of the sphere of the restraint
type: ‘Harmonic’, ‘HarmonicLowerBound’, ‘HarmonicUpperBound’
kforce: weigth of the restraint
radius (nm): radius of the sphere</p>
</dd>
</dl>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a StructuralModels object</p>
</dd>
</dl>
</dd></dl>

<div class="section" id="structuralmodels-class">
<h1>StructuralModels class<a class="headerlink" href="#structuralmodels-class" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="pytadbit.modelling.structuralmodels.load_structuralmodels">
<code class="sig-prename descclassname">pytadbit.modelling.structuralmodels.</code><code class="sig-name descname">load_structuralmodels</code><span class="sig-paren">(</span><em class="sig-param">input_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#load_structuralmodels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.load_structuralmodels" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads <a class="reference internal" href="#pytadbit.modelling.structuralmodels.StructuralModels" title="pytadbit.modelling.structuralmodels.StructuralModels"><code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.modelling.structuralmodels.StructuralModels</span></code></a> from a file
(generated with
<a class="reference internal" href="#pytadbit.modelling.structuralmodels.StructuralModels.save_models" title="pytadbit.modelling.structuralmodels.StructuralModels.save_models"><code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.modelling.structuralmodels.StructuralModels.save_models</span></code></a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>input</strong> – path to the pickled StructuralModels object or dictionary
containing all the information.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a <code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.modelling.imp_model.StructuralModels</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels">
<em class="property">class </em><code class="sig-prename descclassname">pytadbit.modelling.structuralmodels.</code><code class="sig-name descname">StructuralModels</code><span class="sig-paren">(</span><em class="sig-param">nloci</em>, <em class="sig-param">models</em>, <em class="sig-param">bad_models</em>, <em class="sig-param">resolution</em>, <em class="sig-param">original_data=None</em>, <em class="sig-param">zscores=None</em>, <em class="sig-param">clusters=None</em>, <em class="sig-param">config=None</em>, <em class="sig-param">experiment=None</em>, <em class="sig-param">zeros=None</em>, <em class="sig-param">restraints=None</em>, <em class="sig-param">description=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels" title="Permalink to this definition">¶</a></dt>
<dd><p>This class contains three-dimensional models generated from a single Hi-C
data. They can be reached either by their index (integer representing their
rank according to objective function value), or by their IMP random intial
number (as string).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nloci</strong> – number of particles in the selected region</p></li>
<li><p><strong>models</strong> – a dictionary containing the generated
<a class="reference internal" href="reference_modelling_impmodel.html#pytadbit.modelling.impmodel.IMPmodel" title="pytadbit.modelling.impmodel.IMPmodel"><code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.modelling.impmodel.IMPmodel</span></code></a> to be used as ‘best models’</p></li>
<li><p><strong>bad_models</strong> – a dictionary of <a class="reference internal" href="reference_modelling_impmodel.html#pytadbit.modelling.impmodel.IMPmodel" title="pytadbit.modelling.impmodel.IMPmodel"><code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.modelling.impmodel.IMPmodel</span></code></a>,
these model will not be used, just stored in case the set of
‘best models’ needs to be extended later-on (
<a class="reference internal" href="#pytadbit.modelling.structuralmodels.StructuralModels.define_best_models" title="pytadbit.modelling.structuralmodels.StructuralModels.define_best_models"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.modelling.structuralmodels.StructuralModels.define_best_models()</span></code></a>
).</p></li>
<li><p><strong>resolution</strong> – number of nucleotides per Hi-C bin. This will be the
number of nucleotides in each model’s particle.</p></li>
<li><p><strong>original_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a list of list (equivalent to a square matrix) of
the normalized Hi-C data, used to build this list of models.</p></li>
<li><p><strong>clusters</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a dictionary of type
<code class="xref py py-class docutils literal notranslate"><span class="pre">pytadbit.modelling.structuralmodels.ClusterOfModels</span></code></p></li>
<li><p><strong>config</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a dictionary containing the parameter to be used for the
generation of three dimensional models.</p></li>
</ul>
</dd>
</dl>
<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.accessibility">
<code class="sig-name descname">accessibility</code><span class="sig-paren">(</span><em class="sig-param">radius</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">nump=100</em>, <em class="sig-param">superradius=200</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">savedata=None</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">plot=True</em>, <em class="sig-param">error=True</em>, <em class="sig-param">steps=(1</em>, <em class="sig-param">)</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.accessibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.accessibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates a mesh surface around the model (distance equal to input
<strong>radius</strong>) and checks if each point of this mesh could be replaced by
an object (i.e. a protein) of a given <strong>radius</strong></p>
<p>Outer part of the model can be excluded from the estimation of
accessible surface, as the occupancy outside the model is unkown (see
superradius option).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>radius</strong> – radius of the object we want to fit in the model.</p></li>
<li><p><strong>nump</strong> (<em>100</em>) – number of points to draw around a given particle. This
number also sets the number of points drawn around edges, as each
point occupies a given surface (see maths below). <em>Note that this
number is considerably lowered by the occupancy of edges, depending
of the angle formed by the edges surrounding a given particle, only
10% to 50% of the ``nump`` will be drawn in fact.</em></p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path where to save chimera image</p></li>
<li><p><strong>) </strong><strong>steps</strong> (<em>(</em><em>1</em><em>,</em>) – how many particles to group for the
estimation. By default 1 curve is drawn</p></li>
<li><p><strong>superradius</strong> (<em>200</em>) – radius of an object used to exclude outer
surface of the model. Superradius must be higher than radius.</p></li>
</ul>
</dd>
</dl>
<p>This function will first define a mesh around the chromatin,
representing all possible position of the center of the object we want
to fit. This mesh will be at a distance of <em>radius</em> from the chromatin
strand. All dots in the mesh represents an equal area (<em>a</em>), the whole
surface of the chromatin strand being: <span class="math notranslate nohighlight">\(A=n \times a\)</span> (<em>n</em> being
the total number of dots in the mesh).</p>
<p>The mesh consists of spheres around particles of the model, and
cylinders around edges joining particles (no overlap is allowed between
sphere and cylinders or cylinder and cylinder when they are
consecutive).</p>
<p>If we want that all dots of the mesh representing the surface of the
chromatin, corresponds to an equal area (<span class="math notranslate nohighlight">\(a\)</span>)</p>
<div class="math notranslate nohighlight">
\[a = \frac{4\pi r^2}{s} = \frac{2\pi r N_{(d)}}{c}\]</div>
<p>with:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(r\)</span> radius of the object to fit (as the input parameter <strong>radius</strong>)</p></li>
<li><p><span class="math notranslate nohighlight">\(s\)</span> number of points in sphere</p></li>
<li><p><span class="math notranslate nohighlight">\(c\)</span> number of points in circle (as the input parameter <strong>nump</strong>)</p></li>
<li><p><span class="math notranslate nohighlight">\(N_{(d)}\)</span> number of circles in an edge of length <span class="math notranslate nohighlight">\(d\)</span></p></li>
</ul>
<p>According to this, when the distance between two particles is equal
to <span class="math notranslate nohighlight">\(2r\)</span> (<span class="math notranslate nohighlight">\(N=2r\)</span>), we would have <span class="math notranslate nohighlight">\(s=c\)</span>.</p>
<p>As :</p>
<div class="math notranslate nohighlight">
\[2\pi r = \sqrt{4\pi r^2} \times \sqrt{\pi}\]</div>
<p>It is fair to state the number of dots represented along a circle as:</p>
<div class="math notranslate nohighlight">
\[c = \sqrt{s} \times \sqrt{\pi}\]</div>
<p>Thus the number of circles in an edge of length <span class="math notranslate nohighlight">\(d\)</span> must be:</p>
<div class="math notranslate nohighlight">
\[N_{(d)}=\frac{s}{\sqrt{s}\sqrt{\pi}}\times\frac{d}{2r}\]</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of <em>1-</em> the number of dots in the mesh that could be
occupied by an object of the given radius <em>2-</em> the total number of
dots in the mesh <em>3-</em> the estimated area of the mesh (in square
micrometers) <em>4-</em> the area of the mesh of a virtually straight strand
of chromatin defined as
<span class="math notranslate nohighlight">\(contour\times 2\pi r + 4\pi r^2\)</span> (also in
micrometers) <em>5-</em> a list of number of (accessibles, inaccessible) for
each particle (percentage burried can be infered afterwards by
accessible/(accessible+inaccessible) )</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.align_models">
<code class="sig-name descname">align_models</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">in_place=False</em>, <em class="sig-param">reference_model=None</em>, <em class="sig-param">by_cluster=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.align_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.align_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Three-dimensional aligner for structural models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the average model will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the average model only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>in_place</strong> (<em>False</em>) – if True, the result of the alignment will REPLACE
the coordinates of each model. Default is too yield new coordinates of
each model</p></li>
<li><p><strong>reference_model</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – align given model to reference model</p></li>
<li><p><strong>by_cluster</strong> (<em>False</em>) – if True align each cluster individually in case
we align all the models (models=None)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.angle_between_3_particles">
<code class="sig-name descname">angle_between_3_particles</code><span class="sig-paren">(</span><em class="sig-param">parta</em>, <em class="sig-param">partb</em>, <em class="sig-param">partc</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">radian=False</em>, <em class="sig-param">all_angles=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.angle_between_3_particles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.angle_between_3_particles" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the angle between 3 particles.</p>
<p>Given three particles A, B and C, the angle g (angle ACB, shown below):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>      <span class="n">A</span>
     <span class="o">/|</span>
    <span class="o">/</span><span class="n">i</span><span class="o">|</span>
  <span class="n">c</span><span class="o">/</span>  <span class="o">|</span>
  <span class="o">/</span>   <span class="o">|</span>
 <span class="o">/</span>    <span class="o">|</span>
<span class="n">B</span> <span class="p">)</span><span class="n">g</span>  <span class="o">|</span><span class="n">b</span>
 \    <span class="o">|</span>
  \   <span class="o">|</span>
  <span class="n">a</span>\  <span class="o">|</span>
    \<span class="n">h</span><span class="o">|</span>
     \<span class="o">|</span>
      <span class="n">C</span>
</pre></div>
</div>
<p>is given by the theorem of Al-Kashi:</p>
<div class="math notranslate nohighlight">
\[b^2 = a^2 + c^2 - 2ac\cos(g)\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parta</strong> – A particle number</p></li>
<li><p><strong>partb</strong> – A particle number</p></li>
<li><p><strong>partc</strong> – A particle number</p></li>
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the angle will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the angle only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>radian</strong> (<em>False</em>) – if True, return value in radians (in degrees
otherwise)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>an angle, either in degrees or radians. If all_angles is true
returns a list of the angle g, h, i (see picture above)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.average_model">
<code class="sig-name descname">average_model</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.average_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.average_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds and returns an average model representing a given group of models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the average model will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the average model only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>verbose</strong> (<em>False</em>) – prints the distance of each model to average model
(in stderr)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the average model of a given group of models (a new and
ARTIFICIAL model)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.centroid_model">
<code class="sig-name descname">centroid_model</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.centroid_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.centroid_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimates and returns the centroid model of a given group of models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the centroid model will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the centroid model only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>verbose</strong> (<em>False</em>) – prints the distance of each model to average model
(in stderr)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the centroid model of a given group of models (the most model
representative)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.cluster_analysis_dendrogram">
<code class="sig-name descname">cluster_analysis_dendrogram</code><span class="sig-paren">(</span><em class="sig-param">n_best_clusters=None</em>, <em class="sig-param">color=False</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.cluster_analysis_dendrogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.cluster_analysis_dendrogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Representation of the clustering results. The length of the leaves if
proportional to the final objective function value of each model. The
branch widths are proportional to the number of models in a given
cluster (or group of clusters, if it is an internal branch).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_best_clusters</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – number of clusters to represent (by
default all clusters will be shown)</p></li>
<li><p><strong>color</strong> (<em>False</em>) – color the dendrogram based on the significance of
the clustering (basically it depends of the internal branch lengths)</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>width_factor</strong> (<em>10.0</em>) – multiplicator for the width of the line
representing the number of models in a given cluster.</p></li>
<li><p><strong>fontsize</strong> (<em>8</em>) – size of the smallest font represented in the plot</p></li>
<li><p><strong>figsize</strong> (<em>(</em><em>8</em><em>,</em><em>8</em><em>)</em>) – a tuple of width and height, to set the size of
the plot.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.cluster_models">
<code class="sig-name descname">cluster_models</code><span class="sig-paren">(</span><em class="sig-param">fact=0.75</em>, <em class="sig-param">dcutoff=None</em>, <em class="sig-param">method='mcl'</em>, <em class="sig-param">mcl_bin='mcl'</em>, <em class="sig-param">tmp_file=None</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">n_cpus=1</em>, <em class="sig-param">mclargs=None</em>, <em class="sig-param">external=False</em>, <em class="sig-param">what='score'</em>, <em class="sig-param">region=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.cluster_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.cluster_models" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>This function performs a clustering analysis of the generated models
based on structural comparison. The result will be stored in
StructuralModels.clusters</p>
<p>Clustering is done according to a score of pairwise comparison
calculated as:</p>
</div></blockquote>
<div class="math notranslate nohighlight">
\[score_i = eqvs_i \times \frac{dRMSD_i / max(dRMSD)}
                                {RMSD_i / max(RMSD)}\]</div>
<p>where <span class="math notranslate nohighlight">\(eqvs_i\)</span> is the number of equivalent position for the ith
pairwise model comparison.</p>
<blockquote>
<div><dl class="field-list simple">
<dt class="field-odd">param 0.75 fact</dt>
<dd class="field-odd"><p>factor to define the percentage of equivalent
positions to be considered in the clustering</p>
</dd>
<dt class="field-even">param None dcutoff</dt>
<dd class="field-even"><p>distance threshold (nm) to determine if two
particles are in contact, default is 1.5 times resolution times scale</p>
</dd>
<dt class="field-odd">param ‘mcl’ method</dt>
<dd class="field-odd"><p>clustering method to use, which can be either
‘mcl’ or ‘ward’. MCL method is recommended. WARD method uses a scipy
implementation of this hierarchical clustering, and selects the best
number of clusters using the
<a class="reference internal" href="reference_utils.html#pytadbit.utils.tadmaths.calinski_harabasz" title="pytadbit.utils.tadmaths.calinski_harabasz"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.tadmaths.calinski_harabasz()</span></code></a> function.</p>
</dd>
<dt class="field-even">param ‘mcl’ mcl_bin</dt>
<dd class="field-even"><p>path to the mcl executable file, in case of the
‘mcl is not in the PATH’ warning message</p>
</dd>
<dt class="field-odd">param None tmp_file</dt>
<dd class="field-odd"><p>path to a temporary file created during
the clustering computation. Default will be created in /tmp/ folder</p>
</dd>
<dt class="field-even">param True verbose</dt>
<dd class="field-even"><p>same as print StructuralModels.clusters</p>
</dd>
<dt class="field-odd">param 1 n_cpus</dt>
<dd class="field-odd"><p>number of cpus to use in MCL clustering</p>
</dd>
<dt class="field-even">param mclargs</dt>
<dd class="field-even"><p>list with any other command line argument to be passed
to mcl (i.e,: mclargs=[‘-pi’, ‘10’, ‘-I’, ‘2.0’])</p>
</dd>
<dt class="field-odd">param False external</dt>
<dd class="field-odd"><p>if True returns the cluster found instead of
storing it as StructuralModels.clusters</p>
</dd>
<dt class="field-even">param ‘score’ what</dt>
<dd class="field-even"><p>Statistic used for clustering. Can be one of
‘score’, ‘rmsd’, ‘drmsd’ or ‘eqv’.</p>
</dd>
<dt class="field-odd">param None region</dt>
<dd class="field-odd"><p>coordinates of the region to base the clustering.
Can be either one chromosome name, or the coordinate in
the form: “chr3:110000000-120000000”</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.contact_map">
<code class="sig-name descname">contact_map</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">cutoff=None</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">savedata=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.contact_map"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.contact_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots a contact map representing the frequency of interaction (defined
by a distance cutoff) between two particles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the contact map will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the contact map only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – distance cutoff (nm) to define whether two particles
are in contact or not, default is 2 times resolution, times scale.</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the contact map data
generated, in three columns format (particle1, particle2, percentage
of models where these two particles are in contact)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.correlate_with_real_data">
<code class="sig-name descname">correlate_with_real_data</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">cutoff=None</em>, <em class="sig-param">off_diag=1</em>, <em class="sig-param">plot=False</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">corr='spearman'</em>, <em class="sig-param">midplot='hexbin'</em>, <em class="sig-param">log_corr=True</em>, <em class="sig-param">contact_matrix=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.correlate_with_real_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.correlate_with_real_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the result of a correlation between a given group of models and
original Hi-C data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the correlation will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the correlation only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – distance cutoff (nm) to define whether two particles
are in contact or not, default is 2 times resolution, times scale.</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>plot</strong> (<em>False</em>) – to display the plot</p></li>
<li><p><strong>log_corr</strong> (<em>True</em>) – log plot for correlation</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>contact_matrix</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – input a contact matrix instead of computing
it from the models</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>correlation coefficient rho, between the two
matrices. A rho value greater than 0.7 indicates a very good
correlation</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.deconvolve">
<code class="sig-name descname">deconvolve</code><span class="sig-paren">(</span><em class="sig-param">fact=0.75</em>, <em class="sig-param">dcutoff=None</em>, <em class="sig-param">method='mcl'</em>, <em class="sig-param">mcl_bin='mcl'</em>, <em class="sig-param">tmp_file=None</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">n_cpus=1</em>, <em class="sig-param">mclargs=None</em>, <em class="sig-param">what='dRMSD'</em>, <em class="sig-param">n_best_clusters=10</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">represent_models=False</em>, <em class="sig-param">figsize=(11</em>, <em class="sig-param">11)</em>, <em class="sig-param">clusters=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.deconvolve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.deconvolve" title="Permalink to this definition">¶</a></dt>
<dd><p>This function performs a deconvolution analysis of a given froup of models.
It first clusters models based on structural comparison (dRMSD), and
then, performs a differential contact map between each possible pair
of cluster.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Clusters defined here are different from the one defined when using
<a class="reference internal" href="#pytadbit.modelling.structuralmodels.StructuralModels.cluster_models" title="pytadbit.modelling.structuralmodels.StructuralModels.cluster_models"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.modelling.structuralmodels.StructuralModels.cluster_models()</span></code></a>.
They are also not stored into StructuralModels.clusters</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fact</strong> (<em>0.75</em>) – factor to define the percentage of equivalent
positions to be considered in the clustering</p></li>
<li><p><strong>dcutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – distance threshold (nm) to determine if two
particles are in contact, default is 1.5 times resolution times scale</p></li>
<li><p><strong>method</strong> (<em>'mcl'</em>) – clustering method to use, which can be either
‘mcl’ or ‘ward’. MCL method is recommended. WARD method uses a scipy
implementation of this hierarchical clustering, and selects the best
number of clusters using the
<a class="reference internal" href="reference_utils.html#pytadbit.utils.tadmaths.calinski_harabasz" title="pytadbit.utils.tadmaths.calinski_harabasz"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.tadmaths.calinski_harabasz()</span></code></a> function.</p></li>
<li><p><strong>mcl_bin</strong> (<em>'mcl'</em>) – path to the mcl executable file, in case of the
‘mcl is not in the PATH’ warning message</p></li>
<li><p><strong>tmp_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a temporary file created during
the clustering computation. Default will be created in /tmp/ folder</p></li>
<li><p><strong>verbose</strong> (<em>True</em>) – same as print StructuralModels.clusters</p></li>
<li><p><strong>n_cpus</strong> (<em>1</em>) – number of cpus to use in MCL clustering</p></li>
<li><p><strong>mclargs</strong> – list with any other command line argument to be passed
to mcl (i.e,: mclargs=[‘-pi’, ‘10’, ‘-I’, ‘2.0’])</p></li>
<li><p><strong>n_best_clusters</strong> (<em>10</em>) – number of clusters to represent</p></li>
<li><p><strong>clusters</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – provide clusters as a dictionary with keys=cluster
number, or name, and values list of model numbers.</p></li>
<li><p><strong>represent_models</strong> (<em>False</em>) – To generate an interactive visualization
of a representative model for each cluster. Representative model
depends on the value passed to this option, it can be either
‘centroid’ or ‘best’ (this last standing for the model with lowest
IMP objective function value).</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>figsize</strong> (<em>(</em><em>11</em><em>,</em><em>11</em><em>)</em>) – dimension of the plot</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.define_best_models">
<code class="sig-name descname">define_best_models</code><span class="sig-paren">(</span><em class="sig-param">nbest</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.define_best_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.define_best_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Defines the number of top models (based on the objective function) to
keep. If keep_all is set to True in
<code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.modelling.imp_model.generate_3d_models()</span></code> or in
<a class="reference internal" href="reference_experiment.html#pytadbit.experiment.Experiment.model_region" title="pytadbit.experiment.Experiment.model_region"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.experiment.Experiment.model_region()</span></code></a>, then the full set
of models (n_models parameter) will be used, otherwise only the n_keep
models will be available.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>nbest</strong> – number of top models to keep (usually 20% of the
generated models).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.density_plot">
<code class="sig-name descname">density_plot</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">steps=(1</em>, <em class="sig-param">2</em>, <em class="sig-param">3</em>, <em class="sig-param">4</em>, <em class="sig-param">5)</em>, <em class="sig-param">interval=1</em>, <em class="sig-param">use_mass_center=False</em>, <em class="sig-param">error=False</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">savedata=None</em>, <em class="sig-param">plot=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.density_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.density_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the number of nucleotides per nm of chromatin vs the modeled
region bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the density plot will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the density plot only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>2</strong><strong>, </strong><strong>3</strong><strong>, </strong><strong>4</strong><strong>, </strong><strong>5</strong><strong>) </strong><strong>steps</strong> (<em>(</em><em>1</em><em>,</em>) – how many particles to group for the
estimation. By default 5 curves are drawn</p></li>
<li><p><strong>error</strong> (<em>False</em>) – represent the error of the estimates</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>interval</strong> (<em>1</em>) – distance are measure with this given interval
between two bins.</p></li>
<li><p><strong>use_mass_center</strong> (<em>False</em>) – if interval is higher than one, calculates the
distance between the center of mass of the particles <em>n</em> to
<em>n+interval</em> and the center of mass of the particles <em>n+interval</em> and
<em>n+2interval</em></p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the density data
generated (1 column per step + 1 for particle number).</p></li>
<li><p><strong>plot</strong> (<em>True</em>) – e.g. if False, only saves data. No plotting done</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.dihedral_angle">
<code class="sig-name descname">dihedral_angle</code><span class="sig-paren">(</span><em class="sig-param">pa</em>, <em class="sig-param">pb</em>, <em class="sig-param">pc</em>, <em class="sig-param">pd</em>, <em class="sig-param">pe</em>, <em class="sig-param">models</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.dihedral_angle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.dihedral_angle" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Calculates the dihedral angle between 2 planes formed by 5 particles</dt><dd><p>(one common to both planes).</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the angle will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the angle only for the models in the
cluster number ‘cluster’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.fetch_model_by_rand_init">
<code class="sig-name descname">fetch_model_by_rand_init</code><span class="sig-paren">(</span><em class="sig-param">rand_init</em>, <em class="sig-param">all_models=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.fetch_model_by_rand_init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.fetch_model_by_rand_init" title="Permalink to this definition">¶</a></dt>
<dd><p>Models are stored according to their objective function value (first
best), but in order to reproduce a model, we need its initial random
number. This method helps to fetch the model corresponding to a given
initial random number stored under
StructuralModels.models[N][‘rand_init’].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rand_init</strong> – the wanted rand_init number.</p></li>
<li><p><strong>all_models</strong> (<em>False</em>) – whether or not to use ‘bad’ models</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index of 3d model</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.get_contact_matrix">
<code class="sig-name descname">get_contact_matrix</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">cutoff=None</em>, <em class="sig-param">distance=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.get_contact_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.get_contact_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a matrix with the number of interactions observed below a given
cutoff distance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the contact matrix will be
computed using all the models. A list of numbers corresponding to a
given set of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the contact matrix only for the models in
the cluster number ‘cluster’</p></li>
<li><p><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – distance cutoff (nm) to define whether two particles
are in contact or not, default is 2 times resolution, times scale.
Cutoff can also be a list of values, in wich case the returned object
will be a dictionnary of matrices (keys being square cutoffs)</p></li>
<li><p><strong>distance</strong> (<em>False</em>) – returns the distance matrix of all_angles against
all_angles particles instead of a contact_map matrix using the cutoff</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>matrix frequency of interaction</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.get_persistence_length">
<code class="sig-name descname">get_persistence_length</code><span class="sig-paren">(</span><em class="sig-param">begin=0</em>, <em class="sig-param">end=None</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">savedata=None</em>, <em class="sig-param">plot=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.get_persistence_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.get_persistence_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the persistence length (Lp) of given section of the model.
Persistence length is calculated according to <a class="reference internal" href="../biblio.html#bystricky2004" id="id1"><span>[Bystricky2004]</span></a> :</p>
<div class="math notranslate nohighlight">
\[&lt;R^2&gt; = 2 \times Lp^2 \times (\frac{Lc}{Lp} - 1 + e^{\frac{-Lc}{Lp}})\]</div>
<p>with the contour length as <span class="math notranslate nohighlight">\(Lc = \frac{d}{c}\)</span> where <span class="math notranslate nohighlight">\(d\)</span> is
the genomic dstance in bp and <span class="math notranslate nohighlight">\(c\)</span> the linear mass density of the
chromatin (in bp/nm).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>begin</strong> (<em>0</em>) – starting particle of the region to consider</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – ending particle of the region to consider</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(float) persistence length</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.infer_unrestrained_particle_coords">
<code class="sig-name descname">infer_unrestrained_particle_coords</code><span class="sig-paren">(</span><em class="sig-param">xcoords</em>, <em class="sig-param">ycoords</em>, <em class="sig-param">zcoords</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.infer_unrestrained_particle_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.infer_unrestrained_particle_coords" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>if a given particle (and direct neighbors) have no restraints. Infer</dt><dd><p>the coordinates by linear interpolation using closest particles with
restraints.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xcoords</strong> – list of x coordinates</p></li>
<li><p><strong>ycoords</strong> – list of y coordinates</p></li>
<li><p><strong>zcoords</strong> – list of z coordinates</p></li>
</ul>
</dd>
<dt class="field-even">Transforms</dt>
<dd class="field-even"><p>three input lists corresponding to the x, y and z
coordinates of the FULL model</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.interactions">
<code class="sig-name descname">interactions</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">cutoff=None</em>, <em class="sig-param">steps=(1</em>, <em class="sig-param">2</em>, <em class="sig-param">3</em>, <em class="sig-param">4</em>, <em class="sig-param">5)</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">error=False</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">savedata=None</em>, <em class="sig-param">average=True</em>, <em class="sig-param">plot=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.interactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.interactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots, for each particle, the number of interactions (particles closer
than the given cut-off). The value given is the average for all models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the contact map will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the contact map only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – distance cutoff (nm) to define whether two particles
are in contact or not, default is 2 times resolution, times scale.</p></li>
<li><p><strong>2</strong><strong>, </strong><strong>3</strong><strong>, </strong><strong>4</strong><strong>, </strong><strong>5</strong><strong>) </strong><strong>steps</strong> (<em>(</em><em>1</em><em>,</em>) – how many particles to group for the
estimation. By default 5 curves are drawn</p></li>
<li><p><strong>error</strong> (<em>False</em>) – represent the error of the estimates</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the contact map data
generated, in three columns format (particle1, particle2, percentage
of models where these two particles are in contact)</p></li>
<li><p><strong>average</strong> (<em>True</em>) – calculate average interactions along models,
otherwise, the median.</p></li>
<li><p><strong>plot</strong> (<em>True</em>) – e.g. only saves data. No plotting done</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.median_3d_dist">
<code class="sig-name descname">median_3d_dist</code><span class="sig-paren">(</span><em class="sig-param">part1</em>, <em class="sig-param">part2</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">plot=True</em>, <em class="sig-param">median=True</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.median_3d_dist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.median_3d_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the median distance between two particles over a set of models</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>part1</strong> – number corresponding to the first particle</p></li>
<li><p><strong>part2</strong> – number corresponding to the second particle</p></li>
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the distance will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the distance only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>plot</strong> (<em>True</em>) – if True, display a histogram and a box-plot of the
distribution of the calculated distances. If False, return either
the full list of the calculated distances or their median value</p></li>
<li><p><strong>median</strong> (<em>True</em>) – return either the full list of the calculated
distances (False) or their median value (True), when ‘plot’ is set
to False</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>if ‘plot’ is False, return either the full list of the
calculated distances or their median value distances, either the
list of distances.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.model_consistency">
<code class="sig-name descname">model_consistency</code><span class="sig-paren">(</span><em class="sig-param">cutoffs=None</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">savedata=None</em>, <em class="sig-param">plot=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.model_consistency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.model_consistency" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the particle consistency, over a given set of models, vs the
modeled region bins. The consistency is a measure of the variability
(or stability) of the modeled region (the higher the consistency value,
the higher stability).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cutoffs</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – list of distance cutoffs (nm) used to compute the
consistency. Two particle are considered consistent if their distance
is less than the given cutoff, default is a tuple of 0.5, 1, 1.5 and
2 times resolution, times scale.</p></li>
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the consistency will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the consistency only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>tmp_path</strong> (<em>'/tmp/tmp_cons'</em>) – location of the input files for
TM-score program</p></li>
<li><p><strong>tmsc</strong> (<em>''</em>) – path to the TMscore_consistency script (assumed to be
installed by default)</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the consistency data
generated (1 column per cutoff + 1 for particle number).</p></li>
<li><p><strong>plot</strong> (<em>True</em>) – e.g. only saves data. No plotting done</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.objective_function_model">
<code class="sig-name descname">objective_function_model</code><span class="sig-paren">(</span><em class="sig-param">model</em>, <em class="sig-param">log=False</em>, <em class="sig-param">smooth=True</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.objective_function_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.objective_function_model" title="Permalink to this definition">¶</a></dt>
<dd><p>This function plots the objective function value per each Monte-Carlo
step</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – the number of the model to plot</p></li>
<li><p><strong>log</strong> (<em>False</em>) – log plot</p></li>
<li><p><strong>smooth</strong> (<em>True</em>) – curve smoothing</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.particle_coordinates">
<code class="sig-name descname">particle_coordinates</code><span class="sig-paren">(</span><em class="sig-param">part</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.particle_coordinates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.particle_coordinates" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the mean coordinate of a given particle in a group of models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>part</strong> – the index number of a particle</p></li>
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the angle will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the angle only for the models in the
cluster number ‘cluster’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.save_models">
<code class="sig-name descname">save_models</code><span class="sig-paren">(</span><em class="sig-param">outfile</em>, <em class="sig-param">minimal=()</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.save_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.save_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves all the models in pickle format (python object written to disk).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_f</strong> – path where to save the pickle file</p></li>
<li><p><strong>minimal</strong> (<em>(</em><em>)</em>) – list of items to exclude from save. Options:
- ‘restraints’: used for modeling common to all models
- ‘zscores’: used generate restraints common to all models
- ‘original_data’: used generate Z-scores common to all models
- ‘log_objfun’: generated during modeling model specific</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.view_centroid">
<code class="sig-name descname">view_centroid</code><span class="sig-paren">(</span><em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.view_centroid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.view_centroid" title="Permalink to this definition">¶</a></dt>
<dd><p>shortcut for
view_models(tool=’plot’, show=’highlighted’, highlight=’centroid’)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>kwargs</strong> – any parameters to be passed to view_models (i.e.:
view_centroid(azimuth=30, elevation=10, show_axe=True, label=True))</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.view_models">
<code class="sig-name descname">view_models</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">tool='chimera'</em>, <em class="sig-param">show='all'</em>, <em class="sig-param">highlight='centroid'</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">cmd=None</em>, <em class="sig-param">color='index'</em>, <em class="sig-param">align=True</em>, <em class="sig-param">smooth=False</em>, <em class="sig-param">particle_size=50</em>, <em class="sig-param">lw_main=3</em>, <em class="sig-param">alpha_part=0.5</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.view_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.view_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize a selected model in the three dimensions (either with Chimera
or through matplotlib).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the visualization will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the visualization only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>tool</strong> (<em>'chimera'</em>) – path to the external tool used to visualize the
model. Can also be ‘plot’, to use matplotlib.</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image OR movie
generated (depending on the extension; accepted formats are png, mov
and webm). if set to None, the image or movie will be shown using
the default GUI.</p></li>
<li><p><strong>alpha</strong> (<em>False</em>) – only for matplotlib (‘plot’ option), transparency of
particles.</p></li>
<li><p><strong>smooth</strong> (<em>False</em>) – only for matplotlib (‘plot’ option), spline smoothing
(by default smoothing is 0.001).</p></li>
<li><p><strong>particle_size</strong> (<em>50</em>) – only for matplotlib (‘plot’ option), redefine
size of particles. If None, resolution times scale is used.</p></li>
<li><p><strong>color</strong> (<em>'index'</em>) – <p>can be:</p>
<ul>
<li><dl class="simple">
<dt>a string as:</dt><dd><ul>
<li><p>’<strong>index</strong>’ to color particles according to their position in the
model (<a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.color_residues" title="pytadbit.utils.extraviews.color_residues"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.color_residues()</span></code></a>)</p></li>
<li><p>’<strong>tad</strong>’ to color particles according to the TAD they belong to
(<a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.tad_coloring" title="pytadbit.utils.extraviews.tad_coloring"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.tad_coloring()</span></code></a>)</p></li>
<li><p>’<strong>border</strong>’ to color particles marking borders. Color according to
their score (<a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.tad_border_coloring" title="pytadbit.utils.extraviews.tad_border_coloring"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.tad_border_coloring()</span></code></a>)
coloring function like.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>a function, that takes as argument a model and any other parameter
passed through the kwargs.</p></li>
<li><p>a list of (r, g, b) tuples (as long as the number of particles).
Each r, g, b between 0 and 1.</p></li>
</ul>
</p></li>
<li><p><strong>highlight</strong> (<em>'centroid'</em>) – higlights a given model, or group of models.
Can be either ‘all’, ‘centroid’ or ‘best’ (‘best’ being the model
with the lowest IMP objective function value</p></li>
<li><p><strong>show</strong> (<em>'all'</em>) – models to be displayed. Can be either ‘all’, ‘grid’
or ‘highlighted’.</p></li>
<li><p><strong>cmd</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – <p>list of commands to be passed to the viewer. The chimera list is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">focus</span>
<span class="nb">set</span> <span class="n">bg_color</span> <span class="n">white</span>
<span class="n">windowsize</span> <span class="mi">800</span> <span class="mi">600</span>
<span class="n">bonddisplay</span> <span class="n">never</span> <span class="c1">#0</span>
<span class="n">represent</span> <span class="n">wire</span>
<span class="n">shape</span> <span class="n">tube</span> <span class="c1">#0 radius 5 bandLength 100 segmentSubdivisions 1 followBonds on</span>
<span class="n">clip</span> <span class="n">yon</span> <span class="o">-</span><span class="mi">500</span>
<span class="o">~</span><span class="n">label</span>
<span class="nb">set</span> <span class="n">subdivision</span> <span class="mi">1</span>
<span class="nb">set</span> <span class="n">depth_cue</span>
<span class="nb">set</span> <span class="n">dc_color</span> <span class="n">black</span>
<span class="nb">set</span> <span class="n">dc_start</span> <span class="mf">0.5</span>
<span class="nb">set</span> <span class="n">dc_end</span> <span class="mi">1</span>
<span class="n">scale</span> <span class="mf">0.8</span>
</pre></div>
</div>
<p>Followed by the movie command to record movies:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">movie</span> <span class="n">record</span> <span class="n">supersample</span> <span class="mi">1</span>
<span class="n">turn</span> <span class="n">y</span> <span class="mi">3</span> <span class="mi">120</span>
<span class="n">wait</span> <span class="mi">120</span>
<span class="n">movie</span> <span class="n">stop</span>
<span class="n">movie</span> <span class="n">encode</span> <span class="n">output</span> <span class="n">SAVEFIG</span>
</pre></div>
</div>
<p>Or the copy command for images:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">copy</span> <span class="n">file</span> <span class="n">SAVEFIG</span> <span class="n">png</span>
</pre></div>
</div>
<p>Passing as the following list as ‘cmd’ parameter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;focus&#39;</span><span class="p">,</span> <span class="s1">&#39;set bg_color white&#39;</span><span class="p">,</span> <span class="s1">&#39;windowsize 800 600&#39;</span><span class="p">,</span> <span class="s1">&#39;bonddisplay never #0&#39;</span><span class="p">,</span> <span class="s1">&#39;shape tube #0 radius 10 bandLength 200 segmentSubdivisions 100 followBonds on&#39;</span><span class="p">,</span> <span class="s1">&#39;clip yon -500&#39;</span><span class="p">,</span> <span class="s1">&#39;~label&#39;</span><span class="p">,</span> <span class="s1">&#39;set subdivision 1&#39;</span><span class="p">,</span> <span class="s1">&#39;set depth_cue&#39;</span><span class="p">,</span> <span class="s1">&#39;set dc_color black&#39;</span><span class="p">,</span> <span class="s1">&#39;set dc_start 0.5&#39;</span><span class="p">,</span> <span class="s1">&#39;set dc_end 1&#39;</span><span class="p">,</span> <span class="s1">&#39;scale 0.8&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>will return the default image (other commands can be passed to
modified the final image/movie).</p>
</p></li>
<li><p><strong>align</strong> (<em>True</em>) – show aligned models</p></li>
<li><p><strong>radius</strong> (<em>15</em>) – radius for the chimera particles</p></li>
<li><p><strong>kwargs</strong> – see <a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.plot_3d_model" title="pytadbit.utils.extraviews.plot_3d_model"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.plot_3d_model()</span></code></a> or
<a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.chimera_view" title="pytadbit.utils.extraviews.chimera_view"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.chimera_view()</span></code></a> for other arguments
to pass to this function. See also coloring function</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.walking_angle">
<code class="sig-name descname">walking_angle</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">steps=(1</em>, <em class="sig-param">3)</em>, <em class="sig-param">signed=True</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">savedata=None</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">plot=True</em>, <em class="sig-param">error=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.walking_angle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.walking_angle" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the angle between successive loci in a given model or set of
models. In order to limit the noise of the measure angle is calculated
between 3 loci, between each are two other loci. E.g. in the scheme
bellow, angle are calculated between loci A, D and G.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) all models will be used for
computation. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the angle only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>3</strong><strong>) </strong><strong>steps</strong> (<em>(</em><em>1</em><em>,</em>) – how many particles to group for the estimation.
By default 2 curves are drawn</p></li>
<li><p><strong>signed</strong> (<em>True</em>) – whether to compute the sign of the angle according
to a normal plane, or not.</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>error</strong> (<em>False</em>) – represent the error of the estimates</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>plot</strong> (<em>True</em>) – e.g. if False, only saves data. No plotting done</p></li>
<li><p><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the angle data
generated (1 column per step + 1 for particle number).</p></li>
</ul>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                      <span class="n">C</span><span class="o">..........</span><span class="n">D</span>
                   <span class="o">...</span>            <span class="o">...</span>
                <span class="o">...</span>                 <span class="o">...</span>
             <span class="o">...</span>                       <span class="o">...</span>
  <span class="n">A</span><span class="o">..........</span><span class="n">B</span>                            <span class="o">.</span><span class="n">E</span>
 <span class="o">..</span>                                        <span class="o">.</span>
<span class="o">.</span>                                          <span class="o">.</span>
                                           <span class="o">.</span>                 <span class="o">.</span>
                                           <span class="o">.</span>                <span class="o">.</span>
                                           <span class="n">F</span><span class="o">...............</span><span class="n">G</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.walking_dihedral">
<code class="sig-name descname">walking_dihedral</code><span class="sig-paren">(</span><em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">steps=(1</em>, <em class="sig-param">3)</em>, <em class="sig-param">span=(-2</em>, <em class="sig-param">1</em>, <em class="sig-param">0</em>, <em class="sig-param">1</em>, <em class="sig-param">3)</em>, <em class="sig-param">error=False</em>, <em class="sig-param">plot=True</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">axe=None</em>, <em class="sig-param">savedata=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.walking_dihedral"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.walking_dihedral" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the dihedral angle between successive planes. A plane is formed by
3 successive loci.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – if None (default) the dihedral angle will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – compute the dihedral angle only for the models in the
cluster number ‘cluster’</p></li>
<li><p><strong>3</strong><strong>) </strong><strong>steps</strong> (<em>(</em><em>1</em><em>,</em>) – how many particles to group for the estimation.
By default 2 curves are drawn</p></li>
<li><p><strong>signed</strong> (<em>True</em>) – whether to compute the sign of the angle according
to a normal plane, or not.</p></li>
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>-1</strong><strong>, </strong><strong>0</strong><strong>, </strong><strong>1</strong><strong>, </strong><strong>2</strong><strong>) </strong><strong>span</strong> (<em>(</em><em>-3</em><em>,</em>) – by default computes angle between a plane
comprising the current residue (<em>0</em>), the residue after (<em>1</em>) and the
residue 3 positions before (<em>-3</em>), and the plane with the current
residue, the residue directly after (<em>1</em>) and the residue 3 positions
after (<em>3</em>). e.g. In the scheme bellow it’s plane ACD vs plane CDF.</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>error</strong> (<em>False</em>) – represent the error of the estimates</p></li>
<li><p><strong>plot</strong> (<em>True</em>) – e.g. if False, only saves data. No plotting done</p></li>
<li><p><strong>savedata</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the angle data
generated (1 column per step + 1 for particle number).</p></li>
</ul>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                      <span class="n">C</span><span class="o">..........</span><span class="n">D</span>
                   <span class="o">...</span>            <span class="o">...</span>
                <span class="o">...</span>                 <span class="o">...</span>
             <span class="o">...</span>                       <span class="o">...</span>
  <span class="n">A</span><span class="o">..........</span><span class="n">B</span>                            <span class="o">.</span><span class="n">E</span>
 <span class="o">..</span>                                        <span class="o">.</span>
<span class="o">.</span>                                          <span class="o">.</span>
                                           <span class="o">.</span>                 <span class="o">.</span>
                                           <span class="o">.</span>                <span class="o">.</span>
                                           <span class="n">F</span><span class="o">...............</span><span class="n">G</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.write_cmm">
<code class="sig-name descname">write_cmm</code><span class="sig-paren">(</span><em class="sig-param">directory</em>, <em class="sig-param">model_num=None</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">color='index'</em>, <em class="sig-param">rndname=True</em>, <em class="sig-param">infer_unrestrained=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.write_cmm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.write_cmm" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a model in the cmm format, read by Chimera
(<a class="reference external" href="http://www.cgl.ucsf.edu/chimera">http://www.cgl.ucsf.edu/chimera</a>).</p>
<p><strong>Note:</strong> If none of model_num, models or cluster parameter are set,
ALL the models will be written.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory</strong> – location where the file will be written (note: the
name of the file will be model_1.cmm if model number is 1)</p></li>
<li><p><strong>model_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – the number of the model to save</p></li>
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a list of numbers corresponding to a given set of
models to save</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – save the models in the cluster number ‘cluster’</p></li>
<li><p><strong>rndname</strong> (<em>True</em>) – If True, file names will be formatted as:
model.RND.cmm, where RND is the random number feed used by IMP to
generate the corresponding model. If False, the format will be:
model_NUM_RND.cmm where NUM is the rank of the model in terms of
objective function value</p></li>
<li><p><strong>color</strong> (<em>'index'</em>) – <p>can be:</p>
<ul>
<li><dl class="simple">
<dt>a string as:</dt><dd><ul>
<li><p>’<strong>index</strong>’ to color particles according to their position in the
model (<a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.color_residues" title="pytadbit.utils.extraviews.color_residues"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.color_residues()</span></code></a>)</p></li>
<li><p>’<strong>tad</strong>’ to color particles according to the TAD they belong to
(<a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.tad_coloring" title="pytadbit.utils.extraviews.tad_coloring"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.tad_coloring()</span></code></a>)</p></li>
<li><p>’<strong>border</strong>’ to color particles marking borders. Color according to
their score (<a class="reference internal" href="reference_utils.html#pytadbit.utils.extraviews.tad_border_coloring" title="pytadbit.utils.extraviews.tad_border_coloring"><code class="xref py py-func docutils literal notranslate"><span class="pre">pytadbit.utils.extraviews.tad_border_coloring()</span></code></a>)
coloring function like.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>a function, that takes as argument a model and any other parameter
passed through the kwargs.</p></li>
<li><p>a list of (r, g, b) tuples (as long as the number of particles).
Each r, g, b between 0 and 1.</p></li>
</ul>
</p></li>
<li><p><strong>kwargs</strong> – any extra argument will be passed to the coloring
function</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.write_json">
<code class="sig-name descname">write_json</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">color='index'</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">title=None</em>, <em class="sig-param">infer_unrestrained=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.write_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.write_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a model in the json format, read by TADkit.</p>
<p><strong>Note:</strong> If none of model_num, models or cluster parameter are set,
ALL the models will be written.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory</strong> – location where the file will be written (note: the
name of the file will be model_1.cmm if model number is 1)</p></li>
<li><p><strong>model_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – the number of the model to save</p></li>
<li><p><strong>rndname</strong> (<em>True</em>) – If True, file names will be formatted as:
model.RND.cmm, where RND is the random number feed used by IMP to
generate the corresponding model. If False, the format will be:
model_NUM_RND.cmm where NUM is the rank of the model in terms of
objective function value</p></li>
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – overide the default file name writing</p></li>
<li><p><strong>kwargs</strong> – any extra argument will be passed to the coloring
function</p></li>
<li><p><strong>infer_unrestrained</strong> (<em>False</em>) – infer unrestrained particle
coordinates by linear interpolation using closest particles with
restraints</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.write_xyz">
<code class="sig-name descname">write_xyz</code><span class="sig-paren">(</span><em class="sig-param">directory</em>, <em class="sig-param">model_num=None</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">get_path=False</em>, <em class="sig-param">rndname=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.write_xyz"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.write_xyz" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a xyz file containing the 3D coordinates of each particle in the
model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If none of model_num, models or cluster parameter are set,
ALL the models will be written.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory</strong> – location where the file will be written (note: the
file name will be model.1.xyz, if the model number is 1)</p></li>
<li><p><strong>model_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – the number of the model to save</p></li>
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a list of numbers corresponding to a given set of
models to be written</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – save the models in the cluster number ‘cluster’</p></li>
<li><p><strong>rndname</strong> (<em>True</em>) – If True, file names will be formatted as:
model.RND.xyz, where RND is the random number feed used by IMP to
generate the corresponding model. If False, the format will be:
model_NUM_RND.xyz where NUM is the rank of the model in terms of
objective function value</p></li>
<li><p><strong>get_path</strong> (<em>False</em>) – whether to return, or not, the full path where
the file has been written</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.write_xyz_babel">
<code class="sig-name descname">write_xyz_babel</code><span class="sig-paren">(</span><em class="sig-param">directory</em>, <em class="sig-param">model_num=None</em>, <em class="sig-param">models=None</em>, <em class="sig-param">cluster=None</em>, <em class="sig-param">get_path=False</em>, <em class="sig-param">rndname=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.write_xyz_babel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.write_xyz_babel" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a xyz file containing the 3D coordinates of each particle in the
model using a file format compatible with babel
(<a class="reference external" href="http://openbabel.org/wiki/XYZ_%28format%29">http://openbabel.org/wiki/XYZ_%28format%29</a>).
.. note:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">If</span> <span class="n">none</span> <span class="n">of</span> <span class="n">model_num</span><span class="p">,</span> <span class="n">models</span> <span class="ow">or</span> <span class="n">cluster</span> <span class="n">parameter</span> <span class="n">are</span> <span class="nb">set</span><span class="p">,</span>
<span class="n">ALL</span> <span class="n">the</span> <span class="n">models</span> <span class="n">will</span> <span class="n">be</span> <span class="n">written</span><span class="o">.</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory</strong> – location where the file will be written (note: the
file name will be model.1.xyz, if the model number is 1)</p></li>
<li><p><strong>model_num</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – the number of the model to save</p></li>
<li><p><strong>models</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a list of numbers corresponding to a given set of
models to be written</p></li>
<li><p><strong>cluster</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – save the models in the cluster number ‘cluster’</p></li>
<li><p><strong>rndname</strong> (<em>True</em>) – If True, file names will be formatted as:
model.RND.xyz, where RND is the random number feed used by IMP to
generate the corresponding model. If False, the format will be:
model_NUM_RND.xyz where NUM is the rank of the model in terms of
objective function value</p></li>
<li><p><strong>get_path</strong> (<em>False</em>) – whether to return, or not, the full path where
the file has been written</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pytadbit.modelling.structuralmodels.StructuralModels.zscore_plot">
<code class="sig-name descname">zscore_plot</code><span class="sig-paren">(</span><em class="sig-param">axe=None</em>, <em class="sig-param">savefig=None</em>, <em class="sig-param">do_normaltest=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytadbit/modelling/structuralmodels.html#StructuralModels.zscore_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.modelling.structuralmodels.StructuralModels.zscore_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate 3 plots. Two heatmaps of the Z-scores used for modeling, one
of which is binary showing in red Z-scores higher than upper cut-off;
and in blue Z-scores lower than lower cut-off. Last plot is an histogram
of the distribution of Z-scores, showing selected regions. Histogram
also shows the fit to normal distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axe</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a matplotlib.axes.Axes object to define the plot
appearance</p></li>
<li><p><strong>savefig</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</p></li>
<li><p><strong>do_normaltest</strong> (<em>False</em>) – to display the result of a test of normality
(D’Agostino and Pearson’s test, that combines skew and kurtosis).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">TADbit tools</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_mapping.html">From FASTQ to matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impoptimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_modelling_impmodel.html">IMPmodel class</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">StructuralModels class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_utils.html">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_modelling_impmodel.html"
                        title="previous chapter">IMPmodel class</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_utils.html"
                        title="next chapter">Utilities</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="reference_utils.html" title="Utilities"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_modelling_impmodel.html" title="IMPmodel class"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Tadbit 0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Reference Guide</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Last updated on Feb 07, 2020.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.3.1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>