<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Utilities &mdash; Tadbit 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/forkme_nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/TADbit_icon.ico"/>
    <link rel="top" title="Tadbit 0.1 documentation" href="../index.html" />
    <link rel="up" title="Reference Guide" href="index.html" />
    <link rel="next" title="Bibliography" href="../biblio.html" />
    <link rel="prev" title="StructuralModels class" href="reference_imp_structuralmodels.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../biblio.html" title="Bibliography"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_imp_structuralmodels.html" title="StructuralModels class"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="utilities">
<h1>Utilities<a class="headerlink" href="#utilities" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="pytadbit.utils.tadmaths.calinski_harabasz">
<tt class="descclassname">pytadbit.utils.tadmaths.</tt><tt class="descname">calinski_harabasz</tt><big>(</big><em>scores</em>, <em>clusters</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/tadmaths.html#calinski_harabasz"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.tadmaths.calinski_harabasz" title="Permalink to this definition">¶</a></dt>
<dd><p>Implementation of the CH score <a class="reference internal" href="../biblio.html#calinskiharabasz1974" id="id1">[CalinskiHarabasz1974]</a>, that has shown to be
one the most accurate way to compare clustering methods
<a class="reference internal" href="../biblio.html#milligancooper1985" id="id2">[MilliganCooper1985]</a> <a class="reference internal" href="../biblio.html#tibshirani2001" id="id3">[Tibshirani2001]</a>.</p>
<p>The CH score is:</p>
<div class="math">
\[CH(k) = \frac{B(k) / (k-1)}{W(k)/(n - k)}\]</div>
<p>Where <span class="math">\(B(k)\)</span> and <span class="math">\(W(k)\)</span> are the between and within cluster sums
of squares, with <span class="math">\(k\)</span> clusters, and <span class="math">\(n\)</span> the total number of
elements (models in this case).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>scores</strong> &#8211; a dict with, as keys, a tuple with a pair of models; and, as
value, the distance between these models.</li>
<li><strong>clusters</strong> &#8211; a dict with, as key, the cluster number, and as value a
list of models</li>
<li><strong>nmodels</strong> &#8211; total number of models</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the CH score</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.plot_3d_optimization_result">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">plot_3d_optimization_result</tt><big>(</big><em>result</em>, <em>axes=('scale'</em>, <em>'maxdist'</em>, <em>'upfreq'</em>, <em>'lowfreq')</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/extraviews.html#plot_3d_optimization_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.extraviews.plot_3d_optimization_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Displays a three dimensional scatter plot representing the result of the
optimization.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>result</strong> &#8211; 3D numpy array contating correlation values</li>
<li><strong>axes</strong> (<em>&#8216;scale&#8217;,&#8217;maxdist&#8217;,&#8217;upfreq&#8217;,&#8217;lowfreq&#8217;</em>) &#8211; tuple of axes to
represent. The order will define which parameter will be placed on the
w, z, y or x axe.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.plot_2d_optimization_result">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">plot_2d_optimization_result</tt><big>(</big><em>result</em>, <em>axes=('scale'</em>, <em>'maxdist'</em>, <em>'upfreq'</em>, <em>'lowfreq')</em>, <em>show_best=0</em>, <em>skip=None</em>, <em>savefig=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/extraviews.html#plot_2d_optimization_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.extraviews.plot_2d_optimization_result" title="Permalink to this definition">¶</a></dt>
<dd><p>A grid of heatmaps representing the result of the optimization.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>result</strong> &#8211; 3D numpy array contating correlation values</li>
<li><strong>axes</strong> (<em>&#8216;scale&#8217;,&#8217;maxdist&#8217;,&#8217;upfreq&#8217;,&#8217;lowfreq&#8217;</em>) &#8211; tuple of axes to
represent. The order will define which parameter will be placed on the
w, z, y or x axe.</li>
<li><strong>show_best</strong> (<em>0</em>) &#8211; number of best correlation value to identifie.</li>
<li><strong>skip</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a dict can be passed here in order to fix a given axe,
e.g.: {&#8216;scale&#8217;: 0.001, &#8216;maxdist&#8217;: 500}</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.compare_models">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">compare_models</tt><big>(</big><em>sm1</em>, <em>sm2</em>, <em>cutoff=150</em>, <em>models1=None</em>, <em>cluster1=None</em>, <em>models2=None</em>, <em>cluster2=None</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/extraviews.html#compare_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.extraviews.compare_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the difference of contact maps of two group of structural models.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sm1</strong> &#8211; a StructuralModel</li>
<li><strong>sm2</strong> &#8211; a StructuralModel</li>
<li><strong>dcutoff</strong> (<em>150</em>) &#8211; distance threshold (nm) to determine if two
particles are in contact</li>
<li><strong>models</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; if None (default) the contact map will be computed
using all the models. A list of numbers corresponding to a given set
of models can be passed</li>
<li><strong>cluster</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; compute the contact map only for the models in the
cluster number &#8216;cluster&#8217;</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.color_residues">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">color_residues</tt><big>(</big><em>x</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/extraviews.html#color_residues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.extraviews.color_residues" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to color residues from blue to red.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>model</strong> &#8211; a given <a class="reference internal" href="reference_imp_model.html#pytadbit.imp.impmodel.IMPmodel" title="pytadbit.imp.impmodel.IMPmodel"><tt class="xref py py-class docutils literal"><span class="pre">pytadbit.imp.impmodel.IMPmodel</span></tt></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a list of rgb tuples (red, green, blue), each between 0 and 1.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.tad_coloring">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">tad_coloring</tt><big>(</big><em>x</em>, <em>mstart=None</em>, <em>mend=None</em>, <em>tads=None</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#pytadbit.utils.extraviews.tad_coloring" title="Permalink to this definition">¶</a></dt>
<dd><p>Colors TADs from blue to red (first to last TAD). TAD borders are displayed
in scale of grey, from light to dark grey (again first to last border)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>model</strong> &#8211; a given <a class="reference internal" href="reference_imp_model.html#pytadbit.imp.impmodel.IMPmodel" title="pytadbit.imp.impmodel.IMPmodel"><tt class="xref py py-class docutils literal"><span class="pre">pytadbit.imp.impmodel.IMPmodel</span></tt></a></li>
<li><strong>tads</strong> &#8211; a dictionary of TADs, Experiments.tads can be directly passed</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a list of rgb tuples (red, green, blue), each between 0 and 1.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.tad_border_coloring">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">tad_border_coloring</tt><big>(</big><em>x</em>, <em>mstart=None</em>, <em>mend=None</em>, <em>tads=None</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#pytadbit.utils.extraviews.tad_border_coloring" title="Permalink to this definition">¶</a></dt>
<dd><p>Colors TAD borders from blue to red (bad to good score). TAD are displayed
in scale of grey, from light to dark grey (first to last particle in the
TAD)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>model</strong> &#8211; a given <a class="reference internal" href="reference_imp_model.html#pytadbit.imp.impmodel.IMPmodel" title="pytadbit.imp.impmodel.IMPmodel"><tt class="xref py py-class docutils literal"><span class="pre">pytadbit.imp.impmodel.IMPmodel</span></tt></a></li>
<li><strong>tads</strong> &#8211; a dictionary of TADs, Experiments.tads can be directly passed</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a list of rgb tuples (red, green, blue), each between 0 and 1.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.plot_3d_model">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">plot_3d_model</tt><big>(</big><em>x</em>, <em>y</em>, <em>z</em>, <em>label=False</em>, <em>axe=None</em>, <em>thin=False</em>, <em>savefig=None</em>, <em>show_axe=False</em>, <em>azimuth=-90</em>, <em>elevation=0.0</em>, <em>color='index'</em>, <em>**kwargs</em><big>)</big><a class="headerlink" href="#pytadbit.utils.extraviews.plot_3d_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a 3 lists of coordinates (x, y, z) plots a three-dimentional model
using matplotlib</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>model</strong> &#8211; a <a class="reference internal" href="reference_imp_model.html#pytadbit.imp.impmodel.IMPmodel" title="pytadbit.imp.impmodel.IMPmodel"><tt class="xref py py-class docutils literal"><span class="pre">pytadbit.imp.impmodel.IMPmodel</span></tt></a></li>
<li><strong>label</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; show labels</li>
<li><strong>axe</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; a matplotlib.axes.Axes object to define the plot
appearance</li>
<li><strong>thin</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; draw a thin black line instead of representing particles
and edges</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save the image generated;
if None, the image will be shown using matplotlib GUI (the extension
of the file name will determine the desired format).</li>
<li><strong>azimuth</strong> (<em>-90</em>) &#8211; angle to rotate camera along the y axis</li>
<li><strong>elevation</strong> (<em>0</em>) &#8211; angle to rotate camera along the x axis</li>
<li><strong>color</strong> (<em>&#8216;index&#8217;</em>) &#8211; <p>can be:</p>
<ul>
<li><dl class="first docutils">
<dt>a string as:</dt>
<dd><ul class="first last">
<li>&#8216;<strong>index</strong>&#8216; to color particles according to their position in the
model (<a class="reference internal" href="#pytadbit.utils.extraviews.color_residues" title="pytadbit.utils.extraviews.color_residues"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.utils.extraviews.color_residues()</span></tt></a>)</li>
<li>&#8216;<strong>tad</strong>&#8216; to color particles according to the TAD they belong to
(<a class="reference internal" href="#pytadbit.utils.extraviews.tad_coloring" title="pytadbit.utils.extraviews.tad_coloring"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.utils.extraviews.tad_coloring()</span></tt></a>)</li>
<li>&#8216;<strong>border</strong>&#8216; to color particles marking borders. Color according to
their score (<a class="reference internal" href="#pytadbit.utils.extraviews.tad_border_coloring" title="pytadbit.utils.extraviews.tad_border_coloring"><tt class="xref py py-func docutils literal"><span class="pre">pytadbit.utils.extraviews.tad_border_coloring()</span></tt></a>)
coloring function like.</li>
</ul>
</dd>
</dl>
</li>
<li>a function, that takes as argument a model and any other parameter
passed through the kwargs.</li>
<li>a list of (r, g, b) tuples (as long as the number of particles).
Each r, g, b between 0 and 1.</li>
</ul>
</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.extraviews.chimera_view">
<tt class="descclassname">pytadbit.utils.extraviews.</tt><tt class="descname">chimera_view</tt><big>(</big><em>cmm_files</em>, <em>chimera_bin='chimera'</em>, <em>chimera_cmd=None</em>, <em>savefig=None</em>, <em>center_of_mass=False</em>, <em>gyradius=False</em>, <em>align=True</em>, <em>grid=False</em>, <em>highlight='all'</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/extraviews.html#chimera_view"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.extraviews.chimera_view" title="Permalink to this definition">¶</a></dt>
<dd><p>Open a list of .cmm files with Chimera (<a class="reference external" href="http://www.cgl.ucsf.edu/chimera">http://www.cgl.ucsf.edu/chimera</a>)
to view models.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cmm_files</strong> &#8211; list of .cmm files</li>
<li><strong>chimera_bin</strong> (<em>&#8216;chimera&#8217;</em>) &#8211; path to chimera binary</li>
<li><strong>chimera_cmd</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; list of command lines for chimera</li>
<li><strong>savefig</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#None" title="(in Python v2.7)"><em>None</em></a>) &#8211; path to a file where to save generated image</li>
<li><strong>center_of_mass</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; if True, draws the center of mass</li>
<li><strong>gyradius</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; increment the radius of the center of mass by the
value of the radius of girations given.</li>
<li><strong>align</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; align models</li>
<li><strong>grid</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; tile models</li>
<li><strong>higlight</strong> (<em>&#8216;all&#8217;</em>) &#8211; higlights a given model, or group of models.
Can be either &#8216;all&#8217;, or a model number</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.hic_filtering.hic_filtering_for_modelling">
<tt class="descclassname">pytadbit.utils.hic_filtering.</tt><tt class="descname">hic_filtering_for_modelling</tt><big>(</big><em>matrx</em>, <em>method='mean'</em>, <em>silent=False</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/hic_filtering.html#hic_filtering_for_modelling"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.hic_filtering.hic_filtering_for_modelling" title="Permalink to this definition">¶</a></dt>
<dd><p>Main filtering function, to remove artefactual columns in a given Hi-C
matrix</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>matrx</strong> &#8211; Hi-C matrix of a given experiment</li>
<li><strong>method</strong> (<em>mean</em>) &#8211; method to use for filtering Hi-C columns. Aims to
remove columns with abnormally low count of interactions</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the indexes of the columns not to be considered for the
calculation of the z-score</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.three_dim_stats.calc_eqv_rmsd">
<tt class="descclassname">pytadbit.utils.three_dim_stats.</tt><tt class="descname">calc_eqv_rmsd</tt><big>(</big><em>models</em>, <em>nloci</em>, <em>dcutoff=200</em>, <em>one=False</em>, <em>what='score'</em>, <em>normed=True</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/three_dim_stats.html#calc_eqv_rmsd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.three_dim_stats.calc_eqv_rmsd" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the RMSD, dRMSD, the number of equivalent positions and a score
combining these three measures. The measure are done between a group of
models in a one against all manner.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>nloci</strong> &#8211; number of particles per model</li>
<li><strong>dcutoff</strong> (<em>200</em>) &#8211; distance in nanometer from which it is considered
that two particles are separated.</li>
<li><strong>fact</strong> (<em>0.75</em>) &#8211; Factor for equivalent positions</li>
<li><strong>one</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#False" title="(in Python v2.7)"><em>False</em></a>) &#8211; if True assumes that only two models are passed, and
returns the rmsd of their comparison</li>
<li><strong>what</strong> (<em>&#8216;score&#8217;</em>) &#8211; values to return. Can be one of &#8216;score&#8217;, &#8216;rmsd&#8217;,
&#8216;drmsd&#8217; or &#8216;eqv&#8217;</li>
<li><strong>normed</strong> (<a class="reference external" href="http://docs.python.org/library/constants.html#True" title="(in Python v2.7)"><em>True</em></a>) &#8211; normalize result by maximum value (only applies to rmsd
and drmsd)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><p>a score of each pairwise comparison according to:</p>
<div class="math">
\[score_i = eqvs_i \times \frac{dRMSD_i / max(dRMSD)}
                                {RMSD_i / max(RMSD)}\]</div>
<p>where <span class="math">\(eqvs_i\)</span> is the number of equivalent position for the ith
pairwise model comparison.</p>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.three_dim_stats.square_distance">
<tt class="descclassname">pytadbit.utils.three_dim_stats.</tt><tt class="descname">square_distance</tt><big>(</big><em>part1</em>, <em>part2</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/three_dim_stats.html#square_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.three_dim_stats.square_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the square distance between two particles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>part1</strong> &#8211; coordinate (dict format with x, y, z keys)</li>
<li><strong>part2</strong> &#8211; coordinate (dict format with x, y, z keys)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">square distance between two points in space</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.three_dim_stats.distance">
<tt class="descclassname">pytadbit.utils.three_dim_stats.</tt><tt class="descname">distance</tt><big>(</big><em>part1</em>, <em>part2</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/three_dim_stats.html#distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.three_dim_stats.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the distance between two particles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>part1</strong> &#8211; coordinate in list format (x, y, z)</li>
<li><strong>part2</strong> &#8211; coordinate in list format (x, y, z)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">distance between two points in space</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.three_dim_stats.angle_between_3_points">
<tt class="descclassname">pytadbit.utils.three_dim_stats.</tt><tt class="descname">angle_between_3_points</tt><big>(</big><em>point1</em>, <em>point2</em>, <em>point3</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/three_dim_stats.html#angle_between_3_points"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.three_dim_stats.angle_between_3_points" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the angle between 3 particles</p>
<p>Given three particles A, B and C, the angle g (angle ACB, shown below):</p>
<div class="highlight-python"><pre>      A
     /|
    /i|
  c/  |
  /   |
 /    |
B )g  |b
 \    |
  \   |
  a\  |
    \h|
     \|
      C</pre>
</div>
<p>is given by the theorem of Al-Kashi:</p>
<div class="math">
\[b^2 = a^2 + c^2 - 2ac\cos(g)\]</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>point1</strong> &#8211; list of 3 coordinate for x, y and z</li>
<li><strong>point2</strong> &#8211; list of 3 coordinate for x, y and z</li>
<li><strong>point3</strong> &#8211; list of 3 coordinate for x, y and z</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">angle in radians</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.three_dim_stats.dihedral">
<tt class="descclassname">pytadbit.utils.three_dim_stats.</tt><tt class="descname">dihedral</tt><big>(</big><em>a</em>, <em>b</em>, <em>c</em>, <em>d</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/three_dim_stats.html#dihedral"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.three_dim_stats.dihedral" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates dihedral angle between 4 points in 3D (array with x,y,z)</p>
</dd></dl>

<dl class="function">
<dt id="pytadbit.utils.three_dim_stats.generate_circle_points">
<tt class="descclassname">pytadbit.utils.three_dim_stats.</tt><tt class="descname">generate_circle_points</tt><big>(</big><em>x</em>, <em>y</em>, <em>z</em>, <em>u</em>, <em>v</em>, <em>w</em>, <em>n</em><big>)</big><a class="reference internal" href="../_modules/pytadbit/utils/three_dim_stats.html#generate_circle_points"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytadbit.utils.three_dim_stats.generate_circle_points" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of 3d coordinates of points on a circle using the
Rodrigues rotation formula.</p>
<p>see <em>Murray, G. (2013). Rotation About an Arbitrary Axis in 3 Dimensions</em>
for details</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>x</strong> &#8211; x coordinate of a point somewhere on the circle</li>
<li><strong>y</strong> &#8211; y coordinate of a point somewhere on the circle</li>
<li><strong>z</strong> &#8211; z coordinate of a point somewhere on the circle</li>
<li><strong>a</strong> &#8211; x coordinate of the center</li>
<li><strong>b</strong> &#8211; y coordinate of the center</li>
<li><strong>c</strong> &#8211; z coordinate of the center</li>
<li><strong>u</strong> &#8211; 1st element of a vector in the same plane as the circle</li>
<li><strong>v</strong> &#8211; 2nd element of a vector in the same plane as the circle</li>
<li><strong>w</strong> &#8211; 3rd element of a vector in the same plane as the circle</li>
<li><strong>n</strong> &#8211; number of points in the circle</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="docutils">
<dt>TODO: try simplification for a=b=c=0 (and do the translation in the main</dt>
<dd>function)</dd>
</dl>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/TADbit_logo_little.png" alt="Logo"/>
            </a></p>
<h3><a href="../index.html">Table Of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installing TADbit on GNU/Linux</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_chromosome.html">Chromosome class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_experiment.html">Experiment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_tadbit.html">TAD detection functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">TAD clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parser.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_aligner.html">Boundary Aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_boundary_alignment.html">Alignment class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_optimizer.html">IMPoptimizer class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_model.html">IMPmodel class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_imp_structuralmodels.html">StructuralModels class</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../biblio.html">Bibliography</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_imp_structuralmodels.html"
                        title="previous chapter">StructuralModels class</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../biblio.html"
                        title="next chapter">Bibliography</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../biblio.html" title="Bibliography"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_imp_structuralmodels.html" title="StructuralModels class"
             >previous</a> |</li>
        <li><a href="../index.html">Tadbit 0.1 documentation</a> &raquo;</li>
          <li><a href="index.html" >Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright .
      Last updated on Mar 31, 2014.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div><style>div.related ul {padding-right:150px;}</style>
        <a href="https://github.com/3DGenomes/tadbit" target="_blank"><img
            style="position: absolute; top: 0; right: 0; border: 0;"
            src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"></a>
  </body>
</html>